{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1e30ed4-8a0e-4a6f-96eb-dfe116dfb2de",
   "metadata": {},
   "source": [
    "#### Directories etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "cf9c6377-5080-4c22-94db-feabe17cc47c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "import random\n",
    "from tqdm import tqdm\n",
    "from Comparative_Analysis import Utilities as util\n",
    "import random\n",
    "from joblib import Parallel, delayed\n",
    "from Comparative_Analysis import Blast_Functions as blastfn\n",
    "import time\n",
    "import os\n",
    "import copy\n",
    "from scipy.stats import chi2, chi2_contingency\n",
    "from Comparative_Analysis import ORF_Functions as orffn\n",
    "from Comparative_Analysis import Alignment as alignfn\n",
    "from random import sample\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import subprocess\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1988a242-f558-4643-b3a9-27272fa7bebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'F:/Project_Data/Intergenic_Region_Comparative_Analysis'\n",
    "seq_dir_1 = 'F:/Datasets/NCBI_Refseq_Mycobacteriaceae_All_Levels/ncbi_dataset/data'\n",
    "seq_dir_2 = 'F:/Datasets/NCBI_Refseq_Actinobacteria_All_Levels/data'\n",
    "sequence_dirs_1 = util.list_dirs(seq_dir_1)\n",
    "sequence_dirs_2 = util.list_dirs(seq_dir_2)\n",
    "tb_species = 'NC_000962.3' \n",
    "tb_annotation_dirname = 'GCF_000195955.2'\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "muscle_exe = 'C:/Users/nicho/Muscle/muscle3.8.31_i86win32.exe'\n",
    "full_build = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ae80b7b2-f785-4540-affe-e69650a62827",
   "metadata": {},
   "outputs": [],
   "source": [
    "cortes_dict = {}\n",
    "utr_category_dict = {}\n",
    "cortes_df = pd.read_csv('F:/Datasets/Data_From_Publications/Cortes_TSS.csv')\n",
    "for i, r in cortes_df.iterrows():\n",
    "    cortes_dict[r['RvNumber']] = r['Leaderless']\n",
    "    utr_category_dict[r['RvNumber']] = r['UTR Category']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "99ce98f9-ebd2-4317-a32a-9b4622000b9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_info_dict = {}\n",
    "for genome_record in SeqIO.parse(seq_dir_1 + '/' + tb_annotation_dirname + '/genomic.gbff', \"genbank\"):\n",
    "    for feature in genome_record.features:\n",
    "        a = feature.qualifiers\n",
    "        if feature.type == 'CDS' and a.get(\"translation\") != None:\n",
    "            locus_tag = a.get(\"locus_tag\")[0]\n",
    "            product = a.get(\"product\")[0]\n",
    "            cds_info_dict[locus_tag] = product  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "888d0443-64e8-49ac-a431-781b44031265",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RvNumber</th>\n",
       "      <th>Synonym</th>\n",
       "      <th>Strand</th>\n",
       "      <th>UTR Category</th>\n",
       "      <th>Leaderless</th>\n",
       "      <th>Operon Category</th>\n",
       "      <th>Exponential RPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>+</td>\n",
       "      <td>SD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rv0002</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>+</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rv0003</td>\n",
       "      <td>recF</td>\n",
       "      <td>+</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>+</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4017</th>\n",
       "      <td>Rv3920c</td>\n",
       "      <td>Rv3920c</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U</td>\n",
       "      <td>16.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4018</th>\n",
       "      <td>Rv3921c</td>\n",
       "      <td>Rv3921c</td>\n",
       "      <td>-</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4019</th>\n",
       "      <td>Rv3922c</td>\n",
       "      <td>Rv3922c</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U</td>\n",
       "      <td>41.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4020</th>\n",
       "      <td>Rv3923c</td>\n",
       "      <td>rnpA</td>\n",
       "      <td>-</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>54.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4021</th>\n",
       "      <td>Rv3924c</td>\n",
       "      <td>rpmH</td>\n",
       "      <td>-</td>\n",
       "      <td>SD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>211.82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4022 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     RvNumber  Synonym Strand UTR Category Leaderless Operon Category  \\\n",
       "0      Rv0001     dnaA      +           SD        NaN             NaN   \n",
       "1      Rv0002     dnaN      +            U        NaN             NaN   \n",
       "2      Rv0003     recF      +            U        NaN             NaN   \n",
       "3      Rv0004   Rv0004      +          NaN        NaN             NaN   \n",
       "4      Rv0005     gyrB      +            U        NaN             NaN   \n",
       "...       ...      ...    ...          ...        ...             ...   \n",
       "4017  Rv3920c  Rv3920c      -          NaN        NaN               U   \n",
       "4018  Rv3921c  Rv3921c      -            U        NaN             NaN   \n",
       "4019  Rv3922c  Rv3922c      -          NaN        NaN               U   \n",
       "4020  Rv3923c     rnpA      -            U        NaN             NaN   \n",
       "4021  Rv3924c     rpmH      -           SD        NaN             NaN   \n",
       "\n",
       "      Exponential RPKM  \n",
       "0                34.29  \n",
       "1                24.17  \n",
       "2                 9.83  \n",
       "3                 4.89  \n",
       "4                85.84  \n",
       "...                ...  \n",
       "4017             16.29  \n",
       "4018             29.16  \n",
       "4019             41.00  \n",
       "4020             54.02  \n",
       "4021            211.82  \n",
       "\n",
       "[4022 rows x 7 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cortes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "e688b6b3-897c-4f6c-991d-3aeccd0bf718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rv0006\n",
      "Rv0046c\n",
      "Rv0055\n",
      "Rv0059\n",
      "Rv0137c\n",
      "Rv0146\n",
      "Rv0179c\n",
      "Rv0198c\n",
      "Rv0224c\n",
      "Rv0236c\n",
      "Rv0248c\n",
      "Rv0269c\n",
      "Rv0337c\n",
      "Rv0376c\n",
      "Rv0462\n",
      "Rv0528\n",
      "Rv0574c\n",
      "Rv0639\n",
      "Rv0652\n",
      "Rv0769\n",
      "Rv0799c\n",
      "Rv0809\n",
      "Rv0955\n",
      "Rv0983\n",
      "Rv1000c\n",
      "Rv1001\n",
      "Rv1079\n",
      "Rv1178\n",
      "Rv1193\n",
      "Rv1206\n",
      "Rv1223\n",
      "Rv1310\n",
      "Rv1321\n",
      "Rv1341\n",
      "Rv1412\n",
      "Rv1473\n",
      "Rv1593c\n",
      "Rv1594\n",
      "Rv1612\n",
      "Rv1691\n",
      "Rv1748\n",
      "Rv1834\n",
      "Rv1843c\n",
      "Rv2062c\n",
      "Rv2134c\n",
      "Rv2291\n",
      "Rv2373c\n",
      "Rv2413c\n",
      "Rv2450c\n",
      "Rv2521\n",
      "Rv2614A\n",
      "Rv2615c\n",
      "Rv2644c\n",
      "Rv2665\n",
      "Rv2706c\n",
      "Rv2709\n",
      "Rv2717c\n",
      "Rv2730\n",
      "Rv2742c\n",
      "Rv2747\n",
      "Rv2771c\n",
      "Rv2774c\n",
      "Rv2861c\n",
      "Rv2913c\n",
      "Rv3001c\n",
      "Rv3032\n",
      "Rv3042c\n",
      "Rv3246c\n",
      "Rv3292\n",
      "Rv3309c\n",
      "Rv3400\n",
      "Rv3556c\n"
     ]
    }
   ],
   "source": [
    "seq_ids = util.list_dirs(project_dir)\n",
    "seq_ids.sort()\n",
    "out_list = []\n",
    "all_pairs = []\n",
    "pair_utr_count = 0\n",
    "pair_other_count = 0\n",
    "non_pair_utr_count = 0\n",
    "non_pair_other_count = 0\n",
    "for id in (seq_ids):\n",
    "    consensus = []\n",
    "    if os.path.exists(project_dir + '/' + str(id) + '/upstream_region_no_insertions'+str(id) + '_alignment_2.sto') and not(os.path.exists(project_dir + '/' + str(id) + '/rscape_output_no_insertions2.cov')):\n",
    "        print(id)\n",
    "        #blastfn.run_rscape(project_dir + '/' + str(id), 'upstream_region_no_insertions'+str(id) + '_alignment_2.sto', 'rscape_output_no_insertions2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "819dffeb-4eb8-4c3e-8369-7218c434aa1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "distance = 100000\n",
    "seq_ids = util.list_dirs(project_dir)\n",
    "seq_ids.sort()\n",
    "out_list = []\n",
    "all_pairs = []\n",
    "data = []\n",
    "pair_utr_count = 0\n",
    "pair_other_count = 0\n",
    "non_pair_utr_count = 0\n",
    "non_pair_other_count = 0\n",
    "for id in (seq_ids):\n",
    "    consensus = []\n",
    "    if os.path.exists(project_dir + '/' + str(id) + '/upstream_region_no_insertions'+str(id) + '_alignment_2.fasta'):\n",
    "        alignment = util.read_fasta_to_array(project_dir + '/' + str(id) + '/upstream_region_no_insertions'+str(id) + '_alignment_2.fasta')\n",
    "        num_sequences = len(alignment[0])\n",
    "        num_characters = len(alignment[1][0])\n",
    "        for i in range(num_characters):\n",
    "            temp = []\n",
    "            for j in range(num_sequences):\n",
    "                temp.append(alignment[1][j][i])\n",
    "            consensus_found = False\n",
    "            for letter in ['A','C','G','T']:\n",
    "                if temp.count(letter) / num_sequences > 0.5:\n",
    "                    consensus.append(letter)\n",
    "                    consensus_found = True\n",
    "            if consensus_found == False:\n",
    "                consensus.append('-')\n",
    "        if os.path.exists(project_dir + '/' + str(id) + '/rscape_output_no_insertions2.cov'):\n",
    "            with open(project_dir + '/'+str(id) +  '/rscape_output_no_insertions2.cov', 'r') as f:  \n",
    "                num_pairs = 0\n",
    "                e_values = []\n",
    "                pairs = []\n",
    "                for l in f:\n",
    "                    if (not ('#' in l)):\n",
    "                        a = l.split()\n",
    "                        if len(a) > 6: \n",
    "                            if num_characters - int(a[1]) <= distance:\n",
    "                                pairs.append(int(a[1]))\n",
    "                            if num_characters - int(a[2]) <= distance:\n",
    "                                pairs.append(int(a[2]))\n",
    "            pairs = (set(pairs))\n",
    "            if len(pairs) > 0:\n",
    "                for i in pairs:\n",
    "                    if consensus[i-1] == '-':\n",
    "                        consensus[i-1] = 'x'\n",
    "                    else:\n",
    "                        consensus[i-1] = consensus[i-1].lower()\n",
    "            \n",
    "            if id in cortes_dict:\n",
    "                if cortes_dict[id] == 'L':\n",
    "                    leaderless = 'Y'\n",
    "                else:\n",
    "                    leaderless = 'N'\n",
    "            else:\n",
    "                leaderless = 'N'\n",
    "            if id in utr_category_dict:\n",
    "                utr_category = utr_category_dict[id]\n",
    "            else:\n",
    "                utr_category = ''\n",
    "            #print(id, len(pairs), leaderless, utr_category, cds_info_dict[id],''.join(consensus)[-1*distance:])\n",
    "            if len(pairs)> 0:\n",
    "                pairs_ind = 1\n",
    "            else:\n",
    "                pairs_ind = 0\n",
    "            if utr_category in ['SD']:\n",
    "                utr_ind = 1\n",
    "            else:\n",
    "                utr_ind = 0\n",
    "            data.append([pairs_ind, utr_ind])\n",
    "output_df = pd.DataFrame(data, columns = ['pairs', 'utr'])\n",
    "            #if len(pairs) > 0:\n",
    "            #    if utr_category in ['SD']:\n",
    "            #        pair_utr_count +=1\n",
    "            #    else:\n",
    "            #        pair_other_count +=1\n",
    "            #else:        \n",
    "            #    if utr_category in ['SD']:\n",
    "            #        non_pair_utr_count +=1\n",
    "            #    else:\n",
    "            #        non_pair_other_count +=1\n",
    "\n",
    "\n",
    "#print(pair_utr_count, pair_other_count, non_pair_utr_count, non_pair_other_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "371587be-25ff-4c18-aa10-ff30a739b263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utr      0    1\n",
      "pairs          \n",
      "0      981  214\n",
      "1      397  145\n",
      "3.267196856916555e-05\n"
     ]
    }
   ],
   "source": [
    "contigency= pd.crosstab(output_df['pairs'], output_df['utr']) \n",
    "# Chi-square test of independence. \n",
    "c, p, dof, expected = chi2_contingency(contigency) \n",
    "print(contigency)\n",
    "# Print the p-value\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ed11372-e465-4519-883e-85b46b6e8cad",
   "metadata": {},
   "source": [
    "#### Extract covariation information to identify potentially significant regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d47c776f-9647-4d9d-97fc-df4e05cb569e",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_ids = util.list_dirs(project_dir)\n",
    "out_list = []\n",
    "all_pairs = []\n",
    "for id in seq_ids:\n",
    "    if os.path.exists(project_dir + '/' + str(id) + '/rscape_output_no_insertions2.cov'):\n",
    "        alignment = util.read_fasta_to_array(project_dir + '/' + str(id) + '/upstream_region_no_insertions'+str(id) + '_alignment_2.fasta')\n",
    "        seq_len = len(alignment[1][0])\n",
    "        with open(project_dir + '/'+str(id) +  '/rscape_output_no_insertions2.cov', 'r') as f:  \n",
    "            num_pairs = 0\n",
    "            e_values = []\n",
    "            pairs = []\n",
    "            for l in f:\n",
    "                if (not ('#' in l)):\n",
    "                    a = l.split()\n",
    "                    if len(a) > 6: # and abs(int(a[2])-int(a[1])) > 3:    # Exclude covarying pairs less than 3 bp apart)\n",
    "                        #if seq_len - int(a[1]) < 30 and seq_len - int(a[1]) < 30:\n",
    "                        pairs.append(seq_len - int(a[1]))\n",
    "                        pairs.append(seq_len - int(a[2]))\n",
    "                        e_values.append(float(a[4]))\n",
    "                        num_pairs +=1\n",
    "        if len(e_values) > 0:\n",
    "            tot = sum([math.log(x) for x  in e_values])\n",
    "            num = len(e_values)\n",
    "            combined_e_value = 1-chi2.cdf(-2*tot, 2*num)\n",
    "        else:\n",
    "            combined_e_value = 999\n",
    "        if combined_e_value < 1e-5:\n",
    "            print(id, num_pairs, combined_e_value, seq_len, sorted(set(pairs)))\n",
    "            for i in set(pairs):\n",
    "                all_pairs.append(i)\n",
    "sns.histplot([x for x in all_pairs if x <= 30], bins =31)\n",
    "plt.xlabel('Relative position upstream of 3rd nucleotide in start codon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "ad18ed86-dffb-4551-8027-7899f1fcbbc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_dirs_1 = util.list_dirs(seq_dir_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "005a4a35-2b50-4547-861f-ce94b142723f",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = []\n",
    "for seq_dir in sequence_dirs_1:\n",
    "    if os.path.exists(seq_dir_1 + '/' + seq_dir + '/genomic.gbff'):\n",
    "        for genome_record in (SeqIO.parse(seq_dir_1 + '/' + seq_dir + '/genomic.gbff', \"genbank\")):\n",
    "            organism = genome_record.annotations['organism']\n",
    "            full_sequence = str(genome_record.seq)\n",
    "            for feature in genome_record.features:\n",
    "                    a = feature.qualifiers\n",
    "                    if feature.type == 'rRNA' and '16S' in a.get(\"product\")[0]:\n",
    "                        (start, stop, strand) = (int(feature.location.start), int(feature.location.end), int(feature.location.strand))\n",
    "                        if (strand == 1 and not(full_sequence[stop-9:stop] == 'ACCTCCTTT')) or (strand == -1 and not (util.reverse_complement(full_sequence[start:start+9]) == 'ACCTCCTTT')):\n",
    "\n",
    "\n",
    "                            if strand == 1:\n",
    "                                temp.append([organism, (start, stop, strand),  a.get(\"product\")[0], full_sequence[stop-20:stop]])\n",
    "                            else:\n",
    "                                temp.append([organism, (start, stop, strand),  a.get(\"product\")[0], util.reverse_complement(full_sequence[start:start+20])])\n",
    "temp_df = pd.DataFrame(temp, columns = ['species','location', 'product', 'sequence'])\n",
    "temp_df.to_csv(project_dir +'/non_standard_ASD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "4c9549f7-c03c-4673-a290-093692a496fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_dirs_2 = util.list_dirs(seq_dir_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "6c4869c5-a784-412a-90d9-3115060995cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1757"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sequence_dirs_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "05b1c2ec-f469-427b-b1c6-f615985861cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "215\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for seq_dir in sequence_dirs_1:\n",
    "    if os.path.exists(seq_dir_1 + '/' + seq_dir + '/genomic.gbff'):\n",
    "        count +=1\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "23a94f00-a736-44eb-b581-32168c44e596",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1757/1757 [20:35<00:00,  1.42it/s]\n",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2560/2560 [00:00<00:00, 73138.19it/s]\n"
     ]
    }
   ],
   "source": [
    "temp = []\n",
    "for seq_dir in tqdm(sequence_dirs_2):\n",
    "    if os.path.exists(seq_dir_2 + '/' + seq_dir + '/genomic.gbff'):\n",
    "        for genome_record in (SeqIO.parse(seq_dir_2 + '/' + seq_dir + '/genomic.gbff', \"genbank\")):\n",
    "            organism = genome_record.annotations['organism']\n",
    "            full_sequence = str(genome_record.seq)\n",
    "            for feature in genome_record.features:\n",
    "                a = feature.qualifiers\n",
    "                if feature.type == 'rRNA' and '16S' in a.get(\"product\")[0]:\n",
    "                    (start, stop, strand) = (int(feature.location.start), int(feature.location.end), int(feature.location.strand))\n",
    "                    if strand == 1:\n",
    "                        temp.append([organism,  full_sequence[stop-20:stop]])\n",
    "                       \n",
    "                    else:\n",
    "                        temp.append([organism,  util.reverse_complement(full_sequence[start:start+20])])\n",
    "                    break\n",
    "\n",
    "util.produce_fasta_file(temp, project_dir +'/30S.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "83360421-c301-45d1-93ee-dca6710cdff6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1727/1727 [00:00<00:00, 246648.15it/s]\n"
     ]
    }
   ],
   "source": [
    "names = []\n",
    "temp_2 = []\n",
    "for x in temp:\n",
    "    name =  x[0].replace(' ','_')[0:50]\n",
    "    if name in names:\n",
    "        continue\n",
    "    else:\n",
    "        temp_2.append([name, x[1]])\n",
    "        names.append(name)\n",
    "util.produce_fasta_file(temp_2, project_dir +'/30S.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "8f9bcf1f-3139-4018-a20d-f174a5318cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "cline = MuscleCommandline(muscle_exe, input = project_dir +'/30S.fasta', out = project_dir +'/30S_alignment.fasta')\n",
    "cline()\n",
    "blastfn.convert_fasta_to_stockholm(project_dir, '30S.fasta', '30S.sto')\n",
    "blastfn.run_rscape(project_dir, '30S.sto', 'rscape_output_30S')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "c3f9a012-fe9d-407d-95a5-79e350bd2972",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.convert_sequence_file_format(project_dir + '/30S_alignment.fasta', project_dir + '/30S_alignment.sto')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "f1b5fe8b-8e30-4929-95b5-c77a8942e9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.run_rscape(project_dir, '30S_alignment.sto', 'rscape_output_30S')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa95113d-bb8a-4b0d-84da-83c6c1522bf5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
