{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1e30ed4-8a0e-4a6f-96eb-dfe116dfb2de",
   "metadata": {},
   "source": [
    "#### Directories etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf9c6377-5080-4c22-94db-feabe17cc47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nicho\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\ete3-3.1.2-py3.7.egg\\ete3\\evol\\parser\\codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "C:\\Users\\nicho\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\ete3-3.1.2-py3.7.egg\\ete3\\evol\\parser\\codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.fft import fft, fftfreq\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "import random\n",
    "from tqdm import tqdm\n",
    "from Comparative_Analysis import Utilities as util\n",
    "import random\n",
    "from joblib import Parallel, delayed\n",
    "from Comparative_Analysis import Blast_Functions as blastfn\n",
    "import time\n",
    "import os\n",
    "import copy\n",
    "from Comparative_Analysis import ORF_Functions as orffn\n",
    "from random import sample\n",
    "import ete3;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "1988a242-f558-4643-b3a9-27272fa7bebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'F:/Project_Data/Project_9'\n",
    "seq_dir = 'F:/Datasets/Actinobacteria_Ref_Rep_Lev_Complete'\n",
    "tb_species = 'NC_000962.3' \n",
    "tb_genome_filename = 'GCF_000195955.2_ASM19595v2_genomic.gbff'\n",
    "comparator_filename = 'GCF_016745295.1_ASM1674529v1_genomic.gbff'\n",
    "tb_reannotation_filename = 'annot.gbk'\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "198ce0ae-2b01-482b-83cb-93c720537916",
   "metadata": {},
   "source": [
    "#### Find all (maximal nested) ORFs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d288591a-d353-4761-8c02-b7356977e0c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_filename_dict(num_subsets, subset_num, id_list):\n",
    "    ids = util.chunk_list(id_list, num_subsets, subset_num)\n",
    "    temp = []\n",
    "    for id in (ids):\n",
    "        genome_record = next(SeqIO.parse(seq_dir + '/' + id, \"genbank\"))\n",
    "        organism_name = genome_record.annotations['organism']\n",
    "        temp.append((organism_name, id))\n",
    "    return temp\n",
    "\n",
    "species_list = util.list_files(seq_dir)\n",
    "filename_dict = {}\n",
    "parallel_output = Parallel(n_jobs=-1)(delayed(create_filename_dict)(num_cores, core_number, species_list) for core_number in core_numbers)\n",
    "temp = [item for sublist in parallel_output for item in sublist]\n",
    "for (species_name, filename) in temp:\n",
    "    filename_dict[species_name] = filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "a3aee9f9-a1d8-475f-84a6-ac5f1a51bd9f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mycobacterium tuberculosis H37Rv GCF_000195955.2_ASM19595v2_genomic.gbff\n",
      "Corynebacterium marinum DSM 44953 GCF_000835165.1_ASM83516v1_genomic.gbff\n",
      "Mycobacterium marinum GCF_016745295.1_ASM1674529v1_genomic.gbff\n"
     ]
    }
   ],
   "source": [
    "for k, v in filename_dict.items():\n",
    "    if 'H37Rv' in k or 'arinum' in k:\n",
    "        print(k,v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "83fac314-21f0-4cb2-a5c3-f10401327c75",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def produce_non_shadowed_orfs(orfs):\n",
    "    a = orfs\n",
    "    a.sort(key = lambda x: x[3], reverse = True)\n",
    "    orf_list = [a[0]]\n",
    "    for x in tqdm(a[1:]):\n",
    "        matched = 0\n",
    "        for v in orf_list:\n",
    "            if x[1] > v[0] and x[0]<=v[1]:\n",
    "                matched = 1\n",
    "                break\n",
    "               \n",
    "        if matched == 0:\n",
    "            orf_list.append(x)\n",
    "    orf_list.sort(key = lambda x: x[0])\n",
    "    return orf_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "5c7cc1e9-23d9-4134-a0d0-0594d77575c6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 63719/63719 [00:07<00:00, 8142.75it/s] \n",
      "100%|██████████| 88938/88938 [00:15<00:00, 5694.50it/s]\n"
     ]
    }
   ],
   "source": [
    "for record in SeqIO.parse(seq_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "    tb_sequence = str(record.seq)\n",
    "for record in SeqIO.parse(seq_dir + '/' + comparator_filename, \"genbank\"):\n",
    "    comparator_sequence = str(record.seq)\n",
    "    \n",
    "ORFFinder = orffn.ORF_Finder(tb_sequence)\n",
    "tb_orfs = ORFFinder.max_orf(0, len(tb_sequence), output_orfs = 'Nested', min_orf_length = 50)\n",
    "tb_orfs_reduced = produce_non_shadowed_orfs(tb_orfs)\n",
    "ORFFinder = orffn.ORF_Finder(comparator_sequence)\n",
    "comparator_orfs = ORFFinder.max_orf(0, len(comparator_sequence), output_orfs = 'Nested', min_orf_length = 50)\n",
    "comparator_orfs_reduced = produce_non_shadowed_orfs(comparator_orfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "84b8df52-5545-44da-a462-2fdcf62276b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "xls = pd.ExcelFile('F:/Datasets/Data_From_Publications/Mycobrowser_Release_4.xlsx')\n",
    "mycobrowser_df = pd.read_excel(xls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "66d30243-e4cf-4250-9e2d-04f60476a42a",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_intergenic_length = 50\n",
    "\n",
    "mycobrowser_inter_feature_orfs = []\n",
    "genbank_cds_boundaries = []\n",
    "mycobrowser_cds_boundaries = []\n",
    "ORFFinder = orffn.ORF_Finder(tb_sequence)\n",
    "\n",
    "for genome_record in SeqIO.parse(project_dir + '/' + tb_reannotation_filename, \"genbank\"):\n",
    "    for feature in genome_record.features:\n",
    "        if feature.type not in ('source', 'gene'):\n",
    "            a = feature.qualifiers  \n",
    "            genbank_cds_boundaries.append((int(feature.location.start), int(feature.location.end), int(feature.location.strand))) \n",
    "\n",
    "for i,r in mycobrowser_df.iterrows():\n",
    "    mycobrowser_cds_boundaries.append((r['Locus'],r['Start'],r['Stop'], r['Strand']))\n",
    "mycobrowser_cds_boundaries.sort(key = lambda x: x[1])\n",
    "max_stop = 0\n",
    "for i, (locus, start, stop, strand) in enumerate(mycobrowser_cds_boundaries):\n",
    "    if i < len(mycobrowser_cds_boundaries) - 1:\n",
    "        max_stop = max(stop, max_stop)\n",
    "        if mycobrowser_cds_boundaries[i+1][1] > max_stop + min_intergenic_length:\n",
    "            a =ORFFinder.max_orf(max_stop-5, mycobrowser_cds_boundaries[i+1][1]+5, output_orfs = 'Nested', min_orf_length = 50)\n",
    "            for x in a:\n",
    "                mycobrowser_inter_feature_orfs.append(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2683970-7a45-419b-875a-0ca44f8b7a66",
   "metadata": {},
   "source": [
    "#### Produce FASTA file with CDS candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "24efd99f-93c4-4ba7-a0bb-8aaa866c40bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 1524, 1),\n",
       " (2051, 3260, 1),\n",
       " (3279, 4437, 1),\n",
       " (4433, 4997, 1),\n",
       " (5239, 7267, 1),\n",
       " (7301, 9818, 1),\n",
       " (9913, 10828, 1),\n",
       " (10886, 10960, 1),\n",
       " (11111, 11184, 1),\n",
       " (11224, 11416, 1),\n",
       " (11554, 11692, -1),\n",
       " (11873, 12311, -1),\n",
       " (12467, 13016, 1),\n",
       " (13132, 13558, -1),\n",
       " (13713, 13995, -1),\n",
       " (14133, 14877, 1),\n",
       " (14913, 15612, 1),\n",
       " (15589, 17470, -1),\n",
       " (17466, 18762, -1),\n",
       " (18758, 20234, -1),\n",
       " (20230, 21640, -1),\n",
       " (21636, 23172, -1),\n",
       " (23269, 23737, -1),\n",
       " (23860, 25444, -1),\n",
       " (25643, 25726, 1),\n",
       " (25912, 26881, -1),\n",
       " (27022, 27442, -1),\n",
       " (27594, 28365, 1),\n",
       " (28361, 29207, 1),\n",
       " (29244, 29607, 1),\n",
       " (29739, 31151, 1),\n",
       " (31188, 31506, 1),\n",
       " (31513, 31819, 1),\n",
       " (31823, 31967, -1),\n",
       " (32056, 33154, 1),\n",
       " (33223, 33553, 1),\n",
       " (34294, 36610, 1),\n",
       " (36606, 36870, 1),\n",
       " (36866, 37262, 1),\n",
       " (37258, 38947, 1),\n",
       " (39055, 39829, -1),\n",
       " (39876, 41202, -1),\n",
       " (41303, 41912, 1),\n",
       " (42003, 42351, -1),\n",
       " (42432, 43365, -1),\n",
       " (43561, 46471, 1),\n",
       " (46580, 47084, -1),\n",
       " (47365, 48100, -1),\n",
       " (48232, 49027, -1),\n",
       " (49042, 49939, -1),\n",
       " (50020, 51124, -1),\n",
       " (51184, 51727, -1),\n",
       " (51827, 52697, -1),\n",
       " (52809, 53244, 1),\n",
       " (53236, 55699, 1),\n",
       " (55695, 57378, 1),\n",
       " (57334, 57973, 1),\n",
       " (58191, 58482, 1),\n",
       " (58585, 59080, 1),\n",
       " (59121, 59376, 1),\n",
       " (59408, 59867, 1),\n",
       " (59895, 60417, 1),\n",
       " (60395, 63020, 1),\n",
       " (63199, 63892, 1),\n",
       " (63908, 64967, 1),\n",
       " (65011, 65350, -1),\n",
       " (65551, 66694, 1),\n",
       " (66922, 68362, 1),\n",
       " (68417, 68579, 1),\n",
       " (68619, 71559, 1),\n",
       " (71588, 71828, 1),\n",
       " (71820, 72222, 1),\n",
       " (72273, 74511, -1),\n",
       " (74628, 75198, -1),\n",
       " (75300, 76212, 1),\n",
       " (76236, 77622, -1),\n",
       " (77618, 78896, -1),\n",
       " (79203, 79294, -1),\n",
       " (79485, 80193, 1),\n",
       " (80325, 80553, 1),\n",
       " (80623, 81673, 1),\n",
       " (81675, 82668, 1),\n",
       " (82747, 83983, 1),\n",
       " (83995, 85168, 1),\n",
       " (85182, 85563, -1),\n",
       " (85635, 86466, -1),\n",
       " (86527, 87133, 1),\n",
       " (87207, 87801, -1),\n",
       " (87797, 88004, -1),\n",
       " (88203, 89025, 1),\n",
       " (89021, 89480, 1),\n",
       " (89574, 89919, 1),\n",
       " (89923, 90403, 1),\n",
       " (90399, 92322, 1),\n",
       " (92327, 93278, 1),\n",
       " (93288, 93951, 1),\n",
       " (93950, 95417, 1),\n",
       " (95413, 96892, 1),\n",
       " (97121, 97601, 1),\n",
       " (97745, 98351, 1)]"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_cds_boundaries[0:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "16dcdc22-a45b-495a-ab68-91f61a13ab41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_orf_protein_fasta_file(species_name, orf_list, full_sequence, output_file):\n",
    "    trans = util.Translator()\n",
    "    temp = []\n",
    "    for x in orf_list:\n",
    "        if x[2] == 1:\n",
    "            prot = trans.translate_sequence(full_sequence[x[0]:x[1]], 1, 0)\n",
    "        else:\n",
    "            prot = trans.translate_sequence(util.reverse_complement(full_sequence[x[0]:x[1]]), 1, 0)\n",
    "        name = species_name + '_' +str(x[0])+'_'+str(x[1])+'_'+str(x[2])\n",
    "        temp.append([name, prot[:-1]])\n",
    "    util.produce_fasta_file(temp, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "34a34d01-5bbe-4223-a768-d86f7bd16d06",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 63720/63720 [00:01<00:00, 53007.66it/s]\n",
      "100%|██████████| 6122/6122 [00:00<00:00, 69502.61it/s]\n",
      "100%|██████████| 88939/88939 [00:02<00:00, 35440.39it/s]\n",
      "100%|██████████| 8163/8163 [00:00<00:00, 24155.07it/s]\n",
      "100%|██████████| 4832/4832 [00:00<00:00, 210043.39it/s]\n"
     ]
    }
   ],
   "source": [
    "produce_orf_protein_fasta_file('M.tb', tb_orfs, tb_sequence, project_dir + '/tb_orfs.fasta')\n",
    "produce_orf_protein_fasta_file('M.tb', tb_orfs_reduced, tb_sequence, project_dir + '/tb_orfs_reduced.fasta')\n",
    "produce_orf_protein_fasta_file('M.marinum', comparator_orfs, comparator_sequence, project_dir + '/comparator_orfs.fasta')\n",
    "produce_orf_protein_fasta_file('M.marinum', comparator_orfs_reduced, comparator_sequence, project_dir + '/comparator_orfs_reduced.fasta')\n",
    "produce_orf_protein_fasta_file('M.tb', mycobrowser_inter_feature_orfs, tb_sequence, project_dir + '/mycobrowser_inter_feature_orfs.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "907ba2dc-aff7-4e09-9a2f-04bed94dccdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.build_blast_db(project_dir, 'comparator_orfs.fasta', 'comparator_orfs', 'F:/Datasets/BLAST/comparator_orfs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "9c4adacf-2387-4952-a93b-1cce31b10f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.build_blast_db(project_dir, 'comparator_orfs_reduced.fasta', 'comparator_orfs_reduced', 'F:/Datasets/BLAST/comparator_orfs_reduced')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "3ad29f43-3a81-4be3-a424-ed8e8542b64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.run_blastp('F:/Datasets/BLAST/comparator_orfs', 'mycobrowser_inter_feature_orfs.fasta', 'comparator_orfs', e_value = 1e-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "adf9a4de-bd5b-40b8-a247-9a907ed4c342",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.run_blastp('F:/Datasets/BLAST/comparator_orfs_reduced', 'tb_orfs_reduced.fasta', 'comparator_orfs_reduced', e_value = 1e-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "7c4eca9c-0af3-480c-8829-5cb1f8c7c08a",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.run_blastp('F:/Datasets/BLAST/comparator_orfs_reduced', 'mycobrowser_inter_feature_orfs.fasta', 'comparator_orfs_reduced', e_value = 1e-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "87201bf1-80de-4eaa-964e-37d6f4dd5d97",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results_df = pd.read_csv('F:/Datasets/BLAST/comparator_orfs/hits.csv', header=None)\n",
    "blast_results_df = blast_results_df.loc[blast_results_df.groupby([0])[13].idxmax()]\n",
    "blast_results_df['Overlap'] = 0\n",
    "blast_results_df['Genbank_Coordinates'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "301b21bf-f533-49ed-baa8-f631fef005ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "265it [00:00, 2558.37it/s]\n"
     ]
    }
   ],
   "source": [
    "for i, r in tqdm(blast_results_df.iterrows()):\n",
    "    a=(r[0].split('_'))\n",
    "    start=int(a[1])\n",
    "    stop=int(a[2])\n",
    "    strand = int(a[3])\n",
    "    highest_overlap = 0\n",
    "    hit = 0\n",
    "    for feature in genbank_cds_boundaries:\n",
    "            if start <= feature[1] and stop >= feature[0]:\n",
    "                overlap = (min(stop, feature[1]) - max(start, feature[0]))/(stop - start)\n",
    "                if overlap > highest_overlap:\n",
    "                    hit = 1\n",
    "                    highest_overlap = overlap\n",
    "                    highest_entry = feature\n",
    "    if hit == 1:\n",
    "        blast_results_df.at[i,'Overlap'] = highest_overlap\n",
    "        blast_results_df.at[i,'Genbank_Coordinates'] = highest_entry\n",
    "\n",
    "blast_results_df.to_csv(project_dir +'/blast_annotation_xref.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "922aeab9-6318-4a3e-bcc4-a4a65c5a5d64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>Mycobrowser_Locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>M.tb_1014393_1014555_1</td>\n",
       "      <td>M.marinum_460982_470066_-1</td>\n",
       "      <td>53</td>\n",
       "      <td>3027</td>\n",
       "      <td>47.059</td>\n",
       "      <td>51</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>53</td>\n",
       "      <td>2863</td>\n",
       "      <td>2913</td>\n",
       "      <td>5.110000e-06</td>\n",
       "      <td>40.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>M.tb_1016799_1016925_1</td>\n",
       "      <td>M.marinum_457189_457375_-1</td>\n",
       "      <td>41</td>\n",
       "      <td>61</td>\n",
       "      <td>58.537</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>20</td>\n",
       "      <td>60</td>\n",
       "      <td>9.540000e-08</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>M.tb_1016819_1017215_-1</td>\n",
       "      <td>M.marinum_456881_457427_1</td>\n",
       "      <td>131</td>\n",
       "      <td>181</td>\n",
       "      <td>62.222</td>\n",
       "      <td>90</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>48</td>\n",
       "      <td>130</td>\n",
       "      <td>48</td>\n",
       "      <td>137</td>\n",
       "      <td>1.860000e-32</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>M.tb_1056785_1057070_-1</td>\n",
       "      <td>M.marinum_398190_399204_1</td>\n",
       "      <td>94</td>\n",
       "      <td>337</td>\n",
       "      <td>56.790</td>\n",
       "      <td>81</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>75</td>\n",
       "      <td>257</td>\n",
       "      <td>337</td>\n",
       "      <td>2.050000e-21</td>\n",
       "      <td>85.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>M.tb_1057042_1057129_1</td>\n",
       "      <td>M.marinum_398864_398963_-1</td>\n",
       "      <td>28</td>\n",
       "      <td>32</td>\n",
       "      <td>84.000</td>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>9.780000e-08</td>\n",
       "      <td>40.8</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>M.tb_918010_918142_1</td>\n",
       "      <td>M.marinum_710211_710343_-1</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>81.395</td>\n",
       "      <td>43</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "      <td>4.960000e-20</td>\n",
       "      <td>72.8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Rv0824c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>M.tb_918148_918268_1</td>\n",
       "      <td>M.marinum_710085_710205_-1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>89.744</td>\n",
       "      <td>39</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>1.490000e-19</td>\n",
       "      <td>71.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Rv0824c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>M.tb_934087_934405_-1</td>\n",
       "      <td>M.marinum_6223244_6225035_-1</td>\n",
       "      <td>105</td>\n",
       "      <td>596</td>\n",
       "      <td>56.061</td>\n",
       "      <td>66</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>88</td>\n",
       "      <td>209</td>\n",
       "      <td>274</td>\n",
       "      <td>4.900000e-18</td>\n",
       "      <td>77.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>M.tb_934292_934481_-1</td>\n",
       "      <td>M.marinum_6223244_6225035_-1</td>\n",
       "      <td>62</td>\n",
       "      <td>596</td>\n",
       "      <td>50.980</td>\n",
       "      <td>51</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>48</td>\n",
       "      <td>155</td>\n",
       "      <td>205</td>\n",
       "      <td>1.760000e-08</td>\n",
       "      <td>47.8</td>\n",
       "      <td>0.000000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>M.tb_975946_976078_1</td>\n",
       "      <td>M.marinum_508948_511024_-1</td>\n",
       "      <td>43</td>\n",
       "      <td>691</td>\n",
       "      <td>63.889</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>382</td>\n",
       "      <td>417</td>\n",
       "      <td>4.200000e-06</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>Rv0877</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>314 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0                             1    2     3       4  \\\n",
       "59   M.tb_1014393_1014555_1    M.marinum_460982_470066_-1   53  3027  47.059   \n",
       "60   M.tb_1016799_1016925_1    M.marinum_457189_457375_-1   41    61  58.537   \n",
       "61  M.tb_1016819_1017215_-1     M.marinum_456881_457427_1  131   181  62.222   \n",
       "62  M.tb_1056785_1057070_-1     M.marinum_398190_399204_1   94   337  56.790   \n",
       "66   M.tb_1057042_1057129_1    M.marinum_398864_398963_-1   28    32  84.000   \n",
       "..                      ...                           ...  ...   ...     ...   \n",
       "52     M.tb_918010_918142_1    M.marinum_710211_710343_-1   43    43  81.395   \n",
       "54     M.tb_918148_918268_1    M.marinum_710085_710205_-1   39    39  89.744   \n",
       "56    M.tb_934087_934405_-1  M.marinum_6223244_6225035_-1  105   596  56.061   \n",
       "57    M.tb_934292_934481_-1  M.marinum_6223244_6225035_-1   62   596  50.980   \n",
       "58     M.tb_975946_976078_1    M.marinum_508948_511024_-1   43   691  63.889   \n",
       "\n",
       "     5   6  7   8    9    10    11            12     13   Overlap  \\\n",
       "59  51  27  0   3   53  2863  2913  5.110000e-06   40.4  0.000000   \n",
       "60  41  17  0   1   41    20    60  9.540000e-08   42.0  0.000000   \n",
       "61  90  27  2  48  130    48   137  1.860000e-32  112.0  0.000000   \n",
       "62  81  19  2  11   75   257   337  2.050000e-21   85.1  0.000000   \n",
       "66  25   4  0   1   25     3    27  9.780000e-08   40.8  0.000000   \n",
       "..  ..  .. ..  ..  ...   ...   ...           ...    ...       ...   \n",
       "52  43   8  0   1   43     1    43  4.960000e-20   72.8  1.000000   \n",
       "54  39   4  0   1   39     1    39  1.490000e-19   71.2  1.000000   \n",
       "56  66  29  0  23   88   209   274  4.900000e-18   77.0  0.000000   \n",
       "57  51  22  1   1   48   155   205  1.760000e-08   47.8  0.000000   \n",
       "58  36  13  0   1   36   382   417  4.200000e-06   40.0  0.022727   \n",
       "\n",
       "   Mycobrowser_Locus  \n",
       "59                    \n",
       "60                    \n",
       "61                    \n",
       "62                    \n",
       "66                    \n",
       "..               ...  \n",
       "52           Rv0824c  \n",
       "54           Rv0824c  \n",
       "56                    \n",
       "57                    \n",
       "58            Rv0877  \n",
       "\n",
       "[314 rows x 16 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3853493-d17b-404f-87f6-12d333067660",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
