{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2eb433f-4303-4b93-9c03-05644379fd98",
   "metadata": {},
   "outputs": [],
   "source": [
    "complement_dict = {'a':'t', 'c':'g', 'g':'c', 't':'a'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "852a539f-1417-46a9-8a35-8f2668f457f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "complement_dict['t']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2bb5b4e-bff4-4eed-866e-ddd516f4a4f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('F:/Datasets/CRYPTIC_DATA/Cryptic_Mutation_Counts/mutation_counts_1000_2000_.pkl', 'rb') as f:\n",
    "    temp = pickle.load(f)\n",
    "print(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432afda4-73d1-4fd4-8191-5492fdacfb8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(5):\n",
    "    print(\"the desktop \" + str(i)  + \" is really impressive\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4e40b00-eecb-4686-92e7-f4f4cfdea797",
   "metadata": {},
   "outputs": [],
   "source": [
    "from jupyter_dash import JupyterDash\n",
    "import dash_bio as dashbio\n",
    "from dash import html\n",
    "import urllib.request as urlreq\n",
    "from dash.dependencies import Input, Output\n",
    "project_dir = 'F:/Project_Data/E_Pump_Project'\n",
    "import dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9762b8b-2c47-474f-8fe9-c49307a61d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "app = JupyterDash(__name__)\n",
    "\n",
    "#app = dash.Dash(__name__)\n",
    "\n",
    "with open(project_dir + '/align_output_AL123456.3@Rv1063c', 'r') as f:\n",
    "    data2 = f.read()\n",
    "\n",
    "app.layout = html.Div([\n",
    "    dashbio.AlignmentChart(\n",
    "        id='my-default-alignment-viewer',\n",
    "        data=data2,\n",
    "        height=1800,\n",
    "        tilewidth=30,\n",
    "    ),\n",
    "    html.Div(id='default-alignment-viewer-output')\n",
    "])\n",
    "\n",
    "@app.callback(\n",
    "    Output('default-alignment-viewer-output', 'children'),\n",
    "    Input('my-default-alignment-viewer', 'eventDatum')\n",
    ")\n",
    "def update_output(value):\n",
    "    if value is None:\n",
    "        return 'No data.'\n",
    "    return str(value)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    #app.run_server(mode='jupyterlab', debug=False)\n",
    "    #app.run_server(\"external\")\n",
    "    app.run_server(debug=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22ea7bd6-2788-4623-8c92-3ea3c4dfab3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(project_dir + '/align_output_NC_000962.3@Rv2846c', 'r') as f:\n",
    "    data2 = f.read()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3407ea3-9580-4239-9bbb-a04647f39c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "29dd413b-0bf9-4d79-b1fb-4107eddda4d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'F:/Project_Data/Project_10'\n",
    "datasets_dir = 'F:/Datasets/CRYPTIC_DATA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6989fa61-c984-453e-bd71-a09734e021d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = pd.read_csv(datasets_dir + \"/VARIANTS.csv\") \n",
    "variant_df_2 = variant_df[['UNIQUEID', 'VARIANT', 'MUTATION_TYPE', 'IS_NULL', 'IS_HET', 'IS_FILTER_PASS', 'IS_SNP', 'REF', 'ALT', 'FRS', 'GT_CONF_PERCENTILE', 'DP', 'GENOME_INDEX']]    \n",
    "variant_df = []\n",
    "with open(project_dir + '/variant_df_2.pkl', 'wb') as f:\n",
    "    pickle.dump(variant_df_2, f)  \n",
    "print(\"Stage_1_Complete\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "191671ba-9824-4e53-8463-ccbb1976aaaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(project_dir + '/variant_df_2.pkl', 'rb') as f:\n",
    "    variant_df_2 = pickle.load(f)  \n",
    "print(len(variant_df_2))\n",
    "genomes_df = pd.read_csv(datasets_dir + '/GENOMES.csv')\n",
    "gpi_genomes_df = genomes_df[['UNIQUEID','BELONGS_GPI']] \n",
    "print(len(gpi_genomes_df))\n",
    "gpi_variant_df = pd.merge(variant_df_2, gpi_genomes_df, how='left', on = 'UNIQUEID')\n",
    "variant_df_2 = []\n",
    "with open(project_dir + '/nick_variant_df.pkl', 'wb') as f:\n",
    "    pickle.dump(gpi_variant_df, f)    \n",
    "print(\"Stage_2_Complete\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "83f12df5-c28e-4848-a253-7eadd1b0be83",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(project_dir + '/nick_variant_df.pkl', 'rb') as f:\n",
    "    variant_df = pickle.load(f)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e0bc97b8-f21d-4ad4-a183-3ec6f63d61aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "gpi_variant_df = variant_df[variant_df['BELONGS_GPI'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2458bf6c-f00c-4f56-8a30-5be38c09e95d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nicho\\AppData\\Local\\Temp/ipykernel_26776/3007769752.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gpi_variant_df.at[i, 'CLOCK_PASS'] = True\n"
     ]
    }
   ],
   "source": [
    "for i, r in gpi_variant_df.iterrows():\n",
    "    if r['FRS'] >= 0.9 and r['DP'] >= 5 and r['GT_CONF_PERCENTILE'] >= 0.5:\n",
    "        gpi_variant_df.at[i, 'CLOCK_PASS'] = True\n",
    "    else:\n",
    "        gpi_variant_df.at[i, 'CLOCK_PASS'] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b79bca5a-e79c-460b-82ca-1518832e647b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLOCK_PASS</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>1552144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>16607233</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CLOCK_PASS     count\n",
       "0       False   1552144\n",
       "1        True  16607233"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gpi_variant_df.groupby(['CLOCK_PASS']).size().reset_index().rename(columns={0:'count'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2591f315-727d-47fa-9dfc-4246b6a14473",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_fail_sample = gpi_variant_df.groupby(['CLOCK_PASS', 'GENOME_INDEX']).size().reset_index().rename(columns={0:'count'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c47eb7dd-54ee-4588-bc8c-21455a70c2a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = filter_fail_sample[filter_fail_sample['CLOCK_PASS'] == True]\n",
    "b = filter_fail_sample[filter_fail_sample['CLOCK_PASS'] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "68192832-0ebc-40c6-bf67-89ae3fca2d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "c = pd.merge(a, b, how = 'outer', on = 'GENOME_INDEX')\n",
    "c['ratio'] = c['count_x']/(c['count_x'] + c['count_y'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "0d05a275-b86d-47c8-9833-46e6f4d396a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLOCK_PASS_x</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>count_x</th>\n",
       "      <th>CLOCK_PASS_y</th>\n",
       "      <th>count_y</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>True</td>\n",
       "      <td>506</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>True</td>\n",
       "      <td>1377</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>True</td>\n",
       "      <td>1549</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>True</td>\n",
       "      <td>1591</td>\n",
       "      <td>152.0</td>\n",
       "      <td>False</td>\n",
       "      <td>46.0</td>\n",
       "      <td>0.767677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>True</td>\n",
       "      <td>2422</td>\n",
       "      <td>12.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528946</th>\n",
       "      <td>True</td>\n",
       "      <td>4411307</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528949</th>\n",
       "      <td>True</td>\n",
       "      <td>4411317</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528993</th>\n",
       "      <td>True</td>\n",
       "      <td>4411455</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529000</th>\n",
       "      <td>True</td>\n",
       "      <td>4411478</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529001</th>\n",
       "      <td>True</td>\n",
       "      <td>4411501</td>\n",
       "      <td>49.0</td>\n",
       "      <td>False</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.753846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14028 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CLOCK_PASS_x  GENOME_INDEX  count_x CLOCK_PASS_y  count_y     ratio\n",
       "90             True           506      1.0        False      1.0  0.500000\n",
       "265            True          1377      1.0        False      1.0  0.500000\n",
       "309            True          1549      4.0        False      1.0  0.800000\n",
       "315            True          1591    152.0        False     46.0  0.767677\n",
       "512            True          2422     12.0        False      2.0  0.857143\n",
       "...             ...           ...      ...          ...      ...       ...\n",
       "528946         True       4411307      1.0        False      1.0  0.500000\n",
       "528949         True       4411317      1.0        False      1.0  0.500000\n",
       "528993         True       4411455      1.0        False      1.0  0.500000\n",
       "529000         True       4411478      1.0        False      1.0  0.500000\n",
       "529001         True       4411501     49.0        False     16.0  0.753846\n",
       "\n",
       "[14028 rows x 6 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c[c['ratio'] <0.9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "62d05623-f501-4379-85eb-40053069f4eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLOCK_PASS_x</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>count_x</th>\n",
       "      <th>CLOCK_PASS_y</th>\n",
       "      <th>count_y</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>529004</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1566</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529006</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529007</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553141</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4410066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553142</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4410663</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553143</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4410754</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553144</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4411436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553145</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4411518</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24142 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CLOCK_PASS_x  GENOME_INDEX  count_x CLOCK_PASS_y  count_y  ratio\n",
       "529004          NaN          1165      NaN        False      1.0    NaN\n",
       "529005          NaN          1566      NaN        False      1.0    NaN\n",
       "529006          NaN          1611      NaN        False      1.0    NaN\n",
       "529007          NaN          1640      NaN        False      1.0    NaN\n",
       "529008          NaN          2495      NaN        False      1.0    NaN\n",
       "...             ...           ...      ...          ...      ...    ...\n",
       "553141          NaN       4410066      NaN        False      1.0    NaN\n",
       "553142          NaN       4410663      NaN        False      1.0    NaN\n",
       "553143          NaN       4410754      NaN        False      1.0    NaN\n",
       "553144          NaN       4411436      NaN        False      1.0    NaN\n",
       "553145          NaN       4411518      NaN        False      1.0    NaN\n",
       "\n",
       "[24142 rows x 6 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c[c['CLOCK_PASS_x'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ae603b7-9d50-4f3f-8364-bf1f527f2f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "compass_df = pd.read_csv('F:/Datasets/CRYPTIC_DATA/COMPASS.csv', names = ['species', 'start', 'stop','count'])\n",
    "compass_dict = {}\n",
    "for i, r in compass_df.iterrows():\n",
    "    for k in range(r['start'], r['stop']):     #  Should te start-1 but seems to ve inconsistency with mutations!!\n",
    "        compass_dict[k] = 'Y'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
