{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eef24ce4-4729-44b6-8d43-fbdf47b9033a",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Set up packages and directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9105f9b4-46be-44fd-b947-d3c47ce2fa01",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.fft import fft, fftfreq\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "import random\n",
    "from tqdm.auto import tqdm\n",
    "from Comparative_Analysis import Sequence_Analysis_Routines as sar\n",
    "from Comparative_Analysis import HMM as hmm\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from Comparative_Analysis import Alignment_HMM as alignment_hmm\n",
    "from Comparative_Analysis import Alignment_Analysis as alignment_analysis\n",
    "from Comparative_Analysis import Alignment as align\n",
    "from Comparative_Analysis import Master_Alignment_HMM as master_alignment_hmm\n",
    "from Comparative_Analysis import Multi_Species_Master_Alignment_HMM as multi_species_master_alignment_hmm\n",
    "from Comparative_Analysis import Arneson_Ernst_HMM as ae_hmm\n",
    "import random\n",
    "import copy\n",
    "from joblib import Parallel, delayed\n",
    "import os\n",
    "import subprocess\n",
    "from Bio.Blast import NCBIWWW, NCBIXML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "1988a242-f558-4643-b3a9-27272fa7bebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'D:/Project_Data/Project_6'\n",
    "genome_datasets_dir = project_dir + '/Datasets/NCBI_Datasets'\n",
    "literature_datasets_dir = project_dir + '/Datasets/Data_From_Publications'\n",
    "cryptic_output_path = \"D:/Project_Data/CRYPTIC_DATA/Cryptic_Data_Analysis\"\n",
    "cryptic_input_path = \"D:/Project_Data/CRYPTIC_DATA/\"\n",
    "output_dir = project_dir + '/Output'\n",
    "refseq_dir = 'D:/NCBI_Dataset_Mycobacteria'\n",
    "protein_fasta_output_loc = output_dir + '/Protein_Sequences'\n",
    "outgroup_protein_fasta_output_loc = output_dir + '/Protein_Sequences_With_Outgroup'\n",
    "sonic_paranoid_run_name = 'Run_Without_Outgroup'\n",
    "outgroup_sonic_paranoid_run_name = 'Run_With_Outgroup'\n",
    "sonic_paranoid_output_loc = output_dir + '/Sonic_Paranoid_Output'\n",
    "ortholog_dir = sonic_paranoid_output_loc + '/runs/' + sonic_paranoid_run_name + '/ortholog_groups'\n",
    "outgroup_ortholog_dir = sonic_paranoid_output_loc + '/runs/' + outgroup_sonic_paranoid_run_name + '/ortholog_groups'\n",
    "non_cds_output_dir = output_dir + '/Multiple_Alignment_Data/Downstream_Non_CDS'\n",
    "upstream_non_cds_output_dir = output_dir + '/Multiple_Alignment_Data/Upstream_Non_CDS'\n",
    "cds_output_dir = output_dir + '/Multiple_Alignment_Data/CDS'\n",
    "extended_cds_output_dir = output_dir + '/Multiple_Alignment_Data/Extended_CDS'\n",
    "extended_non_cds_regions_output_dir = output_dir + '/Multiple_Alignment_Data/Extended_Non_CDS'\n",
    "outgroup_cds_output_dir = output_dir + '/Multiple_Alignment_Data/CDS_With_Outgroup'\n",
    "outgroup_concatenated_cds_output_dir = output_dir + '/Multiple_Alignment_Data/CDS_With_Outgroup_Concatenated'\n",
    "hmm_parameters_output_dir = output_dir +'/HMM_Model_Parameters'\n",
    "conservation_analysis_output_dir = output_dir + '/Conservation_Analysis'\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "abef54a8-1fd0-4a05-94de-bc9e2f647844",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "205\n"
     ]
    }
   ],
   "source": [
    "species_list = util.list_dirs(refseq_dir)\n",
    "print(len(species_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "017f28e8-a111-4d47-a7d6-9fbe2e08e2d4",
   "metadata": {},
   "source": [
    "##### Translation function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3cdf9a31-18a2-457e-9fc5-fc6c56d5d757",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_dict = {}\n",
    "with open('D:/Project_Data/Project_3/Datasets/Reference_Tables/Standard_Code.txt') as f:\n",
    "    for l in f:\n",
    "        codon_dict[str(l[1:4])] = l[5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94a2ed1e-1303-4248-9b29-728cce1b9485",
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_sequence(input_seq, strand, rf):\n",
    "    output_seq = ''\n",
    "    if strand == 1:\n",
    "        seq = input_seq[rf:]\n",
    "    else:\n",
    "        seq = align.reverse_complement(input_seq)[rf:]\n",
    "    for i in range(0,len(seq)-2,3):\n",
    "        if seq[i:(i+3)] in codon_dict:\n",
    "            output_seq += codon_dict[seq[i:(i+3)]]\n",
    "        else:\n",
    "            output_seq += 'X'\n",
    "    return output_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f975dff-4e95-44f1-978b-5c46eee3e950",
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_orf(start, end, strand):\n",
    "    return translate_sequence(full_sequence[start:end], strand, 0)              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "32a04e0b-16f9-4f02-a425-63a64f4b3408",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_genome(num_subsets, subset_num, id_list):\n",
    "    ids = util.chunk_list(id_list, num_subsets, subset_num)\n",
    "    temp = []\n",
    "    for id in ids:\n",
    "        temp_1 = []\n",
    "        genome_record = next(SeqIO.parse(refseq_dir + '/' + id + '/genomic.gbff', \"genbank\"))\n",
    "        organism_name = genome_record.annotations['organism']\n",
    "        full_sequence = str(genome_record.seq)\n",
    "\n",
    "        for feature in genome_record.features:\n",
    "            a = feature.qualifiers\n",
    "            if a.get(\"protein_id\") != None and a.get(\"locus_tag\")!= None and a.get(\"translation\")!= None:\n",
    "                if a.get(\"note\")!= None:\n",
    "                    note = a.get(\"note\")[0]\n",
    "                else:\n",
    "                    note = ''\n",
    "                if a.get(\"product\")!= None:\n",
    "                    product = a.get(\"product\")[0]\n",
    "                else:\n",
    "                    product = ''\n",
    "                locus_tag = a.get(\"locus_tag\")[0]\n",
    "                protein_id = a.get(\"protein_id\")[0]\n",
    "                temp_1.append([feature.type, int(feature.location.start)+1, int(feature.location.end), int(feature.location.strand), note, product, locus_tag, protein_id])\n",
    "        temp.append([id, temp_1])\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "2a23c5b8-1f54-4804-9862-41b3f38f6f4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "parallel_output = Parallel(n_jobs=-1)(delayed(parse_genome)(num_cores, core_number, species_list) for core_number in core_numbers)\n",
    "temp = [item for sublist in parallel_output for item in sublist]\n",
    "all_genome_feature_dict = {}\n",
    "for entry in temp:\n",
    "    all_genome_feature_dict[entry[0]] = entry[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "831256d0-bdca-47d9-b965-aa562b7321d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_annotation(species, start, end):\n",
    "    info = all_genome_feature_dict[species]\n",
    "    a = ''\n",
    "    overlap = 0\n",
    "    in_frame = False\n",
    "    for items in info:\n",
    "        if min(items[2], end) - max(items[1], start)  > 0.4 * (items[2] - items[1]):\n",
    "            a = items\n",
    "            overlap = (min(items[2], end) - max(items[1], start))/(items[2] - items[1])\n",
    "            if (min(start, end) - items[1])%3 == 0:\n",
    "                in_frame = True\n",
    "            else:\n",
    "                in_frame = False\n",
    "            break\n",
    "    return a, overlap, in_frame\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "46022996-1b55-46e0-83d1-107b53cba396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CDS', 4434, 4997, 1, '', 'hypothetical protein', 'Rv0004', 'NP_214518.1']"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genome_feature_dict['GCF_000195955.2'][3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "e8556a12-0d6f-435d-b8ea-194983495d50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['CDS', 7301, 9818, 1, '', 'Rv0006', 'NP_214520.1'], 0.6738180373460468, True)"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_annotation('GCF_000195955.2', 7304,  9000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc300757-df8c-42d8-a7a0-352213a45d3f",
   "metadata": {},
   "source": [
    "##### Output nt sequences in FASTA format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab243605-4807-411c-8195-401b1a404c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_nt_blast_file(record_list, output_filename):\n",
    "    with open(output_filename, 'w',  newline='') as outfile:\n",
    "        line_length = 60\n",
    "        for record in tqdm(record_list):\n",
    "            sequence = record[1]\n",
    "            lines = []\n",
    "            sequence_length = len(sequence)\n",
    "            number_of_lines = math.ceil(sequence_length / line_length)\n",
    "            lines.append(\">\" +record[0]+ \"\\n\")\n",
    "            for i in range(number_of_lines):\n",
    "                subsequence = sequence[i*line_length:(i+1)*line_length]\n",
    "                lines.append(subsequence + \"\\n\")\n",
    "            outfile.write(''.join(lines))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab3eb090-60d0-4a2d-8f83-14b2b6889726",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Mycobrowser is used for CDS demarcation, sequence is taken from GCF_000195955.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "75949818-7e31-4bdf-a730-bfcb11740ce6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genome_record = next(SeqIO.parse(genome_datasets_dir + '/GCF_000195955.2/genomic.gbff', \"genbank\"))\n",
    "full_sequence = genome_record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "91ea0227-7cbc-49a1-8517-1b753754da71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "locus_info_dict = {}\n",
    "tb_cds_dict = {}\n",
    "tb_all_feature_dict = {}\n",
    "actual_cds_boundaries = []\n",
    "mycobrowser_df = pd.read_excel(literature_datasets_dir+'/Mycobrowser_Release_4.xlsx')\n",
    "temp_cds = mycobrowser_df[mycobrowser_df['Feature'] == 'CDS'][['Locus','Start','Stop','Strand','Function']]      #Just CDS\n",
    "temp_all_features = mycobrowser_df[['Locus','Start','Stop','Strand','Function']]     #All features\n",
    "actual_cds_boundaries = []\n",
    "actual_all_feature_boundaries = []\n",
    "for i, r in temp_cds.iterrows():\n",
    "    if r['Strand'] == '+':\n",
    "        strand = 1\n",
    "    else:\n",
    "        strand = -1\n",
    "    actual_cds_boundaries.append((r['Locus'],r['Start']-1, r['Stop'], strand))\n",
    "    tb_cds_dict[r['Locus']] = (r['Start']-1, r['Stop'], strand)\n",
    "    locus_info_dict[r['Locus']] = r['Function']\n",
    "for i, r in temp_all_features.iterrows():\n",
    "    if r['Strand'] == '+':\n",
    "        strand = 1\n",
    "    else:\n",
    "        strand = -1\n",
    "    actual_all_feature_boundaries.append((r['Locus'],r['Start']-1, r['Stop'], strand))\n",
    "    tb_all_feature_dict[r['Locus']] = (r['Start']-1, r['Stop'], strand)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beacebf7-ca45-4acd-ad98-b9377ec368da",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Load variant dataset and create variant dictionary (0 indexed for genome position - whereas CRyPtiC data uses 1 start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c8830643-ee4a-42ff-a2f1-39d0154817a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_count_df = pd.read_csv(cryptic_output_path + '/filtered_variant_summary_df.csv')\n",
    "variant_count_df = variant_count_df[variant_count_df['MUTATION_PCT'] < 0.1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8a5ac217-24ab-4e70-ac95-b7c24903ae5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = variant_count_df.groupby(['GENOME_INDEX'])[['MYKROBE_LINEAGE_NAME_2']].count().reset_index()\n",
    "temp_dict = dict(zip(temp.GENOME_INDEX, temp.MYKROBE_LINEAGE_NAME_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c8c05206-85d2-4eee-9040-db84f07b4773",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e0c812fd15c545c0ba3c0e36be3bf6bc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4411532 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mutation_counts_dict = {}\n",
    "for i in tqdm(range(len(full_sequence))):\n",
    "    if (i+1) in temp_dict:\n",
    "        mutation_counts_dict[i] = temp_dict[(i+1)]\n",
    "    else:\n",
    "        mutation_counts_dict[i] = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec65c3b9-c53e-4f86-8e21-0a4836d413c2",
   "metadata": {},
   "source": [
    "##### Define binomial probabilities for testing mutation counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f9245c0-e613-4a48-b1d8-f85b601f1a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bin_formula(max_bin_counts, tot_bin_counts, in_frame = False):\n",
    "    return 1- binom.cdf(max_bin_counts-1, tot_bin_counts,1/3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "48c726de-af9f-4d31-904c-606698271a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutation_bin_probability(start, end, strand):\n",
    "    mutations = []\n",
    "    for i in range(start,end):\n",
    "        for j in range(mutation_counts_dict[i]):\n",
    "            mutations.append(i)\n",
    "    bin_counts = [0,0,0]\n",
    "    for m in mutations:\n",
    "        if strand == 1:\n",
    "            bin_counts[(m-(start))%3] +=1\n",
    "        else:\n",
    "            bin_counts[((end-1)-m)%3] +=1\n",
    "    if sum(bin_counts) == 0:\n",
    "        return (2)\n",
    "    else:\n",
    "        return (bin_formula(bin_counts[2], sum(bin_counts)))  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cc97354-4eec-4741-b8b0-58b7be52aa9b",
   "metadata": {},
   "source": [
    "##### Function to find maximal open reading frame between two co-ordinates with mutation probability less than defined p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "de0966b0-0b35-46f6-9a47-c4dc0272f30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def max_orf(seq_start, seq_stop, p_value, output_all_orfs = False, min_orf_length = 0):\n",
    "    max_len = 0\n",
    "    orfs_found = []\n",
    "    start_pos = -999\n",
    "    end_pos = -999\n",
    "    for frame in ['Forward', 'Reverse']:\n",
    "        if frame == 'Forward':\n",
    "            temp = (full_sequence[seq_start: seq_stop])\n",
    "        else:\n",
    "            temp = align.reverse_complement(full_sequence[seq_start: seq_stop])\n",
    "        seq_len = len(temp)\n",
    "        for i in range(seq_len - 2):\n",
    "            orf_length = 0\n",
    "            test_codon = temp[i: i+3] \n",
    "            if test_codon in ['ATG','GTG','TTG']:  #Missed out CTG as doesn't seem to be used very much at all\n",
    "                for j in range(i + 3, seq_len - 2, 3):\n",
    "                    test_codon_2 = temp[j: j+3] \n",
    "                    if test_codon_2 in ['TAG','TGA','TAA']:\n",
    "                        orf_length = j - i\n",
    "                        break\n",
    "                if orf_length > 0:\n",
    "                    if frame == 'Forward':\n",
    "                        orf_start =  seq_start + i\n",
    "                        orf_end = seq_start + j+3\n",
    "                        orf_strand = 1\n",
    "                    else:\n",
    "                        orf_start =  seq_start + seq_len-(j+3)\n",
    "                        orf_end = seq_start + seq_len-i\n",
    "                        orf_strand = -1\n",
    "                    prob = mutation_bin_probability(orf_start, orf_end, orf_strand)\n",
    "                    if prob < p_value and orf_length >= min_orf_length:\n",
    "                        orfs_found.append((orf_start, orf_end, orf_strand, orf_length, prob))\n",
    "                    \n",
    "                if orf_length > max_len and prob< p_value and orf_length >= min_orf_length:                                           \n",
    "                    max_len = orf_length\n",
    "                    start_pos = orf_start\n",
    "                    end_pos = orf_end\n",
    "                    strand = orf_strand \n",
    "    \n",
    "    if output_all_orfs == True:\n",
    "        sorted_orfs = sorted(orfs_found, key=lambda x: x[3], reverse=True)\n",
    "        return sorted_orfs                \n",
    "    elif start_pos == -999:\n",
    "        return(0,0,0)\n",
    "    else:\n",
    "        return(start_pos, end_pos, strand, max_len, mutation_bin_probability(start_pos, end_pos, strand))   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "68732332-e99e-4df8-91d7-a2934a07acaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_orf_list(orfs):\n",
    "    filtered_list = []\n",
    "    #Find largest and add\n",
    "    to_remove = [1,2]\n",
    "    iter = 0\n",
    "    while len(orfs) > 0:\n",
    "        to_remove = []\n",
    "        max_len = 0\n",
    "        max_pos = 0\n",
    "        for i, orf in enumerate(orfs):\n",
    "            if orf[3] > max_len:\n",
    "                max_pos = i\n",
    "                max_len = orf[3]\n",
    "        filtered_list.append(orfs[max_pos])\n",
    "        for i, orf in enumerate(orfs):\n",
    "            if orf[2] == orfs[max_pos][2] and orf[0]%3 == orfs[max_pos][0]%3 and not(orf[1] <orfs[max_pos][0] or orf[0] >= orfs[max_pos][1]):\n",
    "                to_remove.append(orf)\n",
    "        if len(to_remove) > 0:\n",
    "            for orf in to_remove:\n",
    "                orfs.remove(orf)\n",
    "    return filtered_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0373163-fe39-44e7-99f5-9c0e8fb34301",
   "metadata": {},
   "source": [
    "##### Take \"inter-feature\" regions and find maximal ORFs in all frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "21025ccf-bb9c-4bf1-8b95-6c5cb2f4c00b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2b9dcef00d454c6283c2326d2b1c8cfd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1986 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "putative_protein_dict = {}\n",
    "putative_proteins = []\n",
    "non_included_boundaries = []\n",
    "for i, (locus,start,stop,strand) in enumerate(actual_all_feature_boundaries):\n",
    "    if i==0 or stop >= actual_all_feature_boundaries[i-1][2]:\n",
    "        non_included_boundaries.append(actual_all_feature_boundaries[i])\n",
    "for i, (locus,start,stop,strand) in enumerate(non_included_boundaries):\n",
    "        if i < len(non_included_boundaries) - 1:\n",
    "            orf_details_list = max_orf(stop, non_included_boundaries[i+1][1], 3, output_all_orfs = True, min_orf_length = 150)\n",
    "            #filtered_orf_details_list = filter_orf_list(orf_details_list)\n",
    "            filtered_orf_details_list = orf_details_list\n",
    "            if len(filtered_orf_details_list) > 0:\n",
    "                last_locus = 'A'\n",
    "                locus_suffix = 0\n",
    "                for orf_details in filtered_orf_details_list:\n",
    "                    a = translate_orf(orf_details[0], orf_details[1], orf_details[2])\n",
    "                    \n",
    "                    if locus == last_locus:\n",
    "                        locus_suffix +=1\n",
    "                    else:\n",
    "                        locus_suffix = 0\n",
    "                    last_locus = locus\n",
    "                    putative_proteins.append([locus+'_'+str(locus_suffix), a[:-1]])\n",
    "                    putative_protein_dict[locus+'_'+str(locus_suffix)] = orf_details\n",
    "produce_nt_blast_file(putative_proteins, refseq_dir + '/putative_protein.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "0298d165-e798-48b5-8c4f-d777faa2297f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if 1==1:\n",
    "    w_d = os.getcwd()\n",
    "    os.chdir(\"D:/\")\n",
    "    #subprocess.run('cd d:\\\\BLAST\\\\myco_refseq_nt & tblastn -query d:\\\\NCBI_Dataset_Mycobacteria\\\\putative_protein.faa -db myco_refseq_nt -out tblastn_testresults.xml -evalue 0.001 -outfmt 5 -num_threads 16', shell=True, capture_output = True)      #-seg=no \n",
    "    subprocess.run('cd d:\\\\BLAST\\\\myco_refseq_nt & tblastn -query d:\\\\NCBI_Dataset_Mycobacteria\\\\putative_protein.faa -db myco_refseq_nt -out tblastn_results.xml -evalue 0.001 -seg=no  -outfmt 10 -num_threads 16', shell=True, capture_output = True)      #-seg=no \n",
    "    os.chdir(w_d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "92beb7d2-bde1-4bb1-bbca-8daba0f8d9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results = pd.read_csv(\"D:\\\\BLAST\\\\myco_refseq_nt\\\\tblastn_results.xml\", header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "b4e84804-bbbe-4838-b85e-5743d0a0f501",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results.columns = ['query_accession_ver', 'subject_accession_ver', 'percent_identical_matches','alignment_length', 'number_mismatches', 'number_of_gap_openings', 'query_start_alignment', 'query_end_alignment', 'subject_start_alignment', 'subject_end_alignment', 'e_value', 'bit_score']\n",
    "blast_results = blast_results[~blast_results['subject_accession_ver'].isin(['GCF_000195955.2', 'GCF_000253375.1'])]\n",
    "#blast_results = blast_results [blast_results.percent_identical_matches > 50]\n",
    "blast_results = blast_results.groupby('query_accession_ver').filter(lambda x: x['query_accession_ver'].count() >= 10)\n",
    "blast_results['info']=  blast_results['query_accession_ver'].map(putative_protein_dict)\n",
    "blast_results[['start_coord', 'end_coord', 'strand', 'q_length', 'prob']] = pd.DataFrame(blast_results['info'].tolist(), index=blast_results.index)\n",
    "blast_results['alignment_ratio'] = blast_results['alignment_length'] / blast_results['q_length'] * 3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "dbf0bc6a-ff4d-432c-a297-e722840735f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6f06ebc211f5419b8e39db9ef527f0ad",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/24456 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for i, r in tqdm(blast_results.iterrows(), total=blast_results.shape[0]):\n",
    "    a, overlap, in_frame =  find_annotation(r['subject_accession_ver'], min(r['subject_start_alignment'],r['subject_end_alignment']), max(r['subject_start_alignment'],r['subject_end_alignment']))\n",
    "    blast_results.at[i, 'annotation_info'] = a\n",
    "    blast_results.at[i, 'overlap'] = overlap\n",
    "    blast_results.at[i, 'in_frame'] = in_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "51a1e0f9-a1ce-4208-aa6f-f891b5ab3a7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results.to_csv(project_dir + '/temp.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdd6d07-3797-408e-bc72-dee72066eb40",
   "metadata": {},
   "source": [
    "##### Graph p-values for annotated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ef5ea97f-6b6e-4cf0-a335-c4693ecb71c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4031 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "annotated_cds_probabilities = []\n",
    "for (locus, start, stop, strand) in tqdm(actual_cds_boundaries):\n",
    "    annotated_cds_probabilities.append((locus, mutation_bin_probability(start, stop, strand)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "739b7a97-8bf5-4bfd-a76f-154ac634e463",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'number of genes - Mycobrowser R3 annotation')"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAEPCAYAAAC+35gCAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAhsElEQVR4nO3de5hcVZnv8e+PAAEMIJcEQy4kgYAHGEBoLl4HRQFvhBlBo0eNiJPDZQDxqJBhZsB5JkcU7zrgiYJEQTCiSLyDUVFHICYYLgEjbRKgJZIACggYSPPOH3s1FE1V9eqq2t076d/neeqpvVftXetd3dBv9l5rr6WIwMzMLMdmwx2AmZltPJw0zMwsm5OGmZllc9IwM7NsThpmZpbNScPMzLI5aZiZWbbNcw6SNAHYrfb4iPhFWUGZmVk1DZg0JH0MeBtwB9CbigNw0jAzG2E00BPhklYA+0XE+qEJyczMqiqnT2MlsEXZgZiZWfXl9Gk8DiyTtAh45mojIk4vLSozM6uknKSxML3MzGyEG7BPA0DSlsCeaXdFRDxValRmZlZJOR3hhwPzgdWAgEnArKoPud15551jypQpwx2GmdlGZenSpQ9ExNhGn+fcnvokcGRErACQtCdwBXBQZ0Isx5QpU1iyZMlwh2FmtlGRdHezz3NGT23RlzAAIuL3ZIymknSJpLWSbq/z2QclhaSda8rmSOqWtELSUTXlB0m6LX32OUnKiNnMzEqQkzSWSLpY0uHp9SVgacZ5lwJH9y+UNAl4HXBPTdnewExgn3TOhZJGpY8vAmYD09Pred9pZmZDIydpnAwsB04HzqB4MvykgU5KfR4P1fno08CHKZ4q7zMDuDIi1kfEKqAbOETSeGC7iLghis6XrwLHZsRsZmYlGLBPIz0J/qn0aoukY4A/RsQt/e4yTQBurNnvSWVPpe3+5WZmNgwaJg1JCyLirZJu47lXBQBExH6DqUjSNsA5wJH1Pq5TFk3KG9Uxm+JWFpMnTx5MeGZmlqHZlcYZ6f1NHaprd2Aq0HeVMRG4WdIhFFcQk2qOnQjcl8on1imvKyLmAfMAurq6Bn4AxczMBqVhn0ZErEmbp0TE3bUv4JTBVhQRt0XEuIiYEhFTKBLCgRHxJ4onzmdKGi1pKkWH9+IUw6OSDkujpt4NXDPYus3MrDNyOsJfV6fs9QOdJOkK4AZgL0k9kk5sdGxELAcWUHSy/wg4NSL6pmE/GfgyRef4H4AfZsRsZmYlaNancTLFFcU0SbfWfLQt8N8DfXFEvH2Az6f0258LzK1z3BJg34Hq65Te3l66u7uf2d9jjz0YNWpUkzPMzEaOZn0aX6f4V/1HgbNryh+NiHpDaTcJ3d3dzP6v7/OCnXflsQfuY96pb2SvvfYa7rDMzCqhYdKIiIeBh4G3A0gaB2wFjJE0JiLuaXTuxu4FO+/Kdi/abbjDMDOrnAH7NCS9WdJdwCrgeoqJC92vYGY2AuV0hP8ncBjw+4iYChxBRp+GmZltenKSxlMR8SCwmaTNIuJnwAHlhmVmZlWUMzX6XySNAX4BXC5pLbCh3LDMzKyKcq40ZlCsE34mxTMUf6BzT4mbmdlGJCdp/HtEPB0RGyJifkR8Djir7MDMzKx6Snsi3MzMNj2lPRFuZmabHj8RbmZm2bKeCJe0P/DK9NEvqb8in5mZbeJyngg/HbgcGJdel0k6rezAzMysenKe03gfcGhEPAYg6WMUU55/vszAzMysenJGTwnordnvpf4yrGZmtonLudL4CnCTpKvT/rHAxaVFZGZmlTVg0oiIT0m6Hng5xRXGCRHx29IjMzOzysm50gBYBqzpO17S5E15PQ0zM6tvwKSRRkqdC9zPs/0ZAexXbmhmZlY1OVcaZwB7penRzcxsBMsZPXUvxUN+ZmY2wuUkjZXAzyXNkfSBvtdAJ0m6RNJaSbfXlF0g6XeSbpV0taQX1nw2R1K3pBWSjqopP0jSbemzz0nycF8zs2GSkzTuAa4DtqSYrLDvNZBLgaP7lV0H7BsR+wG/B+YASNobmAnsk865UNKodM5FwGxgenr1/04zMxsiOUNuP9LKF0fELyRN6Vd2bc3ujcBxaXsGcGVErAdWSeoGDpG0GtguIm4AkPRViudEfthKTGZm1p6c0VN7Ah8EptQeHxGvabPu9wLfSNsTKJJIn55U9lTa7l9uZmbDIGf01DeBLwJf5rnTibRM0jkU64xf3ldU57BoUt7oe2dT3Mpi8uTJbUZpZmb95SSNDRFxUacqlDSLYo3xIyKiLwH0AJNqDpsI3JfKJ9Yprysi5gHzALq6uhomFzMza01OR/h3JZ0iabykHfterVQm6WiK9cWPiYjHaz5aCMyUNFrSVIoO78URsQZ4VNJhadTUu4FrWqnbzMzal3OlMSu9f6imLIBpzU6SdAVwOLCzpB6Kp8rnAKOB69LI2Rsj4qSIWC5pAXAHxW2rUyOi71bYyRQjsbam6AB3J7iZ2TDJGT01tZUvjoi31yluODtuRMwF5tYpXwLs20oMZmbWWVkTFkraF9gb2KqvLCK+WlZQZmZWTTlDbs+luM20N/AD4PXArwAnDTOzESanI/w44AjgTxFxArA/Rb+EmZmNMDlJ44mIeBrYIGk7YC0DdIKbmdmmKadPY0maWPBLwFLgr8DiMoMyM7Nqyhk9dUra/KKkH1HMBXVruWGZmVkV5S73CkBErC4pDjMz2wjk9GmYmZkBThpmZjYIg0oarc45ZWZmm4aGSUPSyyXdKWm5pEMlXUcxkupeSS8dwhjNzKwimnWEfxp4KzAG+D5wbET8StKBwOeBlw9BfGZmViHNksYWEXEbgKR1EfErgIi4WdLWQxKdmZlVSrM+jdrP5vT7bMsSYjEzs4prljT+TdI2ABHxnb5CSbvjyQrNzEakhrenImJhg/I/SPpGeSGZmVlVNR1yK+mlko6TNC7t7yfp6xRTo5uZ2QjTbMjtBcAlwFuA76d1Na4DbqJYw9vMzEaYZqOn3gi8JCL+JmkH4D5gv4i4a2hCMzOzqml2e+qJiPgbQET8GVjhhGFmNrI1u9LYXVJtZ/iU2v2IOKa8sMzMrIqaJY0Z/fY/OZgvlnQJ8CZgbUTsm8p2BL4BTAFWA29NVzFImgOcCPQCp0fEj1P5QcClwNYUa5SfERExmFjMzKwzGt6eiojrm70yvvtS4Oh+ZWcDiyJiOrAo7SNpb2AmsE8650JJo9I5FwGzKTrfp9f5TjMzGyKlTY0eEb8AHupXPAOYn7bnA8fWlF8ZEesjYhXQDRwiaTzFSoE3pKuLr9acY2ZmQ2yo19PYJSLWAKT3cal8AnBvzXE9qWxC2u5fbmZmw2Cgh/tGpec1yqY6ZdGkvP6XSLMlLZG0ZN26dR0LzszMCk2TRkT0AgdJqvfHuxX3p1tOpPe1qbwHmFRz3ESK50J60nb/8kbxzouIrojoGjt2bIdCNjOzPjm3p34LXCPpXZL+se/VYn0LgVlpexZwTU35TEmjJU2l6PBenG5hPSrpsJS43l1zjpmZDbFmQ2777Ag8CLympiyAbzc7SdIVwOHAzpJ6gHOB84EFkk4E7gGOB4iI5ZIWAHcAG4BT01UOwMk8O+T2h+llZmbDYMCkEREntPLFEfH2Bh8d0eD4ucDcOuVLgH1bicHMzDprwNtTkvaUtEjS7Wl/P0n/Wn5oZmZWNTl9Gl+iWLnvKYCIuJXiQTwzMxthcpLGNhGxuF/ZhjKCMTOzastJGg+kJV4DQNJxwJpSozIzs0rKGT11KjAPeLGkPwKrgHeWGpWZmVVSzuiplcBrJb0A2CwiHi0/LDMzq6Kc0VNnSNoOeBz4tKSbJR1ZfmhmZlY1OX0a742IR4AjKSYYPIHiIT0zMxthcpJG37xTbwC+EhG3UH8iQTMz28TlJI2lkq6lSBo/lrQt8HS5YZmZWRXljJ46ETgAWBkRj0vaieIWlZmZjTA5SeNS4JcUHeF/iYgHKSYwNDOzESbn9tSlwHjg85L+IOlbks4oNywzM6uinOc0firpeuBg4NXAScA+wGdLjs3MzCpmwKQhaRHwAuAGittUB0fE2uZnmZnZpijn9tStwJMUa1rsB+wraetSozIzs0rKuT11JoCkMRSjpr4CvAgYXW5oZmZWNTm3p/4ZeCVwEHA3cAnFbSozMxthcobcbg18ClgaEV5Hw8xsBMu5PXWBpP2BkyQB/DJNJWJmZiNMziy3pwOXU0xWOA64TNJpZQdmZmbVk3N76n3AoRHxGICkj1EMv/18mYGZmVn15M5y21uz30ubs9xKOlPSckm3S7pC0laSdpR0naS70vsONcfPkdQtaYWko9qp28zMWpeTNC4BbpJ0nqTzgBuBi1utUNIE4HSgKyL2BUYBM4GzgUURMR1YlPaRtHf6fB/gaOBCSaNard/MzFrXNGlI2gy4ieL5jIeAPwMnRMRn2qx3c2BrSZsD2wD3ATOA+enz+cCxaXsGcGVErI+IVUA3cEib9ZuZWQua9mlExNOSPhkRLwVu7kSFEfFHSZ8A7gGeAK6NiGsl7RIRa9IxaySNS6dMoLi66dOTyszMbIjl3J66VtJblMbbtiv1VcwApgK7Ai+Q9M5mp9QpiwbfPVvSEklL1q1b136wZmb2HDlJ4wPAN4EnJT2aXo+0UedrgVURsS4ingK+DbwMuF/SeID03jcpYg8wqeb8iRS3s54nIuZFRFdEdI0dO7aNEM3MrJ4Bk0ZEbBsRm0XEFml724jYro067wEOk7RNuno5ArgTWAjMSsfMAq5J2wuBmZJGS5oKTAcWt1G/mZm1KOc5DST9I/AKittCv4yI77RaYUTcJOkqij6SDcBvgXnAGGCBpBMpEsvx6fjlkhYAd6TjT42I3rpfbmZmpcqZsPBCYA/gilR0kqTXRcSprVYaEecC5/YrXk9x1VHv+LnA3FbrMzOzzsi50vh7YN+ICABJ84HbSo3KzMwqKacjfAUwuWZ/EsXCTGZmNsI0vNKQ9F2KPoztgTsl9XU+HwL8eghiMzOziml2e+oTQxaFmZltFBomjYi4vm9b0i7AwWl3cUSsrX+WmZltynLW03grxXMRxwNvpZi88LiyAzMzs+rJGT11DnBw39WFpLHAT4CrygzMzMyqJ2f01Gb9bkc9mHmemZltYnKuNH4k6cc8+3Df24AflheSmZlV1YBJIyI+VDONiIB5EXF16ZGZmVnl5EwjMhX4QUR8O+1vLWlKRKwuOzgzM6uWnL6JbwJP1+z3pjIzMxthcpLG5hHxZN9O2t6yvJDMzKyqcpLGOknH9O1ImgE8UF5IZmZWVTmjp04GLpP0hbTfA7yrvJDMzKyqcpLGqog4TNIYQBHxaNlBmZlZNeXcnuqWdAEwyQnDzGxky0ka+wG/By6WdKOk2ZLaWSPczMw2UgMmjYh4NCK+FBEvAz5MsUzrGknzJe1ReoRmZlYZObPcjpJ0jKSrgc8CnwSmAd8FflByfGZmViE5HeF3AT8DLoiI2hX7rpL0qnLCMjOzKsrq04iIE/slDAAi4vRWKpX0QklXSfqdpDslvVTSjpKuk3RXet+h5vg5krolrZB0VCt1mplZ+5qtEf65mu3nfd5qwkg+C/woIo6TtCWwDfAvwKKIOF/S2cDZwFmS9gZmAvsAuwI/kbRnRPS2Ub+ZmbWg2e2pk4DbgQXAfRQz3LYtjbx6FfAeeGZakifTk+aHp8PmAz8HzgJmAFdGxHpglaRu4BDghk7EY2Zm+ZoljfEUS7y+DdgAfAP4VkT8uc06pwHrgK9I2h9YCpwB7BIRawAiYo2kcen4CcCNNef3pDIzMxtiDfs0IuLBiPhiRLya4qrghcBySe1OIbI5cCBwUUS8BHiM4lZUI/WucKLugcUzJEskLVm3bl2bYZqZWX85Q24PBN4PvJNixb6lbdbZA/RExE1p/yqKJHK/pPGpzvHA2prjJ9WcP5HidtnzRMS8iOiKiK6xY8e2GaaZmfXXMGlI+oikpcAHgOuBrjSK6o52KoyIPwH3StorFR0B3AEsBGalslnANWl7ITBT0ui0INR0YHE7MZiZWWua9Wn8G7AS2D+9/l8aRSUgImK/Nuo9Dbg8jZxaCZxAkcAWSDoRuIeiP4WIWC5pAUVi2QCc6pFTZmbDo1nSmFpWpRGxDOiq89ERDY6fC8wtKx4zM8vTMGlExN1DGYiZmVVfzhPhZmZmgJOGmZkNwqCSRhp+a2ZmI9RgrzS+XEoUZma2URhs0ujI/FNmZrZxGmzS+EgpUZiZ2UZhUEkjIr5TUhxmZrYR8OgpMzPL5qRhZmbZcma53V3S6LR9uKTTJb2w9MjMzKxycq40vgX0StoDuJhiTqqvlxqVmZlVUk7SeDoiNgD/AHwmIs6kWNXPzMxGmJyk8ZSkt1OscfG9VLZFeSGZmVlV5SSNE4CXAnMjYlVaCOmycsMyM7MqaraeBgARcYeks4DJaX8VcH7ZgZmZWfXkjJ56M7AM+FHaP0DSwpLjMjOzCsq5PXUecAjwF3hm1b3SVvUzM7PqykkaGyLi4X5lUUYwZmZWbQP2aQC3S3oHMErSdOB04NflhmVmZlWUc6VxGrAPsB64AngEeH+JMZmZWUUNmDQi4vGIOCciDo6IrrT9t3YrljRK0m8lfS/t7yjpOkl3pfcdao6dI6lb0gpJR7Vbt5mZtSZn9NSekuZJulbST/teHaj7DODOmv2zgUURMR1YlPaRtDcwk+Jq52jgQkmjOlC/mZkNUk6fxjeBL1Is9drbiUolTQTeCMwFPpCKZwCHp+35wM+Bs1L5lRGxHlglqZtiNNcNnYjFzMzy5SSNDRFxUYfr/QzwYWDbmrJdImINQESskTQulU8Abqw5rieVmZnZEMvpCP+upFMkjU/9DjtK2rHVCiW9CVgbEUtzT6lTVnfIr6TZkpZIWrJu3bpWQzQzswZyrjRmpfcP1ZQFMK3FOl8OHCPpDcBWwHaSLgPulzQ+XWWMB9am43uASTXnTwTuq/fFETEPmAfQ1dXlZ0nMzDosZ/TU1DqvVhMGETEnIiZGxBSKDu6fRsQ7gYU8m6BmAdek7YXATEmj02SJ04HFrdZvZmatG/BKQ9I2FJ3VkyNidnrAb6+I+N4Apw7W+cACSScC9wDHA0TEckkLgDuADcCpEdGRDnkzMxucnNtTXwGWAi9L+z0UI6raThoR8XOKUVJExIPAEQ2Om0sx0srMzIZRTkf47hHxceApgIh4gvqd02ZmtonLSRpPStqaNGJJ0u4UU4qYmdkIk3N76lyKtTQmSbqcYvTTe8oMyszMqiln5b7rJN0MHEZxW+qMiHig9MjMzKxyckZPHZg216T3yZK2B+6OiA2lRWZmZpWTc3vqQuBA4FaKK4190/ZOkk6KiGtLjM/MzCokpyN8NfCSNC36QcBLgNuB1wIfLzE2MzOrmJyk8eKIWN63ExF3UCSRleWFZWZmVZRze2qFpIuAK9P+24DfSxpNenbDzMxGhpwrjfcA3RRLvJ4JrExlTwGvLikuMzOroJwht08An0yv/v7a8YjMzKyycq40zMzMACcNMzMbhIZJQ9LX0vsZQxeOmZlVWbMrjYMk7Qa8V9IOtUu9trPcq5mZbbyadYR/kWKiwmkU62nUTofeznKvZma2kWp4pRERn4uI/wVcEhHTOrXcq5mZbbxyhtyeLGl/4JWp6BcRcWu5YZmZWRUNOHpK0unA5cC49Lpc0mllB2ZmZtWTM43I+4BDI+IxAEkfA24APl9mYGZmVj05z2kI6K3Z78VrhJuZjUg5VxpfAW6SdHXaPxa4uLSIzMyssga80oiITwEnAA8BfwZOiIjPtFqhpEmSfibpTknL+x4eTM9/XCfprvS+Q805cyR1S1oh6ahW6zYzs/bkXGkQETcDN3eozg3A/42ImyVtCyyVdB3FzLmLIuJ8SWcDZwNnSdobmAnsA+wK/ETSnhHR2+D7zcysJEM+91RErElJiIh4FLgTmADMAOanw+ZT3AYjlV8ZEesjYhXFNO2HDGnQZmYGDPOEhZKmUCwfexOwS0SsgSKxUAzvhSKh3FtzWk8qMzOzIdY0aUgaJeknZVQsaQzwLeD9EfFIs0PrlEWD75wtaYmkJevWretEmGZmVqNp0kj9Bo9L2r6TlUragiJhXB4R307F90sanz4fD6xN5T3ApJrTJwL3NYh3XkR0RUTX2LFjOxmymZmR1xH+N+C21Fn9WF9hRJzeSoWSRDFk9840MqvPQmAWcH56v6am/OuSPkXRET4dWNxK3WZm1p6cpPH99OqUlwPvokhEy1LZv1AkiwWSTgTuAY4HiIjlkhYAd1CMvDrVI6fMzIZHzoSF8yVtDUyOiBXtVhgRv6LxE+VHNDhnLjC33brNzKw9ORMWvhlYRrG2BpIOkLSw5LjMzKyCcobcnkfxXMRfACJiGTC1tIjMzKyycpLGhoh4uF9Z3SGvZma2acvpCL9d0juAUZKmA6cDvy43LDMzq6KcK43TKOZ9Wg9cATwCvL/EmMzMrKJyRk89DpyTFl+KNF+UmZmNQDmjpw6WdBtwK8WzFbdIOqj80MzMrGpy+jQuBk6JiF8CSHoFxcJM+5UZmJmZVU9On8ajfQkDnnk4z7eozMxGoIZXGpIOTJuLJf1/ik7wAN4G/Lz80MzMrGqa3Z76ZL/9c2u2/ZyGmdkI1DBpRMSrhzIQMzOrvgE7wiW9EHg3MKX2+FanRjczs41XzuipHwA3ArcBT5cbjpmZVVlO0tgqIj5QeiQVFE8/zcqVK5/Z32OPPRg1atQwRmRmNrxyksbXJP0T8D2KqUQAiIiHSouqIh576E+c95272WnCwzz2wH3MO/WN7LXXXsMdlpnZsMlJGk8CFwDn8OyoqQCmlRVUlbxgp13Z7kW7DXcYZmaVkJM0PgDsEREPlB2MmZlVW84T4cuBx8sOxMzMqi/nSqMXWCbpZzy3T2NEDbl1p7iZWV7S+E56jWi1neJ/XdvDnDfuw7RpRbeOE4iZjRQ562nMH4pABiLpaOCzwCjgyxFx/lDH0Ncp/tcH7uO87yx7XgLp7e0FeCaBOJmY2aYm54nwVdSZayoihmz0lKRRwH8BrwN6gN9IWhgRdwxVDP3VSyDr7lrGZttsz04TpjZMJu1sQ14i6u3tpbu7+5n9VpNXp77HzDYdObenumq2twKOB3YsJ5yGDgG6I2IlgKQrgRnAsCWNWrUJZPMxOzVNJu1s978t1ii5rFy5ko/+4E7GjN21reSV8z3wbDKpTTKdSpStJNayvrdq7Rwontpj+stpw2Db3Oj4dn52jdrQqfgH8zMbqO7Btr+dtrX7D8t25NyeerBf0Wck/Qr493JCqmsCcG/Nfg9waFmVPfbAfQA88ee1bLZ+PY9sNbq17W2272hcTzz8AGdd/CO232UCAA/d/TtGbbUt2+8y4XnbYyZMZ0y/c5od3+r3PP7ntfznu45g2rRprFy5kn/92iK22WFcS3V1Ynugn0ur21VrZ048tcf0l9OGwba50fHt/OwataFT8ff/76XZz2ygugfb/nba1qzNXz/3n0p9CFkRzWc5r1lXA4ohul3AyRGxf2lRPT+G44GjIuJ9af9dwCERcVq/42YDs9PuXsCKFqvcGRhpz6WMxDbDyGy32zxytNLu3SJibKMPc25P1a6rsQFYDbx1kEG0qweYVLM/Ebiv/0ERMQ+Y125lkpZERNfAR246RmKbYWS2220eOcpod87tqSqsq/EbYLqkqcAfgZnAO4Y3JDOzkSdn9NRo4C08fz2N/ygvrOeKiA2S/hn4McWQ20siYvlQ1W9mZoWc21PXAA8DS6l5InyoRcQPKNb2GApt3+LaCI3ENsPIbLfbPHJ0vN05HeG3R8S+na7YzMw2PjkTFv5a0t+VHomZmVVeTtJ4BbBU0gpJt0q6TdKtZQfWKZKOTrF3Szq7zueS9Ln0+a21Q4wbnStpR0nXSborve8wVO3JVVK7L5D0u3T81Wn9+Mooo801n39QUkjauex2DEZZbZZ0WvpsuaSPD0VbBqOk/74PkHSjpGWSlkg6ZKjak6PNNl8iaa2k2/udM/i/ZRHR9AXsVu810HlVeFF0mv+BYsGoLYFbgL37HfMG4IeAgMOAmwY6F/g4cHbaPhv42HC3dYjafSSwedr+WJXaXVab0+eTKAZh3A3sPNxtHYLf86uBnwCj0/644W7rELX7WuD1Nef/fLjb2ok2p89eBRwI3N7vnEH/LRvwSiMi7q73Gui8inhm+pGIeBLom36k1gzgq1G4EXihpPEDnDsD6JvIcT5wbMntGKxS2h0R10bEhnT+jRTPy1RFWb9rgE8DH6bOHGzDrKw2nwycHxHrASJi7VA0ZhDKancA26Xt7anzLNgwaqfNRMQvgHpLdA/6b1nO7amNWb3pRyZkHtPs3F0iYg1Aeh/XwZg7oax213ovxb9qqqKUNks6BvhjRNzS6YA7oKzf857AKyXdJOl6SQd3NOr2ldXu9wMXSLoX+AQwp3Mht62dNjcz6L9lm3rSUJ2y/v9abHRMzrlVVWq7JZ1DMTvA5S1FV46Ot1nSNsA5DO08a4NR1u95c2AHilscHwIWSKp3/HApq90nA2dGxCTgTODiliPsvHba3FGbetLImX6k0THNzr2/77IvvVft8r2sdiNpFvAm4H9HuhFaEWW0eXdgKnCLpNWp/GZJL+po5K0r6/fcA3w73eZYDDxNMYdRVZTV7lnAt9P2NyluCVVFO21uZvB/y4a7g6fMF8W/mFZS/I/f13m0T79j3shzO48WD3QucAHP7Tz6+HC3dYjafTTFdPRjh7uNQ9Xmfuevplod4WX9nk8C/iNt70lxy0PD3d4haPedwOFp+whg6XC3tRNtrvl8Cs/vCB/037Jh/2EMwQ/7DcDvKUYenJPKTgJOStuiWODpD8BtQFezc1P5TsAi4K70vuNwt3OI2t2d/oAsS68vDnc7y25zv+9fTYWSRom/5y2By4DbgZuB1wx3O4eo3a+gmPniFuAm4KDhbmcH23wFsAZ4iuKK5MRUPui/ZQM+EW5mZtZnU+/TMDOzDnLSMDOzbE4aZmaWzUnDzMyyOWmYmVk2Jw2zYSTpPZK+MNxxmOVy0jAzs2xOGmaZJE1J64nMT+sVXJXmp+r7fDNJq2vXGUlrG+wi6c1pAsDfSvqJpF3qfP+lko6r2f9rzfaHJP0m1fuREptp1pSThtng7AXMi4j9gEeAU/o+iIingWuAfwCQdCiwOiLuB34FHBYRL6GY1vrDuRVKOhKYTjEX0gHAQZJe1ZHWmA2Sk4bZ4NwbEf+dti+jmHqi1jeAt6XtmWkfisnjfizpNoqZY/cZRJ1HptdvKab1eDFFEjEbck4aZoPTf96d7dPyoMvS2hs3AHtIGkuxoE3frKmfB74QEX8H/B9gqzrfvYH0/2SainzLVC7goxFxQHrtERFVmrbbRhAnDbPBmSzppWn77cD3av6YL4xiMrergU8Bd0bEg+nY7YE/pu1ZDb57NXBQ2p4BbJG2fwy8V9IYAEkTJFVt4S8bIZw0zAbnTmCWpFuBHYGL6hzzDeCdPHtrCuA84JuSfgk80OC7vwT8vaTFwKHAY1Asswt8Hbgh3d66Cti2/aaYDZ5nuTXLJGkKxZXFvsMdi9lw8ZWGmZll85WGmZll85WGmZllc9IwM7NsThpmZpbNScPMzLI5aZiZWTYnDTMzy/Y/A71uS4TWAj8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "annotation_probs = [x[1] for x in annotated_cds_probabilities]\n",
    "annotation_probs = [x for x in annotation_probs if x<0.01]\n",
    "sns.histplot(annotation_probs, bins = 100)\n",
    "plt.xlabel(\"p-value\")\n",
    "plt.ylabel(\"number of genes - Mycobrowser R3 annotation\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ebafc8a-e6a9-47ef-ad3f-17aad81e739a",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Identify function percentage in Unknown between high and low p-value regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b398cf17-b8fc-489d-8aca-f29fd6b16346",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "376 526\n",
      "2310 4031\n",
      "0.714828897338403 0.5517831669044223\n"
     ]
    }
   ],
   "source": [
    "locus_exceptions = [x[0] for x in annotated_cds_probabilities if (x[1] > 0.5 and x[1] < 1.1)]\n",
    "num_unknown_exceptions = 0\n",
    "for ex in locus_exceptions:\n",
    "    if 'nknown' in locus_info_dict[ex]:\n",
    "        num_unknown_exceptions += 1\n",
    "tot_locus = [x[0] for x in annotated_cds_probabilities]\n",
    "num_unknown_tot = 0\n",
    "for ex in tot_locus:\n",
    "    if ex in locus_info_dict:\n",
    "        if len(str(locus_info_dict[ex])) > 4:\n",
    "            if 'nknown' in locus_info_dict[ex]:\n",
    "                num_unknown_tot += 1\n",
    "print(num_unknown_exceptions, len(locus_exceptions))\n",
    "print(num_unknown_tot, len(tot_locus))\n",
    "print(num_unknown_exceptions/len(locus_exceptions), (num_unknown_tot-num_unknown_exceptions)/(len(tot_locus)-len(locus_exceptions)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
