{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "848e9933-2a1f-43e1-8190-941b3c674907",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import re\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import math\n",
    "from tqdm import tqdm\n",
    "import copy\n",
    "from collections import defaultdict\n",
    "import ete3\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from joblib import Parallel, delayed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7f589c6c-3a7c-4127-bba1-17ec7eae4316",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir =  'F:/Project_Data/Project_11'\n",
    "datasets_dir = project_dir + '/Datasets'\n",
    "tb_genome_filename = 'GCF_000195955.2_ASM19595v2_genomic.gbff'\n",
    "dictionary_dir = project_dir + '/Dictionaries'\n",
    "mutation_count_dir = project_dir + '/Mutation_Counts'\n",
    "chunk_variant_dict_dir = project_dir + '/Chunk_Variant_Dictionaries'\n",
    "tb_tree_filename = 'tb_gpi.nwk'\n",
    "full_run = True\n",
    "\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "timeout=99999\n",
    "\n",
    "for record in SeqIO.parse(datasets_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "    full_sequence = str(record.seq)\n",
    "tb_length = len(full_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8f458bd-b618-427f-9a8a-9c2818372813",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  CREATE VARIANT DATASETS FROM CRYPTIC RAW DATA\n",
    "\n",
    "if full_run == True:\n",
    "#if 1==0:\n",
    "    variant_df = pd.read_csv(datasets_dir + \"/VARIANTS.csv\") \n",
    "    with open(project_dir + '/variant_df.pkl', 'wb') as f:\n",
    "        pickle.dump(variant_df[['UNIQUEID', 'VARIANT', 'MUTATION_TYPE', 'IS_NULL', 'IS_HET', 'IS_FILTER_PASS', 'IS_SNP', 'REF', 'ALT', 'GENOME_INDEX']], f)    \n",
    "\n",
    "print(\"Stage_1_Complete\")\n",
    "if full_run == True:\n",
    "#if 1==0:\n",
    "    with open(project_dir + '/variant_df.pkl', 'rb') as f:\n",
    "        full_variant_df = pickle.load(f) \n",
    "    print(len(full_variant_df))\n",
    "    genomes_df = pd.read_csv(datasets_dir + '/GENOMES.csv')\n",
    "    gpi_genomes_df = genomes_df[genomes_df['BELONGS_GPI']==True][['UNIQUEID']] \n",
    "    print(len(gpi_genomes_df))\n",
    "    gpi_variant_df = pd.merge(full_variant_df, gpi_genomes_df, how='inner', on = 'UNIQUEID')\n",
    "    print(len(gpi_variant_df))\n",
    "    with open(project_dir + '/gpi_variant_df.pkl', 'wb') as f:\n",
    "        pickle.dump(gpi_variant_df, f)    \n",
    "print(\"Stage_2_Complete\")\n",
    "if full_run == True:\n",
    "#if 1==1:\n",
    "    complement_dict = {'a':'t', 'c':'g', 'g':'c', 't':'a'}\n",
    "    position_dict = {}\n",
    "    variant_dict = {}\n",
    "    id_dict = {}\n",
    "    with open(project_dir + '/gpi_variant_df.pkl', 'rb') as f:\n",
    "        variant_df = pickle.load(f) \n",
    "        unique_ids = variant_df.UNIQUEID.unique()\n",
    "        for i, unique_id in enumerate(unique_ids):\n",
    "            id_dict[unique_id] = i\n",
    "\n",
    "        for i, r in variant_df.iterrows():\n",
    "            if r['IS_NULL'] == False and r['IS_FILTER_PASS'] == True and r['IS_HET'] == False and r['IS_SNP'] == True :\n",
    "                genome_index = r['GENOME_INDEX']\n",
    "                unique_id = r['UNIQUEID']\n",
    "                ref_sequence_entry = full_sequence[genome_index - 1].lower()   # Cryptic is 1 indexed and lower case\n",
    "                cryptic_ref_sequence_entry = r['REF']\n",
    "                cryptic_alt_sequence_entry = r['ALT']\n",
    "                if cryptic_ref_sequence_entry == ref_sequence_entry:\n",
    "                    alt = cryptic_alt_sequence_entry\n",
    "                elif complement_dict[cryptic_ref_sequence_entry] == ref_sequence_entry:\n",
    "                    alt = complement_dict[cryptic_alt_sequence_entry]\n",
    "                else:\n",
    "                    print(\"Something strange at position \", genome_index, cryptic_ref_sequence_entry, ref_sequence_entry)\n",
    "                    alt = cryptic_alt_sequence_entry\n",
    "                \n",
    "                id_ref = id_dict[unique_id]\n",
    "                if id_ref in variant_dict:\n",
    "                    variant_dict[id_ref].append((genome_index, alt))\n",
    "                else:\n",
    "                    variant_dict[id_ref] = [(genome_index, alt)]\n",
    "\n",
    "                if genome_index in position_dict:\n",
    "                    position_dict[genome_index].append((id_ref, alt))\n",
    "                else:\n",
    "                    position_dict[genome_index] = [ref_sequence_entry, (id_ref, alt)]    # If first entry also include reference value for info\n",
    "\n",
    "    with open(project_dir + '/id_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(id_dict, f)\n",
    "    with open(project_dir + '/variant_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(variant_dict, f) \n",
    "    with open(project_dir + '/position_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(position_dict, f) \n",
    "\n",
    "print(\"Stage_3_Complete\")\n",
    "\n",
    "# CONSTRUCT DISTANCE MATRIX\n",
    "\n",
    "def generate_distances(ind_1, ind_2):\n",
    "    with open(project_dir + '/snp_pos_dict_'+str(ind_1)+'.pkl', 'rb') as f:\n",
    "        snp_pos_dict_1 = pickle.load(f)\n",
    "    with open(project_dir + '/snp_pos_dict_'+str(ind_2)+'.pkl', 'rb') as f:\n",
    "        snp_pos_dict_2 = pickle.load(f)\n",
    "    distance_dict = {}\n",
    "    for k1, v1 in snp_pos_dict_1.items():\n",
    "        for k2, v2 in snp_pos_dict_2.items():\n",
    "            if k2 >= k1:\n",
    "                continue\n",
    "            sd = v1.symmetric_difference(v2)\n",
    "            temp = {x[:-1] for x in sd}   # Only count variants with different nt in SNP as one mutation\n",
    "            #f = len(temp)/tb_length\n",
    "            #d = -3/4 * math.log(1 - 4*f /3)\n",
    "            #distance_dict[(k1, k2)] = d\n",
    "            distance_dict[(k1, k2)] = len(temp)\n",
    "    return(distance_dict)\n",
    "\n",
    "print(\"Stage_4_Complete\")\n",
    "\n",
    "if full_run == True:\n",
    "#if 1==1:\n",
    "    with open(project_dir + '/variant_dict.pkl', 'rb') as f:\n",
    "        temp_variant_dict = pickle.load(f) \n",
    "    print(len(temp_variant_dict))\n",
    "    temp_dict = {}\n",
    "    for (k, v) in temp_variant_dict.items():\n",
    "        v.sort(key = lambda x: x[0])\n",
    "        keystring = ''.join([str(pos) + snp for (pos, snp) in v])\n",
    "        temp_dict[keystring] = k\n",
    "    variant_dict = {}\n",
    "    for (k, v) in temp_dict.items():\n",
    "    \tvariant_dict[v] = temp_variant_dict[v] \n",
    "    print(len(variant_dict))\n",
    "    temp_variant_dict = {} \n",
    "    for core in tqdm(core_numbers):\n",
    "        snp_pos_dict = {}\n",
    "        for n, (k, v) in enumerate(variant_dict.items()):\n",
    "            if k%num_cores + 1 == core:\n",
    "                snp_pos_dict[k] = set([str(pos) + snp for (pos, snp) in v])\n",
    "        with open(project_dir + '/snp_pos_dict_'+str(core)+'.pkl', 'wb') as f:\n",
    "            pickle.dump(snp_pos_dict, f) \n",
    "    variant_dict = {}\n",
    "\n",
    "print(\"Stage_5_Complete\")\n",
    "if full_run == True:\n",
    "#if 1==1:\n",
    "    for core_1 in tqdm(core_numbers):\n",
    "        master_dict = {}\n",
    "        parallel_output = Parallel(n_jobs=-1, timeout = timeout)(delayed(generate_distances)(core_1, core_2) for (core_2) in core_numbers)\n",
    "        for output_dict in parallel_output:\n",
    "            for (k, v) in output_dict.items():\n",
    "                master_dict[k] = v\n",
    "        with open(project_dir + '/master_distance_dict_'+str(core_1)+'.pkl', 'wb') as f:\n",
    "            pickle.dump(master_dict, f) \n",
    "\n",
    "# COMBINE DICTIONARIES GENERATED INTO SINGLE DICTIONARY\n",
    "\n",
    "if full_run == True:\n",
    "#if 1==1:\n",
    "    master_dict = {}\n",
    "    for core_1 in tqdm(core_numbers):\n",
    "        with open(project_dir + '/master_distance_dict_'+str(core_1)+'.pkl', 'rb') as f:\n",
    "            temp_master_dict = pickle.load(f) \n",
    "        k1_vals = []\n",
    "        k2_vals = []\n",
    "        for (k1, k2), v in temp_master_dict.items():\n",
    "            k1_vals.append(k1)\n",
    "            k2_vals.append(k2)\n",
    "        k1_vals = list(set(k1_vals))\n",
    "        k2_vals = list(set(k2_vals))\n",
    "        k1_vals.sort()\n",
    "        k2_vals.sort()\n",
    "\n",
    "        for k1 in k1_vals:\n",
    "            for k2 in k2_vals:\n",
    "                if k2 >= k1:\n",
    "                    continue\n",
    "                if k1 in master_dict:\n",
    "                    master_dict[k1].append(str(abs(temp_master_dict[(k1, k2)])))\n",
    "                else:\n",
    "                    master_dict[k1] = [str(abs(temp_master_dict[(k1, k2)]))]\n",
    "    ids = []\n",
    "    for k, v in master_dict.items():     # Note that seq_0 does not appear on LHS of distance matrix and therefore needs to be included separately in final output\n",
    "        ids.append(k)\n",
    "    ids.append(0)\n",
    "    ids = list(set(ids))   \n",
    "    ids.sort()\n",
    "\n",
    "\n",
    "    #FINAL OUTPUT\n",
    "    print(\"Writing data\")\n",
    "    #with open(project_dir + '/master_distance_dict.pkl', 'wb') as f:\n",
    "    #    pickle.dump(master_dict, f) \n",
    "    with open(project_dir + '/ids.pkl', 'wb') as f:\n",
    "        pickle.dump(ids, f)\n",
    "    print(\"Writing Phylip file\") \n",
    "    with open(project_dir+'/tb_seq_distances.phy', 'w') as f:\n",
    "        f.write('%d\\n' % len(ids))\n",
    "        for n,idref in enumerate(ids):\n",
    "            f.write('seq_'+str(idref))\n",
    "            if idref in master_dict:\n",
    "                for location in master_dict[idref]:\n",
    "                    f.write('\\t%s' % location)\n",
    "            f.write('\\n')\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "982f94b3-60a7-4935-834e-38c3c810db01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "506022\n",
      "960 528 506022\n",
      "Stage_6 Complete\n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526, 527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 551, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567, 568, 569, 570, 571, 572, 573, 574, 575, 576, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 613, 614, 615, 616, 617, 618, 619, 620, 621, 622, 623, 624, 625, 626, 627, 628, 629, 630, 631, 632, 633, 634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 651, 652, 653, 654, 655, 656, 657, 658, 659, 660, 661, 662, 663, 664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 676, 677, 678, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 707, 708, 709, 710, 711, 712, 713, 714, 715, 716, 717, 718, 719, 720, 721, 722, 723, 724, 725, 726, 727, 728, 729, 730, 731, 732, 733, 734, 735, 736, 737, 738, 739, 740, 741, 742, 743, 744, 745, 746, 747, 748, 749, 750, 751, 752, 753, 754, 755, 756, 757, 758, 759, 760, 761, 762, 763, 764, 765, 766, 767, 768, 769, 770, 771, 772, 773, 774, 775, 776, 777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787, 788, 789, 790, 791, 792, 793, 794, 795, 796, 797, 798, 799, 800, 801, 802, 803, 804, 805, 806, 807, 808, 809, 810, 811, 812, 813, 814, 815, 816, 817, 818, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 831, 832, 833, 834, 835, 836, 837, 838, 839, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857, 858, 859, 860, 861, 862, 863, 864, 865, 866, 867, 868, 869, 870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895, 896, 897, 898, 899, 900, 901, 902, 903, 904, 905, 906, 907, 908, 909, 910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934, 935, 936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949, 950, 951, 952, 953, 954, 955, 956, 957, 958]\n",
      "959\n",
      "max  506021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 959/959 [00:11<00:00, 86.85it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Built dictionaries\n",
      "Stage_7 Complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 30/30 [27:27<00:00, 54.90s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage_8 Complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 959/959 [4:16:32<00:00, 16.05s/it]  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage_9 Complete\n"
     ]
    }
   ],
   "source": [
    "num_iterations = 30\n",
    "num_cores = 32\n",
    "core_numbers = list(range(0, num_cores))\n",
    "timeout=99999\n",
    "\n",
    "\n",
    "def list_files(dir):\n",
    "    r = []\n",
    "    s = []\n",
    "    for root, dirs, files in os.walk(dir):\n",
    "        for name in files:\n",
    "            if name == '.ipynb_checkpoints':\n",
    "                pass\n",
    "            else:\n",
    "                s.append(name)\n",
    "    return s\n",
    "\n",
    "def generate_variant_sequence_dict(chunk):\n",
    "    if os.path.exists(chunk_variant_dict_dir + '/chunk_variant_dictionary_'+str(chunk) + '.pkl'):    #  Number of chunk dictionaries generated can be lower given rounding of chunk size        \n",
    "        with open(chunk_variant_dict_dir + '/chunk_variant_dictionary_'+str(chunk) + '.pkl', 'rb') as f:\n",
    "            chunk_variant_dict = pickle.load(f) \n",
    "        with open(project_dir + '/ids.pkl', 'rb') as f:\n",
    "            distinct_sequence_names = pickle.load(f) \n",
    "\n",
    "        full_tb_variant_sequence = ''.join([full_sequence[pos] for pos in sorted_variant_position_list])\n",
    "        start_pos = chunk * chunk_size\n",
    "        end_pos = min(tb_variants_sequence_length, start_pos + chunk_size)\n",
    "        variant_sequence_dict = {}\n",
    "        seq_chunk = [{x} for x in full_tb_variant_sequence[start_pos:end_pos]]\n",
    "        for seq_id in distinct_sequence_names:\n",
    "            temp = copy.copy(seq_chunk)\n",
    "            if seq_id in chunk_variant_dict:\n",
    "                for (pos, snp) in chunk_variant_dict[seq_id]:\n",
    "                    temp[pos] = {snp}\n",
    "            variant_sequence_dict['seq_'+str(seq_id)] = temp\n",
    "        with open(dictionary_dir + '/variant_dictionary_'+str(start_pos)+'_'+str(end_pos)+'_' + '.pkl', 'wb') as f:\n",
    "            pickle.dump(variant_sequence_dict, f)\n",
    "\n",
    "def fitch_1(list_1, list_2):\n",
    "    res =[]\n",
    "    for i, j in zip(list_1, list_2):\n",
    "        a = i.intersection(j)\n",
    "        if len(a) == 0:\n",
    "            a = i.union(j)\n",
    "        res.append(a)\n",
    "    return res\n",
    "\n",
    "def fitch_2(parent_list, child_list):\n",
    "    res = []\n",
    "    mutations = []\n",
    "    for i, j in zip(parent_list, child_list):\n",
    "        mutation = 0\n",
    "        a = i.intersection(j)\n",
    "        if len(a) == 0:\n",
    "            a = set(list(j)[0])\n",
    "            mutation = 1\n",
    "        res.append(a)\n",
    "        if mutation == 1:\n",
    "            mutations.append(1)\n",
    "        else:\n",
    "            mutations.append(0)\n",
    "    return (res, mutations)\n",
    "\n",
    "def generate_mutation_counts(filename, core_number):    \n",
    "    seq_length = 22    #22 * 24 is length of chunks output in previous step\n",
    "    a = filename.split('_')\n",
    "    start = a[-3]\n",
    "    stop = a[-2]\n",
    "    with open(filename, 'rb') as f:\n",
    "        sequence_to_score_dict = pickle.load(f)\n",
    "    master_tree2= ete3.Tree(project_dir + '/' + tb_tree_filename)\n",
    "    for node in master_tree2.traverse(\"postorder\"):\n",
    "        if node.is_leaf():\n",
    "            node.add_features(seq = sequence_to_score_dict[node.name][core_number * seq_length: (core_number+1) * seq_length])\n",
    "        else:\n",
    "            children = node.children\n",
    "            node.add_features(seq = fitch_1(children[0].seq, children[1].seq))\n",
    "    #for k, v in sequence_to_score_dict.items():              \n",
    "        #seq_length = len(v)\n",
    "        #break\n",
    "    \n",
    "    mutation_counts = [0 for i in range(seq_length)]\n",
    "    for node in master_tree2.traverse(\"preorder\"):\n",
    "        if node.is_leaf():\n",
    "            continue\n",
    "        if node.is_root():\n",
    "            node.seq = [{list(x)[0]} for x in node.seq]\n",
    "        children = node.children\n",
    "        mutations = []\n",
    "        child_sequences = []\n",
    "        for child in children:\n",
    "            (temp_1, temp_2) = fitch_2(node.seq ,child.seq)\n",
    "            child_sequences.append(temp_1)\n",
    "            child.seq = temp_1\n",
    "            mutations.append(temp_2)\n",
    "        temp = []\n",
    "        for n, (h, i, j) in enumerate(zip(mutation_counts, mutations[0], mutations[1])):\n",
    "            if i + j == 0:\n",
    "                temp.append(h+0)\n",
    "            elif i + j == 1:\n",
    "                temp.append(h+1)\n",
    "            else:\n",
    "                if child_sequences[0][i] == child_sequences[1][i]:\n",
    "                    temp.append(h+1)\n",
    "                else:\n",
    "                    temp.append(h+2)\n",
    "            \n",
    "        mutation_counts = temp     \n",
    "    return (start, stop, mutation_counts)\n",
    "\n",
    "#  IMPORT REFERENCE SEQUENCE\n",
    "\n",
    "for record in SeqIO.parse(datasets_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "    full_sequence = str(record.seq)\n",
    "tb_length = len(full_sequence)\n",
    "\n",
    "#  LOAD VARIANT DICTIONARY AND IDS IN TREE\n",
    "\n",
    "#if full_run == True:\n",
    "if 1==1:\n",
    "    with open(project_dir + '/variant_dict.pkl', 'rb') as f:\n",
    "        variant_dict = pickle.load(f) \n",
    "    with open(project_dir + '/ids.pkl', 'rb') as f:\n",
    "        distinct_sequence_names = pickle.load(f) \n",
    "\n",
    "    variant_positions = []\n",
    "    for k, v in variant_dict.items():\n",
    "        if k in distinct_sequence_names:\n",
    "            for (pos, snp) in v:\n",
    "                variant_positions.append(pos-1)     #Cryptic is 1 indexed\n",
    "    sorted_variant_position_list = list(set(variant_positions))\n",
    "    sorted_variant_position_list.sort()\n",
    "    pos_id_dict = dict(zip(sorted_variant_position_list, range(len(set(variant_positions)))))\n",
    "    id_pos_dict = dict(zip(range(len(set(variant_positions))), sorted_variant_position_list))\n",
    "    with open(project_dir + '/pos_id_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(pos_id_dict, f) \n",
    "    with open(project_dir + '/id_pos_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(id_pos_dict, f) \n",
    "    tb_variants_sequence_length = len(pos_id_dict)\n",
    "    num_chunks = num_cores * num_iterations   \n",
    "    chunk_size = math.ceil(tb_variants_sequence_length/num_chunks)\n",
    "    print(tb_variants_sequence_length)\n",
    "    print(num_chunks, chunk_size, len(sorted_variant_position_list))\n",
    "\n",
    "print(\"Stage_6 Complete\")\n",
    "#if full_run == True:\n",
    "if 1==1:\n",
    "    chunk_variant_dict = defaultdict(lambda: defaultdict(list))\n",
    "    for k, v in variant_dict.items():\n",
    "        if k in distinct_sequence_names:\n",
    "            for (pos, snp) in v:\n",
    "                chunk = int(pos_id_dict[pos-1]/chunk_size)\n",
    "                position_in_chunk = pos_id_dict[pos-1] % chunk_size\n",
    "                chunk_variant_dict[chunk][k].append((position_in_chunk,snp.upper()))\n",
    "    a =list(chunk_variant_dict.keys())\n",
    "    a.sort()\n",
    "    print(a)\n",
    "    print(len(list(chunk_variant_dict.keys())))\n",
    "    max_id = 0\n",
    "    for k, v in pos_id_dict.items():\n",
    "        if v > max_id:\n",
    "            max_id = v\n",
    "    print(\"max \", v)\n",
    "    for chunk in tqdm(list(chunk_variant_dict.keys())):\n",
    "        with open(chunk_variant_dict_dir + '/chunk_variant_dictionary_'+str(chunk) + '.pkl', 'wb') as f:\n",
    "            pickle.dump(chunk_variant_dict[chunk], f)\n",
    "    print(\"Built dictionaries\")\n",
    "print(\"Stage_7 Complete\")\n",
    "#if full_run == True:\n",
    "if 1==1:\n",
    "    for core_1 in tqdm(range(num_iterations)):\n",
    "        parallel_output = Parallel(n_jobs=-1, timeout = timeout)(delayed(generate_variant_sequence_dict)(core_1 * num_cores + core_2) for (core_2) in core_numbers)\n",
    "\n",
    "print(\"Stage_8 Complete\")\n",
    "#if full_run==True:  \n",
    "if 1==1:\n",
    "    res = []\n",
    "    filename_list = list_files(dictionary_dir)\n",
    "    for filename in tqdm(filename_list):\n",
    "        temp_2 = filename.split('_')\n",
    "        start_pos = int(temp_2[2])\n",
    "        end_pos = int(temp_2[3])\n",
    "        parallel_output = Parallel(n_jobs=-1)(delayed(generate_mutation_counts)(dictionary_dir+'/' + filename, core_number) for core_number in range(24))\n",
    "        temp = []\n",
    "        for x in parallel_output:\n",
    "            temp+=x[2]\n",
    "        res.append((int(parallel_output[0][0]), int(parallel_output[0][1]), temp))\n",
    "        with open(mutation_count_dir + '/mutation_counts_'+str(start_pos)+'_'+str(end_pos)+'_' + '.pkl', 'wb') as f:\n",
    "                   pickle.dump((int(parallel_output[0][0]), int(parallel_output[0][1]), temp), f) \n",
    "    res.sort(key = lambda x: x[0])\n",
    "    with open(mutation_count_dir + '/all_mutation_counts.pkl', 'wb') as f:\n",
    "        pickle.dump(res, f) \n",
    "print(\"Stage_9 Complete\")\n",
    "#if full_run==True:  \n",
    "if 1==1:\n",
    "    with open(mutation_count_dir + '/all_mutation_counts.pkl', 'rb') as f:\n",
    "        res = pickle.load(f) \n",
    "    non_zero_mutation_counts = []\n",
    "    for x in res:\n",
    "        non_zero_mutation_counts += x[2]\n",
    "    zero_and_non_zero_mutation_counts = []\n",
    "    for i in range(len(full_sequence)):\n",
    "        if i in pos_id_dict:\n",
    "            zero_and_non_zero_mutation_counts.append(non_zero_mutation_counts[pos_id_dict[i]])\n",
    "        else:\n",
    "            zero_and_non_zero_mutation_counts.append(0)\n",
    "    with open(mutation_count_dir + '/zero_and_non_zero_mutation_counts.pkl', 'wb') as f:\n",
    "        pickle.dump(zero_and_non_zero_mutation_counts, f)  \n",
    "    mutation_df = pd.DataFrame(zero_and_non_zero_mutation_counts, columns = ['Num_Mutations'])\n",
    "    mutation_df.to_csv(project_dir + '/mutation_df.csv')   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "26c96e22-3dcf-4367-bc3f-7775369811ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4411532"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(zero_and_non_zero_mutation_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "95fa29e6-3c08-4154-8b2f-41846f321873",
   "metadata": {},
   "outputs": [],
   "source": [
    "nick = copy.deepcopy(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6ad44fd4-da79-43cf-9e0d-f17736a7b59e",
   "metadata": {},
   "outputs": [],
   "source": [
    "nick.sort(key = lambda x: x[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cd085cfc-bab4-44fb-8dec-a37f4642b738",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nick[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f0b4e2e4-4941-41c2-9f97-c02a4c426928",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage_9 Complete\n"
     ]
    }
   ],
   "source": [
    "res.sort(key = lambda x: x[0])\n",
    "with open(mutation_count_dir + '/all_mutation_counts.pkl', 'wb') as f:\n",
    "    pickle.dump(res, f) \n",
    "print(\"Stage_9 Complete\")\n",
    "#if full_run==True:  \n",
    "if 1==1:\n",
    "    with open(mutation_count_dir + '/all_mutation_counts.pkl', 'rb') as f:\n",
    "        res = pickle.load(f) \n",
    "    non_zero_mutation_counts = []\n",
    "    for x in res:\n",
    "        non_zero_mutation_counts += x[2]\n",
    "    zero_and_non_zero_mutation_counts = []\n",
    "    for i in range(len(full_sequence)):\n",
    "        if i in pos_id_dict:\n",
    "            zero_and_non_zero_mutation_counts.append(non_zero_mutation_counts[pos_id_dict[i]])\n",
    "        else:\n",
    "            zero_and_non_zero_mutation_counts.append(0)\n",
    "    with open(mutation_count_dir + '/zero_and_non_zero_mutation_counts.pkl', 'wb') as f:\n",
    "        pickle.dump(zero_and_non_zero_mutation_counts, f)  \n",
    "    mutation_df = pd.DataFrame(zero_and_non_zero_mutation_counts, columns = ['Num_Mutations'])\n",
    "    mutation_df.to_csv(project_dir + '/mutation_df.csv')   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5e2c2cb1-ff6c-432b-8926-68e80cfe0474",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6c220c5-a335-4906-a497-bec8467e49ed",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
