{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e36786c5-8a4d-4b87-9f58-d6a718cc37c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "from tqdm.auto import tqdm\n",
    "from Comparative_Analysis import Sequence_Analysis_Routines as sar\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from Comparative_Analysis import Alignment as align\n",
    "from Comparative_Analysis import Blast_Functions as blastfn\n",
    "from Comparative_Analysis import ORF_Functions as orffn\n",
    "from goatools import obo_parser\n",
    "import random\n",
    "import copy\n",
    "from joblib import Parallel, delayed\n",
    "import os\n",
    "import wget\n",
    "import shutil\n",
    "import subprocess\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import re\n",
    "import shutil\n",
    "from ftplib import FTP\n",
    "import Bio.UniProt.GOA as GOA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1c8f9536-cd01-45c5-abdb-ef4e9a9dba3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_run = True\n",
    "project_dir = 'D:/Project_Data/Project_8'\n",
    "datasets_dir = project_dir + '/Datasets'\n",
    "output_dir = project_dir + '/Output'\n",
    "r_scape_output_loc = project_dir + '/RScape_Run_Thoth_3'\n",
    "wsl_output_loc = util.wslname(output_dir)\n",
    "seq_dir = 'D:/Actinobacteria_Ref_Rep_Lev_Complete'\n",
    "blast_dir = 'D:/BLAST/actinobacteria_ref_rep_comp'\n",
    "blast_db_name = 'actinobacteria_ref_rep_comp'\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "reference_species_filename = 'GCF_000195955.2_ASM19595v2_genomic.gbff'\n",
    "species_list = util.list_files(seq_dir)\n",
    "reference_species_name = 'Mycobacterium tuberculosis H37Rv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e105d41e-77b1-4010-9cc5-7e146cac95b3",
   "metadata": {},
   "source": [
    "##### Load data for orthologs, RFAM, etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "93cca258-f866-4457-a46f-99420194ac58",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_dir + '/Blast_Output/reciprocal_best_hits.pkl', 'rb') as f:\n",
    "    rbh_results = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "18c8ebd7-9419-4270-8ed1-6e6e492f273d",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_dir + '/Blast_Output/all_to_ref_best_hits.pkl', 'rb') as f:\n",
    "    bh_results = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6ab59b26-ac5f-44c8-a320-d7c80bd70702",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_ortholog_dict = {}\n",
    "for i, r in rbh_results.iterrows():\n",
    "    ref_ortholog_dict[r['target_ref']] = r['query_ref'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2ad86e94-e80c-4873-8d14-5d7fdb1f5ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_bh_dict = {}\n",
    "for i, r in bh_results.iterrows():\n",
    "    ref_bh_dict[r['query_ref']] = r['target_ref'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "13acd23b-f536-497a-ab1e-e96a66a0bed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_dir + '/gene_info_dict.pkl', 'rb') as f:\n",
    "    gene_info_dict = pickle.load(f) \n",
    "with open(output_dir + '/names_dict.pkl', 'rb') as f:\n",
    "    names_dict = pickle.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "acec513d-c0d6-4f12-87b8-dbfc647b6da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rfam_df = pd.read_excel(datasets_dir+'/RFAM_Hits_H37Rv_sorted.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0fe3bfcd-2e82-46ad-b48d-776fba191578",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ortholog_in_reference(accession_locus, ref_ortholog_dict):\n",
    "    if accession_locus in ref_ortholog_dict:\n",
    "        return ref_ortholog_dict[accession_locus].split('@')[1]\n",
    "    else:\n",
    "        return 'NO_ORTHOLOG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8b22418f-3cca-44f2-a865-f0d16e413c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def orthologs(reference_locus):\n",
    "    temp = rbh_results[rbh_results.query_ref == 'NC_000962.3@'+ reference_locus]\n",
    "    orths = []\n",
    "    for i, r in temp.iterrows():\n",
    "        orths.append(r['target_ref'])\n",
    "    return orths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "33efe30b-28e1-46e3-a9c7-6b98ba410f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "mycobrowser_df = pd.read_excel(datasets_dir+'/Mycobrowser_Release_4.xlsx')\n",
    "features = []\n",
    "for i, r in mycobrowser_df.iterrows():\n",
    "    features.append([r['Locus'],r['Start'], r['Product']])\n",
    "features.sort(key=lambda x: x[1])\n",
    "downstream_dict = {}\n",
    "for i, feature in enumerate(features):\n",
    "    if i + 1 < len(features):\n",
    "        downstream_dict[feature[0]] = (features[i+1][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "68c21521-1a88-4e64-90e4-1b25021255fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_downstream_dict = {}\n",
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "features = []\n",
    "for feature in genome_record.features:\n",
    "    if not(feature.type == 'gene'):\n",
    "        a = feature.qualifiers\n",
    "        if a.get(\"locus_tag\")!= None:\n",
    "            features.append((a.get(\"locus_tag\")[0], feature.type))\n",
    "for i,feature in enumerate(features):\n",
    "    if i+1 < len(features):\n",
    "        next_feature = features[i+1][1]\n",
    "        refseq_downstream_dict[feature[0]] = next_feature\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ca6bc272-3ca8-4103-9941-169788fb8096",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "product_dict = {}\n",
    "for feature in genome_record.features:\n",
    "    if not(feature.type == 'gene'):\n",
    "        a = feature.qualifiers\n",
    "        if a.get(\"locus_tag\")!= None and a.get(\"product\")!=None:\n",
    "            product_dict[a.get(\"locus_tag\")[0]] = a.get(\"product\")[0]\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "11e5c731-6dc2-481a-9bc0-29ea3a98b193",
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic_regions = util.list_dirs(r_scape_output_loc)\n",
    "temp_rows = []\n",
    "for id in intergenic_regions:\n",
    "    if os.path.isfile(r_scape_output_loc + '/' + id + '/rscape_3.cacofold.power'):\n",
    "        with open(r_scape_output_loc + '/' + id + '/rscape_3.cacofold.power', 'r') as f:\n",
    "            num_covarying = ''\n",
    "            num_expected = ''  \n",
    "            for l in f:\n",
    "                if 'BPAIRS observed' in l:\n",
    "                    num_covarying = int(l.split('covary ')[1])\n",
    "                if 'BPAIRS expected' in l:\n",
    "                    num_expected = (l.split('covary ')[1])\n",
    "                \n",
    "            with open(r_scape_output_loc + '/' + id + '/summary.txt', 'r') as f:  \n",
    "                rfam_overlaps = []\n",
    "                feature_overlaps = []\n",
    "                region_type_list = []\n",
    "                upstream_ortholog_list = []\n",
    "                downstream_ortholog_list = []\n",
    "                ref_count = 0\n",
    "                tot_count = 0\n",
    "                for l in f:\n",
    "                    if (not ('#' in l)):\n",
    "                        a = l.split()\n",
    "                        species = a[0]\n",
    "                        seq_from = int(a[7])\n",
    "                        seq_to = int(a[8])\n",
    "                        seq_strand = a[9]\n",
    "                        significant = a[16]\n",
    "                        start = min(seq_from,seq_to)\n",
    "                        stop = max(seq_from,seq_to)\n",
    "                        if (significant == '!'):\n",
    "                            tot_count+=1\n",
    "                            feature_list = gene_info_dict[species]  \n",
    "                            feature_list.sort(key=lambda x: x[1])\n",
    "                            max_pct_cover = -1\n",
    "                            max_feature = []\n",
    "                            for feature in feature_list:\n",
    "                                if feature[1] > stop:      #No point to continue as subsequent features are all after the stop \n",
    "                                    continue\n",
    "                                if feature[1] < (stop - 1) and feature[2] >= (start - 1):\n",
    "                                    len_feature = feature[2] - feature[1]\n",
    "                                    pct_cover = (min(feature[2], stop) - max(feature[1], start)) / (stop - start)\n",
    "                                    if pct_cover > 0.05 and len_feature < 100000:\n",
    "                                        if pct_cover > max_pct_cover:\n",
    "                                            max_feature = feature\n",
    "                                            max_pct_cover = pct_cover\n",
    "                                    if len(max_feature) > 0:\n",
    "                                        if max_feature[0][-2:] == 'IG':\n",
    "                                            region_type = 'IG'\n",
    "                                            previous_gene = max_feature[3]\n",
    "                                            next_gene = max_feature[4]\n",
    "                                            upstream_genes = []\n",
    "                                            downstream_genes = []\n",
    "                                            upstream_orthologs = []\n",
    "                                            downstream_orthologs = []\n",
    "                                            if previous_gene[1] == 1:\n",
    "                                                upstream_genes.append(previous_gene[0])\n",
    "                                            else:\n",
    "                                                downstream_genes.append(previous_gene[0])\n",
    "                                            if next_gene[1] == 1:\n",
    "                                                downstream_genes.append(next_gene[0])\n",
    "                                            else:\n",
    "                                                upstream_genes.append(next_gene[0])\n",
    "                                            upstream_orthologs = [ortholog_in_reference(species+ '@' + x, ref_ortholog_dict) for x in upstream_genes]\n",
    "                                            downstream_orthologs = [ortholog_in_reference(species + '@' + x, ref_ortholog_dict) for x in downstream_genes]\n",
    "                                        \n",
    "                                        else:\n",
    "                                            region_type = 'G'\n",
    "                                        \n",
    "                                        region_type_list.append(region_type)\n",
    "                                        if region_type == 'IG':\n",
    "                                            upstream_ortholog_list.append(upstream_orthologs)\n",
    "                                            downstream_ortholog_list.append(downstream_orthologs)\n",
    "                                        \n",
    "                            # RFAM overlaps for reference hits\n",
    "                            if species == 'NC_000962.3':\n",
    "                                ref_count+=1\n",
    "                                for i, r in rfam_df.iterrows():\n",
    "                                    rfam_start = min(int(r['seq from']), int(r['seq to']))\n",
    "                                    rfam_stop = max(int(r['seq from']), int(r['seq to']))\n",
    "                                    pct_cover = (min(rfam_stop, stop) - max(rfam_start, start)) / (rfam_stop - rfam_start)\n",
    "                                    if pct_cover > 0.1:\n",
    "                                        rfam_overlaps.append((r.accession, r['description of target'], pct_cover))\n",
    "            \n",
    "            # Near covarying pairs - suggests not structural, possibly CDS\n",
    "            with open(r_scape_output_loc + '/' + id + '/rscape_3.cacofold.R2R.sto', 'r') as f: \n",
    "                r2r_near_pairs = 0\n",
    "                for l in f:\n",
    "                    if '#=GC SS_cons ' in l:\n",
    "                        for i in range(len(l)-2):\n",
    "                            if l[i:(i+2)] == '<>':\n",
    "                                r2r_near_pairs +=1\n",
    "        \n",
    "        if id in refseq_downstream_dict:\n",
    "            refseq_downstream = refseq_downstream_dict[id]\n",
    "        else:\n",
    "            refseq_downstream = ''\n",
    "        if len(region_type_list) > 0:\n",
    "            intergenic_pct = region_type_list.count('IG') / len(region_type_list)\n",
    "        else:\n",
    "            intergenic_pct = 0\n",
    "        temp_rows.append([id, downstream_dict[id],  refseq_downstream , rfam_overlaps, num_covarying, num_expected, ref_count, tot_count, intergenic_pct, r2r_near_pairs, region_type_list, [], []])\n",
    "    else:\n",
    "        temp_rows.append([id,downstream_dict[id],  refseq_downstream , rfam_overlaps, 0 , 0, 0, 0, 0, 0,[],[],[]])\n",
    "results_df = pd.DataFrame(temp_rows, columns = ['Downstream_of','Feature_Downstream','Refseq_Downstream','RFAM_Overlaps','num_BP_covarying', 'num_Expected', 'num_reference_hits','tot_hits','intergenic_region_pct', 'num_r2r_near_pairs','region_types','upstream_orthologs', 'downstream_orthologs'])\n",
    "results_df.to_csv(r_scape_output_loc + '/intergenic_regions_covariation_summary.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a1729e-cc34-45a8-9403-86db042a7612",
   "metadata": {},
   "source": [
    "####   Print out all hits for particular region  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d143e744-a076-4317-8225-80d46e0fb440",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upstream counts\n",
      "Rv0816c\n",
      "GTGGTCAACTGGTTAGGCGGGACGGGGAGTGTTACTGCTTGACTGCTCCTTGGGACGTCTGTTGCACAGAAACGGCGGGCGACACGCTACGGTGGGGCTCCTAGGCTGCTCTAAGTGCTGCGCGGACGTGCGCGGCTACTCAGCAGCTACAGCAACAGCAACAACCCGCTAGGCGGCACAGATCAACTGCGCGACGCTTGGTGAGCATGGGCTCGATGCGGGCTGACACGTCGGACAGCTTACCCAA\n",
      "[('Mycobacterium noviomagense', 1135097, 1134854), ('Mycobacterium malmoense', 4711803, 4711542), ('Mycobacterium heckeshornense', 695142, 695384), ('Mycobacterium mantenii', 5317902, 5318168), ('Mycolicibacterium gilvum Spyr1', 1051397, 1051648), ('Mycobacterium haemophilum DSM 44634', 3598610, 3598339), ('Mycobacterium branderi', 2143565, 2143800), ('Mycobacterium rufum', 4825600, 4825353), ('Mycobacterium parmense', 4966026, 4966289), ('Mycolicibacterium psychrotolerans', 1778568, 1778819), ('Mycobacterium canettii CIPT 140010059', 917065, 917311), ('Mycobacterium tuberculosis variant bovis BCG str. Pasteur 1173P2', 939775, 940021), ('Mycobacterium tuberculosis variant bovis AF2122/97', 910183, 910429), ('Mycobacterium tuberculosis H37Rv', 909353, 909599), ('Mycolicibacterium anyangense', 3284125, 3283869), ('Mycobacterium marseillense', 641042, 641309), ('Mycolicibacterium gadium', 3595885, 3595621), ('Mycobacterium intracellulare ATCC 13950', 658546, 658813), ('Mycobacterium paraintracellulare', 830702, 830969), ('Mycolicibacter sinensis', 3750409, 3750159), ('Mycobacterium novum', 2735270, 2735520), ('Mycobacterium heidelbergense', 2783005, 2783268), ('Mycobacterium simiae', 3293406, 3293667), ('Mycolicibacterium moriokaense', 1396812, 1397090), ('Mycobacterium cookii', 2689015, 2688759), ('Mycobacterium florentinum', 4433614, 4433875), ('Mycobacterium lepromatosis', 2854174, 2854422), ('Mycolicibacterium sarraceniae', 893937, 893680), ('Mycolicibacterium aurum', 5116591, 5116337), ('Mycolicibacter hiberniae', 3075573, 3075324), ('Mycolicibacterium celeriflavum', 4080878, 4081139), ('Mycolicibacterium mucogenicum DSM 44124', 570959, 571200), ('Mycolicibacterium phocaicum', 3175894, 3175653), ('Mycobacterium grossiae', 397492, 397248), ('Mycolicibacter terrae', 3712271, 3712022), ('Mycolicibacterium confluentis', 357402, 357644), ('Mycobacterium koreense', 3778597, 3778837), ('Mycobacterium stomatepiae', 4702124, 4702386), ('Mycobacterium paraterrae', 673137, 673385), ('Mycolicibacterium boenickei', 2260509, 2260773), ('Mycolicibacterium arabiense', 4716130, 4716371), ('Mycolicibacterium chitae', 707104, 707359), ('Mycolicibacterium parafortuitum', 2390906, 2391157), ('Mycolicibacterium sediminis', 48100, 47857), ('Mycolicibacterium fortuitum', 1110807, 1111071), ('Mycolicibacterium litorale', 5283576, 5283829), ('Mycobacterium dioxanotrophicus', 6282962, 6282708), ('Mycobacterium conspicuum', 5120525, 5120263), ('Mycolicibacterium baixiangningiae', 4944674, 4944407), ('Mycolicibacterium monacense', 5830189, 5830441), ('Mycolicibacterium helvum', 3131172, 3131427), ('Mycobacterium spongiae', 4891755, 4891494), ('Mycobacterium crocinum', 5198000, 5197741), ('Mycobacterium pallens', 5151839, 5151580), ('Mycobacterium virginiense', 4111148, 4110895), ('Mycobacterium senegalense', 5113155, 5112891), ('Mycobacteroides immunogenum', 858932, 859161), ('Mycolicibacterium farcinogenes', 1391695, 1391959), ('Mycobacterium shinjukuense', 1003358, 1003585), ('Mycolicibacterium alvei', 4597239, 4597503), ('Mycolicibacterium aichiense', 734381, 734639), ('Mycobacteroides salmoniphilum', 652666, 652895), ('Mycolicibacterium mageritense', 1278299, 1278049), ('Mycolicibacterium hassiacum DSM 44199', 3826183, 3825905), ('Mycobacterium shigaense', 4637298, 4637032), ('Mycolicibacterium madagascariense', 4921688, 4921443), ('Mycolicibacter minnesotensis', 1103992, 1103756), ('Mycobacteroides chelonae CCUG 47445', 698538, 698764), ('Mycolicibacterium mengxianglii', 4870219, 4869968), ('[Mycobacterium] stephanolepidis', 690446, 690672), ('Mycobacterium goodii', 5718458, 5718197), ('Mycobacteroides abscessus', 2945187, 2945413), ('Mycobacteroides saopaulense', 654176, 654414), ('Mycolicibacterium tokaiense', 2534555, 2534809), ('Mycolicibacterium pulveris', 3095834, 3096088), ('Mycobacterium diernhoferi', 5234720, 5234479), ('Mycolicibacterium neoaurum', 1504649, 1504892), ('Rhodococcus qingshengii', 5029885, 5029603), ('Mycolicibacillus parakoreensis', 584639, 584864), ('Mycobacterium leprae', 2546813, 2546558), ('Rhodococcus erythropolis R138', 1920364, 1920645), ('Mycobacterium holsaticum', 807500, 807778), ('Rhodococcus globerulus', 5487923, 5488228), ('Hoyosella subflava DQS3-9A1', 780155, 780401), ('Nocardia huaxiensis', 980053, 980350), ('Mycobacterium doricum', 3234648, 3234403), ('Rhodococcus koreensis', 5858008, 5858283), ('Nocardia seriolae', 7415657, 7415381), ('Rhodococcus opacus PD630', 1392254, 1392530), ('Nocardia farcinica IFM 10152', 614633, 614924), ('Nocardia tengchongensis', 689066, 688784), ('Rhodococcus fascians D188', 825861, 825594), ('Rhodococcus triatomae', 349472, 349746), ('Nocardia yamanashiensis', 6116941, 6116641), ('Rhodococcus hoagii', 989355, 989657), ('Rhodococcus pyridinivorans SB3094', 4270032, 4270306), ('Rhodococcus aetherivorans', 4348394, 4348111), ('Mycolicibacterium insubricum', 4364744, 4364514), ('Nocardia gipuzkoensis', 3591289, 3591588), ('Nocardia wallacei', 648762, 649050), ('Nocardia cyriacigeorgica GUH-2', 649124, 649423), ('Nocardia mangyaensis', 610031, 610317), ('Nocardia iowensis', 767745, 768044), ('Nocardia asteroides', 1409611, 1409890), ('Nocardia otitidiscaviarum', 6266611, 6266313), ('Rhodococcus ruber', 1366017, 1366314), ('Nocardia brasiliensis', 5822930, 5822631), ('Mycolicibacterium fallax', 3658268, 3658540), ('Rhodococcus coprophilus', 2096527, 2096201), ('Nocardia terpenica', 817305, 817612), ('Tomitella fengzijianii', 1149431, 1149623), ('Tomitella gaofuii', 1319029, 1319254), ('Skermania piniformis', 3593924, 3593628), ('Actinomadura graeca', 5875760, 5875558), ('Thermomonospora curvata DSM 43183', 670131, 669926), ('Thermomonospora amylolytica', 800883, 800685), ('Amycolatopsis japonica', 8580791, 8580586), ('Amycolatopsis keratiniphila', 8572088, 8571883), ('Amycolatopsis orientalis', 381230, 381438), ('Kutzneria albida DSM 43870', 9534139, 9533910), ('Pseudonocardia autotrophica', 534564, 534742), ('Kibdelosporangium phytohabitans', 11437726, 11437526), ('Amycolatopsis aidingensis', 7299870, 7299636), ('Nocardioides euryhalodurans', 3065715, 3065455), ('Amycolatopsis mediterranei RB', 9874247, 9874014), ('Amycolatopsis albispora', 6862404, 6862654), ('Allosaccharopolyspora coralli', 4606588, 4606438), ('Saccharopolyspora pogona', 6837390, 6837572), ('Nocardioides okcheonensis', 2183853, 2184117), ('Streptomonospora sp. M2', 378958, 378720)]\n",
      "\n",
      "Downstream counts\n"
     ]
    }
   ],
   "source": [
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "full_sequence = str(genome_record.seq)\n",
    "with open(r_scape_output_loc + '/' + 'Rv0815c' + '/summary.txt', 'r') as f:  \n",
    "                hit_dict = {}\n",
    "                ref_count = 0\n",
    "                tot_count = 0\n",
    "                for l in f:\n",
    "                    if (not ('#' in l)):\n",
    "                        a = l.split()\n",
    "                        accession = a[0]\n",
    "                        species = names_dict[a[0]]\n",
    "                        seq_from = int(a[7])\n",
    "                        seq_to = int(a[8])\n",
    "                        seq_strand = a[9]\n",
    "                        significant = a[16]\n",
    "                        start = min(seq_from,seq_to)\n",
    "                        stop = max(seq_from,seq_to)\n",
    "                        if (significant == '!'):\n",
    "                            tot_count+=1\n",
    "                            feature_list = gene_info_dict[accession]  \n",
    "                            feature_list.sort(key=lambda x: x[1])\n",
    "                            max_pct_cover = -1\n",
    "                            max_feature = []\n",
    "                            for feature in feature_list:\n",
    "                                if feature[1] > stop:      #No point to continue as subsequent features are all after the stop \n",
    "                                    continue\n",
    "                                if feature[1] < (stop - 1) and feature[2] >= (start - 1):\n",
    "                                    len_feature = feature[2] - feature[1]\n",
    "                                    pct_cover = (min(feature[2], stop) - max(feature[1], start)) / (stop - start)\n",
    "                                    if pct_cover > 0.05 and len_feature < 100000:\n",
    "                                        if pct_cover > max_pct_cover:\n",
    "                                            max_feature = feature\n",
    "                                            max_pct_cover = pct_cover\n",
    "                            if len(max_feature) > 0:\n",
    "                                if max_feature[0][-2:] == 'IG':\n",
    "                                    region_type = 'IG'\n",
    "                                    previous_gene = max_feature[3]\n",
    "                                    next_gene = max_feature[4]\n",
    "                                    output = []\n",
    "                                    if previous_gene[1] == 1:\n",
    "                                        output.append(('Upstream gene', ortholog_in_reference(accession + '@' + previous_gene[0], ref_ortholog_dict), a[15], max_pct_cover, seq_from, seq_to)) \n",
    "                                    else:\n",
    "                                         output.append(('Downstream gene', ortholog_in_reference(accession + '@' + previous_gene[0], ref_ortholog_dict), a[15], max_pct_cover, seq_from, seq_to))  \n",
    "\n",
    "                                    if next_gene[1] == 1:\n",
    "                                        output.append(('Downstream gene', ortholog_in_reference(accession + '@' + next_gene[0], ref_ortholog_dict), a[15], max_pct_cover, seq_from, seq_to)) \n",
    "                                    else:\n",
    "                                        output.append(('Upstream gene', ortholog_in_reference(accession + '@' + next_gene[0], ref_ortholog_dict), a[15], max_pct_cover, seq_from, seq_to))\n",
    "                                    \n",
    "                                                                    \n",
    "                                    \n",
    "                                    if species in hit_dict:\n",
    "                                        #hit_dict[species].append(output)\n",
    "                                        hit_dict[species] = hit_dict[species] + output\n",
    "                                    else:\n",
    "                                        hit_dict[species] = output\n",
    "\n",
    "upstream_count_dict = {}                                                \n",
    "downstream_count_dict = {}\n",
    "upstream_species_dict = {}                                                \n",
    "downstream_species_dict = {}\n",
    "for k, v in hit_dict.items():\n",
    "    #if 'H37' in k:\n",
    "    if 1==1:\n",
    "        #print(k)\n",
    "        #print(' ')\n",
    "        product_list = []\n",
    "        for x in v:\n",
    "            if x[0] == 'Upstream gene':\n",
    "                if x[1] in upstream_count_dict:\n",
    "                    upstream_count_dict[x[1]] += 1\n",
    "                    upstream_species_dict[x[1]].append((k, x[4],x[5])) \n",
    "                else:\n",
    "                    upstream_count_dict[x[1]] = 1\n",
    "                    upstream_species_dict[x[1]] = [(k, x[4],x[5])]\n",
    "            else:\n",
    "                if x[1] in downstream_count_dict:\n",
    "                    downstream_count_dict[x[1]] += 1\n",
    "                    downstream_species_dict[x[1]].append((k, x[4],x[5]))\n",
    "                else:\n",
    "                    downstream_count_dict[x[1]] = 1\n",
    "                    downstream_species_dict[x[1]] = [(k, x[4],x[5])]\n",
    "            if x[1] in product_dict:\n",
    "                product_list.append((x, product_dict[x[1]]))\n",
    "            else:\n",
    "                product_list.append((x, ''))\n",
    "        \n",
    "        #print(len([x for x in product_list if x[0][0] == 'Upstream gene']))\n",
    "        #print(sorted([x[0][1] for x in product_list if x[0][0] == 'Upstream gene']))\n",
    "        #print(' ')   \n",
    "print('Upstream counts')\n",
    "#print (dict(sorted(upstream_count_dict.items(), key=lambda item: item[1], reverse=True)))\n",
    "for k, v in dict(sorted(upstream_species_dict.items())).items():\n",
    "    if not(k == 'NO_ORTHOLOG') and reference_species_name in [x[0] for x in v]:\n",
    "        print(k)\n",
    "        for x in v:\n",
    "            if reference_species_name in x[0]:\n",
    "                if x[1] < x[2]:\n",
    "                    print(full_sequence[x[1]-1:x[2]])\n",
    "                else:\n",
    "                    print(util.reverse_complement(full_sequence[x[2]-1:x[1]]))\n",
    "        \n",
    "        print(v)\n",
    "        print()\n",
    "\n",
    "print('Downstream counts')\n",
    "for k, v in dict(sorted(downstream_species_dict.items())).items():\n",
    "    if not(k == 'NO_ORTHOLOG') and reference_species_name in [x[0] for x in v]:\n",
    "        print(k)\n",
    "        print(v)\n",
    "        print(k)\n",
    "        for x in v:\n",
    "            if reference_species_name in x[0]:\n",
    "                if x[1] < x[2]:\n",
    "                    print(full_sequence[x[1]-1:x[2]])\n",
    "                else:\n",
    "                    print(util.reverse_complement(full_sequence[x[2]-1:x[1]]))\n",
    "        \n",
    "        print(v)\n",
    "        print()\n",
    "#print (dict(sorted(downstream_count_dict.items(), key=lambda item: item[1], reverse=True)))\n",
    "                                        \n",
    "                           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "50b7c414-1206-475d-a5f8-b5c2a492aecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "features = []\n",
    "for feature in genome_record.features:\n",
    "      if not(feature.type in ['CDS','source']):\n",
    "        if 'note' in feature.qualifiers:\n",
    "            a = feature.qualifiers['note']\n",
    "        else:\n",
    "            a = ''\n",
    "        features.append((feature.type, a, int(feature.location.start), int(feature.location.end)))\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8ce0dbc9-635e-4fff-8dc6-1ba55ce884eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '3351086', '3351216', '+', 'no', '1', '0.76', '11.4', '123.4', '2.4e-29', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1907449', '1907582', '+', 'no', '1', '0.63', '0.1', '99.7', '2e-22', '!', '-']\n",
      "('repeat_region', ['56 bp direct repeat 1, AGTCGGGTGACGATGCGGGCCGGTGTGGTCCGAGGAGGAGCCCGACAATTTAAGCT'], 1907459, 1907515)\n",
      "('repeat_region', ['56 bp direct repeat 2, AGTCGGGTGACGATGCGGGCCGGTGTGGTCCGAGGAGGAGCCCGACAATTTAAGCT'], 1907515, 1907571)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '3690943', '3691077', '+', 'no', '1', '0.67', '1.8', '97.5', '8.9e-22', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '12', '115', '3594452', '3594349', '-', 'no', '1', '0.70', '2.3', '89.0', '2.7e-19', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '4087618', '4087487', '-', 'no', '1', '0.73', '5.6', '88.5', '3.6e-19', '!', '-']\n",
      "('gene', '', 4087609, 4088188)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '2401937', '2401808', '-', 'no', '1', '0.68', '2.1', '85.6', '2.6e-18', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '2069077', '2068948', '-', 'no', '1', '0.71', '2.9', '85.3', '3.2e-18', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '2703892', '2704026', '+', 'no', '1', '0.76', '14.7', '83.6', '1e-17', '!', '-']\n",
      "('gene', '', 2704008, 2704437)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '9', '119', '3820391', '3820504', '+', 'no', '1', '0.72', '3.7', '82.6', '1.9e-17', '!', '-']\n",
      "('gene', '', 3818041, 3820402)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '10', '118', '2461325', '2461434', '+', 'no', '1', '0.70', '9.8', '80.0', '1.1e-16', '!', '-']\n",
      "('gene', '', 2459677, 2461327)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '2074433', '2074549', '+', 'no', '1', '0.78', '20.0', '79.7', '1.3e-16', '!', '-']\n",
      "('gene', '', 2073942, 2074437)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1895459', '1895590', '+', 'no', '1', '0.73', '9.9', '77.0', '8.5e-16', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '595459', '595323', '-', 'no', '1', '0.72', '7.1', '73.6', '8.1e-15', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1282026', '1281888', '-', 'no', '1', '0.73', '9.2', '71.0', '4.7e-14', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '917742', '917609', '-', 'no', '1', '0.70', '6.5', '68.0', '3.6e-13', '!', '-']\n",
      "('gene', '', 916476, 917646)\n",
      "('gene', '', 917733, 918750)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '4110667', '4110825', '+', 'no', '1', '0.71', '12.1', '67.1', '6.6e-13', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1267273', '1267139', '-', 'no', '1', '0.69', '4.1', '63.6', '6.9e-12', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '569814', '569935', '+', 'no', '1', '0.80', '20.4', '58.9', '1.6e-10', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '2372425', '2372560', '+', 'no', '1', '0.76', '19.4', '58.8', '1.8e-10', '!', '-']\n",
      "('gene', '', 2370904, 2372569)\n",
      "('repeat_region', ['56 bp direct repeat 1, GCCCGCCGACGATGCGGGCCGCGCAGCGGGCCGCTGAGGAGGCGGGCATCAAGCAA'], 2372436, 2372492)\n",
      "('repeat_region', ['56 bp direct repeat 2, GCCCGCCGACGATGCGGGCCGCGCAGCGGGCCGCTGAGGAGGCGGGCATCAAGCAA'], 2372493, 2372549)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '4120908', '4121042', '+', 'no', '1', '0.73', '9.2', '58.5', '2e-10', '!', '-']\n",
      "('gene', '', 4119794, 4120955)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1305490', '1305635', '+', 'no', '1', '0.75', '11.9', '56.1', '1.1e-09', '!', '-']\n",
      "('gene', '', 1302930, 1305501)\n",
      "('repeat_region', ['62 bp direct repeat copy 1, GGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGCGGCCCGTTGAGGAGCGGGGCA ATCT'], 1305494, 1305556)\n",
      "('repeat_region', ['62 bp direct repeat copy 2, GGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGCGGCCCGTTGAGGAGCGGGGCA ATCT'], 1305556, 1305618)\n",
      "('repeat_region', ['62 bp direct repeat partial copy 3 (43/62 bp),GGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGGGGCCCG'], 1305618, 1305661)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '12', '116', '4053136', '4053297', '+', 'no', '1', '0.70', '14.7', '56.0', '1.1e-09', '!', '-']\n",
      "('repeat_region', ['111 bp direct repeat unit 4, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG'], 4053105, 4053216)\n",
      "('repeat_region', ['111 bp direct repeat unit 3, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG'], 4053216, 4053327)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1955684', '1955569', '-', 'no', '1', '0.69', '1.0', '56.0', '1.1e-09', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '16', '74', '4247298', '4247240', '-', 'no', '1', '0.71', '0.0', '53.6', '5.7e-09', '!', '-']\n",
      "('gene', '', 4246513, 4249810)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '1443493', '1443358', '-', 'no', '1', '0.67', '1.1', '52.7', '1e-08', '!', '-']\n",
      "('gene', '', 1442766, 1443399)\n",
      "('gene', '', 1443481, 1445047)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '3291511', '3291380', '-', 'no', '1', '0.73', '10.1', '50.3', '5.2e-08', '!', '-']\n",
      "('gene', '', 3291502, 3296353)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '12', '112', '577286', '577390', '+', 'no', '1', '0.67', '1.2', '48.6', '1.7e-07', '!', '-']\n",
      "('gene', '', 576786, 577338)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '154123', '154249', '+', 'no', '1', '0.66', '0.4', '47.0', '4.6e-07', '!', '-']\n",
      "('gene', '', 152323, 154129)\n",
      "('repeat_region', ['53 bp Mycobacterial Interspersed Repetitive Unit,Class III (see Supply et al., 1997)'], 154072, 154125)\n",
      "('repeat_region', ['53 bp Mycobacterial Interspersed Repetitive Unit,Class II'], 154125, 154178)\n",
      "('repeat_region', ['53 bp Mycobacterial Interspersed Repetitive Unit,Class II'], 154178, 154231)\n",
      "('gene', '', 154231, 155599)\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '2115761', '2115620', '-', 'no', '1', '0.71', '8.5', '46.5', '6.4e-07', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '17', '73', '2464869', '2464924', '+', 'no', '1', '0.79', '1.2', '43.8', '4e-06', '!', '-']\n",
      "['NC_000962.3', '-', 'rscape_2', '-', 'cm', '1', '133', '577391', '577519', '+', 'no', '1', '0.63', '0.3', '43.3', '5.6e-06', '!', '-']\n"
     ]
    }
   ],
   "source": [
    "with open(r_scape_output_loc + '/' + 'Rv1668c' + '/summary.txt', 'r') as f:  \n",
    "                for l in f:\n",
    "                    if (not ('#' in l)):\n",
    "                        a = l.split()\n",
    "                        accession = a[0]\n",
    "                        species = names_dict[a[0]]\n",
    "                        if species == reference_species_name:\n",
    "                            seq_from = int(a[7])\n",
    "                            seq_to = int(a[8])\n",
    "                            seq_strand = a[9]\n",
    "                            significant = a[16]\n",
    "                            start = min(seq_from,seq_to)\n",
    "                            stop = max(seq_from,seq_to)\n",
    "                            if (significant == '!'):\n",
    "                                print(a)\n",
    "                                for feature in features:\n",
    "                                    if feature[2] < stop and feature[3] > start:\n",
    "                                        print(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "879fc175-a098-4c82-839c-a166335c9894",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('Mycobacterium shinjukuense', 4), ('Rhodococcus koreensis', 4), ('Mycobacterium noviomagense', 2), ('Mycobacterium malmoense', 2), ('Mycobacterium heckeshornense', 2), ('Mycobacterium mantenii', 2), ('Mycolicibacterium gilvum Spyr1', 2), ('Mycobacterium haemophilum DSM 44634', 2), ('Mycobacterium branderi', 2), ('Mycobacterium rufum', 2), ('Mycobacterium parmense', 2), ('Mycolicibacterium psychrotolerans', 2), ('Mycobacterium canettii CIPT 140010059', 2), ('Mycobacterium tuberculosis variant bovis BCG str. Pasteur 1173P2', 2), ('Mycobacterium tuberculosis variant bovis AF2122/97', 2), ('Mycobacterium tuberculosis H37Rv', 2), ('Mycolicibacterium anyangense', 2), ('Mycobacterium marseillense', 2), ('Mycolicibacterium gadium', 2), ('Mycobacterium intracellulare ATCC 13950', 2), ('Mycobacterium paraintracellulare', 2), ('Mycolicibacter sinensis', 2), ('Mycobacterium novum', 2), ('Mycobacterium heidelbergense', 2), ('Mycobacterium seoulense', 2), ('Mycobacterium simiae', 2), ('Mycolicibacterium moriokaense', 2), ('Mycobacterium cookii', 2), ('Mycobacterium florentinum', 2), ('Mycobacterium kansasii ATCC 12478', 2), ('Mycobacterium lepromatosis', 2), ('Mycolicibacterium sarraceniae', 2), ('Mycolicibacterium aurum', 2), ('Mycobacterium ostraviense', 2), ('Mycolicibacter hiberniae', 2), ('Mycolicibacterium celeriflavum', 2), ('Mycolicibacterium mucogenicum DSM 44124', 2), ('Mycolicibacterium phocaicum', 2), ('Mycobacterium grossiae', 2), ('Mycolicibacter terrae', 2), ('Mycolicibacterium confluentis', 2), ('Mycobacterium paraseoulense', 2), ('Mycobacterium pseudoshottsii JCM 15466', 2), ('Mycobacterium marinum', 2), ('Mycobacterium koreense', 2), ('Mycobacterium stomatepiae', 2), ('Mycobacterium paraterrae', 2), ('Mycolicibacterium boenickei', 2), ('Mycolicibacterium arabiense', 2), ('Mycolicibacterium chitae', 2), ('Mycolicibacterium parafortuitum', 2), ('Mycolicibacterium sediminis', 2), ('Mycolicibacterium fortuitum', 2), ('Mycobacterium vicinigordonae', 2), ('Mycolicibacterium litorale', 2), ('Mycobacterium shottsii', 2), ('Mycobacterium dioxanotrophicus', 2), ('Mycobacterium conspicuum', 2), ('Mycolicibacterium baixiangningiae', 2), ('Mycolicibacterium monacense', 2), ('Mycolicibacterium helvum', 2), ('Mycobacterium spongiae', 2), ('Mycobacterium crocinum', 2), ('Mycobacterium pallens', 2), ('Mycobacterium ulcerans', 2), ('Mycobacterium virginiense', 2), ('Mycobacterium senegalense', 2), ('Mycobacteroides immunogenum', 2), ('Mycolicibacterium farcinogenes', 2), ('Mycobacterium basiliense', 2), ('Mycolicibacterium alvei', 2), ('Mycobacterium lacus', 2), ('Mycolicibacterium aichiense', 2), ('Mycobacteroides salmoniphilum', 2), ('Mycolicibacterium mageritense', 2), ('Mycolicibacterium hassiacum DSM 44199', 2), ('Mycobacterium shigaense', 2), ('Mycolicibacterium madagascariense', 2), ('Mycolicibacter minnesotensis', 2), ('Mycobacteroides chelonae CCUG 47445', 2), ('Mycolicibacterium mengxianglii', 2), ('[Mycobacterium] stephanolepidis', 2), ('Mycobacterium goodii', 2), ('Mycobacteroides abscessus', 2), ('Mycobacteroides saopaulense', 2), ('Mycolicibacterium tokaiense', 2), ('Mycolicibacterium pulveris', 2), ('Mycobacterium diernhoferi', 2), ('Mycolicibacterium neoaurum', 2), ('Mycobacterium paragordonae', 2), ('Rhodococcus qingshengii', 2), ('Mycolicibacillus parakoreensis', 2), ('Mycobacterium leprae', 2), ('Rhodococcus erythropolis R138', 2), ('Mycobacterium holsaticum', 2), ('Rhodococcus globerulus', 2), ('Hoyosella subflava DQS3-9A1', 2), ('Nocardia huaxiensis', 2), ('Mycobacterium doricum', 2), ('Nocardia seriolae', 2), ('Rhodococcus opacus PD630', 2), ('Mycolicibacterium thermoresistibile', 2), ('Nocardia farcinica IFM 10152', 2), ('Nocardia tengchongensis', 2), ('Rhodococcus fascians D188', 2), ('Rhodococcus triatomae', 2), ('Nocardia yamanashiensis', 2), ('Rhodococcus hoagii', 2), ('Rhodococcus pyridinivorans SB3094', 2), ('Rhodococcus aetherivorans', 2), ('Mycolicibacterium insubricum', 2), ('Nocardia gipuzkoensis', 2), ('Nocardia wallacei', 2), ('Nocardia cyriacigeorgica GUH-2', 2), ('Nocardia mangyaensis', 2), ('Nocardia iowensis', 2), ('Nocardia asteroides', 2), ('Nocardia otitidiscaviarum', 2), ('Rhodococcus ruber', 2), ('Nocardia brasiliensis', 2), ('Mycolicibacterium fallax', 2), ('Rhodococcus coprophilus', 2), ('Nocardia terpenica', 2), ('Tomitella fengzijianii', 2), ('Tomitella gaofuii', 2), ('Skermania piniformis', 2), ('Gordonia rubripertincta', 2), ('Gordonia otitidis', 2), ('Gordonia amicalis', 2), ('Gordonia amarae', 2), ('Gordonia jinghuaiqii', 2), ('Gordonia hongkongensis', 2), ('Gordonia zhaorongruii', 2), ('Actinomadura graeca', 2), ('Gordonia insulae', 2), ('Thermobispora bispora', 2), ('Gordonia polyisoprenivorans VH2', 2), ('Gordonia pseudoamarae', 2), ('Gordonia bronchialis DSM 43247', 2), ('Gordonia alkanivorans', 2), ('Thermomonospora curvata DSM 43183', 2), ('Thermomonospora amylolytica', 2), ('Tsukamurella paurometabola DSM 20162', 2), ('Amycolatopsis japonica', 2), ('Gordonia phthalatica', 2), ('Amycolatopsis keratiniphila', 2), ('Actinoalloteichus hoggarensis', 2), ('Amycolatopsis orientalis', 2), ('Acidothermus cellulolyticus 11B', 2), ('Actinoalloteichus fjordicus', 2), ('Nonomuraea coxensis DSM 45129', 2), ('Salinispora tropica CNB-440', 2), ('Actinoalloteichus hymeniacidonis', 2), ('Kutzneria albida DSM 43870', 2), ('Streptosporangium roseum DSM 43021', 2), ('Pseudonocardia autotrophica', 2), ('Nonomuraea gerenzanensis', 2), ('Natronosporangium hydrolyticum', 2), ('Nonomuraea nitratireducens', 2), ('Amycolatopsis acidiphila', 2), ('Actinoplanes friuliensis DSM 7358', 2), ('Kibdelosporangium phytohabitans', 2), ('Nocardioides baekrokdamisoli', 2), ('Polymorphospora rubra', 2), ('Lentzea guizhouensis', 2), ('Actinoplanes teichomyceticus ATCC 31121', 2), ('Pseudonocardia broussonetiae', 2), ('Saccharothrix espanaensis DSM 44229', 2), ('Phytohabitans suffuscus', 2), ('Phytohabitans flavus', 2), ('Actinoplanes ianthinogenes', 2), ('Dactylosporangium vinaceum', 2), ('Micromonospora craniellae', 2), ('Micromonospora aurantiaca ATCC 27029', 2), ('Actinoplanes missouriensis 431', 2), ('Kitasatospora aureofaciens', 2), ('Catenulispora acidiphila DSM 44928', 2), ('Micromonospora tulbaghiae', 2), ('Nocardioides panacisoli', 2), ('Micromonospora maris AB-18-032', 2), ('Streptomyces coeruleorubidus', 2), ('Micromonospora terminaliae', 2), ('Streptomyces sp. ZYC-3', 2), ('Streptomyces dengpaensis', 2), ('Nocardioides mesophilus', 2), ('Streptomyces seoulensis', 2), ('Streptomyces pristinaespiralis', 2), ('Amycolatopsis aidingensis', 2), ('Streptomyces cadmiisoli', 2), ('Streptomyces broussonetiae', 2), ('Micromonospora carbonacea', 2), ('Nocardioides dongkuii', 2), ('Streptomyces fungicidicus', 2), ('Streptomyces sp. CRXT-Y-14', 2), ('Nocardioides euryhalodurans', 2), ('Actinosynnema mirum DSM 43827', 2), ('Amycolatopsis mediterranei RB', 2), ('Streptomyces tirandamycinicus', 2), ('Streptomyces cyaneochromogenes', 2), ('Streptomyces aquilus', 2), ('Streptomyces kanamyceticus', 2), ('Frankia alni ACN14a', 2), ('Streptomyces aurantiacus', 2), ('Nocardiopsis gilva YIM 90087', 2), ('Amycolatopsis albispora', 2), ('Streptomyces chromofuscus', 2), ('Saccharothrix syringae', 2), ('Saccharopolyspora gregorii', 2), ('Streptomyces pratensis', 2), ('Streptomyces avermitilis MA-4680 = NBRC 14893', 2), ('Streptomyces hundungensis', 2), ('Actinosynnema pretiosum', 2), ('Streptomyces dangxiongensis', 2), ('Allosaccharopolyspora coralli', 2), ('Streptomyces murinus', 2), ('Streptomyces tuirus', 2), ('Streptomyces ferrugineus', 2), ('Thermobifida alba', 2), ('Streptomyces calvus', 2), ('Streptomyces cyanogenus', 2), ('Streptomyces rectiverticillatus', 2), ('Streptomyces qaidamensis', 2), ('Streptomyces griseofuscus', 2), ('Streptomyces ficellus', 2), ('Saccharopolyspora spinosa', 2), ('Streptomyces gardneri', 2), ('Saccharopolyspora pogona', 2), ('Streptacidiphilus bronchialis', 2), ('Streptomyces nodosus', 2), ('Streptomyces genisteinicus', 2), ('Frankia inefficax', 2), ('Streptomyces cattleya NRRL 8057 = DSM 46488', 2), ('Streptomyces galilaeus', 2), ('Streptomyces griseorubiginosus', 2), ('Streptomyces qinzhouensis', 2), ('Nocardioides okcheonensis', 2), ('Streptomyces hawaiiensis', 2), ('Streptomonospora sp. M2', 2), ('Kitasatospora albolonga', 2), ('Streptomyces tsukubensis', 2), ('Streptomyces anulatus', 2), ('Streptomyces bathyalis', 2), ('Streptomyces clavuligerus', 2), ('Streptomyces solisilvae', 2), ('Streptomyces chartreusis', 2), ('Streptomyces californicus', 2), ('Streptomyces venezuelae', 2), ('Streptomyces violaceoruber', 2), ('Thermobifida halotolerans', 2), ('Streptomyces formicae', 2), ('Streptomyces olivaceus', 2), ('Streptomyces buecherae', 2), ('Streptomyces alfalfae', 2), ('Nocardioides aquaticus', 2), ('Streptomyces viridosporus T7A', 2), ('Streptomyces griseoviridis', 2), ('Streptomyces vinaceus', 2), ('Streptomyces mobaraensis', 2), ('Streptomyces pluripotens', 2), ('Streptomyces spectabilis', 2), ('Nocardioides coralli', 2), ('Streptomyces olivoreticuli subsp. olivoreticuli', 2), ('Nocardioides rotundus', 2), ('Streptomyces rimosus subsp. rimosus ATCC 10970', 2), ('Kribbella qitaiheensis', 2), ('Streptomyces sp. 1_25', 2), ('Streptomyces fradiae ATCC 10745 = DSM 40063', 2), ('Streptomyces microflavus', 2), ('Streptomyces griseochromogenes', 2), ('Streptomyces xinghaiensis S187', 2)]\n"
     ]
    }
   ],
   "source": [
    "hits = []\n",
    "for k, v in hit_dict.items():\n",
    "    hits.append((k, len(v)))\n",
    "print(sorted(hits, key= lambda item: item[1], reverse = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "369d7a0a-b660-4115-8521-d612771b8ba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Downstream gene', 'Rv3117', '8.6e-53', 1.0, 939775, 940021),\n",
       " ('Upstream gene', 'Rv0816c', '8.6e-53', 1.0, 939775, 940021)]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hit_dict['Mycobacterium tuberculosis variant bovis BCG str. Pasteur 1173P2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "5507d973-84d8-4e58-bd4a-0c2b400bbc23",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "full_sequence = str(genome_record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "38acbf8e-f1fa-4e34-a94d-81d88c52c54d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TCGGCGACGATGCGCCCCGGGTAACGGGGTGAGGAGGAGCCGGGCAATCAAATCGAGCTCGGCGACGATGCGCCCCGGGTAACGGGGTGAGGAGGAGCCAGGCAAT'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "util.reverse_complement(full_sequence[2401818:2401924])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "88e71cc1-71ab-4165-aa70-8a4805a895eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCGAGCAGACGCAGAATCGCACGCGCGAGGTCCGCGCCGTGCGATTCTGCGTCTGCTC'"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_sequence[2522172:2522230]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7399b2c1-3bbf-46b4-aa1b-57d653cf1284",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCCATTGACTACCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCC'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "util.reverse_complement(full_sequence[960173:960263])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fbc8541-4b19-4cd1-8610-2e76374460e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.reverse_complement(full_sequence[2314661:2314825]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d5b6e5-f38a-425e-b8b0-41d891e0399a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tr = util.Translator()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02a3d868-e545-4db8-9355-9268c20ccd94",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e1b53ec-86da-4581-9ad8-5bb6d54c09ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "tr.translate_sequence(full_sequence[2314660:2314825],-1,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6e9dcc7-7f0e-4eac-8cd4-cf865faf5711",
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_finder = orffn.ORF_Finder()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fe9e8eb5-33f5-4185-ac8d-fb8fa7741ae1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(123860, 123983, -1, 120, 0.05438814266846015)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf_finder.max_orf(123700, 124000, 100) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0475a8-4154-43c0-9b87-201ac3dae3ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_dict = {}\n",
    "for i, r in rbh_results.iterrows():\n",
    "    if r['query_ref'] == 'NC_000962.3@Rv0794c':\n",
    "        print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c9924947-be34-4541-8f08-1fdb2c6b07d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NZ_LT985188.1@MPLG2_RS04790'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_dict['NC_000962.3@Rv0794c']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "474c768c-486b-4a38-9664-ba2089493143",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000962.3@Rv0001\n",
      "NC_000962.3@Rv0002\n",
      "NC_000962.3@Rv0003\n",
      "NC_000962.3@Rv0004\n",
      "NC_000962.3@Rv0005\n",
      "NC_000962.3@Rv0006\n",
      "NC_000962.3@Rv0007\n",
      "NC_000962.3@Rv0008c\n",
      "NC_000962.3@Rv0009\n",
      "NC_000962.3@Rv0010c\n",
      "NC_000962.3@Rv0011c\n"
     ]
    }
   ],
   "source": [
    "i =0\n",
    "for k,v in temp_dict.items():\n",
    "    print(k)\n",
    "    i+=1\n",
    "    if i>10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2c0c3fea-dbf9-4c8d-a930-604b1c9d29c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['101 bp Mycobacterial Interspersed Repetitive Unit,Class I. See Supply et al. (1997) Molecular Microbiology 26, 991-1003']\n",
      "GTGACCCGCGCTGGCGACGATGCAGAGCGCAGAGAAGCGGTGGGGGCGCGCCCCCACAAGTGGGGGGTACCCCCACCCGCTTGCGGGGGAGAGTGGCGCGCG\n",
      " \n",
      "['5 x 9 bp GTGGACCCG repeats']\n",
      "GGTGGACCCGGTGGACCCGGTGGACCCGGTGGACCCGGTGGACCCG\n",
      " \n",
      "[\"(MTV030.15), len: 315 nt. Probable REP'-1 pseudogene fragment\"]\n",
      "AAAGTCGCACGTCCGGTTCGAAGGGCGGCCACGGGAAACGGACCCGCAGCAACGCGGGCACCGCACCCATGGTCGACCCAACTGCCACGCACCCGGTGACCGGTGCGAAGTCCACCATATCGACCAGTGGGCAACCGGCGGCTCAACCGATATCGACAAACTCACCTTCACCTGCACACCCAACCACAAGCTAGTCGGGAAAGGCTGGCAGACAAGGAAACGGTCCGACGGCCAAACGGAATGGATCCCGCCACCCCACCTCGACCGCGGTGCCCACACCAACGACTACCACCACCCCGAACGCCTCTTCGACCAC\n",
      " \n",
      "['REP-2, len: 1503 nt. REP251, member of REP13E12 family.']\n",
      "GTGCGGTACCTTCCCGTATCAACTCGTAGGATCTGGGTAAATCCCTTGTGTCACTTCAGTTTCACGGTTATCAGCGGGGCGCTCTTTGTCAGTGCCCGACGTTATGATTCGAACATGTTAGCGAATAGCCGGGAGGAGCTTGTCGAGGTCTTCGACGCGCTGGATGCCGACCTGGACCGCTTGGACGAGGTGTCCTTTGAGGTGCTGAGCACCCCGGAACGGCTGCGGTCTCTGGAACGTCTGGAATGCTTGGCGCGCCGGCTACCGGCGGCCCAGCACACGTTGATCAACCAACTCGACACCCAAGCCAGCGAGGAAGAACTGGGCGGCACGCTGTGCTGCGCGCTGGCCAACCGGCTGCGCATCACCAAGCCCGAAGCCGGCCGACGCAGCGCCGAAGCCAAGCCTTAGGGCAACGTCGGGCGCTCACTGGCGAACCGTTGGCCCCGCAGTTGACCGCCACCGCCACCGCCCAACGCCAGGGCCTGATCGGCGAGGCGCACGTCAAAGTGATTCGCGCCCTTTTTCGCCCACCTGCCCGCCGCGGTGGATGTGTCCACCCGCCAGGCCGCCGAAGCCGACCTGGCCGGCAAAGCCGCTCAATATCGTCCCGACGAGCTGGCCCGCTACGCCCAGCGGGTCATGGACTGGCTACACCCCGACGGCGACCTCACCGACACCGAACGCGCCCGCAAACGCGGCATCACCCTGAGCAACCAGCAATACGACGGCATGTCACGGCTAAGTGGCTACCTGACCCCCCAAGCGCGGGCCACCTTTGAAGCCGTGCTAGCCAAACTGGCCGCCCCCGGCGCGACCAACCCCGACGACCACACCCCGGTCATCGACACCACCCCCGATGCGGCCGCCATCGACCGCGACACCCGCAGCCAAGCCCAACGCAACCACGACGGGCTGCTGGCCGGGCTGCGCGCGCTGATCGCCTCCGGGAAACTGGGCCAACACAACGGTCTTCCCGTCTCGATCGTGGTCACCACCACCCTGACCGACCTGCAAACCGGCGCCGGCAAGGGCTTCACCGGCGGCGGCACCCTGCTACCCATGGCCGATGTGATCCGCATGACCAGCCACGCCCACCACTACTCCCCCGCAAGCGGGAGGTACCCCCAGGCGATCTTCGACCACGGCACACCCCTGGCGCTGTATCACACCAAACGCCTAGCCTCCCCGGCCCAGCGGATCATGCTGTTCGCCAACGACCGCGGCTGCACCAAACCCGGCTGTGACGCACCGGCCTACCACAGCCAAGCCCACCACGTCACCGCCTGGACCAGCACCGGACGCACCGACATCACCGAGCTGACCCTGGCCTGCGGCCCCGACAACCGACTCGCCGAAAAAGGCTGGACCACCCACAACAACACCCACGGCCACACCGAATGGCTACCACCACCCCACCTCGACCACGGCCAACCCCGCACCAACACCTTCCACCACCCCGAACGATTCCTCCACAACCAAGACGACGACGACAAACCCGATT\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class III (see Supply et al., 1997)']\n",
      "GTTGACCACACATGAGGTGGGGGCACCTCCCACTTGCGGGGGAGAGCGGCGCCT\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TATGACTCGCGCCGGCGACGATGCACAGCGAAGCGATGAGGAGGAGCGGCGCCT\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TATGACTCGCGCCAGCGACGATGCACAGCGAAGCGATGAGGAGGAGCGGCGCCT\n",
      " \n",
      "['39 bp direct repeat 1,AGGTGAAGGCGGCGGATTCGGCGGAATCTGACGCCGGAG']\n",
      "GAGGTGAAGGCGGCGGATTCGGCGGAATCTGACGCCGGAG\n",
      " \n",
      "['39 bp direct repeat 2,AGGTGAAGGCGGCGGATTCGGCGGAATCTGACGCCGGAG']\n",
      "CAGGTGAAGGCGGCGGATTCGGCGGAATCTGACGCCGGAG\n",
      " \n",
      "['101 bp Mycobacterial Interspersed Repetitive Unit,class III']\n",
      "GTGAACCGCGCCGGCGACGATGCAGTGGGGGCACCGCCCGCTTGCGGGGGACGAAGCGATGAGGTGGGGGCACCACCCGCTTGCGGGGGAGAGCGGCGCAGG\n",
      " \n",
      "['3 copies of a 10 bp near-perfect direct repeat,ATTACTACCTATTACTACGTATTACTATCT']\n",
      "ATTACTACCTATTACTACGTATTACTATCTG\n",
      " \n",
      "['77 bp Mycobacterial Interspersed Repetitive Unit,Class I (see Supply et al., 1997)']\n",
      "GATGACCTGCGCCGACGACGATGCAGAGCGTAGCGATGAGGTGGGGGCACCACCCGCTTGCGGGGGAGAGTGGCGCTG\n",
      " \n",
      "['77 bp Mycobacterial Interspersed Repetitive Unit,Class I (see Supply et al., 1997)']\n",
      "GATGACCTGCGCCGACGACGATGCAGAGCGTAGCGATGAGGTGGGGGCACCACCCGCTTGCGGGGGAGAGTGGCGCTG\n",
      " \n",
      "['77 bp Mycobacterial Interspersed Repetitive Unit,Class I (see Supply et al., 1997)']\n",
      "GATGACCTGCGCCGACGACGATGCAGAGCGTAGCGATGAGGTGGGGGCACCACCCGCTTGCGGGGGAGAGTGGCGCTG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TTTGACCGCGCCGGCGACGATGCAGAGCAGAGCGACCAGAAGGAGCGGCGCT\n",
      " \n",
      "['58 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "ATGAGCACCGCCGGCGACGATGCAGTGGGGGTACCACCCGCTTGCGGGGGACGAAGCGA\n",
      " \n",
      "['123 bp imperfect direct repeat 2, 92/103 bp identical to first copy at 709425..709548, AGCCCCGGCTCGACGCGGCATAGGGTGGCCACCGTGGCCGAAGCGTTCCATGCGACCG TGCCGTGGCGAGGATCCCGGCCGAACATGGCCCATTGAACGAGGACGTCATCGCACGA CGCCTGC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "AGCCCCGGCTCGACGCGGCATAGGGTGGCCACCGTGGCCGAAGCGTTCCATGCGACCGTGCCGTGGCGAGGATCCCGGCCGAACATGGCCCATTGAACGAGGACGTCATCGCACGACGCCTGCA\n",
      " \n",
      "['74 bp imperfect direct repeat 2, 64/73 bp identical to first copy at 706790..706863, CACAGCGGACACCACAAAGCCCGCCGCTGCCACCGGATCGTCGGAACGAAAATAGTCG TACCCGTGAGCCTCGC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CACAGCGGACACCACAAAGCCCGCCGCTGCCACCGGATCGTCGGAACGAAAATAGTCGTACCCGTGAGCCTCGCA\n",
      " \n",
      "['74 bp imperfect direct repeat 1, 64/73 bp identical to second copy at 703912..703985, CACATCGGACACGACGAAACCCGCCGCTGCCACCGGATCGTCGGAGCGGAAGTAGTCG TACCCGTCGGCCTCGC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CACATCGGACACGACGAAACCCGCCGCTGCCACCGGATCGTCGGAGCGGAAGTAGTCGTACCCGTCGGCCTCGCG\n",
      " \n",
      "['123 bp imperfect direct repeat 1, 92/103 bp identical to second copy at 701247..701369, AGCCTCGGCTGGCCGCGGCATAAGGTGGCCACCGTGGCCGAAGCGTTCGATGCGACCC AAGCCGTGGCGAGAATCCTGGCCGAACATGGCCCATTGAGCGAGGACGACATCGCACG ACGCCTGC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "AGCCTCGGCTGGCCGCGGCATAAGGTGGCCACCGTGGCCGAAGCGTTCGATGCGACCCAAGCCGTGGCGAGAATCCTGGCCGAACATGGCCCATTGAGCGAGGACGACATCGCACGACGCCTGCT\n",
      " \n",
      "['79 bp imperfect direct repeat 1, 73/78 bp identical to second copy at 711624..711702, TAGGGTTCGGCGTTGTGACGGCGCCGACGCGGTGGACCCTGGCCGACGGACGTGAGCT GCTGTTCTTTTCGCTGCCCGG. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CTAGGGTTCGGCGTTGTGACGGCGCCGACGCGGTGGACCCTGGCCGACGGACGTGAGCTGCTGTTCTTTTCGCTGCCCGG\n",
      " \n",
      "['79 bp imperfect direct repeat 2, 73/78 bp identical to first copy at 709585..709663, TAGGGTTCTGCGTTGTGACGGCGCCGACGCGGTGGACCCTGGCCGATGGCCGTGACCT GCTGTTCTTTTCGCTGCCCGG']\n",
      "GTAGGGTTCTGCGTTGTGACGGCGCCGACGCGGTGGACCCTGGCCGATGGCCGTGACCTGCTGTTCTTTTCGCTGCCCGG\n",
      " \n",
      "['52 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CATGACCGCGCCGGCGACGATGCAGAGCGAAGCGATGAGGAGAAGCGGCGCAG\n",
      " \n",
      "['49 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CGGCCTAGCCCGGCGACGATGCAGAGCGAAGCGATGAGGAGGAGCAGGGC\n",
      " \n",
      "['87 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "TATGACCGCGCTGACGACGATGCAGTGGGGGTACCCCCAGACGCGCAGCGGCGAGGGGGCCGCAAGCGATGAGGAGGAGTAGCGCTCG\n",
      " \n",
      "['54 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GATGACCGCGCTGACGACGATGCAGAGCGCAAGCGATGAGGAGGAGTAGCGCTCG\n",
      " \n",
      "['101 bp Mycobacterial Interspersed Repetitive Unit,Class I']\n",
      "CATGACCGGCGCCGGCGCCGATGCAGAGCGGGGCGATGAGGTGGGGGCACGCCCCCACAAGTGGGAGGTACCCCCATCCGCTGGCGGGGGAGAGCGGCGCTC\n",
      " \n",
      "['4 bp direct repeat: GAGG, at the right end of IS6110']\n",
      "GGAGG\n",
      " \n",
      "['28 bp inverted repeat at the left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "GTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at the right end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "['4 bp direct repeat: GAGG, at the left end of IS6110']\n",
      "AGAGG\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TTGACTACCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCCAT\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TTGACTACCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCCAT\n",
      " \n",
      "['55 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGACCCCCGCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCCAT\n",
      " \n",
      "['15 bp inverted repeat, ATTCGGTGTAAGTGG, at the left end of IS1554 element']\n",
      "TATTCGGTGTAAGTGG\n",
      " \n",
      "['15 bp inverted repeat, ATTCGGTGTAAGTGG, at the right end of IS1554 element']\n",
      "ATTCGGTGTAAGTGGT\n",
      " \n",
      "['16 bp inverted repeat, TTGAGTGTGTTTTAGT, at the left end of IS element IS1535']\n",
      "GTTGAGTGTGTTTTAGT\n",
      " \n",
      "['16 bp inverted repeat, TTGAGTGTGTTTTAGT, at the right end of IS element IS1535']\n",
      "TTGAGTGTGTTTTAGTT\n",
      " \n",
      "['18 bp inverted repeat at the left end of IS-LIKE element, CTAGGGCGTGTCTCCCAA. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "ACTAGGGCGTGTCTCCCAA\n",
      " \n",
      "['18 bp inverted repeat at the right end of a IS-LIKE element, CTAGGGCGTGTCTCCCAA. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "CTAGGGCGTGTCTCCCAAA\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TATGAGCCGCGCCGGCGACGATGCGGAACGAAGTGATGAGGAGGAGCGGCGC\n",
      " \n",
      "['REP-3, len: 1325 nt. REP22G8, member of REP13E12 family.']\n",
      "CGGGTCGCCTTCGTCGTCTGCCAGGAGCTTTTCCGGGTGATGGAACGTATTGACTCGAGGTTGGCCGTGGTCGAGATGTGGCGGCGGTAGCCACTCGGTGTCGCCGTGGGCGTTCTTGCGGGTCGTCCAGCCACGTTCGGCTAACGGATGATGGCCACCGCAGCCGAGTGTCAGGTCATTGACGTCGGTGTTGCGGCACTGGGCGTACGGCGTGACATGATGGACTTCACAGTAATAGCCGGGCACGTCGCAACCAGGTGCGCTGCAGCCACTGTCCTTGGCGTACAACATAATTCGCTGCGCCGGGGAGGCCAGGCGCTTGGTGTGGTAGAGCGCCAGGGCCTTGCCTCGATCGAATATCGCGAGGTAGTGGTTTGCGTGGCGGGCCAGCCGGATCACATCCGATATGGGCAAGATCGTACCCCCGCCGGTGAGGCCCGCGCCGGCCGCGGCCTCCAAGTCCTTCAGCGTGGTGGTCACGATGATGCTGGCCGGTAATCCGTTGTGCTGGCCCAGATTGCCACTTGTCAACAAACTACGTAATTCGGCGTTGAGCGCGTCGTGGTTCCGCTGTGGGCAGCTGCGGGTGTCTCGCCGCGCCTGCTCCTTCGAGGGCGCGCCGTTCACACACGGTGCCTTCTGCTCGGGGTTGCACATACCCGGGGCGGCCAGCTTGGCCCACACCGCCTCGATAGTGGCGCGCAGCTCGGGGGTCACATATCCGCTGAGCCGCGACATCCCATCGACATCTTGCTTTCCTAACGTCAAGCCGCGGCGGCGGGCGCGGTCCTCGTCGGTGTAGTCGCCATCGGGGTTGAGGCAGTCCATGATCCGCGCGGCCAATTTGGCCAGCTGGTCGGGACGGTACTGGGTGGCCTGCTTAGCCAAGTCCCGTTCGGCCTTCTCCAGGGTCTTGAGGTCTACCCAGGATGGTAGGCGGTGCACGAAAGCACGGATTACTTCAACATGGCCGTCACCAATTAACCCGTGGCGCTGTGCCTTTGCGGTGGCGGTGAGTAGCGGTGGCAGCGGCTCGCCGGTCAGCGCACGGCGCTGGCCAAGGTCGGCGGCCTCGGCCACTCGCCGCTTGGCCTCGCTGCGGGTGATGCGCAACCGGTCGGCCAGCGTCAATCCCAGCTTGCCGCCCAGCTCCTCCTCGGTGGATTGTTCGCCGATCTGATTGATCAACGTGTGTTCGACGCTGGGCAGCTGGCGTCGCGCGGTCTCGCAGTGCTCCAGCAGCGCCAGGCGCTCCGGGGTGGTCAATGCGTCAAAGGTCAGCCCCAGCACGCGGGACAGCGCGGTAGCCAATGACGCGAAGGCCTC\n",
      " \n",
      "['REP-4, len: 1348 nt. REP165, member of REP13E12 family.']\n",
      "GGGTGTCAATGTGGTTCGTCGTCGTCTTGTTCGCACAGGATTTTCGCGGGGTGGTGGTATCGATTTATTCGCGGTTGGCCGTGGTCGAGGTGTGGTGGTGGTAGCCATTCGGTGTCGCCGTGGGCGTTTTTGCGGGTCTTCCAGCCTTTTTCGACAAGGCGATTGTCGGGGCCGCAGGCCAGCGTGAGGTCGTTGATGTCGGTACGGTGGGTGGTTGTCCACGGCGTTACGTGGTGGACCTCACTGTGGTAGGCCGGGGCGTCGCAACCCGGCCTGGAGCAGCCACGATCCTTCGCGTACAACATGATTCGCTGCGCCGGGGAAGCTAACCGCTTGGTGTGATACAACGCCAACGGCTTAGCGCCGTCAAACAATGCCAGATAGTGGTTGGCGTGGCTCGCCATCCGGATAAGGTCCGACATCGGCACCCGCGAACCACCACCGGTTACCCCCTTGCCGGTGGCGGCTTCCAGCTCCTTTAGCGTGGTGCTCACCACGATCGTTACCGGCAGCCCCTTGTGTTGGCCCAGCTCACCGGAGGCCAACAGGCCCCGCAGCGCGGCCAAAAACGCATCATGATTGCGTTGCGCCTGGCTGCGGGTGTCGCGGCGCACCGCGTCCGCATCCGGTGTGTCATCCACGAGCGGGGTCTGGTCATCGGGGTTGCACGCCCCCGGTGCGGCCAGTTTGGCCAACACCGCCTCGATGGTGGCCCGCAACTCCGGGGTCAGCAGACCGCTGATACGTGACATCCCGTCAAATTCCTGCTTACCCATCGTGATGCCGCGCTTGCGGGCACGCTCCTGGTCGGAAAAGTTGCCGTCGGGGTGCAGCCAGTCCATCAGCTGCGTGGCCAGGCCATGCAGGTGATCGGGACGCCGACTGGTGGCCAGTTCGGCCAGCTGGGCCTCGGCGGCCTCGCGGATACCCAGATCCACCGCGGCGGACAACTCCTTGAAGAAGGCCTGGATCTCCTTAATGTGTTCTCGGCCGATCTTGCCCTCACGTTGAGCGGCCGCGGTCGCGGTCAACTGCGCTGGCAGCGGTTCACCGGTCAGGGCGCGGCGCTCACCGAGGTCTTCGGCTTCGGCGATGCGGCGGCTGGCCTCACCGGGAGTGATGTGTAGCCGGTTGGCCAACGCCGTGCGCAGCGTCCCGCCGAGCTCTTCCTCGCAGGCTTGCCCAGCGAGTTGGTTGATCAAGGCGTGCTCGGCGGCGCCCTGGCGGCGCCGTTCGACCTCGAGTCGCTGCAAACAGGCCAGCAATTCCGGGGTGGTCAACGCATCGCACTTGAGATCGAGCACCCGCGACAACGAGGCGTGGTAGGCATCCAACGCCGCGGAGATCTC\n",
      " \n",
      "['4 bp direct repeat, CTAG, generated by IS element on insertion. Proposed by Mariani et al. 1993. J. Gen. Microbiol., 139: 1767-1772. Note that as motif palindromic could be part of inverted repeat itself.']\n",
      "ACTAG\n",
      " \n",
      "['17 bp Inverted repeat at the left end of putative IS-LIKE-2 element : GGCGTGTCTCCCAAATT. Proposed by Mariani et al. 1993. J. Gen. Microbiol. 139: 1767-1772.']\n",
      "GGGCGTGTCTCCCAAATT\n",
      " \n",
      "['17 bp Inverted repeat at the right end of putative IS-LIKE-2 element :GGCGTGTCTCCCAATTT. Proposed by Mariani et al. 1993. J. Gen. Microbiol. 139: 1767-1772']\n",
      "GGCGTGTCTCCCAATTTC\n",
      " \n",
      "['4 bp direct repeat, CTAG generated by IS element on insertion. Proposed by Mariani et al. 1993. J. Gen. Microbiol. 139: 1767-1772. Note that as motif palindromic could be part of inverted repeat itself.']\n",
      "CCTAG\n",
      " \n",
      "['62 bp direct repeat copy 1, GGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGCGGCCCGTTGAGGAGCGGGGCA ATCT']\n",
      "CGGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGCGGCCCGTTGAGGAGCGGGGCAATCT\n",
      " \n",
      "['62 bp direct repeat copy 2, GGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGCGGCCCGTTGAGGAGCGGGGCA ATCT']\n",
      "TGGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGCGGCCCGTTGAGGAGCGGGGCAATCT\n",
      " \n",
      "['62 bp direct repeat partial copy 3 (43/62 bp),GGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGGGGCCCG']\n",
      "TGGCCTAGCCCCGGCGACGATGCCGGGTCGCGGGATGGGGCCCG\n",
      " \n",
      "['43 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "CGATGCGGTGGGGGAACCTCCCGCTTGCGGGGGAGAGCGGCACA\n",
      " \n",
      "['52 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "TGTGACGGCGATGAGGTGGGGGCACCTCCCGCTTGCAGGGGAGAACGGCGCCA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATTGACCCGCGCCAGCGACGATGCAGAGCGAAGCGATGAGGAGAAGCGGCGCCA\n",
      " \n",
      "['19 bp inverted repeat, GCAGACGCAAAAGCCCCCA, at the left end of IS1557']\n",
      "AGCAGACGCAAAAGCCCCCA\n",
      " \n",
      "['19 bp inverted repeat, GCAGACGCGAAAGCCCCCA, at the right end of IS1557. Single base difference at 3-end.']\n",
      "GCAGACGCGAAAGCCCCCAA\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CCTGACCGCGCCGCGACGATGCAGAGCGGAGCGATGAGGAGGAGTGGCGCTT\n",
      " \n",
      "[\"3 bp direct repeat, CGG, at 3' end of IS6110 target sequence\"]\n",
      "AGCC\n",
      " \n",
      "[\"3 bp direct repeat, CGG, at 5' end of IS6110 target sequence\"]\n",
      "AGCC\n",
      " \n",
      "['21 bp imperfect direct repeat 5,GGCGCACCGGTACCGGTACCC']\n",
      "GGCGCACCGGTACCGGTACCCG\n",
      " \n",
      "['21 bp imperfect direct repeat 4,GGCGGACCGGTACCGATACCG']\n",
      "GGCGGACCGGTACCGATACCGG\n",
      " \n",
      "['21 bp imperfect direct repeat 3,GGCGCACCGGTACCAATCCCC']\n",
      "GGCGCACCGGTACCAATCCCCG\n",
      " \n",
      "['21 bp imperfect direct repeat 2,GGCGCACCGGTACCGATACCG']\n",
      "GGCGCACCGGTACCGATACCGG\n",
      " \n",
      "['21 bp imperfect direct repeat 1,GGCGCACCGGTACCAATCCCT']\n",
      "GGCGCACCGGTACCAATCCCTG\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GTGATCCGCGCCGGCGACGATGCAGAGCGAAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['1260 bp imperfect direct repeat 2, first copy at 1637133..1638392']\n",
      "CCGCAATGCCACGAGCCGATAGTAGGGCGAATGTGCACTGTGCACAGCAGATTCGTGCTGGTTATGGGGATTTTGACCTGGAATCGGACAAGGTGGCGGCGCGACCCTTATCTGGCAGGAGTGTGACCTGCCGAAGGGGCAGGGGCGTGGCGGGAGGTCTGTGATGTCGTTGGTGATCGTGGCCCCGGAGACGGTGGCGGCCGCAGCCTTAGATGTGGCGCGCATCGGGTCATCGATCGGCGCGGCCAATGCGGCGGCGGCGGGGTCGACCACCAGCGTGCTGGCCGCGGGCGCCGATGAGGTGTCGGCGGCGATCGCGACGCTGTTTGGCAGCCATGCTCGGGAGTATCAGGCGATCAGCACGCAGGTGGCGGCGTTTCATGACCGATTTGCGCAGACGTTAAGCGCCGCGGTCGGCTCGTATGTCAGCGCGGAGGCGACCAACGCCGCACCGTTGGCGACGCTGGAGCACAACGTGCTCAATGCCCTCAATGCGCCCACCCAGGCGTTGCTGGGTCGCCCGTTGATCGGTGATGGGGCGGCTGGAGCACCCGGCACCGGGCAGGCCGGCGGGGCCGGCGGGATCTTGTGGGGCAACGGTGGGGCCGGCGGCTCGGGCGCGCCCGGCCAAGTCGGCGGGGCCGGCGGGGCCGCCGGGTTGTTCGGCACCGGCGGGGCCGGTGGGGCCGGTGGGGCCGGCGCCGCCGGTGGGGCCGGGGGTAGCGGCGGCTGGCTGCTGGGCAATGGTGGAGTCGGCGGGGCCGGCGGGCAGAGCCTGCTGGGCGGGGCAACCGGCGGGGCCGGCGGCAACGCCGGACTGTTCGGGGTCGGCGGAACCGGCGGGCCCGGCGGGCCCGGCGGGCCCGGCGGGGTCGGCGGTACCGGTGGTGCCGGTGGCCTGGGCGGGACCCTCTACGGGGCCGGCGGGCACGGAGGTGCCGGCGGGCCCGGCCCGATCGGTGGTGTGGGTGGACACGGCGGTGTCGGGGGTGCGGCCGGGCTGTTGGGCGTGGGCGGGCACGGTGGTGCCGGCGGACACGGCGCGGAGGGTGTGGCCGGCGCAGCCGGTGAGGACTTGTCCCCGCACGGTACGTCCGGTGGGGTCGGCGGCGACGCCGGCGATGGCGGCACCGGAGGGCGGGGCGGCTGGCTGGCCGGCGCCGGTGGGGCCGGCGGGGCCGGTGGGGTTGGCGGGACCGGCGGGGCCGGCGGGGCCGGATTTTCTCGTGCCTTGATTGTCGCTGGGGATAACGGCGGTGAC\n",
      " \n",
      "['1260 bp imperfect direct repeat 1, second copy at 1633531..1634790']\n",
      "CCGCAATGCCACGAGCCGATAGTAGGGCGAATGTGCACTGTGCACAGCAGATTCGTGCTGGTTATGGGGATTTTGACCTGGAATCGGACAAGGTGGCGGCGCGACCCTTATCTGGCAGGAGTGTGACCTGCCGAAGGGGCAGGGGCGTGGCGGGTGGACTGTGATGTCGTTGGTGATCGTAACCCCGGAGACGGTGGCCGCGGCGGCCTCGGATGTGGCGCGCATCGGGTCATCGATCGGTGTAGCCAACAGCGCGGCGGCGGGGTCAACCACCAGCGTGCTGGCCGCGGGCGCCGATGAGGTGTCGGCGGCGATCGCGACGCTGTTTGGCAGCCATGCTCGGGAGTATCAGGCGATCAGCACGCAGGTGGCAGCGTTTCATGACCGATTTGCGCAGACGTTAAGCGCCGCGGTCGGCTCGTATGTCAGCGCGGAGGCGACCAACGCCGCACCGTTGGCGACGCTGGAGCACAACGTGCTCAATGCCCTCAATGCGCCCACCCAGGCGTTGCTGGGTCGCCCGTTGATCGGTGATGGGGCGGCTGGAGCACCCGGCACCGGGCAGGCCGGCGGGGCCGGCGGGATCTTGTGGGGCAACGGTGGGGCCGGCGGCTCGGGCGCGCCCGGCCAAGTCGGCGGGGCCGGCGGGGCCGCCGGGTTGTTCGGCACCGGCGGGGCCGGTGGGGCCGGTGGGGCCGGCGCCGCCGGTGGGGCCGGGGGTAGCGGCGGCTGGCTGCTGGGCAATGGTGGAGTCGGCGGGGCCGGCGGGCAGAGCCTGCTGGGCGGGGCAACCGGCGGGGCCGGCGGCAACGCCGGACTGTTCGGGGTCGGCGGAACCGGCGGGCCCGGCGGGCCCGGCGGGCCCGGCGGGGTCGGCGGTACCGGTGGTGCCGGTGGCCTGGGCGGGACCCTCTACGGGGCCGGCGGGCACGGAGGTGCCGGCGGGCCCGGCCCGATCGGTGGTGTGGGTGGACACGGCGGTGTCGGGGGTGCGGCCGGGCTGTTGGGCGTGGGCGGGCACGGTGGTGCCGGCGGACACGGCGCGGAGGGTGTGGCCGGCGCAGCCGGTGAGGACTTGTCCCCGCACGGTACGTCCGGTGGGGTCGGCGGCGACGCCGGCGATGGCGGCACCGGAGGGCGGGGCGGCTGGCTGGCCGGCGCCGGTGGGGCCGGCGGGGCCGGTGGGGTTGGCGGGACCGGCGGGGCCGGCGGGGCCGGATTTTCTCGTGCCTTGATTGTCGCTGGGGATAACGGCGGTGAT\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GTGAACCGCGCCGACGACGATGCAGAGCGAAGCGATGAGGAGGAGCGGCGCTCG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GAGCCGCGCCGACGACGATGCAGAGCGAAGCGATGAGGAGGAGCGGCGCTCG\n",
      " \n",
      "['12 bp direct repeat 1, ccacggccaacc, flanking phage-like element, second site at 1788514..1788525. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CACGGCCAACCCC\n",
      " \n",
      "['89 bp direct repeat 2, first copy at 1780485..1780573, GGGTTGCGTTGTCGATTCGTTTGAGCCGCCGGTAGGTGCCGGCGGAGATGCCGAGGGC TG CGCCGATAGCAGTGTCTGTTTTCGTCGAA. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "GGGTTGCGTTGTCGATTCGTTTGAGCCGCCGGTAGGTGCCGGCGGAGATGCCGAGGGCTGCGCCGATAGCAGTGTCTGTTTTCGTCGAAT\n",
      " \n",
      "['89 bp direct repeat 1, second copy at 1779959..1780047, GGGTTGCGTTGTCGATTCGTTTGAGCCGCCGGTAGGTGCCGGCGGAGATGCCGAGGGC TG CGCCGATAGCAGTGTCTGTTTTCGTCGAA. Many repeats, both direct and inverted, in this region. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "GGGTTGCGTTGTCGATTCGTTTGAGCCGCCGGTAGGTGCCGGCGGAGATGCCGAGGGCTGCGCCGATAGCAGTGTCTGTTTTCGTCGAAC\n",
      " \n",
      "['REP-5, len: 1298 nt. REP336, member of REP13E12 family. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "TTGTCGAGGTCTTTGATGCGCTGGATGCCGAGCTGGACCGCTTGGACGAGGTGTCTTTTGAGGTGTTGACCACCCCGGAACGGCTGCGGTCTCTGGAACGTCTGGAATGCTTGGTGCGCCGGCTACCGGCGGTGGGTCACGCGTTGATCAACCAACTTGACGCCCAAGCCAGCGAGGAAGAACTGGGCGGCACGCTGTGCTGCGCGCTGGCCAACCGGTTACGCATCACCAAGCCCGACGCCGCCCGGCGCATCGCCGACGCCGCCGATCTCGGACCTCGTCGAGCACTCACCGGTGAACCGCTAGCCCCACAGTTGACCGCCACCGCCACCGCCCAACGCCAGGGCCTGATCGGCGAGGCGCACGTCAAAGTGATTCGCGCCCTTTTTCGCCCACCTGCCCGCCGCGGTGGATGTGTCCACCCGCCAGGCCGCCGAAGCCGACCTGGCCGGCAAAGCCGCTCAATATCGTCCCGACGAGCTGGCCCGCTACGCCCAGCGGGTCATGGACTGGCTACACCCCGACGGCGACCTCACCGACACCGAACGCGCCCGCAAACGCGGCATCACCCTGAGCAACCAGCAATACGACGGCATGTCACGGCTAAGTGGCTACCTGACCCCCCAAGCGCGGGCCACCTTTGAAGCCGTGCTAGCCAAACTGGCCGCCCCCGGCGCGACCAACCCCGACGACCACACCCCGGTCATCGACACCACCCCCGATGCGGCCGCCATCGACCGCGACACCCGCAGCCAAGCCCAACGCAACCACGACGGGCTGCTGGCCGGGCTGCGCGCGCTGATCGCCTCCGGGAAACTGGGCCAACACAACGGTCTTCCCGTCTCGATCGTGGTCACCACCACCCTGACCGACCTGCAAACCGGCGCCGGCAAGGGCTTCACCGGCGGCGGCACCCTGCTACCCATGGCCGATGTGATCCGCATGACCAGCCACGCCCACCACTACTCCCCCGCAAGCGGGAGGTACCCCCAGGCGATCTTCGACCACGGCACACCCCTGGCGCTGTATCACACCAAACGCCTAGCCTCCCCGGCCCAGCGGATCATGCTGTTCGCCAACGACCGCGGCTGCACCAAACCCGGCTGTGACGCACCGGCCTACCACAGCCAAGCCCACCACGTCACCGCCTGGACCAGCACCGGACGCACCGACATCACCGAGCTGACCCTGGCCTGCGGCCCCGACAACCGACTCGCCGAAAAAGGCTGGACCACCCACAACAACACCCACGGCCACACCGAATGGCTACCACCACCCCACCTCGACCACGGCCAACCG\n",
      " \n",
      "['12 bp direct repeat 2, ccacggccaacc, flanking phage-like element, first site at 1779266..1779277. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CCACGGCCAACCG\n",
      " \n",
      "['52 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CATCGCGCGCCGGCGACGATGCAGAGCGCAGCGATGCTGAGGAGCGGCGCCGA\n",
      " \n",
      "['56 bp direct repeat 1, AGTCGGGTGACGATGCGGGCCGGTGTGGTCCGAGGAGGAGCCCGACAATTTAAGCT']\n",
      "CAGTCGGGTGACGATGCGGGCCGGTGTGGTCCGAGGAGGAGCCCGACAATTTAAGCT\n",
      " \n",
      "['56 bp direct repeat 2, AGTCGGGTGACGATGCGGGCCGGTGTGGTCCGAGGAGGAGCCCGACAATTTAAGCT']\n",
      "TAGTCGGGTGACGATGCGGGCCGGTGTGGTCCGAGGAGGAGCCCGACAATTTAAGCT\n",
      " \n",
      "['REP-6, len: 1372 nt. REPI125, member of REP13E12 family.']\n",
      "ATCGTCTTCGTCGCTGTCGTGCAGGAGTTTTTCGGGGTGAAAGTAACTGTTGGTGCGGGGTTGTCCATGGTCGAGGTGGGCTGGGGGAAGCCATTCGGTGGTGCCGTCTTTGCGTTTGCGGGTGATCCAGCCCCCGGTGGTGGCCAGTTGGTGGTGGGGGCCGCAGCCCTGGGTGAGTTCGTTGATGTCGGTTTCTTGGCATTGGGCGAAGTCCGTCACATGATGCACCTCGGTGAGATAGCCCGGTACGTCGCAGTTGGGGAACGAGCAGCCGCGGTCCTTGGCGTAGAGGACGATTCGCTGTCCGGGTGAGGCTAGCCGCTTGGTGTGATAGAGGGCCAGCTCGCGGCCGTGGTCGAAGATACGTAGGTAGTGGTTGGCGTGGCTGGCCAGCCGGATCACGTCGCTCATGGGCAGCAGGGTGCCGCCGCCGGTCAGCGCGTGGCCGGCGCGTGATTGCAGTTCGGTCAGGCTGGTGGACACGATGATGGCCGCGGGTAGCCCGTTGTGTTGGCCCAGCTCCCCCGAGCACAGCAGGGCCCGCAGCGCGGCCAGCAGGCCGTCGTGGTGGCGTTGGCCGGCGCTGCGGGTGTCGGCCTCGATCGCGGCCTGTGACGGGGTGCCGGCCAGGCAGGGGGTGTCATCGGCGGGGTTGGCCATGCCGGGGGCGGCCAGCTTGGCCAACACGGCGTCGACGGTGGCGCGGGCTTCGGGGGTCAGGTAGCCGCTGATCGCCGACATGCCGTCGGGGCCTTGGTTGCCCAGGATGATGCTGCGGCGGCGGGCGCGGTCGGTGTCGTTGTAGTTGCCGTCGGGGTTCAAACAGTCGGCGAGTTTGGTGGCTAGTTTGTGTAGTTGGTCGGGGCGAAACCGGCCGCCTAGGGTGGCCAGCTCGGCTTCGGCTTTCTCCCGGGTGGGTAGGTCCACATGGTGGGGTAGCTGGTGCAGGAAGCAGCGGATGACCTGCACGTGGGCGGGGCCGAGGTGGCCGGCGCGTTGGGCGGCGGCGGTGGCGGTCAGCAACGGGGGCAGGGGTTGGCCGGTTAGCGTGCGGCGTGGGCCCAGGTCGGCGGCTTCATGGATGCGTCGGGATGCTTCGCCGCGGCTGATGTGTAGCCGTTCGGCCAGGGCGAAGGGTAGTTTGCCGCCCAGTTCGGTTTGGTCGGTTTGGTCGGCGAGTTTGTTGATGAAGGGGTGTTCGGCGGCGGGTAGGCGCCGACGGATCTTTTCGCAGCGCTGCAGCATTGCCAGGCATTCCGGGATGGTCAGGTCGTCAGGGGAGACCTTCAGGACCCGGTTAAGGGCGGTGTCGAGGTTGTCGAACGCGGCGACGGCCTCCTCCCGGCTACTCGAATACATGTTCGAATACTATC\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GGTTTCGTGGACCGGTGACGATGCAGAGCGCAGCGATGAGAAGGAGCGGCGCTT\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GATGAGCTGCACCGGCGACGATGCAGAGCGAAGCGATGCTGAGGAGCGGTGCGA\n",
      " \n",
      "['78 bp imperfect direct repeat 6, CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACC ACACCCGCCAACATCACCGT']\n",
      "CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACCACACCCGCCAACATCACCGTA\n",
      " \n",
      "['78 bp imperfect direct repeat 5, CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACC ACACCAGCCAACATCACCGT']\n",
      "CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACCACACCAGCCAACATCACCGTC\n",
      " \n",
      "['78 bp imperfect direct repeat 4, CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACC ACACCAGCCAACATCACCGT']\n",
      "CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACCACACCAGCCAACATCACCGTC\n",
      " \n",
      "['78 bp imperfect direct repeat 3, GGGTGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACC ACACCAGCCAACATCACCGT']\n",
      "GGGTGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACCACACCAGCCAACATCACCGTC\n",
      " \n",
      "['78 bp imperfect direct repeat 2, CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACC ACACCAGCCAACATCACCGT']\n",
      "CGGCGCCTTCAGCCTGCCCGGGTTGACGTTGCCGTCGTTGAACATCCCGGCCGCCACCACACCAGCCAACATCACCGTG\n",
      " \n",
      "['78 bp imperfect direct repeat 1, TCCCGCCTTCAGTCTGCCGGCAATAACGCTGCCGTCGCTGAACATCCCGGCCGCCACC ACACCGGCCAACATCACCGT']\n",
      "TCCCGCCTTCAGTCTGCCGGCAATAACGCTGCCGTCGCTGAACATCCCGGCCGCCACCACACCGGCCAACATCACCGTC\n",
      " \n",
      "['28 bp inverted repeat at the left end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC.']\n",
      "GTGAACCGCCCCGGTGAGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at the right end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGCATGTCCGGAGACTCC\n",
      " \n",
      "['28 bp inverted repeat at the left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "CTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at the right end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "[\"14 bp Inverted repeat at the left end of ISB9',ATCACCCCGCAAAG\"]\n",
      "CATCACCCCGCAAAG\n",
      " \n",
      "['14 bp Inverted repeat at the right end of ISB9,ATCACCCCGGCAAG']\n",
      "ATCACCCCGGCAAGC\n",
      " \n",
      "['58 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "AATGAGCCGCGCCGGCGACGATGCAGTGGGGGTACCGCCCGCTTGCGGGGGACGAAGCG\n",
      " \n",
      "['58 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "AATGAGCCGCGCCGGCGACGATGCAGTGGGGGTACCGCCCGCTTGCGGGGGACGAAGCG\n",
      " \n",
      "['69 bp imperfect direct repeat 3, TTAATCCGTTTGGGTTGAATGTTCCGTTGAGCGGGGGCACGAGCCCGGTTACGATCCC CGGCTTCACCAT']\n",
      "TTAATCCGTTTGGGTTGAATGTTCCGTTGAGCGGGGGCACGAGCCCGGTTACGATCCCCGGCTTCACCATT\n",
      " \n",
      "['69 bp imperfect direct repeat 2, TTAATCCGTTTGGGTTGAATGTTCCGTTGAGCGGGGGCACGAGCCCGGTTACGATCCC TGGTTTCGCGA']\n",
      "TTAATCCGTTTGGGTTGAATGTTCCGTTGAGCGGGGGCACGAGCCCGGTTACGATCCCTGGTTTCGCGAT\n",
      " \n",
      "['69 bp imperfect direct repeat 1, TTAATCCGTTCGGTTTGAATATTCCGCTGAGCGGTGCTACCAACGCTGTCACGATCCC TGGTTTCGCGA']\n",
      "TTAATCCGTTCGGTTTGAATATTCCGCTGAGCGGTGCTACCAACGCTGTCACGATCCCTGGTTTCGCGAT\n",
      " \n",
      "['69 bp imperfect direct repeat 5, TCGGTCCGATTGTGGTGCCTGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGC GCTGGGTGGTG']\n",
      "TCGGTCCGATTGTGGTGCCTGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGCGCTGGGTGGTGT\n",
      " \n",
      "['69 bp imperfect direct repeat 4, TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGC GCTGGGTGGTG']\n",
      "TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGCGCTGGGTGGTGT\n",
      " \n",
      "['69 bp imperfect direct repeat 3, TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGC GCTGGGTGGTG']\n",
      "TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGCGCTGGGTGGTGT\n",
      " \n",
      "['69 bp imperfect direct repeat 2, TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGC GCTGGGTGGTG']\n",
      "TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGCGCTGGGTGGTGT\n",
      " \n",
      "['69 bp imperfect direct repeat 1, TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGC GCTGGGTGGTG']\n",
      "TCGGTCCGATTGTGGTGCCGGATATTACTATTCCTGGTATTCCGTTGAGCCTGAACGCGCTGGGTGGTGT\n",
      " \n",
      "['REP-7, len: 1362 nt. REP09F9, member of the REP13E12 family. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CATGAGGTCGGACACGCGCGAGGAGATCTCCGCGGCGTTGGATGCCTACCACGCCTCGTTGTCGCGGGTGCTCGATCTCAAGTGCGATGCGTTGACCACCCCGGAATTGCTGGCCTGTTTGCAGCGACTCGAGGTCGAACGGCGCCGCCAGGGCGCCGCCGAGCACGCCTTGATCAACCAACTCGCTGGGCAAGCCTGCGAGGAAGAGCTCGGCGGGACGCTGCGCACGGCGTTGGCCAACCGGCTACACATCACTCCCGGTGAGGCCAGCCGCCGCATCGCCGAAGCCGAAGACCTCGGTGAGCGCCGCGCCCTGACCGGTGAACCGCTGCCAGCGCAGTTGACCGCGACCGCGGCCGCTCAACGTGAGGGCAAGATCGGCCGAGAACACATTAAGGAGATCCAGGCCTTCTTCAAGGAGTTGTCCGCCGCGGTGGATCTGGGTATCCGCGAGGCCGCCGAGGCCCAGCTGGCCGAACTGGCCACCAGTCGGCGTCCCGATCACCTGCATGGCCTGGCCACGCAGCTGATGGACTGGCTGCACCCCGACGGCAACTTTTCCGACCAGGAGCGTGCCCGCAAGCGCGGCATCACGATGGGTAAGCAGGAATTTGACGGGATGTCACGTATCAGCGGTCTGCTGACCCCGGAGTTGCGGGCCACCATCGAGGCGGTGTTGGCCAAACTGGCCGCACCGGGGGCGTGCAACCCCGATGACCAGACCCCGGTCGTGGATGACACACCGGATGCGGACGCGGTGCGCCGCGACACCCGCAGCCAAGCCCAACGACACCATGACGGTTTACTGGCCGGGCTGCGCGGGTTGTTGGCCTCCGGTGAGCTAGGGCAGCATCGGGGGTTGCCGGTGACCGTCGTGGTGAGCACCACGCTTAAAGAGCTGGAAGCCGCCACCGGCAAGGGGGTAACCGGTGGTGGTTCGCGGGTGCCGATGTCGGACCTTATCCGGATGGCGAGCAACGCGCACCACTATCTGGCATTGTTTGACGGCGCTAAGCCGTTGGCGTTGTATCACACCAAGCGGTTAGCTTCCCCGGCGCAGCGAATCATGTTGTACGCCAAGGATCGTGGCTGCTCCAGGCCGGGTTGCGACGCCCCGGCCTACCACAGTGAGGTCCACCACGTAACGCCGTGGACAACCACCCACCGTACCGACATCAACGACCTCACGCTGGCCTGCGGCCCCGACAATCGCCTTGTCGAAAAAGGCTGGAAAACCCGCAAGAACGCCAAAGGCGACACTGAATGGCTACCGCCGGCCCACTTGGACCATGGCCAACCACGCATCAATCGATACCACCACCCCGAGAAAATCCTGTGCGAACCCGACGACGACGAACCACAT\n",
      " \n",
      "['79 bp Mycobacterial Interspersed Repetitive Unit,Class I']\n",
      "ATGCCGCGATGATCCGCGACGATGCAGAGCGCAGCGATGCGGTGGGGGTACCGCCCGCTTGCGGGGGAGAGCGGCGCCGA\n",
      " \n",
      "['28bp inverted repeat at the left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "ATGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28bp inverted repeat at the right end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "['56 bp direct repeat 1, GCCCGCCGACGATGCGGGCCGCGCAGCGGGCCGCTGAGGAGGCGGGCATCAAGCAA']\n",
      "GGCCCGCCGACGATGCGGGCCGCGCAGCGGGCCGCTGAGGAGGCGGGCATCAAGCAA\n",
      " \n",
      "['56 bp direct repeat 2, GCCCGCCGACGATGCGGGCCGCGCAGCGGGCCGCTGAGGAGGCGGGCATCAAGCAA']\n",
      "CGCCCGCCGACGATGCGGGCCGCGCAGCGGGCCGCTGAGGAGGCGGGCATCAAGCAA\n",
      " \n",
      "['28 bp Inverted repeat at the left end of IS6110; GAGTCTCCGGACTCACCGGGGCGGTTCA']\n",
      "GAGTCTCCGGACTCACCGGGGCGGTTCAC\n",
      " \n",
      "['28 bp Inverted repeat at the right end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGCATGTCCGGAGACTCC\n",
      " \n",
      "['58 bp Mycobacterial Interspersed Repetitive Unit,Class II I. Overlaps Rv2195 suggesting alternative GTG start at 2458 468 may be used']\n",
      "CATGAGCCGCGCCGACGACGATGCAGTGGGGGTACCACCCACTTGCGGGGGACGAAGCG\n",
      " \n",
      "[\"18 bp inverted repeat between 3' end of MTCY427.04c and 5' end of MTCY427.03c\"]\n",
      "GTTGCGGGGCGGCCCGCAA\n",
      " \n",
      "[\"58 bp inverted repeat near 3'end of MTCY427.28\"]\n",
      "CGCGAGCAGACGCAGAATCGCACGCGCGAGGTCCGCGCCGTGCGATTCTGCGTCTGCTC\n",
      " \n",
      "[\"53 bp inverted repeat between 3' ends of MTCY427.29 and MT CY427.31c\"]\n",
      "TGAGCAGACGCAGAGTCGCACTGCGGCCGGCGCAGTGCGACTCTGCGTCTGCTC\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGAGTCGCGCCGGTGACGATGCAGAGCGAAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGAGTCGCGCCGGTGACGATGCAGAGCGAAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGAGTCGCGCCGGTGACGATGCAGAGCGAAGCGATGAGCAGGAGCGGCGCAGA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGAGTCGCGCCGGTGACGATGCAGAGCGAAGCGATGAGCAGGAGCGGCGCAGA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGAGTCGCGCCGGTGACGATGCAGAGCGAAGCGATGAGCAGGAGCGGCGCAGA\n",
      " \n",
      "['50 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGCGCGCCGGTGACGATGCAGAGCGAAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['3 bp direct repeat, ccg, flanking IS6110']\n",
      "ACCG\n",
      " \n",
      "['28 bp inverted repeat at the left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "GTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at the right end of IS6110,GAGTCTCCGGACTCACCGGGGCGGTTCA']\n",
      "AGAGTCTCCGGACTCACCGGGGCGGTTCA\n",
      " \n",
      "['3 bp direct repeat, ccg, flanking IS6110']\n",
      "ACCG\n",
      " \n",
      "['28 bp Inverted repeat, TGAACCGCCCCGGCATGTCCGGAGACTC,at the left end of IS6110']\n",
      "CTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp Inverted repeat, TGAACCGCCCCGGTGAGTCCGGAGACTC,at the right end of IS6110']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "['52 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CGTGACCGCGCCGGCGACGATGCAGAGCGCAGCGATGCTGAGGAGCGGCGCCA\n",
      " \n",
      "['78 bp Mycobacterial Interspersed Repetitive Unit,Class I']\n",
      "ATCGAAGCACCGCCGGCGACGATGCAGAGCGCAGCGATGCGGTGGGGGCACCTCCCGCTTGCGGAGGAGAGCGGCACCA\n",
      " \n",
      "['77 bp Mycobacterial Interspersed Repetitive Unit,Class I']\n",
      "ATGCCCCGCGCCGGCGACGATGCAGAGCGAAGCGATGAGGTGGGGGCACCTCCCGCTTGCGGGGGAGAGCGGCGCCGG\n",
      " \n",
      "['13 bp inverted repeat, GCAGTCG(C)AAAAG, at the left end of IS1558']\n",
      "AGCAGTCGCAAAAG\n",
      " \n",
      "['13 bp inverted repeat, GCAGTCG(T)AAAAG, at the right end of IS1558']\n",
      "GCAGTCGTAAAAGA\n",
      " \n",
      "['300 bp direct repeat copy 1']\n",
      "ACGACGGCTGATGGATCAACACCCGCGCATTCGGCAGCGCCATGCGCTTGCCCGGTGTTCCGGCGGCCAGCAGCACCGCAGCCGCCGAGGCGGCCTGGCCCAGACACACCGTCTGGATATCGGCCCGCACGTATTGCATGGTGTCGTAGATCGCCATCAGCGAGGTGAACCCACCGCCCGGCGAGTTGATGTACATGGTGATATCGCGGTCGGGATCCAACGACTCCAACACCAGCAACTGTGCCATGATGTCGTTCGCCGACGCGTCGTCGACCTGGACGCCGAGGAAGATGATGCGTTC\n",
      " \n",
      "['300 bp direct repeat copy 2']\n",
      "CCAACGGCTGGTGCATCAGGATGCGAGCATGCGGCAGCGCGTAGCGCTTGCCCTTGGTACCTGCCGCCAGCAGGAACTCGCCCATCGAGGCGGCCATGCCCATCGCGTAGGTGGCGATGTCACAGGGCGCCAGCACCATGGTGTCGTAGATCGCCATGCCGGCGCTGATCGATCCACCCGGCGAATTGATGTAGAGGCTGATGTCCTTGCTGGCGTCTTCGGCGGCCAGCAGCAGAATCTGAGCGCATAACCGGTTGGCGATCTCGTCGTTCACCTCCGAGCCCAGGAAGATGATGCGCTC\n",
      " \n",
      "['29 bp Inverted repeat at the left end of IS6110,GTGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "AGTGAACCGCCCCGGTGAGTCCGGAGACTC\n",
      " \n",
      "['29 bp Inverted repeat at the right end of IS6110,GTGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "GTGAACCGCCCCGGCATGTCCGGAGACTCC\n",
      " \n",
      "['248 bp direct repeat 2']\n",
      "GCAGGTGGGTCGCGCCCTCGATATCGGCCAGCAGCAACGTCACCGTGCCCGTCGGTACAAGCTCGCTCACACCATCTGCGCTCCAGTCCACAGGTACCACGTCGACGCCGGGGTGAATCTTGCTCATGCTAGCCAGCATCGAGCCAGCGCGTAGCGCATTACATCGGCACCTGCGCCTAGATTGCTCGAAATCTCTTGGCCGCCGGTCCATGTGTTCTACGCGCTTTAGTCGATGCATTCGGCGACCGG\n",
      " \n",
      "['258 bp direct repeat 2. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "CCAGGTGGGTCGAACCCACTCTGTCGGCCAGCAGCAATGCCACCGAACCCGCCGCCACCGGATTGCCCACGTCGTCTTTGCTGACTTTGCTGCAGTCCAGAGGTGCCACGTCGGCACCCGGGTCAATCTCGCGCATGCCAGCCAGCATCGAGCCAGCGGGCACCGCAATACATCAGCACGTGTGACTAGATTGCTCCGAATTCTGTCGAACGCGGGTCCGCGTGATCTGCGCGTTTCGGTCGATGCTTTCGGCAGCCCG\n",
      " \n",
      "[\"3 bp direct repeat: TCG at 5'-end of IS6110\"]\n",
      "TTCG\n",
      " \n",
      "['28 bp inverted repeat: TGAACCGCCCCGGCATGTCCGGAGACTC at the left end of IS6110']\n",
      "GTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat, TGAACCGCCCCGGTGAGTCCGGAGACTC,at the right end of IS6110.']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "[\"3 bp direct repeat: TCG at 3'-end of IS6110\"]\n",
      "ATCG\n",
      " \n",
      "['15 bp Inverted repeat at the left end of IS1081:TCGCGTGATCCTTCG, right end copy is missing']\n",
      "ATCGCGTGATCCTTCG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TGTGACCTCCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGC\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CTTGACCTCCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGC\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CTTGACCTCCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGC\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GTGACCCGCGCCGGCGACGATGCAGAGCAAAGCGATGGGGAGGAGCGGCGCCAG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GTGACCCGCGCCGGCGACGATGCAGAGCAAAGCGATGGGGAGGAGCGGCGCCAG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "TTGACCCGCGCCGGCGACGATGCAGAGCAAAGCGATGGGGAGGAGCGGCGCCAG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class I']\n",
      "TGTGACCCCGGCGACGATGCAAAGGAGGGCGCGATGAGGTGGGGGTACCACCCGCTTACGGGGGAGAGCGGCGCTCG\n",
      " \n",
      "['51 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "CATGAGCGCGCGCCGCGACGATGCAGTGGGGGTACCACCCGCTTGCGGGGGACGAAGCG\n",
      " \n",
      "['4392 bp direct repeat region. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTAAAACCGTGTTGCACTGCAACCCGGAATTCTTGCACGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCATAGAGGGTCGCCGGCTCTGGATCACGCTCCCCTAGTCGTGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTTTTGCCTCATGCTTGGGCGACAGCTTTTGACCAAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTCGCAAGCGCCGTGCTTCCAGTGATCGCCTTCTAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACAACACCTCAGTAGCACGTCATACGCCGACCAATCATCAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTTTCTGACCACTTGTGCGGGATTAGCGGGCTTAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACACCAATGCGTCGTCATTTCCGGCTTCAATTTCAGCCTGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCTGAGGAGAGCGAGTACTCGGGGCTGCCGTCTGCGCTGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACGCGTGAAACCGCCCCCAGCCTCGCCGGGGCCGCCTAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACACTCGGAATCCCATGTGCTGACAGCGGATTCGCATGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCGGGCAGCGTTCGACACCCGCTCTAGTTGACTTCCGGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCAGGTGAGCAACGGCGGCGGCAACCTGGCGGCCACGGGTCGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACATGGGATATCTGCTGCCCGCCCGGGGAGATGCTGTCCGAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTCGTCGACCATCATTGCCATTCCCTCTCCCCACGTGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTGCGCCAACCCTTTCGGTGTGATGCGGATGGTCGGCTCGGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCTTGAATAACGCGCAGTGAATTTCGCGGATCAGACCCAAAACCCCGAGAGGGGACGGAAACATTCGCACGAGTTCCCGTCAGCGTCGTAAATCGCCAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCCGGCAACAATCGCGCCGGCCCGCGCGGATGACTCCGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCGCATGGACCCGGGCGAGCTGCAGATGGTCCGGGAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTGGATTGCGCTAACTGGCTTGGCGCTGATCCTGGTGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTGGAGCGTGTCACCGCAGACGGCACGATTGAGACAAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCCTCAGCTCAGCATCGCTGATGCGGTCCAGCTCGTCCGTGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCCAACCTCACCGCCTGCTGGGTGAGACGTGCTCGCCGCGAGTCGTCAGACCCAAAACCCTGAACCGCCCCGGCATGTCCGGAGACTCCAGTTCTTGGAAAGGATGGGGTCATGTCAGGTGGTTCATCGAGGAGGTACCCGCCGGAGCTGCGTGAGCGGGCGGTGCGGATGGTCGCAGAGATCCGCGGTCAGCACGATTCGGAGTGGGCAGCGATCAGTGAGGTCGCCCGTCTACTTGGTGTTGGCTGCGCGGAGACGGTGCGTAAGTGGGTGCGCCAGGCGCAGGTCGATGCCGGCGCACGGCCCGGGACCACGACCGAAGAATCCGCTGAGCTGAAGCGCTTGCGGCGGGACAACGCCGAATTGCGAAGGGCGAACGCGATTTTAAAGACCGCGTCGGCTTTCTTCGCGGCCGAGCTCGACCGGCCAGCACGCTAATTACCCGGTTCATCGCCGATCATCAGGGCCACCGCGAGGGCCCCGATGGTTTGCGGTGGGGTGTCGAGTCGATCTGCACACAGCTGACCGAGCTGGGTGTGCCGATCGCCCCATCGACCTACTACGACCACATCAACCGGGAGCCCAGCCGCCGCGAGCTGCGCGATGGCGAACTCAAGGAGCACATCAGCCGCGTCCACGCCGCCAACTACGGTGTTTACGGTGCCCGCAAAGTGTGGCTAACCCTGAACCGTGAGGGCATCGAGGTGGCCAGATGCACCGTCGAACGGCTGATGACCAAACTCGGCCTGTCCGGGACCACCCGCGGCAAAGCCCGCAGGACCACGATCGCTGATCCGGCCACAGCCCGTCCCGCCGATCTCGTCCAGCGCCGCTTCGGACCACCAGCACCTAACCGGCTGTGGGTAGCAGACCTCACCTATGTGTCGACCTGGGCAGGGTTCGCCTACGTGGCCTTTGTCACCGACGCCTACGCTCGCAGGATCCTGGGCTGGCGGGTCGCTTCCACGATGGCCACCTCCATGGTCCTCGACGCGATCGAGCAAGCCATCTGGACCCGCCAACAAGAAGGCGTACTCGACCTGAAAGACGTTATCCACCATACGGATAGGGGATCTCAGTACACATCGATCCGGTTCAGCGAGCGGCTCGCCGAGGCAGGCATCCAACCGTCGGTCGGAGCGGTCGGAAGCTCCTATGACAATGCACTAGCCGAGACGATCAACGGCCTATACAAGACCGAGCTGATCAAACCCGGCAAGCCCTGGCGGTCCATCGAGGATGTCGAGTTGGCCACCGCGCGCTGGGTCGACTGGTTCAACCATCGCCGCCTCTACCAGTACTGCGGCGACGTCCCGCCGGTCGAACTCGAGGCTGCCTACTACGCTCAACGCCAGAGACCAGCCGCCGGCTGAGGTCTCAGATCAGAGAGTCTCCGGACTCACCGGGGCGGTTCACCCCGAGAGGGGACGGAAACTCGGGGAGCCGATCAGCGACCACCGCACCCTGTCAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCTTCAGCACCACCATCATCCGGCGCCTCAGCTCAGCATGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCCTTCGACGCCGGATTCGTGATCTCTTCCCGCGGATAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTGCCCCGGCGTTTAGCGATCACAACACCAACTAATGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCAGCGAAATACAGGCTCCACGACACGACCACAACGCGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTCTTGACGATGCGGTTGCCCCGCGCCCTTTTCCAGCCGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACAGGTTCGCGTCAGACAGGTTCGCGTCGATCAAGTCCGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTTTATCACTCCCGACCAAATAGGTATCGGCGTGTTCAAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACATTTTGAGCGCGAACTCGTCCACAGTCCCCCTTTCAGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACGCCCCGTGGATGGCGGATGCGTTGTGCGCGCAAGTGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCCGACGATGGCCAGTAAATCGGCGTGGGTAACCGATCCGGGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACTAGTACGCCATCTGTGCCTCATACAGGTCCAGTGCCCTGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCTGACGGCACGGAGCTTTCCGGCTTCTATCAGGTAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACCCTCATGGTGGGACATGGACGAGCGCGACTATCGGGGTCGTCGGACCCAAAACCCCGAGAGGGGACGGAAACTGGACGCAGAATCGCACCGGGTGCGGGAGGTGCAGCAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACGCATATCGCCCGCCACACCACAGCCACGCTACTGCTCCATGTCCTCAGACCCAAAACCCCGAGAGGGGACGGAAACACACCGCCGATGACAGCTATGTCCGAGTGACATCCTCCCAGTCGTCAGACCCAAAACCACGAGAGGGGACGGAAACTTGAACCGCCCTTCGCGCGGTGTTTCGGCCGTGCCCGAGTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 35 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCACGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 35 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCCTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACG\n",
      " \n",
      "['36 bp direct repeat, 35 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCGGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACG\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['20 bp partial direct repeat, CCCCGAGAGGGGACGGAAAC,of sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['16 bp partial direct repeat, GTCGTCAGACCCAAAA, of sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAAC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 32 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CGGATCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACG\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACA\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACC\n",
      " \n",
      "['36 bp direct repeat, 36 out of 36 bp identical to sequence GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAAC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "GTCGTCAGACCCAAAACCCCGAGAGGGGACGGAAACT\n",
      " \n",
      "['54 bp direct repeat 4,GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTAGGCTTGGC']\n",
      "GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTAGGCTTGGCG\n",
      " \n",
      "['54 bp direct repeat 3,GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTGGCCCTGAT']\n",
      "GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTGGCCCTGATG\n",
      " \n",
      "['54 bp direct repeat 2,GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTGGCCCTGAT']\n",
      "GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTGGCCCTGATG\n",
      " \n",
      "['54 bp direct repeat 1,GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTGGCCCTGAT']\n",
      "GGTGGCGACCCGCGGCGCCCGGTCCCCGCGCTTGCGATCGCCACTGGCCCTGATG\n",
      " \n",
      "['62 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "ATGACCCGCGCCGGCGACGATACAGTGGGGGTACCGCCCGCTTGCGGGGGAGAGCGGCGCTTA\n",
      " \n",
      "['51 bp direct repeat 1,GTCCGATGGTGGCGACCCGCTGCGCCCGGCTTCGCCGCGCTTGCGATCGCC']\n",
      "GTCCGATGGTGGCGACCCGCTGCGCCCGGCTTCGCCGCGCTTGCGATCGCCA\n",
      " \n",
      "['51 bp direct repeat 2,GTCCGATGGTGGCGACCCGCTGCGCCCGGCTTCGCCGCGCTTGCGATCGCC']\n",
      "GTCCGATGGTGGCGACCCGCTGCGCCCGGCTTCGCCGCGCTTGCGATCGCCT\n",
      " \n",
      "['(41 bp) part of 51 bp direct repeat 3,GGCGACCCGCTGCGCCCGGCTTCGCCGCGCTTGCGATCGCC']\n",
      "GGCGACCCGCTGCGCCCGGCTTCGCCGCGCTTGCGATCGCCT\n",
      " \n",
      "['(43 bp) part of 51 bp direct repeat,GTGTCGACCCGCTGCGCCCGGCTTCGCCGTGCTTGCGATCGCC. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "CGTGTCGACCCGCTGCGCCCGGCTTCGCCGTGCTTGCGATCGCC\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGGTTGGCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "ATGGTTGGCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CTGCTTCGCGCCGGCGACGATGCAGAGCGCAGCGATGAGGAGGAGCGGCGCAGA\n",
      " \n",
      "['55 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "ATGACCGCCGGCGACGATGCGGTGGGGGCCGTGCCCGCTTGCGGGGGACGCAGCGA\n",
      " \n",
      "['101 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "ATGAGCACCGCCGGCGACGATGCAGTGGGGGTACCGCCCGCTTGCGGGGGACGAAGCGATGAGGTGGGGGTACCGCCCGCTTGCGAGGGAGAGCGGCGCACT\n",
      " \n",
      "['98 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "CTGACCGCCCCGGCGACGATGCAGTGGGGGTACCGCCTGCTTGCGGGGGACGAAGCGATGAGGTGGGGGTACCGCCCGCTTGCGGAGGAGCGGCGCAGA\n",
      " \n",
      "['6 bp inverted repeat at the left end of IS1538,TGAGTG']\n",
      "GTGAGTG\n",
      " \n",
      "['6 bp inverted repeat at the right end of IS1538,TGAGTG.']\n",
      "TGAGTGG\n",
      " \n",
      "['15 bp Inverted repeat at left end of IS1081:TCGCGTGATCCTTCG']\n",
      "CTCGCGTGATCCTTCG\n",
      " \n",
      "['15 bp Inverted repeat at the right end of IS1081:TCGCGTGATCCTTCG']\n",
      "TCGCGTGATCCTTCGA\n",
      " \n",
      "['15 bp inverted repeat at left end of IS1081: TCGCGTGATCCTTCG. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "ATCGCGTGATCCTTCG\n",
      " \n",
      "['15 bp inverted repeat at right end of IS1081: TCGCGTGATCCTTCG. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "TCGCGTGATCCTTCGG\n",
      " \n",
      "[\"3 bp direct repeat, cga, at 5'-end of IS6110. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).\"]\n",
      "TCGA\n",
      " \n",
      "['28 bp inverted repeat at left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "ATGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at right end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "[\"3 bp direct repeat, cga, at 3'-end of IS6110. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).\"]\n",
      "ACGA\n",
      " \n",
      "[\"3 bp direct repeat, att, at 5'-end of IS6110. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).\"]\n",
      "GATT\n",
      " \n",
      "['28 bp inverted repeat at left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "TTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at right end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC. This region is a possible MT-complex-specific genomic island (See Becq et al.,2007).']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "[\"3 bp direct repeat, att, at 5'-end of IS6110. This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).\"]\n",
      "AATT\n",
      " \n",
      "['77 bp Mycobacterial Interspersed Repetitive Unit,Class I']\n",
      "GTGACTGCGCGGGCGACGACGAAGGGAGAGCACGATGAGGTGGGGGTCCCTGCCGCTTGCGGGGGAGAGCGGCGCCTA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "GTGATCCGCGCCGGCGACGATGCAGAGCGCAGCGATGCTAAGGAGCGGCGCCGA\n",
      " \n",
      "['58 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "CCACACCGCGCCAGCGACGATGCAGTGGGGGTACCACCCGCGTGCGGGGGACGAAGCGA\n",
      " \n",
      "['110 bp Mycobacterial Interspersed Repetitive Unit,Class III']\n",
      "ATCAACGCCGCCGCCGACGATGCAGTGGGGGCACCCCCAGCCGCGAAGCGACGAGGGGGCCGCAGCGATGAGGTGGGGGTACCGCCCGCTTGCGGGGGAGAGGTGGCGTGA\n",
      " \n",
      "['28 bp inverted repeat at left end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "CTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at right end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "['207 bp imperfect direct repeat 1, 199/207 bp identical to second copy at 3769514..3769720']\n",
      "TCTAGCCCGTCCCCAACGGGTCAATGGTGGATGCGATATAGACCATGGCCGCCGCGACCGTCACGGTCGTCACGAAATCGATCCCCTTGCTGCGCACCACCAACAGGCCGGCCCGTTCCTCGGACAACACCAACCGCAGCACCGCCGCCACCCCAACGCCGATACCGATCAGCAGCGCACCACGGCGCCAGAAGTTGACCCCCGCCAG\n",
      " \n",
      "['109 bp imperfect direct repeat 1, 95/109 bp identical to second copy at 3769754..3769862']\n",
      "CCAGGATCGGCCACTGGGCGCCAACAGTGCGCCGCAAAACGGCCCTCACGGTCATCGCCGCTCAGCCAGCTCCACGACACTTGTCAGCAAGGACGCCCGGGGCGAAGGGC\n",
      " \n",
      "['98 bp imperfect direct repeat 1, 82/98 bp identical to the second copy at 3770994..3771091']\n",
      "GGGCGTTCGCCAAGTCTGTAGATGAGCTGCGGGAGATGGCCGACGGCGAGGGTTGAGAAGCGTCAACTTCGATCGTGATGCCTGGGAGGACTTCTTATT\n",
      " \n",
      "['207 bp imperfect direct repeat 2, 199/207 bp identical to first copy at 3743198..3743404']\n",
      "CCTAGCCGGTCCCTAACGGATCGATCGTGGATGCGATGTAGACCATGGCCGCCGCGACCGTCACGGTCGTCACGAAATCGATCCCCTTGCTGCGCACCACCAACAGGCCGGCCCGTTCCTCGGACAACACCAACCGCAGCACCGCCGCCACCCCAACGCCGATACCGATCAGCAGCGCACCACGGCGCCAGAAGTTAGCCCCCGCCAG\n",
      " \n",
      "['109 bp imperfect direct repeat 2, 95/109 bp identical to first copy at 3743402..3743510']\n",
      "GCAGGATCGGCCACTGGGCGCCAACAGTGCGCCGGAAAACGGCCCTCACGGTCATCGCCGCTCAGCCAGCTCCACGACATTGGTCAACAAGAACGCCCGGGTCAACGGGC\n",
      " \n",
      "['98 bp imperfect direct repeat 2, 82/98 bp identical to the first copy at 3743508..3743605']\n",
      "CGGCTTTCACCAAGTCTGTAGATGAGCTGCGGGAGATGGCCGGCGGCGAGGAGTGAGAAGCGTCAACTTCGATCCCGATGCCTGGGAGGACTTCTTGTT\n",
      " \n",
      "['28 bp inverted repeat at the left end of IS6110,TGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "TTGAACCGCCCCGGTGAGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at the right end of IS6110,TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGCATGTCCGGAGACTCC\n",
      " \n",
      "['25 bp inverted repeat at the right end of putative IS1560, TAATTACTAGGACCTGAAAAAGTCG']\n",
      "TTAATTACTAGGACCTGAAAAAGTCG\n",
      " \n",
      "['25 bp inverted repeat at the right end of putative IS1560, TAATTACTAAGACCTGAAAAAGTCG']\n",
      "TAATTACTAAGACCTGAAAAAGTCGC\n",
      " \n",
      "['This region is a possible MT-complex-specific genomic island (See Becq et al., 2007).']\n",
      "ATTGTCGAGGTCTTTGATGCGCTGGATGCCGAGCTGGACCGCTTGGACGAGGTGTCTTTTGAGGTGTTGACCACCCCAGAACGGCTGCGGTCTCTGGAACGTCTGGAATGCTTGGTGCGCCGGCTACCGGCGGTGGGTCACGCGTTGATCAACCAACTTGACGCCCAAGCCAGCGAGGAAGAACTGGGCGGCACGCTGTGCTGCGCGCTGGCCAACCGGTTACGCATCACCAAGCCCGACGCCGCCCGGCGCATCGCCGACGCCGCCGATCTCGGACCTCGTCGAGCACTCACCGGTGAACCGCTAGCCCCACAGTTGACCGCCACCGCCACCGCCCAACGCCAGGGCCTGATCGGCGAGGCGCACGTCAAAGTGATTCGCGCCCTTTTTCGCCCACCTGCCCGCCGCGGTGGATGTGTCCACCCGCCAGGCCGCCGAAGCCGACCTGGCCGGCAAAGCCGCTCAATATCGTCCCGACGAGCTGGCCCGCTACGCCCAGCGGGTCATGGACTGGCTACACCCCGACGGCGACCTCACCGACACCGAACGCGCCCGCAAACGCGGCATCACCCTGAGCAACCAGCAATACGACGGCATGTCACGGCTAAGTGGCTACCTGACCCCCCAAGCGCGGGCCACCTTTGAAGCCGTGCTAGCCAAACTGGCCGCCCCCGGCGCGACCAACCCCGACGACCACACCCCGGTCATCGACACCACCCCCGATGCGGCCGCCATCGACCGCGACACCCGCAGCCAAGCCCAACGCAACCACGACGGGCTGCTGGCCGGGCTGCGCGCGCTGATCGCCTCCGGGAAACTGGGCCAACACAACGGTCTTCCCGTCTCGATCGTGGTCACCACCACCCTGACCGACCTGCAAACCGGCGCCGGCAAGGGCTTCACCGGCGGCGGCACCCTGCTACCCATGGCCGATGTGATCCGCATGACCAGCCACGCCCACCACTACTCCCCCGCAAGCGGGAGGTACCCCCAGGCGATCTTCGACCACGGCACACCCCTGGCGCTGTATCACACCAAACGCCTAGCCTCCCCGGCCCAGCGGATCATGCTGTTCGCCAACGACCGCGGCTGCACCAAACCCGGCTGTGACGCACCGGCCTACCACAGCCAAGCCCACCACGTCACCGCCTGGACCAGCACCGGACGCACCGACATCACCGAGCTGACCCTGGCCTGCGGCCCCGACAACCGACTCGCCGAAAAAGGCTGGACCACCCACAACAACACCCACGGCCACACCGAATGGCTACCACCACCCCACCTCGACCACGGCCAACCCCGCACCAACACCTTCCACCACCCCGAACGATTCCTCCACAACCAAGACGACGACGACAAACCCGATTGACCCC\n",
      " \n",
      "['28 bp inverted repeat at left end of IS6110 :TGAACCGCCCCGGCATGTCCGGAGACTC']\n",
      "CTGAACCGCCCCGGCATGTCCGGAGACTC\n",
      " \n",
      "['28 bp inverted repeat at right end of IS6110 :TGAACCGCCCCGGTGAGTCCGGAGACTC']\n",
      "TGAACCGCCCCGGTGAGTCCGGAGACTCT\n",
      " \n",
      "['500 bp perfect direct repeat 2; second copy at 3950830..3951329.']\n",
      "TCGCCAAGAAGGGCAACATTCCCGTCGGCTACTACAAGGACGAGAAGAAGACGGCCGAGACGTTCCGGACGATTAACGGTGTGCGCTACGCCATTCCGGGGGACTACGCCCAAGTCGAGGAGGACGGCACGGTCACCATGCTGGGTCGTGGCTCGGTGTCGATCAACAGCGGCGGCGAGAAGGTCTACCCCGAAGAGGTCGAGGCCGCTTTGAAGGGGCATCCCGACGTGTTCGACGCTCTGGTGGTCGGGGTGCCCGATCCGCGTTACGGCCAGCAGGTGGCCGCCGTGGTGCAGGCCCGGCCGGGTTGTCGGCCGTCGTTGGCCGAGTTGGACTCGTTCGTGCGCTCCGAGATCGCGGGCTACAAAGTGCCGCGCAGTCTGTGGTTTGTCGACGAGGTGAAGCGTTCGCCCGCCGGTAAGCCGGACTACCGCTGGGCCAAGGAGCAGACCGAGGCGCGGCCAGCCGATGACGTGCATGCGGGACACGTGACGAGCGGTT\n",
      " \n",
      "['500 bp perfect direct repeat 1; second copy at 3945098..3945597.']\n",
      "TCGCCAAGAAGGGCAACATTCCCGTCGGCTACTACAAGGACGAGAAGAAGACGGCCGAGACGTTCCGGACGATTAACGGTGTGCGCTACGCCATTCCGGGGGACTACGCCCAAGTCGAGGAGGACGGCACGGTCACCATGCTGGGTCGTGGCTCGGTGTCGATCAACAGCGGCGGCGAGAAGGTCTACCCCGAAGAGGTCGAGGCCGCTTTGAAGGGGCATCCCGACGTGTTCGACGCTCTGGTGGTCGGGGTGCCCGATCCGCGTTACGGCCAGCAGGTGGCCGCCGTGGTGCAGGCCCGGCCGGGTTGTCGGCCGTCGTTGGCCGAGTTGGACTCGTTCGTGCGCTCCGAGATCGCGGGCTACAAAGTGCCGCGCAGTCTGTGGTTTGTCGACGAGGTGAAGCGTTCGCCCGCCGGTAAGCCGGACTACCGCTGGGCCAAGGAGCAGACCGAGGCGCGGCCAGCCGATGACGTGCATGCGGGACACGTGACGAGCGGTG\n",
      " \n",
      "['58 bp Mycobacterial Interspersed Repetitive Unit,Class III.']\n",
      "CTCATAGCCACCGCCGACAATGCAGTGGGGGTACCACCCGCTTGCAGGGGACGAAGCGA\n",
      " \n",
      "['18 bp direct repeat 2, GGGTTTGCGATCGCCACG']\n",
      "GGGTTTGCGATCGCCACGA\n",
      " \n",
      "['(24 bp) part of 111 bp direct repeat unit 6,GTGGCGACCCGCTGCACCCGGCTC']\n",
      "GTGGCGACCCGCTGCACCCGGCTCC\n",
      " \n",
      "['111 bp direct repeat unit 5, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG']\n",
      "GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTTTGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATGG\n",
      " \n",
      "['18 bp direct repeat 1, GGGTTTGCGATCGCCACG']\n",
      "GGGTTTGCGATCGCCACGG\n",
      " \n",
      "['111 bp direct repeat unit 4, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG']\n",
      "GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTTTGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATGG\n",
      " \n",
      "['111 bp direct repeat unit 3, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG']\n",
      "GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTTTGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATGG\n",
      " \n",
      "['111 bp direct repeat unit 2, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG']\n",
      "GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTTTGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATGG\n",
      " \n",
      "['111 bp direct repeat unit 1, GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTT TGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATG']\n",
      "GTGGCGACCCGCTGCACCCGGCTCTGGGGTGATTGCCTGGCTCCTCCTCGGCCCGTTTTGCGGGCCGCATTGTCGCCAGGCGCGGGGTTTGCGATCGCCACGGGGCTGATGG\n",
      " \n",
      "['16 bp inverted repeat at the left end of putative IS element IS1534; GAAAATTGACCAGCTT.']\n",
      "CACCGAGATGATCTTGG\n",
      " \n",
      "['16 bp inverted repeat at the right end of putative IS element IS1534; GAAAATTGACCAGCTT.']\n",
      "GAAAATTGACCAGCTTG\n",
      " \n",
      "['13 bp inverted repeat at the right end of putative IS element IS1553; GAGTTCGTCGGTG.']\n",
      "TGAGTTCGTCGGTG\n",
      " \n",
      "['13 bp inverted repeat at the left end of putative IS element IS1553; GAGATCGTCGGTG.']\n",
      "GAGATCGTCGGTGT\n",
      " \n",
      "['125 bp Mycobacterial Interspersed Repetitive Unit,Class III.']\n",
      "TGTGATCGGCGCCGGCGACGATGCAGTGGGGGCACCACCCGCTTGCGGGGGACGAAGCGATGGGGTGGGGGTACGCCCCCACAAGTGGGAGGTACCCCCACCCGCTTGCGGGGGAGAGCGGCGCAG\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "CGTGACCCGCGCCGGCGACGATGCAAAGCGCAGCGATGAGGAGGAGCGGCGCCA\n",
      " \n",
      "['53 bp Mycobacterial Interspersed Repetitive Unit,Class II']\n",
      "AACGGCCCGCGCCGGCGACGATGCAAAGCGCAGCGATGAGGAGGAGCGGCGCGC\n",
      " \n",
      "['51 bp imperfect direct repeat 1,GAACCGGCCGCATCTAAACCACCCACACCCCCCATGCCCATCGCCGGACCC']\n",
      "GGAACCGGCCGCATCTAAACCACCCACACCCCCCATGCCCATCGCCGGACCC\n",
      " \n",
      "['51 bp imperfect direct repeat 2,GAACCGGCCCCACCCAAACCACCCACACCCCCCATGCCCATCGCCGGACCC']\n",
      "CGAACCGGCCCCACCCAAACCACCCACACCCCCCATGCCCATCGCCGGACCC\n",
      " \n",
      "['51 bp imperfect direct repeat 3,GAACCGGCCCCACCCAAACCACCCACACCTCCGATGCCCATCGCCGGACCT']\n",
      "CGAACCGGCCCCACCCAAACCACCCACACCTCCGATGCCCATCGCCGGACCT\n",
      " \n"
     ]
    }
   ],
   "source": [
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "features = []\n",
    "for feature in genome_record.features:\n",
    "    if not(feature.type in ['gene','CDS','source']):\n",
    "        if 'repeat' in feature.type:\n",
    "            #print (feature)\n",
    "            #print(feature.location.strand)\n",
    "            print (feature.qualifiers['note'])\n",
    "            if int(feature.location.strand) == 1:\n",
    "                print (full_sequence[feature.location.start - 1: feature.location.end])\n",
    "            else:\n",
    "                print(util.reverse_complement(full_sequence[feature.location.start - 1: feature.location.end]))\n",
    "            print(\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "234d826e-28dc-408b-8472-da2e5bd8fdc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(183291, 183230, '-', '!'), (234448, 234506, '+', '!'), (234504, 234446, '-', '!'), (253604, 253665, '+', '!'), (253663, 253602, '-', '!'), (279535, 279595, '+', '!'), (279594, 279534, '-', '!'), (456202, 456262, '+', '!'), (456261, 456201, '-', '!'), (459447, 459387, '-', '!'), (558819, 558879, '+', '!'), (558877, 558817, '-', '!'), (663390, 663451, '+', '!'), (663449, 663388, '-', '!'), (736233, 736293, '+', '!'), (736292, 736232, '-', '!'), (755266, 755327, '+', '!'), (755325, 755265, '-', '!'), (767325, 767385, '+', '!'), (767384, 767323, '-', '!'), (786020, 786087, '+', '!'), (786085, 786019, '-', '!'), (806173, 806233, '+', '!'), (806231, 806171, '-', '!'), (815637, 815698, '+', '!'), (815696, 815635, '-', '!'), (829707, 829768, '+', '!'), (829766, 829705, '-', '!'), (842303, 842364, '+', '!'), (842362, 842301, '-', '!'), (958458, 958519, '+', '!'), (958517, 958456, '-', '!'), (1000772, 1000832, '+', '!'), (1000831, 1000771, '-', '!'), (1006619, 1006680, '+', '!'), (1006678, 1006617, '-', '!'), (1064053, 1064114, '+', '!'), (1064112, 1064051, '-', '!'), (1133267, 1133326, '+', '!'), (1133324, 1133264, '-', '!'), (1182321, 1182382, '+', '!'), (1182380, 1182319, '-', '!'), (1205239, 1205299, '+', '!'), (1205297, 1205237, '-', '!'), (1224310, 1224370, '+', '!'), (1224369, 1224308, '-', '!'), (1313334, 1313394, '+', '!'), (1313392, 1313332, '-', '!'), (1345183, 1345243, '+', '!'), (1345242, 1345182, '-', '!'), (1357065, 1357127, '+', '!'), (1357123, 1357063, '-', '!'), (1363439, 1363498, '+', '!'), (1363498, 1363438, '-', '!'), (1371709, 1371768, '+', '!'), (1371766, 1371706, '-', '!'), (1488085, 1488146, '+', '!'), (1488144, 1488084, '-', '!'), (1497161, 1497221, '+', '!'), (1497219, 1497159, '-', '!'), (1568047, 1568109, '+', '!'), (1568105, 1568045, '-', '!'), (1828800, 1828861, '+', '!'), (1828859, 1828798, '-', '!'), (1872570, 1872630, '+', '!'), (1872629, 1872568, '-', '!'), (2075533, 2075593, '+', '!'), (2075591, 2075531, '-', '!'), (2317103, 2317164, '+', '!'), (2317163, 2317101, '-', '!'), (2451179, 2451240, '+', '!'), (2451238, 2451177, '-', '!'), (2466966, 2467026, '+', '!'), (2467024, 2466964, '-', '!'), (2522173, 2522233, '+', '!'), (2522232, 2522171, '-', '!'), (2523182, 2523240, '+', '!'), (2523238, 2523181, '-', '!'), (2546436, 2546496, '+', '!'), (2546495, 2546435, '-', '!'), (2554866, 2554926, '+', '!'), (2554925, 2554864, '-', '!'), (2700469, 2700529, '+', '!'), (2700528, 2700468, '-', '!'), (2751597, 2751658, '+', '!'), (2751656, 2751595, '-', '!'), (2839476, 2839537, '+', '!'), (2839535, 2839474, '-', '!'), (2907752, 2907812, '+', '!'), (2907811, 2907751, '-', '!'), (2931696, 2931635, '-', '!'), (3059791, 3059851, '+', '!'), (3059850, 3059791, '-', '!'), (3075379, 3075439, '+', '!'), (3075438, 3075378, '-', '!'), (3104979, 3105039, '+', '!'), (3105036, 3104977, '-', '!'), (3173095, 3173160, '+', '!'), (3173158, 3173094, '-', '!'), (3215656, 3215595, '-', '!'), (3348481, 3348541, '+', '!'), (3348540, 3348481, '-', '!'), (3369888, 3369948, '+', '!'), (3369947, 3369886, '-', '!'), (3387006, 3387067, '+', '!'), (3508001, 3508062, '+', '!'), (3508060, 3508000, '-', '!'), (3560116, 3560176, '+', '!'), (3560175, 3560114, '-', '!'), (3590632, 3590692, '+', '!'), (3590690, 3590630, '-', '!'), (3628087, 3628148, '+', '!'), (3628146, 3628085, '-', '!'), (3707572, 3707633, '+', '!'), (3707631, 3707570, '-', '!'), (3724554, 3724615, '+', '!'), (3724613, 3724552, '-', '!'), (3803854, 3803916, '+', '!'), (3803912, 3803852, '-', '!'), (3835265, 3835204, '-', '!'), (3909759, 3909820, '+', '!'), (3909818, 3909757, '-', '!'), (4008655, 4008716, '+', '!'), (4008714, 4008654, '-', '!'), (4021394, 4021455, '+', '!'), (4021453, 4021392, '-', '!'), (4088259, 4088320, '+', '!'), (4088318, 4088258, '-', '!'), (4237936, 4237996, '+', '!'), (4237994, 4237934, '-', '!'), (4246526, 4246587, '+', '!'), (4246585, 4246524, '-', '!')]\n"
     ]
    }
   ],
   "source": [
    "with open(r_scape_output_loc + '/' + 'Rv1324' + '/summary.txt', 'r') as f:  \n",
    "                hit_dict = {}\n",
    "                ref_count = 0\n",
    "                tot_count = 0\n",
    "                hits = []\n",
    "                for l in f:\n",
    "                    if (not ('#' in l)):\n",
    "                        a = l.split()\n",
    "                        accession = a[0]\n",
    "                        species = names_dict[a[0]]\n",
    "                        seq_from = int(a[7])\n",
    "                        seq_to = int(a[8])\n",
    "                        seq_strand = a[9]\n",
    "                        significant = a[16]\n",
    "                        start = min(seq_from,seq_to)\n",
    "                        stop = max(seq_from,seq_to)\n",
    "                        if (significant == '!') and accession == 'NC_000962.3':\n",
    "                            hits.append((seq_from, seq_to,seq_strand, significant))\n",
    "print(sorted(hits, key=lambda item: item[0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "3276d916-6d63-4150-b111-b4c837d03379",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000962.3/1488085-1488146           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..C.U..A.A.A.C.CC.G.C.A.C.......................G.G.G.U.U.U.A..G.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/4021453-4021392           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..G.A.U.U.UG.A.G.C.U.......................C.A.A.A.U.C.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1828800-1828861           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.CG.U.A.C.C.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/3724613-3724552           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.U.U.UC.G.U.G.C.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3707572-3707633           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..G.A.A.A.AC.C.A.G.U.......................G.G.U.U.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/4021394-4021455           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..G.A.U.U.UG.A.G.C.U.......................C.A.A.A.U.C.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GA\n",
      "\n",
      "NC_000962.3/3707631-3707570           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.A.A.CC.A.C.U.G.......................G.U.U.U.U.C.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3909759-3909820           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..C.C..A.U.U.U.CG.G.C.A.C.......................G.A.A.A.U.U.G..G.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1828859-1828798           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.CG.G.G.U.A.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/1488144-1488084           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..C.U..A.A.A.C.CC.G.U.G.C.......................G.G.G.U.U.U.A..G.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/4246585-4246524           GC.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..A.A.A.A.UC.G.C.C.C.......................G.A.U.U.U.G.G..G.G.U.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3628146-3628085           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.UC.A.U.G.A.......................C.G.A.A.A.A.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/842303-842364             GC.G.A.GC.A.GACGC.AU.A.A.G.C..G.C..C.C..A.A.U.U.UC.G.U.G.U.......................C.G.A.A.A.U.G..G.G.C.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3909818-3909757           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..C.C..A.A.U.U.UC.G.U.G.C.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GA\n",
      "\n",
      "NC_000962.3/1568105-1568045           -C.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..G.C.G.G.AA.G.C.C.U.......................U.U.C.C.G.C.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/279535-279595             GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.U.U.UC.G.U.G.C.......................C.G.A.A.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/4246526-4246587           GC.G.A.GC.A.GACGC.AA.A.A.G.C..A.C..C.C..C.A.A.A.UC.G.G.G.C.......................G.A.U.U.U.U.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/663449-663388             GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..C.A.A.A.AU.C.G.C.U.......................G.A.U.U.U.U.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2751656-2751595           GC.G.A.AA.A.GACGC.AA.A.A.G.C..U.C..C.C..A.A.A.U.CC.G.G.U.C.......................G.G.A.U.U.U.G..G.G.A.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/279594-279534             GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.U.U.CG.G.C.A.C.......................G.A.A.A.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/3803854-3803916           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.U.U.UC.G.U.G.C.......................C.G.A.A.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.GgC.GC\n",
      "\n",
      "NC_000962.3/3105036-3104977           --.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..A.U.U.U.CG.G.G.C.C.......................C.G.A.A.A.U.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.A.C.GC\n",
      "\n",
      "NC_000962.3/4008655-4008716           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.C.A.CG.C.C.C.G.......................C.A.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3628087-3628148           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..U.U.U.U.CG.U.C.A.U.......................G.A.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/3104979-3105039           GU.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..A.U.U.U.CG.G.G.C.C.......................C.G.A.A.A.U.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/3724554-3724615           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.CG.G.C.A.C.......................G.A.A.A.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/1568047-1568109           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..G.C.G.G.AA.A.G.G.C.......................U.U.C.C.G.C.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.GgC.GC\n",
      "\n",
      "NC_000962.3/2317163-2317101           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..G.C.G.C.GG.G.A.C.C.......................C.C.G.C.G.C.A.gU.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/4237936-4237996           -C.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..A.A.U.U.CG.G.G.C.A.......................C.G.A.A.A.U.G..G.G.G.G.C.U.U.UU.ACGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3803912-3803852           -C.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.U.U.CG.G.C.A.C.......................G.A.A.A.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/663390-663451             GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..A.A.A.A.UC.A.G.C.G.......................A.U.U.U.U.G.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1357123-1357063           -C.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..C.C..A.A.U.U.UC.G.U.G.C.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.CU.GUGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2839476-2839537           GC.G.A.GC.A.GACGC.AG.A.G.U.C..G.C..A.U..G.A.A.A.AU.C.C.G.G.......................A.U.U.U.C.G.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1000831-1000771           GC.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.U..A.A.U.A.CC.U.G.C.C.......................G.G.U.U.U.A.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/2751597-2751658           GC.G.A.GC.A.GACGC.AA.A.A.G.C..U.C..C.C..A.A.A.U.CC.G.A.C.C.......................G.G.A.U.U.U.G..G.G.A.G.C.U.U.UU.GCGUC.U.U...UU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/1363498-1363438           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..U.G.A.A.CC.G.U.G.C.......................G.U.U.C.C.A.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/1000772-1000832           GC.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..U.A.A.A.CC.G.G.C.A.......................G.G.U.A.U.U.A..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/4237994-4237934           GC.G.A.GC.A.GACGU.AA.A.A.G.C..C.C..C.C..A.U.U.U.CG.U.G.C.C.......................C.G.A.A.U.U.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/3590632-3590692           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..G.A.A.A.CC.G.A.U.G.......................G.C.U.U.U.C.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...-U...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1006619-1006680           GC.G.A.GC.A.GACAC.AG.A.G.U.C..G.C..A.U..C.A.A.A.AC.C.G.C.C.......................G.A.U.U.U.U.G..U.G.C.G.A.C.U.CU.GUGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/3348481-3348541           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.CG.A.G.A.C.......................G.A.A.A.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/842362-842301             GC.G.A.GC.A.GACGC.AU.A.A.G.C..G.C..C.C..A.U.U.U.CG.A.C.A.C.......................G.A.A.A.U.U.G..G.G.C.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/4008714-4008654           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.UG.C.G.G.G.......................C.G.U.G.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/958458-958519             GC.G.A.GC.A.GACGU.AA.A.A.G.C..C.C..C.C..G.C.A.C.GC.U.C.-.G.......................G.C.G.U.GuC.G..G.G.G.G.C.U.U.UU.ACGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2839535-2839474           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..C.G.A.A.AU.C.C.G.G.......................A.U.U.U.U.C.A..U.G.C.G.A.C.U.CU.GCGUC.U.G...CU...C.G.C.GA\n",
      "\n",
      "NC_000962.3/3590690-3590630           GC.G.A.-C.A.GACGC.AA.A.A.U.C..G.C..C.C..G.A.A.A.GC.C.A.U.C.......................G.G.U.U.U.C.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1872629-1872568           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..G.C.C.G.AC.G.G.G.C.......................G.U.G.G.C.G.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/2700469-2700529           CC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.A.U.UC.G.G.G.C.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/958517-958456             GC.G.A.GC.A.GACGU.AA.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.C.G.A.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.UU.ACGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1182380-1182319           GC.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..A.U.U.U.CG.A.C.A.C.......................G.A.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2700528-2700468           CC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.CG.G.C.C.C.......................G.A.A.U.U.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/2523182-2523240           GU.G.A.GC.A.GACGC.AG.A.G.U.C..G.C..A.C..U.G.C.G.GC.-.-.-.C.......................G.G.C.G.C.A.G..U.G.C.G.A.C.U.CU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1182321-1182382           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.UC.G.U.G.U.......................C.G.A.A.A.U.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/1363439-1363498           CC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..U.G.G.A.AC.G.C.A.C.......................G.G.U.U.C.A.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.-C\n",
      "\n",
      "NC_000962.3/456202-456262             GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..A.C..G.G.U.U.UG.C.G.G.U.......................U.G.A.U.U.C.G..U.G.C.G.A.U.U.UU.GUGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/1006678-1006617           GC.G.A.GC.A.GACAC.AG.A.G.U.C..G.C..A.C..A.A.A.A.UC.G.G.C.G.......................G.U.U.U.U.G.A..U.G.C.G.A.C.U.CU.GUGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/1872570-1872630           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..C.G.C.C.AC.G.C.C.C.......................G.U.C.G.G.C.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/3059791-3059851           GC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.U..C.A.U.U.UC.G.G.C.A.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/1133267-1133326           --.G.A.GC.A.GACGC.AA.A.A.U.C..C.C..C.U..C.G.A.C.AC.G.C.C.G.......................G.U.U.G.C.G.A..G.G.G.G.A.U.U.UU.GCGUC.U.G...CU...C.A.C.GC\n",
      "\n",
      "NC_000962.3/3075379-3075439           CC.A.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..U.A.A.U.CG.C.G.C.C.......................C.G.G.U.U.U.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...U.G.-.GC\n",
      "\n",
      "NC_000962.3/3348540-3348481           CC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.A.U.U.UC.G.U.C.U.......................C.G.A.A.A.U.G..G.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.-C\n",
      "\n",
      "NC_000962.3/1357065-1357127           GC.G.A.GC.A.GACAC.AG.A.A.U.C..G.C..C.C..A.U.U.U.CG.G.C.A.C.......................G.A.A.A.U.U.G..G.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.GgC.GC\n",
      "\n",
      "NC_000962.3/2522232-2522171           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..G.G.C.G.CG.G.A.C.C.......................U.C.G.C.G.C.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/815637-815698             GC.G.A.GC.A.GACGC.AA.A.AuG.C..C.C..C.C..G.A.-.A.AU.G.C.C.C.......................A.U.U.U.U.C.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/829707-829768             GC.G.G.GC.A.GACGC.GU.A.A.G.C..G.C..C.C..A.A.U.G.UC.G.U.G.C.......................C.G.A.A.A.U.G..G.G.C.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2451179-2451240           GU.G.A.GC.A.GACGC.AA.A.G.U.C..G.C..C.C..A.A.A.A.UC.G.C.C.G.......................G.U.U.U.C.C.G..G.G.U.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/2317103-2317164           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..AcU..G.C.G.C.GG.G.G.U.C.......................C.C.G.C.G.C.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/3560175-3560114           GC.G.A.GC.A.GACGC.AG.A.A.G.U..C.G..C.C..G.A.U.U.UC.A.C.C.G.......................G.A.A.A.U.C.G..G.G.G.A.C.U.U.CU.GCGUC.U.G...CU...C.A.C.GG\n",
      "\n",
      "NC_000962.3/4088259-4088320           GC.G.A.GC.A.GACGU.A-.A.A.G.C..C.C..C.C..A.A.A.A.CG.G.A.A.C.......................C.G.U.U.U.U.G..G.G.G.G.C.U.UuUU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/3508001-3508062           GC.G.A.GC.A.GACGU.CA.A.A.G.C..C.C..C.C..G.A.A.U.UU.C.A.G.U.......................G.A.A.A.U.C.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...CU...G.G.C.GC\n",
      "\n",
      "NC_000962.3/1133324-1133264           GU.G.A.GC.A.GACGC.AA.A.A.U.C..C.C..C.U..C.G.C.A.AC.C.G.G.C.......................G.U.G.U.C.G.A..G.G.G.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/558819-558879             -C.G.A.GC.A.GACGC.AA.A.A.G.C..A.C..C.C..U.U.U.U.GC.G.G.C.G.......................C.A.A.A.A.G.U..G.G.C.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3075438-3075378           CC.A.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..A.A.A.C.CG.G.G.C.G.......................C.G.A.U.U.A.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...U.G.-.GC\n",
      "\n",
      "NC_000962.3/2451238-2451177           GC.G.A.GC.A.GACGC.AA.A.A.U.C..A.C..C.C..G.G.A.A.AC.C.G.G.C.......................G.A.U.U.U.U.G..G.G.C.G.A.C.U.UU.GCGUC.U.G...CU...C.A.C.GC\n",
      "\n",
      "NC_000962.3/1064053-1064114           GC.G.A.GC.A.GACGC.AU.A.A.G.C..C.C..C.C..G.C.A.C.GC.U.C.-.G.......................G.C.G.U.GuC.G..G.G.G.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/3560116-3560176           GU.G.A.GC.A.GACGC.AG.A.A.G.U..C.C..C.C..G.A.U.U.UC.C.G.G.U.......................G.A.A.A.U.C.G..G.C.G.A.C.U.U.CU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/1313392-1313332           GC.G.A.GC.A.GACGC.AA.A.A.G.G..C.C..C.C..G.A.A.A.UC.C.G.U.U.......................G.G.A.U.U.C.G..G.G.G.C.C.U.U.UU.GCGUC.U.-...UU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/1064112-1064051           GC.G.A.GC.A.GACGC.AU.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.C.G.A.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/2466966-2467026           GC.G.A.GC.A.GACGC.GG.A.A.U.C..G.C..A.U..C.G.G.C.GG.G.G.C.C.......................C.A.C.A.C.G.G..U.G.C.G.A.U.U.CC.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/815696-815635             GC.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..G.A.A.A.AU.G.G.G.C.......................A.U.U.-.U.C.G..G.G.G.G.CaU.U.UU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/736233-736293             GC.G.A.GC.A.GACGC.AU.A.A.G.C..C.C..C.C..G.C.A.C.GC.A.C.-.G.......................G.C.G.U.GuC.G..G.G.G.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/2522173-2522233           GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..G.C.G.C.GA.G.G.U.C.......................C.G.C.G.C.C.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/736292-736232             GC.G.A.GC.A.GACGC.AU.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.C.G.U.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/1205239-1205299           GC.G.A.GC.A.GACGC.GA.A.A.G.C..A.C..C.C..C.A.A.A.AC.C.G.C.C.......................G.G.U.U.U.G.G..G.G.-.G.C.U.U.CU.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/558877-558817             GC.G.A.GC.A.GACGC.AA.A.A.G.C..G.C..CaC..U.U.U.U.GC.-.G.C.C.......................G.C.A.A.A.A.G..G.G.U.G.C.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/456261-456201             GC.G.A.GC.A.GACAC.AA.A.A.U.C..G.C..A.C..G.A.A.U.CA.A.C.C.G.......................C.A.A.A.C.C.G..U.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/1224369-1224308           GC.A.A.GC.A.GACAC.AG.A.A.U.C..G.C..A.C..G.C.G.C.CG.G.G.-.C.......................C.G.G.C.G.CaG..U.G.C.G.A.U.U.CU.GUGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/1205297-1205237           GC.G.A.GC.A.GACGC.AG.A.A.G.C..-.C..C.C..C.A.A.A.CC.G.G.C.G.......................G.U.U.U.U.G.G..G.G.U.G.C.U.U.UC.GCGUC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/2467024-2466964           -C.G.A.GC.A.GACGC.GG.A.A.U.C..G.C..A.C..C.G.U.G.UG.G.G.C.C.......................C.C.G.C.C.G.A..U.G.C.G.A.U.U.CC.GCGUC.U.G...CU...C.G.C.GA\n",
      "\n",
      "NC_000962.3/1313334-1313394           GC.G.A.A-.A.GACGC.AA.A.A.G.G..C.C..C.C..G.A.A.U.CC.A.A.C.G.......................G.A.U.U.U.C.G..G.G.G.C.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/767384-767323             GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..C.G.C.G.CA.G.G.C.A.......................U.U.U.C.G.C.G..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/3369947-3369886           GC.G.A.GC.A.GACAC.AG.A.C.U.C..G.C..A.U..C.C.G.G.CG.C.C.C.C.......................U.C.G.C.C.C.G..U.G.C.G.A.U.U.CU.GUGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/4088318-4088258           GC.G.A.GC.A.GACGC.AAaA.A.G.C..C.C..C.C..A.A.A.A.CG.G.U.U.C.......................C.G.U.U.U.U.G..G.G.G.G.C.U.U.-U.ACGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/3059850-3059791           CC.G.A.GC.A.GACGC.AA.A.A.U.C..G.C..C.C..A.U.U.U.CG.U.G.C.-.......................C.G.A.A.A.U.G.aG.G.C.G.A.U.U.UU.GCGUC.U.G...CU...C.G.-.-C\n",
      "\n",
      "NC_000962.3/2523238-2523181           GC.G.A.GC.A.GACGC.AG.A.G.U.C..G.C..A.C..U.G.C.G.CC.-.-.-.G.......................G.C.C.G.C.A.G..U.G.C.G.A.C.U.CU.GCGUC.U.G...CU...C.A.C.-C\n",
      "\n",
      "NC_000962.3/1345242-1345182           CC.G.A.GC.A.GACGC.AA.A.A.G.C..A.C..C.C..G.A.U.U.CU.G.G.U.C.......................G.A.A.A.U.A.G..G.G.U.G.G.U.U.UU.GUGUC.U.G...UU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/1224310-1224370           GC.G.A.GC.A.GACAC.AG.A.A.U.C..G.C..A.C.uG.C.G.C.CG.-.G.C.C.......................C.G.G.C.G.C.G..U.G.C.G.A.U.U.CU.GUGUC.U.G...CU...U.G.C.-C\n",
      "\n",
      "NC_000962.3/2907811-2907751           CC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.U..U.G.G.G.CG.G.G.C.U.......................G.A.U.U.U.C.G..U.G.C.G.A.U.U.CU.GUGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/767325-767385             GC.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..A.C..G.C.G.A.AA.U.G.C.C.......................-.U.G.C.G.C.G.aU.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/234504-234446             GC.G.A.GC.A.G--GC.AG.A.A.U.C..G.C..C.C..A.U.U.U.GG.G.A.U.A.......................C.C.A.A.A.U.G..G.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/755266-755327             GC.G.A.GC.A.GACGC.AU.A.A.G.C..C.C..C.C..G.C.A.C.GC.U.C.-.G.......................G.C.G.U.GuC.G..G.G.G.G.C.U.U.AU.GCGAC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/234448-234506             -C.G.A.GC.A.GACGC.AG.A.A.U.C..G.C..C.C..A.U.U.U.GG.U.A.U.C.......................C.C.A.A.A.U.G..G.G.C.G.A.U.U.CU.GC--C.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/806231-806171             GC.G.A.GC.A.GUCGC.AA.A.A.G.C..C.C..C.C..G.C.A.C.GC.-.C.G.C.......................G.C.G.U.GuC.G..G.G.G.G.C.U.U.UU.GCGAC.U.G...CU...C.G.U.-C\n",
      "\n",
      "NC_000962.3/786020-786087             GC.C.A.CC.A.GACGC.AA.A.A.G.A..A.C..A.U..G.A.U.G.GG.C.-.-.-gcac.............cagcg.C.C.C.A.U.C.A..U.G.U.U.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GA\n",
      "\n",
      "NC_000962.3/1497161-1497221           -C.G.A.GA.A.GACGC.AA.A.A.G.C..U.C..C.U..A.A.A.U.CU.G.G.C.C.......................G.A.U.U.U.A.G..U.G.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3508060-3508000           GC.C.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..G.A.U.U.UC.A.C.U.G.......................A.A.A.U.U.C.G..G.G.G.G.C.U.U.UG.ACGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/3173095-3173160           GC.G.A.GU.A.GACGC.GA.A.A.G.A..A.C..A.U..G.A.U.G.GG.C.-.-.-gcu...............ggug.C.C.C.A.U.C.A..U.G.U.U.C.U.U.UU.GCGCC.U.A...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2546495-2546435           GC.G.A.GU.A.GACGC.AA.A.A.G.C..C.C..C.C..G.A.C.A.CU.C.C.G.U.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.UU.GCGUC.U.G...UU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/3835265-3835204           GU.C.A.GC.A.GACGC.AA.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.C.G.A.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.UA.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/806173-806233             -C.G.A.GC.A.GUCGC.AA.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.G.C.G.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.UU.GCGAC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/755325-755265             GC.G.A.GC.A.GUCGC.AU.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.C.G.A.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/829766-829705             GC.G.A.GC.A.GACGC.AU.A.A.G.C..G.C..C.C..A.U.U.U.CG.G.C.A.C.......................G.A.C.A.U.U.G..G.G.C.G.C.U.U.AC.GCGUC.U.G...CC...C.G.C.CC\n",
      "\n",
      "NC_000962.3/1345183-1345243           CC.G.A.AC.A.GACAC.AA.A.A.C.C..A.C..C.C..U.A.U.U.UC.G.A.C.C.......................A.G.A.A.U.C.G..G.G.U.G.C.U.U.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/253604-253665             GC.G.A.GC.A.GACGC.AU.A.A.G.C..C.C..C.C..G.C.A.C.GC.A.C.-.G.......................G.C.G.U.GuC.G..A.G.G.G.C.U.U.UA.GUGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/3215656-3215595           GC.G.A.GC.A.GACGC.GA.A.A.U.C..C.C..C.C..G.A.C.A.CG.C.C.G.A.......................A.C.G.U.G.C.G..G.U.G.G.C.U.U.UU.GCGUC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2554925-2554864           GC.G.A.GC.A.GUCGC.AU.A.A.G.C..C.C..C.C..G.C.A.C.GC.A.C.-.G.......................G.C.G.U.GuC.G..G.G.G.G.C.U.U.UA.GCGAC.U.G...CU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/2546436-2546496           GC.G.A.AC.A.GACGC.AA.A.A.G.C..C.C..C.C..G.C.A.C.GC.A.C.G.G.......................A.G.U.G.U.C.G..G.G.G.G.C.U.U.UU.GCGUC.U.A...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/459447-459387             AC.G.A.GC.A.GACGC.GG.A.A.U.C..G.C..A.C..G.C.G.U.GA.G.G.-.-u...................cc.-.G.C.G.C.A.G..U.G.C.G.A.U.U.CC.UCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/3369888-3369948           GC.G.A.GC.A.GACAC.AG.A.A.U.C..G.C..A.C..G.G.G.C.GA.G.-.-.-ggg................cgc.-.-.-.C.G.G.A..U.G.C.G.A.G.U.CU.GUGUC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/2907752-2907812           CC.G.A.GC.A.GACAC.AG.A.A.U.C..G.C..A.C..G.A.A.A.UC.A.G.C.C.......................C.G.C.C.C.A.A..U.G.C.G.A.U.U.CU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/1497219-1497159           GC.G.A.GC.A.GACGC.AA.A.A.G.C..C.C..A.C..U.A.A.A.UC.G.G.C.C.......................A.G.A.U.U.U.A..G.G.A.G.C.U.U.UU.GCGUC.U.U...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/786085-786019             GC.G.A.GC.A.GACGC.AA.A.A.G.A..A.C..A.U..G.A.U.G.GG.C.-.-.-gcug.............gugcg.C.C.C.A.U.C.A..U.G.U.U.C.U.U.UU.GCGUC.U.G...GU...G.G.C.-C\n",
      "\n",
      "NC_000962.3/183291-183230             GC.G.A.GC.A.GACGC.AA.A.A.C.C..C.C..C.U..U.C.U.C.GU.C.G.G.C.......................G.U.G.U.C.G.A..G.G.G.A.C.U.U.AU.GCGUC.U.G...CU...A.G.C.GC\n",
      "\n",
      "NC_000962.3/2554866-2554926           GC.G.A.GC.A.GUCGC.UA.A.A.G.C..C.C..C.C..GaC.A.C.GC.-.C.G.U.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.AU.GCGAC.U.G...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/2931696-2931635           GU.G.A.GC.A.GUCGC.AA.A.A.G.C..A.C..C.C..G.U.U.U.CA.G.G.C.C.......................G.A.A.A.U.U.U..G.G.A.G.C.U.U.UU.GUGUC.U.G...CU...C.G.U.GC\n",
      "\n",
      "NC_000962.3/3387006-3387067           GU.G.A.GC.A.GGCGC.CA.A.G.G.C..C.C..C.C..C.A.C.G.CC.U.C.G.G.......................C.G.U.G.U.C.G..G.G.G.U.C.U.U.UU.GCGAC.U.G...CU...C.G.C.GC\n",
      "\n",
      "NC_000962.3/2075591-2075531           -C.G.A.GC.A.GACAC.AC.A.A.U.C..G.C..A.U..C.C.G.A.GG.A.G.C.C.......................U.G.C.C.G.C.G..U.G.C.G.A.U.U.CU.GUGUC.U.G...UU...C.G.C.GG\n",
      "\n",
      "NC_000962.3/1371709-1371768           --.G.A.GC.A.GACGC.AA.C.G.G.C..A.C..C.C..A.A.U.U.UC.G.A.C.C.......................A.G.A.A.U.C.G..G.G.U.U.C.U.U.UU.GCGUC.U.G...CU...C.A.C.GC\n",
      "\n",
      "NC_000962.3/3173158-3173094           GC.G.A.GU.A.GGCGC.AA.A.A.G.A..A.C..A.U..G.A.U.G.GG.C.-.-.-acc...............agcg.C.C.C.A.U.C.A..U.G.U.U.C.U.U.UC.GCGUC.U.A...CU...C.G.C.-C\n",
      "\n",
      "NC_000962.3/253663-253602             GC.G.A.GC.A.GACAC.UA.A.A.G.C..C.C..U.C..GaC.A.C.GC.-.C.G.U.......................G.C.G.U.G.C.G..G.G.G.G.C.U.U.AU.GCGUC.U.G...CU...C.G.C.CC\n",
      "\n",
      "NC_000962.3/2075533-2075593           GC.G.A.AC.A.GACAC.AG.A.A.U.C..G.C..A.C..G.C.G.G.CA.G.G.C.U.......................C.C.U.C.G.G.A..U.G.C.G.A.U.U.GU.GUGUC.U.G...CU...C.G.-.GC\n",
      "\n",
      "NC_000962.3/1371766-1371706           GU.G.A.GC.A.GACGC.AA.A.A.G.A..A.C..C.C..G.A.U.U.CU.G.G.U.C.......................G.A.A.A.U.U.G..G.G.U.G.C.C.G.UU.GCGUC.U.G...CU...C.G.-.GC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(r_scape_output_loc + '/' + 'Rv1324' + '/search_3.sto', 'r') as f:  \n",
    "                for l in f:\n",
    "                    if 'NC_000962.3' in l and not ('#' in l):\n",
    "                        print(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4bb27e-0170-486a-9076-6a2d1a14b721",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
