{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9c5181b7-3b50-4aca-84d8-e6ad809c14fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_run = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "898a25bb-a79e-4516-9f4d-7315c4699f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "from tqdm.auto import tqdm\n",
    "from Comparative_Analysis import Sequence_Analysis_Routines as sar\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from Comparative_Analysis import Alignment as align\n",
    "import random\n",
    "import copy\n",
    "from joblib import Parallel, delayed\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "muscle_exe = 'C:/Users/nicho/Muscle/muscle3.8.31_i86win32.exe'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a0b5574-5195-4364-a0c6-0f649d3f8d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'D:/Project_Data/Project_7'\n",
    "literature_datasets_dir = project_dir + '/Data_From_Publications'\n",
    "output_dir = project_dir + '/Output'\n",
    "refseq_dir = project_dir + '/NCBI_Dataset_Mycobacteria'\n",
    "num_cores = 8\n",
    "core_numbers = list(range(1, num_cores+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0450d244-1510-48ff-b251-40510173eee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = util.list_dirs(refseq_dir)\n",
    "reference_species = 'GCF_000195955.2'\n",
    "outgroup_species = 'GCF_000696675.2'\n",
    "#species_list = species_list[4:6]    #For testing\n",
    "species_list_excl_ref = [x for x in species_list if x!= reference_species]\n",
    "num_species = len(species_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "21eaf76c-ebec-4227-9ee2-0ee10c857a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_genome_record = next(SeqIO.parse(refseq_dir + '/'+reference_species+'/genomic.gbff', \"genbank\"))\n",
    "reference_sequence_length = len(str(ref_genome_record.seq))\n",
    "translator = util.Translator()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f307521b-c70e-44d1-a4c3-0a993e290497",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 26.28it/s]\n"
     ]
    }
   ],
   "source": [
    "reference_list = [[reference_species, str(ref_genome_record.seq)]]\n",
    "util.produce_fasta_file(reference_list, 'D:/H37Rv.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a8665e30-5cdc-4d28-9bb4-569344f3e950",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_blast_output(infile_loc, outfile_loc, top_hit_only = False):\n",
    "    blast_results = pd.read_csv(infile_loc, header = None)\n",
    "    blast_results.columns = ['query_ref', 'target_ref', 'query_length', 'subject_length', 'percent_identical_matches','alignment_length', 'number_mismatches', 'number_of_gap_openings', 'query_start_alignment', 'query_end_alignment', 'target_start_alignment', 'target_end_alignment', 'e_value', 'bit_score']\n",
    "    for i, r in blast_results.iterrows():\n",
    "        blast_results.at[i, 'query_species'] = '_'.join(r.query_ref.split('_')[0:2])\n",
    "        blast_results.at[i, 'target_species'] = '_'.join(r.target_ref.split('_')[0:2])\n",
    "    blast_results['query_species_name'] = blast_results['query_species'].map(names_dict)\n",
    "    blast_results['target_species_name'] = blast_results['target_species'].map(names_dict)\n",
    "    if top_hit_only == True:\n",
    "        blast_results = blast_results.loc[blast_results.groupby(['query_ref','target_species'])['bit_score'].idxmax()]\n",
    "    blast_results['species_count'] = blast_results.groupby('query_ref')['query_ref'].transform('size')\n",
    "    with open(outfile_loc, 'wb') as f:\n",
    "        pickle.dump(blast_results, f)\n",
    "    return blast_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a8c8ad6-b9be-4076-a34e-083317a6391e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def keep_reciprocal_best_hits(query_df, reverse_query_df, outfile_loc):\n",
    "    temp_1_dict = {}\n",
    "    temp_2_dict = {}\n",
    "    for i, r in query_df.iterrows():\n",
    "        temp_1_dict[r['query_ref']] = r['target_ref']\n",
    "    for i, r in reverse_query_df.iterrows():\n",
    "        temp_2_dict[r['query_ref']] = r['target_ref']\n",
    "    for i, r in query_df.iterrows():\n",
    "        if temp_1_dict[r['query_ref']] in temp_2_dict and temp_2_dict[temp_1_dict[r['query_ref']]] == r['query_ref']:\n",
    "            query_df.at[i, 'reciprocal_best_hit'] = 'Y'\n",
    "        else:\n",
    "            query_df.at[i, 'reciprocal_best_hit'] = 'N'\n",
    "    output = query_df[query_df.reciprocal_best_hit == 'Y'] \n",
    "    with open(outfile_loc, 'wb') as f:\n",
    "        pickle.dump(output, f)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2616d0d9-b446-472c-a198-5ced3637aa18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_myco_info(num_subsets, subset_num, species_master_list):\n",
    "    output = []\n",
    "    species_list = util.chunk_list(species_master_list, num_subsets, subset_num)\n",
    "    for species in species_list:\n",
    "        features = []\n",
    "        genome_record = next(SeqIO.parse(refseq_dir + '/'+species+'/genomic.gbff', \"genbank\"))\n",
    "        full_sequence = str(genome_record.seq)\n",
    "        if full_sequence.count('A') + full_sequence.count('C') + full_sequence.count('G') + full_sequence.count('T') < len(full_sequence):\n",
    "            continue\n",
    "        organism = genome_record.annotations['organism']\n",
    "        \n",
    "        #  Read feature information\n",
    "        if species == reference_species:\n",
    "            mycobrowser_df = pd.read_excel(literature_datasets_dir+'/Mycobrowser_Release_4.xlsx')\n",
    "            for i, r in mycobrowser_df.iterrows():\n",
    "                if r['Feature'] == 'CDS':\n",
    "                    if r['Strand'] == '+':\n",
    "                        strand = 1\n",
    "                    else:\n",
    "                        strand = -1\n",
    "                    features.append([r['Locus'],r['Start']-1, r['Stop'], strand])\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            for feature in genome_record.features:\n",
    "                    a = feature.qualifiers\n",
    "                    if feature.type == 'CDS' and a.get(\"locus_tag\")!= None and int(feature.location.end) - int(feature.location.start) < 100000:  #  Exclude strange Biopython parsing where starts with complement join and looks like a CDS is full length of genome!   \n",
    "                        locus_tag = a.get(\"locus_tag\")[0]\n",
    "                        features.append([locus_tag, int(feature.location.start), int(feature.location.end), int(feature.location.strand)])\n",
    "        \n",
    "        features.sort(key=lambda x: x[1])\n",
    "        \n",
    "        feature_info = []\n",
    "        for i, feature in enumerate(features):\n",
    "            if feature[1] < feature[2]:  \n",
    "                if feature[3] == 1:\n",
    "                    cds_nt_sequence = full_sequence[feature[1]:feature[2]]\n",
    "                else:\n",
    "                    cds_nt_sequence = util.reverse_complement(full_sequence[feature[1]:feature[2]])\n",
    "                cds_aa_sequence = translator.translate_sequence(full_sequence[feature[1]:feature[2]],feature[3], 0)                                                                                           \n",
    "                if (i + 1)< len(features) and feature[3] == 1 and feature[2] < features[i+1][1]:\n",
    "                    feature_info.append([species, feature[0], cds_aa_sequence, full_sequence[feature[2]: features[i+1][1]], feature, cds_nt_sequence])\n",
    "                elif (i > 0) and feature[3] == -1 and features[i-1][2] < feature[1]:\n",
    "                    feature_info.append([species, feature[0], cds_aa_sequence, util.reverse_complement(full_sequence[features[i-1][2]: feature[1]]), feature, cds_nt_sequence])\n",
    "                else:\n",
    "                    feature_info.append([species, feature[0], cds_aa_sequence, '', feature, cds_nt_sequence])\n",
    "\n",
    "        output.append((species, organism, feature_info))\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "40d5262c-20f8-496f-a241-2d60c488e767",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    myco_info_dict = {}\n",
    "    protein_info_dict = {}\n",
    "    names_dict = {}\n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(generate_myco_info)(num_cores, core_number, species_list) for core_number in core_numbers)\n",
    "    for core_output in parallel_output:\n",
    "        for results in core_output:\n",
    "            myco_info_dict[results[0]] = (results[1], results[2])\n",
    "            for feature in results[2]:\n",
    "                protein_info_dict[feature[0] + '_' + feature[1]] = feature\n",
    "            names_dict[results[0]] = results[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "601b1f9b-8303-4beb-adb2-0b4a0941fd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 412288/412288 [00:01<00:00, 261535.26it/s]\n",
      "100%|██████████| 4031/4031 [00:00<00:00, 236603.87it/s]\n",
      "100%|██████████| 4031/4031 [00:00<00:00, 100779.90it/s]\n",
      "100%|██████████| 412288/412288 [00:03<00:00, 107365.61it/s]\n",
      "100%|██████████| 2544/2544 [00:00<00:00, 424131.86it/s]\n",
      "100%|██████████| 244844/244844 [00:00<00:00, 466573.70it/s]\n"
     ]
    }
   ],
   "source": [
    "if full_run == True:\n",
    "    comparison_protein_list = []\n",
    "    reference_nt_list = []\n",
    "    reference_protein_list = []\n",
    "    reference_utr_list = []\n",
    "    comparison_utr_list = []\n",
    "    comparison_nt_list = []\n",
    "    for species in species_list:\n",
    "        for feature_info in myco_info_dict[species][1]:\n",
    "            comparison_protein_list.append([feature_info[0]+'_'+feature_info[1],feature_info[2][:-1]])\n",
    "            comparison_nt_list.append([feature_info[0]+'_'+feature_info[1],feature_info[5]])\n",
    "            if len(feature_info[3]) > 8:\n",
    "                comparison_utr_list.append([feature_info[0]+'_'+feature_info[1],feature_info[3]])  \n",
    "    for feature_info in myco_info_dict[reference_species][1]:\n",
    "            reference_protein_list.append([feature_info[0]+'_'+feature_info[1],feature_info[2][:-1]])\n",
    "            reference_nt_list.append([feature_info[0]+'_'+feature_info[1],feature_info[5]])\n",
    "            if len(feature_info[3]) > 8:\n",
    "                reference_utr_list.append([feature_info[0]+'_'+feature_info[1],feature_info[3]])\n",
    "    util.produce_fasta_file(comparison_protein_list, 'D:/BLAST/comp_prot/comparison_proteins.faa')\n",
    "    util.produce_fasta_file(reference_protein_list, 'D:/BLAST/ref_prot/reference_proteins.faa')\n",
    "    util.produce_fasta_file(reference_nt_list, 'D:/BLAST/ref_nt/reference_nt.faa')\n",
    "    util.produce_fasta_file(comparison_nt_list, 'D:/BLAST/comp_nt/comparison_nt.faa')\n",
    "    util.produce_fasta_file(reference_utr_list, 'D:/BLAST/ref_utr/reference_utr.faa')\n",
    "    util.produce_fasta_file(comparison_utr_list, 'D:/BLAST/comp_utr/comparison_utr.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3b888e74-a7c0-4a4b-90a8-e27d6215257a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    w_d = os.getcwd()\n",
    "    os.chdir(\"D:/\")\n",
    "    subprocess.run('cd D:/BLAST/comp_prot &  makeblastdb -in comparison_proteins.faa -dbtype prot -out d://BLAST//comp_prot//comp_prot', shell=True, capture_output = True)\n",
    "    subprocess.run('cd D:/BLAST/ref_prot &  makeblastdb -in reference_proteins.faa -dbtype prot -out d://BLAST//ref_prot//ref_prot', shell=True, capture_output = True)\n",
    "    subprocess.run('cd D:/BLAST/ref_nt &  makeblastdb -in reference_nt.faa -dbtype nucl -out d://BLAST//ref_nt//ref_nt', shell=True, capture_output = True)\n",
    "    subprocess.run('cd D:/BLAST/comp_nt &  makeblastdb -in comparison_nt.faa -dbtype nucl -out d://BLAST//comp_nt//comp_nt', shell=True, capture_output = True)\n",
    "    subprocess.run('cd D:/BLAST/comp_utr &  makeblastdb -in comparison_utr.faa -dbtype nucl -out d://BLAST//comp_utr//comp_utr', shell=True, capture_output = True)\n",
    "    os.chdir(w_d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6d726ba4-9d93-4ba4-964b-97f5265233ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    w_d = os.getcwd()\n",
    "    os.chdir(\"D:/\")\n",
    "    subprocess.run('cd d:\\\\BLAST\\\\comp_prot & blastp -query D:/BLAST/ref_prot/reference_proteins.faa -db comp_prot -out ref_comp_hits.csv -evalue 1e-10 -seg no -outfmt  \"10 qaccver saccver qlen slen pident length mismatch gapopen qstart qend sstart send evalue bitscore\" -num_threads 16', shell=True, capture_output = True)\n",
    "    subprocess.run('cd d:\\\\BLAST\\\\ref_prot & blastp -query D:/BLAST/comp_prot/comparison_proteins.faa -db ref_prot -out comp_ref_hits.csv -evalue 1e-10 -seg no -outfmt  \"10 qaccver saccver qlen slen pident length mismatch gapopen qstart qend sstart send evalue bitscore\" -num_threads 16', shell=True, capture_output = True)\n",
    "    subprocess.run('cd d:\\\\BLAST\\\\ref_nt & blastn -query D:/BLAST/ref_utr/reference_utr.faa -db ref_nt -out utr_ref_hits.csv -evalue 1e-10 -outfmt  \"10 qaccver saccver qlen slen pident length mismatch gapopen qstart qend sstart send evalue bitscore\" -num_threads 16', shell=True, capture_output = True)\n",
    "    subprocess.run('cd d:\\\\BLAST\\\\comp_nt & blastn -query D:/BLAST/comp_utr/comparison_utr.faa -db comp_nt -out utr_comp_hits.csv -evalue 1e-10 -outfmt  \"10 qaccver saccver qlen slen pident length mismatch gapopen qstart qend sstart send evalue bitscore\" -num_threads 16', shell=True, capture_output = True)\n",
    "    subprocess.run('cd d:\\\\BLAST\\\\comp_utr & blastn -query D:/BLAST/comp_utr/comparison_utr.faa -db comp_utr -out utr_all_comp_hits.csv -evalue 1e-10 -outfmt  \"10 qaccver saccver qlen slen pident length mismatch gapopen qstart qend sstart send evalue bitscore\" -num_threads 16', shell=True, capture_output = True)\n",
    " \n",
    "    os.chdir(w_d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "55b4d7d6-b618-4c76-a4f6-0b88ecea0ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    blast_results_rc = process_blast_output('D:\\\\BLAST\\\\comp_prot\\\\ref_comp_hits.csv', project_dir + '/blast_results_rc.pkl', True)\n",
    "    blast_results_cr = process_blast_output('D:\\\\BLAST\\\\ref_prot\\\\comp_ref_hits.csv', project_dir + '/blast_results_cr.pkl', True)\n",
    "    blast_results_utr_ref = process_blast_output('D:\\\\BLAST\\\\ref_nt\\\\utr_ref_hits.csv', project_dir + '/utr_results_ref.pkl', False)\n",
    "    blast_results_comp_ref = process_blast_output('D:\\\\BLAST\\\\comp_nt\\\\utr_comp_hits.csv', project_dir + '/utr_results_comp.pkl', False)\n",
    "    blast_results_utr_utr = process_blast_output('D:\\\\BLAST\\\\comp_utr\\\\utr_all_comp_hits.csv', project_dir + '/comp_results_utr_utr.pkl', False)\n",
    "    rbh_results = keep_reciprocal_best_hits(blast_results_rc, blast_results_cr, project_dir + '/rbh_results_temp.pkl')\n",
    "    rbh_results['query_info'] = rbh_results['query_ref'].map(protein_info_dict)\n",
    "    rbh_results['target_info'] = rbh_results['target_ref'].map(protein_info_dict)\n",
    "    organism_names = rbh_results[['target_species','target_species_name']].drop_duplicates().reset_index(drop=True)\n",
    "    organism_dict = {}\n",
    "    for i, r in organism_names.iterrows():\n",
    "        if 'BCG' in r['target_species_name']:\n",
    "            organism_names.at[i,'sname'] = 'M.bovis_BCG'\n",
    "        elif 'AF2122' in r['target_species_name']:\n",
    "            organism_names.at[i,'sname'] = 'M.bovis_AF2122/97'\n",
    "        else:\n",
    "            organism_names.at[i,'sname'] = r['target_species_name'][0] + '.' + r['target_species_name'].split(' ')[1]\n",
    "    for i, r in organism_names.iterrows():\n",
    "        organism_dict[r['target_species']] = r['sname']\n",
    "    rbh_results['target_species_sname'] = rbh_results['target_species'].map(organism_dict)\n",
    "    for i, r in rbh_results.iterrows():\n",
    "        rbh_results.at[i,'query_start'] = r['query_info'][4][1]\n",
    "        rbh_results.at[i,'target_start'] = r['target_info'][4][1]\n",
    "        rbh_results.at[i,'target_utr'] = r['target_info'][3]\n",
    "        rbh_results.at[i,'target_cds'] = r['target_info'][2][:-1]\n",
    "    blast_results_rc.to_csv(project_dir + '/blast_results_rc.csv')\n",
    "    blast_results_cr.to_csv(project_dir + '/blast_results_cr.csv')   \n",
    "    blast_results_utr_ref.to_csv(project_dir + '/blast_results_utr_ref.csv') \n",
    "    blast_results_comp_ref.to_csv(project_dir + '/blast_results_comp_ref.csv') \n",
    "    blast_results_utr_utr.to_csv(project_dir + '/blast_results_utr_utr.csv') \n",
    "    with open(project_dir + '/rbh_results.pkl', 'wb') as f:\n",
    "        pickle.dump(rbh_results, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1ff56aae-5031-4ff6-8f56-e6b570bad030",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results_utr_utr = process_blast_output('D:\\\\BLAST\\\\comp_utr\\\\utr_all_comp_hits.csv', project_dir + '/comp_results_utr_utr.pkl', False)\n",
    "blast_results_utr_utr.to_csv(project_dir + '/blast_results_utr_utr.csv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4f1f140b-2e98-4ca7-b29d-0fa3524a09d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_refs = rbh_results[['query_ref']].drop_duplicates().reset_index(drop=True)['query_ref'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d48b8796-324d-42e4-9095-62e028f1392c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_alignments(num_subsets, subset_num, field_name, directory, ref_master_list):\n",
    "    ref_list = util.chunk_list(ref_master_list, num_subsets, subset_num)\n",
    "    for ref in ref_list:\n",
    "        temp_df = rbh_results[rbh_results.query_ref == ref]\n",
    "        seq_list = []\n",
    "        for i, r in temp_df.iterrows():\n",
    "            if len(r[field_name]) > 0:\n",
    "                seq_list.append([r['target_species_sname'], r[field_name]])\n",
    "        util.produce_fasta_file(seq_list, project_dir+'/testseq_'+str(subset_num)+'.fasta')\n",
    "        cline = MuscleCommandline(muscle_exe, input=project_dir+'/testseq_'+str(subset_num)+'.fasta', out=directory + '/All/align_'+ ref +'.fasta')\n",
    "        try:\n",
    "            stdout, stderr = cline()\n",
    "        except Exception as e:\n",
    "            continue\n",
    "        temp = util.read_fasta_to_array(directory + '/All/align_'+ ref +'.fasta', species_order = [])\n",
    "        num_sequences = len(temp[0])\n",
    "        if num_sequences == num_species:\n",
    "            shutil.copyfile(directory + '/All/align_'+ ref +'.fasta', directory +'/Full_Ortholog/align_'+ ref +'.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0118534b-4a82-4cda-9064-90115486049b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    Parallel(n_jobs=-1)(delayed(run_alignments)(num_cores, core_number, 'target_cds', project_dir + '/CDS_Alignments', query_refs) for core_number in core_numbers)\n",
    "    Parallel(n_jobs=-1)(delayed(run_alignments)(num_cores, core_number, 'target_utr', project_dir + '/UTR_Alignments', query_refs) for core_number in core_numbers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7d20d3b-7240-47e7-9e67-b702f0326e83",
   "metadata": {},
   "source": [
    "##### Build tree from full orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "119b9166-5ff9-4e13-8a47-c22edf3c528b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='cd \\\\users\\\\nicho\\\\IQTree & bin\\\\iqtree2 -q D:/Project_Data/Project_7/CDS_Alignments/Full_Ortholog/ --prefix D:/Project_Data/Project_7/CDS_Alignments/Full_Ortholog_Tree/CDS_Full_Ortholog_Tree -m LG -B 1000 -T AUTO -o R.erythropolis', returncode=0)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run('cd \\\\users\\\\nicho\\\\IQTree & bin\\\\iqtree2 -q ' + project_dir + '/CDS_Alignments/Full_Ortholog/' + ' --prefix '+ project_dir + '/CDS_Alignments/Full_Ortholog_Tree/CDS_Full_Ortholog_Tree -m LG -B 1000 -T AUTO -o ' + organism_dict['GCF_000696675.2'], shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09b7744e-9799-4915-b477-8b25bff5625c",
   "metadata": {},
   "source": [
    "###### Insert blank sequences for display and delete gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4a87f394-00bb-4c34-b007-b4bd7286333e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 78/78 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "temp = util.read_fasta_to_array(project_dir+'/UTR_Alignments/All/align_GCF_000195955.2_Rv0044c.fasta', species_order = [])\n",
    "sequence_length = len(temp[1][0])\n",
    "blank_seq = '-' * sequence_length\n",
    "sequence_names = temp[0]\n",
    "sequences = temp[1]\n",
    "for k, v in organism_dict.items():\n",
    "    if not(v in sequence_names):\n",
    "        sequence_names.append(v)\n",
    "        sequences.append(blank_seq)\n",
    "for i, name in enumerate(sequence_names):\n",
    "    if name == organism_dict[reference_species]:\n",
    "        ref_species_index = i\n",
    "ref_insert_positions = []\n",
    "for i, letter in enumerate(sequences[ref_species_index]):\n",
    "    if letter == '-':\n",
    "        ref_insert_positions.append(i)\n",
    "insert_deleted_sequences = []\n",
    "for sequence in sequences:\n",
    "    temp_letter_list = []\n",
    "    for i,letter in enumerate(sequence):\n",
    "        if i in ref_insert_positions:\n",
    "            continue\n",
    "        else:\n",
    "            temp_letter_list.append(sequence[i])\n",
    "    insert_deleted_sequences.append(''.join(temp_letter_list))\n",
    "sequence_info = []\n",
    "for (sequence_name, sequence) in zip(sequence_names, insert_deleted_sequences):\n",
    "    sequence_info.append([sequence_name, sequence])\n",
    "util.produce_fasta_file(sequence_info,project_dir+ '/UTR_Alignments/All_Deleted_Gaps/align_GCF_000195955.2_Rv0044c.fasta')            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a1d755a3-2e89-493a-a61b-7a7774fd32e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.call ('C:/\"Program Files\"/R/R-4.1.2/bin/x64/Rscript --vanilla D:/Project_Data/Project_7/R_Scripts/Motif_Plots.R', shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e068072d-891d-43ce-95cd-bc7b16bd291a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(12, 12))\n",
    "g = sns.FacetGrid(rbh_results, col='target_species_name', height=5, col_wrap=3)\n",
    "g.map(sns.scatterplot, 'target_start', 'query_start', s=2)\n",
    "g.set_titles(row_template = '{row_name}', col_template = '{col_name}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "0c5e15f5-a9e5-400a-9301-cde2a44691be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'R.erythropolis'"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism_dict['GCF_000696675.2']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d09197f-3410-47f5-bab2-e4251ee5664d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#####  Run Infernal/r-scape process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d26d2487-f555-43d2-bf73-8b650c2521fb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_ref</th>\n",
       "      <th>target_ref</th>\n",
       "      <th>query_length</th>\n",
       "      <th>subject_length</th>\n",
       "      <th>percent_identical_matches</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>number_mismatches</th>\n",
       "      <th>number_of_gap_openings</th>\n",
       "      <th>query_start_alignment</th>\n",
       "      <th>query_end_alignment</th>\n",
       "      <th>...</th>\n",
       "      <th>target_species_name</th>\n",
       "      <th>species_count</th>\n",
       "      <th>reciprocal_best_hit</th>\n",
       "      <th>query_info</th>\n",
       "      <th>target_info</th>\n",
       "      <th>target_species_sname</th>\n",
       "      <th>query_start</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_utr</th>\n",
       "      <th>target_cds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000195955.2_Rv0001</td>\n",
       "      <td>GCF_000009445.1_BCG_RS00165</td>\n",
       "      <td>507</td>\n",
       "      <td>507</td>\n",
       "      <td>99.803</td>\n",
       "      <td>507</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycobacterium tuberculosis variant bovis BCG s...</td>\n",
       "      <td>74</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...</td>\n",
       "      <td>[GCF_000009445.1, BCG_RS00165, LTDDPGSGFTTVWNA...</td>\n",
       "      <td>M.bovis_BCG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29667.0</td>\n",
       "      <td>CACGGCGTGTTCTTCCGACAACGTTCTTAAAAAAACTTCTCTCTCC...</td>\n",
       "      <td>LTDDPGSGFTTVWNAVVSELNGDPKVDDGPSSDANLSAPLTPQQRA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GCF_000195955.2_Rv0001</td>\n",
       "      <td>GCF_000015305.1_MVAN_RS00015</td>\n",
       "      <td>507</td>\n",
       "      <td>494</td>\n",
       "      <td>78.698</td>\n",
       "      <td>507</td>\n",
       "      <td>95</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycolicibacterium vanbaalenii PYR-1</td>\n",
       "      <td>74</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...</td>\n",
       "      <td>[GCF_000015305.1, MVAN_RS00015, LTTDPDPPFVSIWD...</td>\n",
       "      <td>M.vanbaalenii</td>\n",
       "      <td>0.0</td>\n",
       "      <td>248.0</td>\n",
       "      <td>CGGTCAGCTCACCCCACCCTGTTTCCTGACACCTCGCCGCGCCGGC...</td>\n",
       "      <td>LTTDPDPPFVSIWDNVVTELNGAGEVGNGSLTPQQRAWLKLVKPLV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GCF_000195955.2_Rv0001</td>\n",
       "      <td>GCF_000157895.3_MKAN_RS14130</td>\n",
       "      <td>507</td>\n",
       "      <td>514</td>\n",
       "      <td>86.576</td>\n",
       "      <td>514</td>\n",
       "      <td>62</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycobacterium kansasii ATCC 12478</td>\n",
       "      <td>74</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...</td>\n",
       "      <td>[GCF_000157895.3, MKAN_RS14130, LTDDPGSGFASVWN...</td>\n",
       "      <td>M.kansasii</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3239626.0</td>\n",
       "      <td>ACCCGGTGCATTTCCGGTCGCCTTCTGCAAAAAACTTCTCTCCCGC...</td>\n",
       "      <td>LTDDPGSGFASVWNAVVSELNGELNADGIAHTGTTLTTPLTPQQRA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>GCF_000195955.2_Rv0001</td>\n",
       "      <td>GCF_000184435.1_MSPYR1_RS00010</td>\n",
       "      <td>507</td>\n",
       "      <td>490</td>\n",
       "      <td>78.304</td>\n",
       "      <td>507</td>\n",
       "      <td>93</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycolicibacterium gilvum Spyr1</td>\n",
       "      <td>74</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...</td>\n",
       "      <td>[GCF_000184435.1, MSPYR1_RS00010, LTTDPDPPFVAI...</td>\n",
       "      <td>M.gilvum</td>\n",
       "      <td>0.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>GCCGAACCCCCGTATCCCTCACCCCGTGCGCTGCCCCCGTTGGGCG...</td>\n",
       "      <td>LTTDPDPPFVAIWDNVVTELNGTGGTQNGSLTPQQRAWLKLVRPLV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000195955.2_Rv0001</td>\n",
       "      <td>GCF_000195955.2_Rv0001</td>\n",
       "      <td>507</td>\n",
       "      <td>507</td>\n",
       "      <td>100.000</td>\n",
       "      <td>507</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>74</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...</td>\n",
       "      <td>[GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...</td>\n",
       "      <td>M.tuberculosis</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CACGGCGTGTTCTTCCGACAACGTTCTTAAAAAAACTTCTCTCTCC...</td>\n",
       "      <td>LTDDPGSGFTTVWNAVVSELNGDPKVDDGPSSDANLSAPLTPQQRA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705716</th>\n",
       "      <td>GCF_000195955.2_Rv3924c</td>\n",
       "      <td>GCF_900292015.1_MB901379_RS23850</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>93.617</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycobacterium basiliense</td>\n",
       "      <td>78</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...</td>\n",
       "      <td>[GCF_900292015.1, MB901379_RS23850, VAKGKRTFQP...</td>\n",
       "      <td>M.basiliense</td>\n",
       "      <td>4410785.0</td>\n",
       "      <td>5606962.0</td>\n",
       "      <td>TCGTAACGGCGGGTTTTGCAGCG</td>\n",
       "      <td>VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRSIVSSRRRKGRRALTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705781</th>\n",
       "      <td>GCF_000195955.2_Rv3924c</td>\n",
       "      <td>GCF_900603025.1_MHAS_RS21010</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>85.106</td>\n",
       "      <td>47</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycolicibacterium hassiacum DSM 44199</td>\n",
       "      <td>78</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...</td>\n",
       "      <td>[GCF_900603025.1, MHAS_RS21010, VAKGKRTFQPNNRR...</td>\n",
       "      <td>M.hassiacum</td>\n",
       "      <td>4410785.0</td>\n",
       "      <td>4488368.0</td>\n",
       "      <td>TCGCCGGCTAGATCCGGGCAGGGCCGGCACG</td>\n",
       "      <td>VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRAIVANRRAKGRRRLTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705743</th>\n",
       "      <td>GCF_000195955.2_Rv3924c</td>\n",
       "      <td>GCF_900637205.1_EL338_RS26040</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>87.234</td>\n",
       "      <td>47</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycolicibacterium chitae</td>\n",
       "      <td>78</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...</td>\n",
       "      <td>[GCF_900637205.1, EL338_RS26040, VAKGKRTFQPNNR...</td>\n",
       "      <td>M.chitae</td>\n",
       "      <td>4410785.0</td>\n",
       "      <td>5460869.0</td>\n",
       "      <td>GTCTGCTCGGGCGCCTGAC</td>\n",
       "      <td>VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRAIVTGRRRKGRRSLTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705708</th>\n",
       "      <td>GCF_000195955.2_Rv3924c</td>\n",
       "      <td>LT_708304.1_BQ2027_MB3955C</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>100.000</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycobacterium tuberculosis variant bovis AF212...</td>\n",
       "      <td>78</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...</td>\n",
       "      <td>[LT_708304.1, BQ2027_MB3955C, VTKGKRTFQPNNRRRA...</td>\n",
       "      <td>M.bovis_AF2122/97</td>\n",
       "      <td>4410785.0</td>\n",
       "      <td>4349157.0</td>\n",
       "      <td></td>\n",
       "      <td>VTKGKRTFQPNNRRRARVHGFRLRMRTRAGRSIVSSRRRKGRRTLSA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705772</th>\n",
       "      <td>GCF_000195955.2_Rv3924c</td>\n",
       "      <td>NC_008596.1_MSMEG_RS33455</td>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "      <td>85.106</td>\n",
       "      <td>47</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycolicibacterium smegmatis MC2 155</td>\n",
       "      <td>78</td>\n",
       "      <td>Y</td>\n",
       "      <td>[GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...</td>\n",
       "      <td>[NC_008596.1, MSMEG_RS33455, VAKGKRTFQPNNRRRAR...</td>\n",
       "      <td>M.smegmatis</td>\n",
       "      <td>4410785.0</td>\n",
       "      <td>6985942.0</td>\n",
       "      <td>TCCTGCGCCGGCACAGGCAGGATCCAGCGCG</td>\n",
       "      <td>VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRAIVANRRSKGRRALTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208547 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      query_ref                        target_ref  \\\n",
       "2        GCF_000195955.2_Rv0001       GCF_000009445.1_BCG_RS00165   \n",
       "57       GCF_000195955.2_Rv0001      GCF_000015305.1_MVAN_RS00015   \n",
       "20       GCF_000195955.2_Rv0001      GCF_000157895.3_MKAN_RS14130   \n",
       "50       GCF_000195955.2_Rv0001    GCF_000184435.1_MSPYR1_RS00010   \n",
       "1        GCF_000195955.2_Rv0001            GCF_000195955.2_Rv0001   \n",
       "...                         ...                               ...   \n",
       "705716  GCF_000195955.2_Rv3924c  GCF_900292015.1_MB901379_RS23850   \n",
       "705781  GCF_000195955.2_Rv3924c      GCF_900603025.1_MHAS_RS21010   \n",
       "705743  GCF_000195955.2_Rv3924c     GCF_900637205.1_EL338_RS26040   \n",
       "705708  GCF_000195955.2_Rv3924c        LT_708304.1_BQ2027_MB3955C   \n",
       "705772  GCF_000195955.2_Rv3924c         NC_008596.1_MSMEG_RS33455   \n",
       "\n",
       "        query_length  subject_length  percent_identical_matches  \\\n",
       "2                507             507                     99.803   \n",
       "57               507             494                     78.698   \n",
       "20               507             514                     86.576   \n",
       "50               507             490                     78.304   \n",
       "1                507             507                    100.000   \n",
       "...              ...             ...                        ...   \n",
       "705716            47              47                     93.617   \n",
       "705781            47              47                     85.106   \n",
       "705743            47              47                     87.234   \n",
       "705708            47              47                    100.000   \n",
       "705772            47              47                     85.106   \n",
       "\n",
       "        alignment_length  number_mismatches  number_of_gap_openings  \\\n",
       "2                    507                  1                       0   \n",
       "57                   507                 95                       2   \n",
       "20                   514                 62                       3   \n",
       "50                   507                 93                       3   \n",
       "1                    507                  0                       0   \n",
       "...                  ...                ...                     ...   \n",
       "705716                47                  3                       0   \n",
       "705781                47                  7                       0   \n",
       "705743                47                  6                       0   \n",
       "705708                47                  0                       0   \n",
       "705772                47                  7                       0   \n",
       "\n",
       "        query_start_alignment  query_end_alignment  ...  \\\n",
       "2                           1                  507  ...   \n",
       "57                          1                  507  ...   \n",
       "20                          1                  507  ...   \n",
       "50                          1                  507  ...   \n",
       "1                           1                  507  ...   \n",
       "...                       ...                  ...  ...   \n",
       "705716                      1                   47  ...   \n",
       "705781                      1                   47  ...   \n",
       "705743                      1                   47  ...   \n",
       "705708                      1                   47  ...   \n",
       "705772                      1                   47  ...   \n",
       "\n",
       "                                      target_species_name  species_count  \\\n",
       "2       Mycobacterium tuberculosis variant bovis BCG s...             74   \n",
       "57                    Mycolicibacterium vanbaalenii PYR-1             74   \n",
       "20                      Mycobacterium kansasii ATCC 12478             74   \n",
       "50                         Mycolicibacterium gilvum Spyr1             74   \n",
       "1                        Mycobacterium tuberculosis H37Rv             74   \n",
       "...                                                   ...            ...   \n",
       "705716                           Mycobacterium basiliense             78   \n",
       "705781              Mycolicibacterium hassiacum DSM 44199             78   \n",
       "705743                           Mycolicibacterium chitae             78   \n",
       "705708  Mycobacterium tuberculosis variant bovis AF212...             78   \n",
       "705772                Mycolicibacterium smegmatis MC2 155             78   \n",
       "\n",
       "        reciprocal_best_hit  \\\n",
       "2                         Y   \n",
       "57                        Y   \n",
       "20                        Y   \n",
       "50                        Y   \n",
       "1                         Y   \n",
       "...                     ...   \n",
       "705716                    Y   \n",
       "705781                    Y   \n",
       "705743                    Y   \n",
       "705708                    Y   \n",
       "705772                    Y   \n",
       "\n",
       "                                               query_info  \\\n",
       "2       [GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...   \n",
       "57      [GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...   \n",
       "20      [GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...   \n",
       "50      [GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...   \n",
       "1       [GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...   \n",
       "...                                                   ...   \n",
       "705716  [GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...   \n",
       "705781  [GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...   \n",
       "705743  [GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...   \n",
       "705708  [GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...   \n",
       "705772  [GCF_000195955.2, Rv3924c, VTKGKRTFQPNNRRRARVH...   \n",
       "\n",
       "                                              target_info  \\\n",
       "2       [GCF_000009445.1, BCG_RS00165, LTDDPGSGFTTVWNA...   \n",
       "57      [GCF_000015305.1, MVAN_RS00015, LTTDPDPPFVSIWD...   \n",
       "20      [GCF_000157895.3, MKAN_RS14130, LTDDPGSGFASVWN...   \n",
       "50      [GCF_000184435.1, MSPYR1_RS00010, LTTDPDPPFVAI...   \n",
       "1       [GCF_000195955.2, Rv0001, LTDDPGSGFTTVWNAVVSEL...   \n",
       "...                                                   ...   \n",
       "705716  [GCF_900292015.1, MB901379_RS23850, VAKGKRTFQP...   \n",
       "705781  [GCF_900603025.1, MHAS_RS21010, VAKGKRTFQPNNRR...   \n",
       "705743  [GCF_900637205.1, EL338_RS26040, VAKGKRTFQPNNR...   \n",
       "705708  [LT_708304.1, BQ2027_MB3955C, VTKGKRTFQPNNRRRA...   \n",
       "705772  [NC_008596.1, MSMEG_RS33455, VAKGKRTFQPNNRRRAR...   \n",
       "\n",
       "       target_species_sname query_start target_start  \\\n",
       "2               M.bovis_BCG         0.0      29667.0   \n",
       "57            M.vanbaalenii         0.0        248.0   \n",
       "20               M.kansasii         0.0    3239626.0   \n",
       "50                 M.gilvum         0.0        243.0   \n",
       "1            M.tuberculosis         0.0          0.0   \n",
       "...                     ...         ...          ...   \n",
       "705716         M.basiliense   4410785.0    5606962.0   \n",
       "705781          M.hassiacum   4410785.0    4488368.0   \n",
       "705743             M.chitae   4410785.0    5460869.0   \n",
       "705708    M.bovis_AF2122/97   4410785.0    4349157.0   \n",
       "705772          M.smegmatis   4410785.0    6985942.0   \n",
       "\n",
       "                                               target_utr  \\\n",
       "2       CACGGCGTGTTCTTCCGACAACGTTCTTAAAAAAACTTCTCTCTCC...   \n",
       "57      CGGTCAGCTCACCCCACCCTGTTTCCTGACACCTCGCCGCGCCGGC...   \n",
       "20      ACCCGGTGCATTTCCGGTCGCCTTCTGCAAAAAACTTCTCTCCCGC...   \n",
       "50      GCCGAACCCCCGTATCCCTCACCCCGTGCGCTGCCCCCGTTGGGCG...   \n",
       "1       CACGGCGTGTTCTTCCGACAACGTTCTTAAAAAAACTTCTCTCTCC...   \n",
       "...                                                   ...   \n",
       "705716                            TCGTAACGGCGGGTTTTGCAGCG   \n",
       "705781                    TCGCCGGCTAGATCCGGGCAGGGCCGGCACG   \n",
       "705743                                GTCTGCTCGGGCGCCTGAC   \n",
       "705708                                                      \n",
       "705772                    TCCTGCGCCGGCACAGGCAGGATCCAGCGCG   \n",
       "\n",
       "                                               target_cds  \n",
       "2       LTDDPGSGFTTVWNAVVSELNGDPKVDDGPSSDANLSAPLTPQQRA...  \n",
       "57      LTTDPDPPFVSIWDNVVTELNGAGEVGNGSLTPQQRAWLKLVKPLV...  \n",
       "20      LTDDPGSGFASVWNAVVSELNGELNADGIAHTGTTLTTPLTPQQRA...  \n",
       "50      LTTDPDPPFVAIWDNVVTELNGTGGTQNGSLTPQQRAWLKLVRPLV...  \n",
       "1       LTDDPGSGFTTVWNAVVSELNGDPKVDDGPSSDANLSAPLTPQQRA...  \n",
       "...                                                   ...  \n",
       "705716    VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRSIVSSRRRKGRRALTA  \n",
       "705781    VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRAIVANRRAKGRRRLTA  \n",
       "705743    VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRAIVTGRRRKGRRSLTA  \n",
       "705708    VTKGKRTFQPNNRRRARVHGFRLRMRTRAGRSIVSSRRRKGRRTLSA  \n",
       "705772    VAKGKRTFQPNNRRRARVHGFRLRMRTRAGRAIVANRRSKGRRALTA  \n",
       "\n",
       "[208547 rows x 27 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rbh_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "262d7a7a-c3a7-45b7-b4de-646b7e06e2a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project_dir = 'D:/Project_Data/Project_7'\n",
    "dataset_loc = project_dir + '/NCBI_Dataset_Corynebacteriales'\n",
    "r_scape_output_loc = project_dir + '/R_Scape_Results_2'\n",
    "merged_file_loc = dataset_loc + '/merged_file.txt'\n",
    "intergenic_alignment_loc = project_dir + '/UTR_Alignments/Full_Ortholog'\n",
    "wsl_merged_file_loc = util.wslname(merged_file_loc)\n",
    "\n",
    "#This option creates a file with all genome sequence but may be slow and also open up too many possibilities for alignments - other option generates list from UTRs of orthologues in loop below\n",
    "if 1 == 0:\n",
    "    with open(merged_file_loc, 'w') as outfile:\n",
    "        for dir in util.list_dirs(dataset_loc):\n",
    "            directory = dataset_loc + '/' + dir\n",
    "            for file in util.list_files(directory):\n",
    "                if file.endswith(\"genomic.fna\"):\n",
    "                    with open(directory + '/' + file, encoding=\"utf-8\", errors='ignore') as infile:\n",
    "                        outfile.write(infile.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c1717346-dc76-48e9-b0d4-72858d079f09",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment_ids = ['.'.join(x.split('.')[:-1]) for x in util.list_files(intergenic_alignment_loc)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9782e874-d94a-4dca-9c8b-6536c4b8a61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment_ids = [alignment_ids[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "54ef5a64-373b-4063-90ba-9b7e021bb6da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['align_GCF_000195955.2_Rv0004']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dda2d166-3f28-461f-9fc3-97ecf4b3b6df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1a2b28263e374c7092ac55947576e968",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/28 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "0it [00:00, ?it/s]\u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'D:/Project_Data/Project_7/UTR_Alignments/Full_Ortholog/a.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_22480/1479302329.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     15\u001b[0m         \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmakedirs\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0manalysis_directory\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m     \u001b[0mintergenic_region_alignment\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mintergenic_alignment_loc\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'/'\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mintergenic_file\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m     \u001b[0malignment\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mintergenic_region_alignment\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"fasta\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     18\u001b[0m     \u001b[0mAlignIO\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0manalysis_directory\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'/initial_align.sto'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"stockholm\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m;\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\Bio\\AlignIO\\__init__.py\u001b[0m in \u001b[0;36mread\u001b[1;34m(handle, format, seq_count)\u001b[0m\n\u001b[0;32m    388\u001b[0m     \u001b[0miterator\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mseq_count\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    389\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 390\u001b[1;33m         \u001b[0malignment\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    391\u001b[0m     \u001b[1;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    392\u001b[0m         \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"No records found in handle\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\Bio\\AlignIO\\__init__.py\u001b[0m in \u001b[0;36mparse\u001b[1;34m(handle, format, seq_count)\u001b[0m\n\u001b[0;32m    327\u001b[0m         \u001b[1;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Need integer for seq_count (sequences per alignment)\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    328\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 329\u001b[1;33m     \u001b[1;32mwith\u001b[0m \u001b[0mas_handle\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    330\u001b[0m         \u001b[1;31m# Map the file format to a sequence iterator:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    331\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[1;32min\u001b[0m \u001b[0m_FormatToIterator\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\Bioinformatics\\lib\\contextlib.py\u001b[0m in \u001b[0;36m__enter__\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    117\u001b[0m         \u001b[1;32mdel\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfunc\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    118\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 119\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    120\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    121\u001b[0m             \u001b[1;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"generator didn't yield\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\Bio\\File.py\u001b[0m in \u001b[0;36mas_handle\u001b[1;34m(handleish, mode, **kwargs)\u001b[0m\n\u001b[0;32m     70\u001b[0m     \"\"\"\n\u001b[0;32m     71\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 72\u001b[1;33m         \u001b[1;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhandleish\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     73\u001b[0m             \u001b[1;32myield\u001b[0m \u001b[0mfp\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     74\u001b[0m     \u001b[1;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'D:/Project_Data/Project_7/UTR_Alignments/Full_Ortholog/a.fasta'"
     ]
    }
   ],
   "source": [
    "for alignment_id in tqdm(alignment_ids):\n",
    "    ref = alignment_id[6:]\n",
    "    temp_df = rbh_results[rbh_results.query_ref == ref]\n",
    "    seq_list = []\n",
    "    for i, r in temp_df.iterrows():\n",
    "        if len(r['target_utr']) > 0:\n",
    "            seq_list.append([r['target_species_sname'], r['target_utr']])\n",
    "    util.produce_fasta_file(seq_list, merged_file_loc)\n",
    "    \n",
    "    intergenic_file = alignment_id + '.fasta'\n",
    "    analysis_directory = r_scape_output_loc + '/' + ref\n",
    "    wsl_analysis_directory = util.wslname(analysis_directory)\n",
    "    if not os.path.exists(analysis_directory):\n",
    "        os.makedirs(analysis_directory)\n",
    "    intergenic_region_alignment = intergenic_alignment_loc + '/' + intergenic_file\n",
    "    alignment = AlignIO.read(intergenic_region_alignment, \"fasta\")\n",
    "    AlignIO.write(alignment, analysis_directory + '/initial_align.sto', \"stockholm\");\n",
    "\n",
    "    # Initial run with HMM\n",
    "\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmbuild --noss -F initial_cm.cm initial_align.sto'  , shell=True)\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmsearch -A initial_search.sto initial_cm.cm ' + wsl_merged_file_loc  , shell=True) \n",
    "\n",
    "    # Second run with CM\n",
    "\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmbuild -F interim_cm.cm initial_search.sto'  , shell=True)\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmcalibrate interim_cm.cm', shell= True)\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmsearch -A interim_search.sto interim_cm.cm ' + wsl_merged_file_loc  , shell=True)  \n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; ~/rscape_v2.0.0.g/bin/R-scape --cacofold --outname rscape_interim interim_search.sto'  , shell=True)\n",
    "\n",
    "    # Final run with CaCofold CM\n",
    "\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmbuild -F final_cm.cm rscape_interim.cacofold.sto'  , shell=True)\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmcalibrate final_cm.cm', shell= True)\n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; cmsearch -A final_search.sto final_cm.cm ' + wsl_merged_file_loc  , shell=True)  \n",
    "    subprocess.run('wsl cd ' + wsl_analysis_directory + ' ; ~/rscape_v2.0.0.g/bin/R-scape --cacofold --outname rscape_final final_search.sto'  , shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23da1081-450d-4ad3-9cc3-3de4df98fda6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
