{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1e30ed4-8a0e-4a6f-96eb-dfe116dfb2de",
   "metadata": {},
   "source": [
    "#### Directories etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf9c6377-5080-4c22-94db-feabe17cc47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nicho\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\ete3-3.1.2-py3.7.egg\\ete3\\evol\\parser\\codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "C:\\Users\\nicho\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\ete3-3.1.2-py3.7.egg\\ete3\\evol\\parser\\codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.fft import fft, fftfreq\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "import random\n",
    "from tqdm import tqdm\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from collections import defaultdict\n",
    "import random\n",
    "from joblib import Parallel, delayed\n",
    "from Comparative_Analysis import Blast_Functions as blastfn\n",
    "import time\n",
    "import os\n",
    "import copy\n",
    "from Comparative_Analysis import ORF_Functions as orffn\n",
    "from random import sample\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import ete3;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1988a242-f558-4643-b3a9-27272fa7bebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'F:/Project_Data/Project_9'\n",
    "seq_dir = 'F:/Datasets/Actinobacteria_Ref_Rep_Lev_Complete'\n",
    "dictionary_dir = 'F:/Datasets/CRYPTIC_DATA/Cryptic_Dictionaries'\n",
    "mutation_count_dir = 'F:/Datasets/CRYPTIC_DATA/Cryptic_Mutation_Counts'\n",
    "mutation_count_dir = 'F:/Datasets/CRYPTIC_DATA/Cryptic_Mutation_Counts'\n",
    "test_dir = 'F:/Datasets/Test'\n",
    "tb_species = 'NC_000962.3' \n",
    "tb_genome_filename = 'GCF_000195955.2_ASM19595v2_genomic.gbff'\n",
    "mycobrowser_dir = 'F:/Datasets/Data_From_Publications'\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3235e1e1-a7fa-474b-8a88-72232e020b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in SeqIO.parse(seq_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "    full_sequence = str(record.seq)\n",
    "for record in SeqIO.parse(project_dir + '/annot.gbk', \"genbank\"):\n",
    "    reannotated_sequence = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bba7e94a-f3f3-4b38-b620-c4a62e55efbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mycobrowser_features =[]\n",
    "mycobrowser_df = pd.read_csv(mycobrowser_dir+'/Mycobrowser_Release_4.csv')\n",
    "for i, r in mycobrowser_df.iterrows():\n",
    "    if r['Strand'] == '+':\n",
    "        strand = 1\n",
    "    else:\n",
    "        strand = -1\n",
    "    mycobrowser_features.append([r['Locus'],r['Start']-1, r['Stop'], strand])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5bed02d4-5edf-46c9-bc55-cc30edc8028c",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_df = pd.read_csv(project_dir +'/Barcode.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b88b412-d21e-4178-aef9-84e81e3c0f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_run = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3ee7e9e-6a66-485f-badf-22c81560e77f",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30ebdfd2-85cc-4bc6-86c5-9e3c6c9cf3ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_sequences(position_list, variant_dict, position_dict):\n",
    "    base_sequence = []\n",
    "    for i in position_list:\n",
    "        base_sequence.append(full_sequence[i-1])     # Cryptic are 1 based \n",
    "    output_sequence_dict ={}\n",
    "    for k, v in variant_dict.items():\n",
    "        output_sequence_dict[k] = copy.deepcopy(base_sequence)\n",
    "    for i, pos in enumerate(position_list):\n",
    "        if pos in position_dict:\n",
    "            variant_info = position_dict[pos][1:]  # Miss out initlal \"ref\" record\n",
    "            for (name, alt) in variant_info:\n",
    "                output_sequence_dict[name][i] = alt.upper()\n",
    "    output_sequences = []\n",
    "    for k, v in output_sequence_dict.items():\n",
    "        output_sequences.append(['seq_'+str(k), ''.join(v)])\n",
    "    return output_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fefd2694-16c3-4f59-b516-08d2b64a3082",
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_sequences_to_score(position_list, variant_dict, position_dict, distinct_sequence_names):\n",
    "    base_sequence = []\n",
    "    for i in position_list:\n",
    "        base_sequence.append({full_sequence[i-1]})     # Cryptic are 1 based \n",
    "    output_sequence_dict ={}\n",
    "    for k, v in variant_dict.items():\n",
    "        if k in distinct_sequence_names:\n",
    "            output_sequence_dict[k] = copy.deepcopy(base_sequence)\n",
    "    for i, pos in enumerate(position_list):\n",
    "        if pos in position_dict:\n",
    "            variant_info = position_dict[pos][1:]  # Miss out initlal \"ref\" record\n",
    "            for (name, alt) in variant_info:\n",
    "                if name in output_sequence_dict:\n",
    "                    output_sequence_dict[name][i] = {alt.upper()}\n",
    "    temp_dict = {}\n",
    "    for k, v in output_sequence_dict.items():\n",
    "        temp_dict['seq_'+str(k)] = v\n",
    "    return temp_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9acb43-19f3-45e0-8214-f9481910cfae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitch_1(list_1, list_2):\n",
    "    res =[]\n",
    "    for i, j in zip(list_1, list_2):\n",
    "        a = i.intersection(j)\n",
    "        if len(a) == 0:\n",
    "            a = i.union(j)\n",
    "        res.append(a)\n",
    "    return res\n",
    "\n",
    "def fitch_2(parent_list, child_list):\n",
    "    res = []\n",
    "    mutations = []\n",
    "    for i, j in zip(parent_list, child_list):\n",
    "        mutation = 0\n",
    "        a = i.intersection(j)\n",
    "        if len(a) == 0:\n",
    "            a = set(list(j)[0])\n",
    "            mutation = 1\n",
    "        res.append(a)\n",
    "        if mutation == 1:\n",
    "            mutations.append(1)\n",
    "        else:\n",
    "            mutations.append(0)\n",
    "    return (res, mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "562c1fbb-768e-4ab5-a781-b9a8ae96b6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bin_formula(position_3_counts, tot_bin_counts):\n",
    "    return 1- binom.cdf(position_3_counts-1, tot_bin_counts,1/3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b9eaf090-a2a6-4b2d-b054-602281be9016",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def mutation_bin_probability(mutation_counts):\n",
    "    bin_counts = [0,0,0]\n",
    "    for i, c in enumerate(mutation_counts):\n",
    "        bin_counts[i % 3] += c\n",
    "    if sum(bin_counts) == 0:\n",
    "        return (2)\n",
    "    else:\n",
    "        return (bin_counts, bin_formula(bin_counts[2], sum(bin_counts)))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "461e5a4a-6849-4e60-b1d0-8b6c81e20610",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculation_mutation_significance(start, stop):  \n",
    "    positions = list(range(start,stop))\n",
    "    sequence_to_score_dict = produce_sequences_to_score(positions, variant_dict, position_dict, distinct_sequence_names)\n",
    "\n",
    "    for node in master_tree.traverse(\"postorder\"):\n",
    "        if node.is_leaf():\n",
    "            node.add_features(seq = sequence_to_score_dict[node.name])\n",
    "        else:\n",
    "            children = node.children\n",
    "            node.add_features(seq = fitch_1(children[0].seq, children[1].seq))\n",
    "    for k, v in sequence_to_score_dict.items():              \n",
    "        seq_length = len(v)\n",
    "        break\n",
    "\n",
    "    mutation_counts = [0 for i in range(seq_length)]\n",
    "    for node in master_tree.traverse(\"preorder\"):\n",
    "        if node.is_leaf():\n",
    "            continue\n",
    "        if node.is_root():\n",
    "            node.seq = [{list(x)[0]} for x in node.seq]\n",
    "        children = node.children\n",
    "        mutations = []\n",
    "        for child in children:\n",
    "            (temp_1, temp_2) = fitch_2(node.seq ,child.seq)\n",
    "            child.seq = temp_1\n",
    "            mutations.append(temp_2)\n",
    "        temp = []\n",
    "        for h, i, j in zip(mutation_counts, mutations[0], mutations[1]):\n",
    "            temp.append(h+max(i, j))\n",
    "        mutation_counts = temp        \n",
    "    return mutation_bin_probability(mutation_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95cb8d3e-ef11-4e1a-bf89-cacff5ff6f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_mutation_counts(filename, core_number):    \n",
    "    seq_length = 100\n",
    "    a = filename.split('_')\n",
    "    start = a[-3]\n",
    "    stop = a[-2]\n",
    "    with open(filename, 'rb') as f:\n",
    "        sequence_to_score_dict = pickle.load(f)\n",
    "    master_tree2= ete3.Tree(project_dir + '/' + 'my_tree.nwk')\n",
    "    for node in master_tree2.traverse(\"postorder\"):\n",
    "        if node.is_leaf():\n",
    "            node.add_features(seq = sequence_to_score_dict[node.name][core_number * seq_length: (core_number+1) * seq_length])\n",
    "        else:\n",
    "            children = node.children\n",
    "            node.add_features(seq = fitch_1(children[0].seq, children[1].seq))\n",
    "    #for k, v in sequence_to_score_dict.items():              \n",
    "        #seq_length = len(v)\n",
    "        #break\n",
    "    \n",
    "    mutation_counts = [0 for i in range(seq_length)]\n",
    "    for node in master_tree2.traverse(\"preorder\"):\n",
    "        if node.is_leaf():\n",
    "            continue\n",
    "        if node.is_root():\n",
    "            node.seq = [{list(x)[0]} for x in node.seq]\n",
    "        children = node.children\n",
    "        mutations = []\n",
    "        child_sequences = []\n",
    "        for child in children:\n",
    "            (temp_1, temp_2) = fitch_2(node.seq ,child.seq)\n",
    "            child_sequences.append(temp_1)\n",
    "            child.seq = temp_1\n",
    "            mutations.append(temp_2)\n",
    "        temp = []\n",
    "        for n, (h, i, j) in enumerate(zip(mutation_counts, mutations[0], mutations[1])):\n",
    "            if i + j == 0:\n",
    "                temp.append(h+0)\n",
    "            elif i + j == 1:\n",
    "                temp.append(h+1)\n",
    "            else:\n",
    "                if child_sequences[0][i] == child_sequences[1][i]:\n",
    "                    temp.append(h+1)\n",
    "                else:\n",
    "                    temp.append(h+2)\n",
    "            \n",
    "        mutation_counts = temp     \n",
    "    return (start, stop, mutation_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcc25129-08e0-4ce1-9b17-a64f477c50d8",
   "metadata": {},
   "source": [
    "#### Create variant dictionaries "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5fcbb04-2c85-4719-b420-2651cf1e2cdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    variant_df = pd.read_csv(cryptic_input_path + \"VARIANTS.csv\") \n",
    "    with open(project_dir + '/variant_df.pkl', 'wb') as f:\n",
    "        pickle.dump(variant_df[['UNIQUEID', 'VARIANT', 'MUTATION_TYPE', 'IS_NULL', 'IS_HET', 'IS_FILTER_PASS', 'IS_SNP', 'REF', 'ALT', 'GENOME_INDEX']], f)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77fe61cc-adfe-4aa1-97d1-ba7ef7ac00ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    position_dict = {}\n",
    "    variant_dict = {}\n",
    "    id_dict = {}\n",
    "    with open(project_dir + '/variant_df.pkl', 'rb') as f:\n",
    "        variant_df = pickle.load(f) \n",
    "        unique_ids = variant_df.UNIQUEID.unique()\n",
    "        for i, unique_id in enumerate(unique_ids):\n",
    "            id_dict[unique_id] = i\n",
    "\n",
    "        for i, r in variant_df.iterrows():\n",
    "            if r['IS_NULL'] == False and r['IS_FILTER_PASS'] == True and r['IS_HET'] == False and r['IS_SNP'] == True :\n",
    "                \n",
    "                if id_dict[r['UNIQUEID']] in variant_dict:\n",
    "                    variant_dict[id_dict[r['UNIQUEID']]].append((r['GENOME_INDEX'], r['ALT']))\n",
    "                else:\n",
    "                    variant_dict[id_dict[r['UNIQUEID']]] = [(r['GENOME_INDEX'], r['ALT'])]\n",
    "\n",
    "                if r['GENOME_INDEX'] in position_dict:\n",
    "                    position_dict[r['GENOME_INDEX']].append((id_dict[r['UNIQUEID']], r['ALT']))\n",
    "                else:\n",
    "                    position_dict[r['GENOME_INDEX']] = [r['REF'], (id_dict[r['UNIQUEID']], r['ALT'])]    # If first entry also include reference value for info\n",
    "\n",
    "    with open(project_dir + '/id_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(id_dict, f)\n",
    "    with open(project_dir + '/variant_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(variant_dict, f) \n",
    "    with open(project_dir + '/position_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(position_dict, f) \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9da2b4-26ab-4150-90e6-99fc3b388cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == False:\n",
    "    with open(project_dir + '/id_dict.pkl', 'rb') as f:\n",
    "        id_dict = pickle.load(f)  \n",
    "    with open(project_dir + '/variant_dict.pkl', 'rb') as f:\n",
    "        variant_dict = pickle.load(f)  \n",
    "    with open(project_dir + '/position_dict.pkl', 'rb') as f:\n",
    "        position_dict = pickle.load(f)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af93de5-fc91-4cc9-92c0-17c2bfd9bf59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_distances(snp_pos_dict_1, snp_pos_dict_2):\n",
    "    distance_dict = {}\n",
    "    for k1, v1 in snp_pos_dict_1.items():\n",
    "        for k2, v2 in snp_pos_dict_2.items():\n",
    "            sd = v1.symmetric_difference(v2)\n",
    "            temp = {x[:-1] for x in sd}   # Only count variants with different nt in SNP as one mutation\n",
    "            f = len(temp)/tb_length\n",
    "            d = -3/4 * math.log(1 - 4*f /3)\n",
    "            distance_dict[(k1, k2)] = d\n",
    "            distance_dict[(k2, k1)] = d\n",
    "    return(distance_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53eb345b-8511-431b-924e-012955b705dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tb_length = len(full_sequence)\n",
    "reduced_num_cores = 32\n",
    "reduced_core_numbers = list(range(1, reduced_num_cores+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "881db538-693e-4df7-8a31-56aafc155913",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(project_dir + '/variant_dict.pkl', 'rb') as f:\n",
    "            variant_dict = pickle.load(f)  \n",
    "temp_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1f7315f-97ee-4fd8-aae3-c8a315f51c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "for n,(k, v) in enumerate(variant_dict.items()):\n",
    "    if n > 1000:\n",
    "        break\n",
    "    temp_dict[k] = set([str(pos) + snp for (pos, snp) in v])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad5a0452-8832-452e-9a52-0a783abb8db0",
   "metadata": {},
   "outputs": [],
   "source": [
    "a=generate_distances(temp_dict, temp_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2dde158-68e5-42a1-98e2-78f0e7bd39b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "#if 1==1:\n",
    "    with open(project_dir + '/variant_dict.pkl', 'rb') as f:\n",
    "            variant_dict = pickle.load(f)  \n",
    "    snp_pos_dict = {}\n",
    "    for core in tqdm(reduced_core_numbers):\n",
    "        temp_dict = {}\n",
    "        for (k, v) in variant_dict.items():\n",
    "            if k%reduced_num_cores + 1 == core:\n",
    "                temp_dict[k] = set([str(pos) + snp for (pos, snp) in v])\n",
    "        snp_pos_dict[core] = temp_dict\n",
    "    with open(project_dir + '/snp_pos_dict.pkl', 'wb') as f:\n",
    "            pickle.dump(snp_pos_dict, f) \n",
    "else:\n",
    "    with open(project_dir + '/snp_pos_dict.pkl', 'rb') as f:\n",
    "        snp_pos_dict = pickle.load(f)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c759ced-4bc9-4661-a481-9e8c86bc7219",
   "metadata": {},
   "outputs": [],
   "source": [
    "core_size_to_run = 8\n",
    "pairwise_blocks = []\n",
    "for core in reduced_core_numbers:\n",
    "    for core2 in reduced_core_numbers:\n",
    "        if core2 >= core:\n",
    "            pairwise_blocks.append((core, core2))\n",
    "pairwise_list = []\n",
    "for core in list(range(1, reduced_num_cores*4+1)):\n",
    "    if (core - 1) * core_size_to_run >= len(pairwise_blocks):\n",
    "        continue\n",
    "    else:\n",
    "        pairwise_list.append(pairwise_blocks[(core - 1) * core_size_to_run : core * core_size_to_run])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06e933c9-1a86-4078-893c-21a0e3190059",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(pairwise_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25fb6b90-9f16-4471-b2ef-0f62611cc242",
   "metadata": {},
   "outputs": [],
   "source": [
    "master_dict = {}\n",
    "for comparators in tqdm(pairwise_list):\n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(generate_distances)(snp_pos_dict[pos1], snp_pos_dict[pos2]) for (pos1, pos2) in comparators)\n",
    "    for output_dict in parallel_output:\n",
    "        for (k, v) in output_dict.items():\n",
    "            master_dict[k] = v\n",
    "with open(project_dir + '/master_distance_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(master_dict, f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a76331e-942d-4ee7-8391-9f44fa51ceb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = []\n",
    "for k, v in master_dict.items():\n",
    "    ids.append(k[0])\n",
    "ids = list(set(ids))\n",
    "with open(project_dir+'/tb_seq_distances.phy', 'w') as f:\n",
    "    f.write('%d\\n' % len(ids))\n",
    "    for idref in ids:\n",
    "        f.write('seq_'+str(idref))\n",
    "        for opref in ids:\n",
    "            f.write('\\t%s' % str(abs(master_dict[idref, opref])))\n",
    "        f.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4f26106-6d45-411e-b4bd-e5f796e37793",
   "metadata": {},
   "source": [
    "#### Output full sequences and distinct sequences based on barcode positions for construction of tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c107d29a-0a57-458f-bbc8-1e041dcd0e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = []\n",
    "for k, v in position_dict.items():\n",
    "    if len(v) > 0 and len(v) < 100:\n",
    "        temp.append(len(v))\n",
    "len(temp)\n",
    "sns.histplot(temp, bins =100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a5a5d24-12d7-4e40-a0f8-0443893f9e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_snps = []\n",
    "for i, r in barcode_df.iterrows():\n",
    "    start = r['start'] + 1            # Cryptic file is 1 indexed\n",
    "    lineage_name = r['lineage']\n",
    "    if start in position_dict and 'lineage' in lineage_name:   # Just TB lineages\n",
    "        barcode_snps.append(start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02308c27-c3ad-4ae6-92ec-951cec9a27c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_snps = []\n",
    "for k, v in position_dict.items():\n",
    "    if len(v) >= 10:\n",
    "        barcode_snps.append(k)\n",
    "len(barcode_snps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "935875cd-ea0e-4d02-934e-38727f7c0881",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences = produce_sequences(barcode_snps, variant_dict, position_dict)\n",
    "util.produce_fasta_file(sequences, project_dir + '/' + 'tb_variants.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9468f0da-f0af-48d1-b9ee-2fc476fc3169",
   "metadata": {},
   "outputs": [],
   "source": [
    "distinct_sequences = []\n",
    "distinct_sequence_names = []\n",
    "temp_dict = {}\n",
    "for (ref, seq) in sequences:\n",
    "    temp_dict[seq] = ref\n",
    "for k, v in temp_dict.items():    \n",
    "    distinct_sequences.append([v, k])\n",
    "    distinct_sequence_names.append(int(v.split('_')[1]))\n",
    "util.produce_fasta_file(distinct_sequences, project_dir + '/' + 'distinct_tb_variants.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d3d1ef-4adc-4cab-9e85-d3ac1ac49742",
   "metadata": {},
   "outputs": [],
   "source": [
    "#blastfn.convert_fasta_to_stockholm(project_dir,  'distinct_tb_variants.faa', 'distinct_tb_variants.sto')\n",
    "util.convert_sequence_file_format(project_dir + '/distinct_tb_variants.faa', project_dir + '/distinct_tb_variants2.sto', input_file_format = \"fasta\", output_file_format = \"stockholm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7f0b860-4c57-4de7-997c-c2e7cc074d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run in Windows PowerShell      .\\\\downloads\\fasttree.exe -nt -fastest -noml -nosupport d:\\Project_Data\\Project_9\\distinct_tb_variants.faa > d:\\Project_Data\\Project_9\\my_tree.nwk\n",
    "master_tree= ete3.Tree(project_dir + '/' + 'tb_tree.nwk')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea87943f-fab0-45c5-81eb-c365a0f1da91",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(master_tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38020a2d-eb43-4999-a94a-d887fcbba4a1",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Produce and save mutations per position (first produce files containing blocks of 10,000 nt values for all sequences in tree - will take about 36 hours)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "420ab585-aaff-4ef5-8f31-46e766323fa3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variant_positions = []\n",
    "for k, v in variant_dict.items():\n",
    "    if k in distinct_sequence_names:\n",
    "        for (pos, snp) in v:\n",
    "            variant_positions.append(pos-1)     #Cryptic is 1 indexed\n",
    "sorted_variant_position_list = list(set(variant_positions))\n",
    "sorted_variant_position_list.sort()\n",
    "pos_id_dict = dict(zip(sorted_variant_position_list, range(len(set(variant_positions)))))\n",
    "id_pos_dict = dict(zip(range(len(set(variant_positions))), sorted_variant_position_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ffa22c4-4035-4951-9272-4fe8b1219382",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tb_variants_sequence_length = len(pos_id_dict)\n",
    "chunk_size = 1000\n",
    "num_chunks = math.ceil(tb_variants_sequence_length/chunk_size)\n",
    "chunk_variant_dict = defaultdict(lambda: defaultdict(list))\n",
    "for k, v in variant_dict.items():\n",
    "    if k in distinct_sequence_names:\n",
    "        for (pos, snp) in v:\n",
    "            chunk = int(pos_id_dict[pos-1]/chunk_size)\n",
    "            position_in_chunk = pos_id_dict[pos-1] % chunk_size\n",
    "            chunk_variant_dict[chunk][k].append((position_in_chunk,snp.upper()))\n",
    "print(\"Built dictionary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba4903bc-3222-4018-a92e-cf0f1effc52f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "full_tb_variant_sequence = ''.join([full_sequence[pos] for pos in sorted_variant_position_list])\n",
    "if full_run == True:\n",
    "    for chunk in tqdm(range(num_chunks)):\n",
    "            start_pos = chunk * chunk_size\n",
    "            end_pos = min(tb_variants_sequence_length, start_pos + chunk_size)\n",
    "            variant_sequence_dict = {}\n",
    "            seq_chunk = [{x} for x in full_tb_variant_sequence[start_pos:end_pos]]\n",
    "            for seq_id in distinct_sequence_names:\n",
    "                temp = copy.copy(seq_chunk)\n",
    "                if seq_id in chunk_variant_dict[chunk]:\n",
    "                    for (pos, snp) in chunk_variant_dict[chunk][seq_id]:\n",
    "                        temp[pos] = {snp}\n",
    "                variant_sequence_dict['seq_'+str(seq_id)] = temp\n",
    "            with open(dictionary_dir + '/variant_dictionary_'+str(start_pos)+'_'+str(end_pos)+'_' + '.pkl', 'wb') as f:\n",
    "                   pickle.dump(variant_sequence_dict, f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f7730ef-05b3-46e6-b3ab-a58bdd82c5e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename_list = util.list_files(dictionary_dir)\n",
    "filename_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f7750a2-60f6-41fd-acb6-307438824c3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if full_run==True:   #mutation_count_dir\n",
    "if 1==1:\n",
    "    res = []\n",
    "    filename_list = util.list_files(dictionary_dir)\n",
    "    for filename in tqdm([filename_list[0]]):\n",
    "        temp_2 = filename.split('_')\n",
    "        start_pos = int(temp_2[2])\n",
    "        end_pos = int(temp_2[3])\n",
    "        parallel_output = Parallel(n_jobs=-1)(delayed(generate_mutation_counts)(dictionary_dir+'/' + filename, core_number) for core_number in range(10))\n",
    "        temp = []\n",
    "        for x in parallel_output:\n",
    "            temp+=x[2]\n",
    "        res.append((int(parallel_output[0][0]), int(parallel_output[0][1]), temp))\n",
    "        with open(test_dir + '/mutation_counts_'+str(start_pos)+'_'+str(end_pos)+'_' + '.pkl', 'wb') as f:\n",
    "                   pickle.dump((int(parallel_output[0][0]), int(parallel_output[0][1]), temp), f) \n",
    "    with open(test_dir + '/all_mutation_counts.pkl', 'wb') as f:\n",
    "        pickle.dump(res, f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d789a34c-e491-4d71-851d-ac8230f2f004",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    non_zero_mutation_counts = []\n",
    "    for x in res:\n",
    "        non_zero_mutation_counts += x[2]\n",
    "    zero_and_non_zero_mutation_counts = []\n",
    "    for i in range(len(full_sequence)):\n",
    "        if i in pos_id_dict:\n",
    "            zero_and_non_zero_mutation_counts.append(non_zero_mutation_counts[pos_id_dict[i]])\n",
    "        else:\n",
    "            zero_and_non_zero_mutation_counts.append(0)\n",
    "with open(mutation_count_dir + '/zero_and_non_zero_mutation_counts.pkl', 'wb') as f:\n",
    "        pickle.dump(zero_and_non_zero_mutation_counts, f)     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "958cf361-756d-4c83-a9d4-d7df50a81587",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == False:\n",
    "    with open(mutation_count_dir + '/zero_and_non_zero_mutation_counts.pkl', 'rb') as f:\n",
    "        zero_and_non_zero_mutation_counts = pickle.load(f)  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5186cf5-e207-4c3f-bc83-a8bf5b4f28b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "240d70ef-14d8-4bf2-a360-0e611506a410",
   "metadata": {},
   "source": [
    "#### Calculate probabilites for annotated (and reannotated PGAP) CDS regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3f6b3bf0-ecda-4573-8e66-7fcde8acd18c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_boundaries = []\n",
    "for genome_record in SeqIO.parse(seq_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "    for feature in genome_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            a = feature.qualifiers  \n",
    "            if a.get(\"pseudo\") == None:\n",
    "                pseudo = False\n",
    "            else:\n",
    "                pseudo = True\n",
    "            cds_boundaries.append((a.get(\"locus_tag\")[0], pseudo, a.get(\"product\")[0], int(feature.location.start), int(feature.location.end), int(feature.location.strand)))   \n",
    "reannotated_cds_boundaries = []\n",
    "for genome_record in SeqIO.parse(project_dir + '/annot.gbk', \"genbank\"):\n",
    "    for feature in genome_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            a = feature.qualifiers  \n",
    "            if a.get(\"pseudo\") == None:\n",
    "                pseudo = False\n",
    "            else:\n",
    "                pseudo = True\n",
    "            reannotated_cds_boundaries.append((a.get(\"locus_tag\")[0], pseudo, a.get(\"product\")[0], int(feature.location.start), int(feature.location.end), int(feature.location.strand)))   \n",
    "cds_boundaries.sort(key = lambda x: x[3])\n",
    "reannotated_cds_boundaries.sort(key = lambda x: x[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "471e647b-16e4-4c53-8c59-787dacbd80f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_df = pd.read_csv(project_dir+'/mutation_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9ad47ede-8f39-4d1e-8a59-48ed5618e255",
   "metadata": {},
   "outputs": [],
   "source": [
    "zero_and_non_zero_mutation_counts = []\n",
    "for i, r in mutation_df.iterrows():\n",
    "    zero_and_non_zero_mutation_counts.append(r['Num_Mutations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "34f72d73-f975-4510-9896-aad8e3c4b8b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp =[]\n",
    "for (locus, pseudo, product, start, stop, strand) in cds_boundaries:\n",
    "    if pseudo == False:\n",
    "        if strand == 1:\n",
    "            #temp.append(mutation_bin_probability(old_mutation_count_list[start:stop]))\n",
    "            temp.append(mutation_bin_probability(zero_and_non_zero_mutation_counts[start:stop]))\n",
    "        else:\n",
    "            #temp.append(mutation_bin_probability(reversed(old_mutation_count_list[start:stop])))\n",
    "            temp.append(mutation_bin_probability(reversed(zero_and_non_zero_mutation_counts[start:stop])))\n",
    "scores = []\n",
    "for x in temp:\n",
    "    if x == 2:\n",
    "        scores.append(x)\n",
    "    else:\n",
    "        scores.append(x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e88b9d7e-f827-4b4d-921e-f63b44f50ac3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Count'>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAD4CAYAAAAdIcpQAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAUyklEQVR4nO3dfbBc9X3f8fcHgQHzFLAEFpJsiaDQABNjo1BiMh0c0qLakxHuhFSejGEyTJRS3LEbj6fgzsTuH5r6jzy4uIaMYjNAx4FRa2MUGxxjQuNJzIMvjIwQD+VWPN1IRcJOjSAeUsnf/rFHsL5a3bNX6Oxecd+vmZ09+z3nt/vV6id97nnYvakqJEmayRHjbkCSNPcZFpKkVoaFJKmVYSFJamVYSJJaHTnuBrqycOHCWr58+bjbkKTDysMPP/xSVS2aXn/LhsXy5cuZmJgYdxuSdFhJ8tyguoehJEmtDAtJUivDQpLUyrCQJLUyLCRJrToLiyTHJHkoyQ+SbE3yn5r6KUnuSfJ0c39y35jrkkwmeSrJpX3185NsadZdnyRd9S1J2l+XexavAb9WVe8BzgNWJ7kQuBa4t6pWAvc2j0lyNrAWOAdYDdyQZEHzXDcC64CVzW11h31LkqbpLCyq55Xm4VHNrYA1wC1N/RbgsmZ5DXB7Vb1WVc8Ak8AFSRYDJ1bV/dX7PvVb+8ZIkkag0w/lNXsGDwNnAl+sqgeTnFZVOwCqakeSU5vNlwAP9A2famr/r1meXh/0euvo7YHwrne966B63rt3L5OTk68/PvPMM1mwYMEMIyTpra/TsKiqvcB5SX4OuCPJuTNsPug8RM1QH/R6G4ANAKtWrTqo3+o0OTnJui9+k+MWns6rL21nwzUf4qyzzjqYp5Kkt4yRfN1HVf3fJP+T3rmGF5MsbvYqFgM7m82mgGV9w5YC25v60gH1zhy38HROfOe7u3wJSTqsdHk11KJmj4IkxwK/DjwJbAKubDa7ErizWd4ErE1ydJIV9E5kP9Qcstqd5MLmKqgr+sZIkkagyz2LxcAtzXmLI4CNVfWNJPcDG5NcBTwPXA5QVVuTbAQeB/YA1zSHsQCuBm4GjgXubm6SpBHpLCyq6lHgvQPqPwQuOcCY9cD6AfUJYKbzHZKkDvkJbklSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSq87CIsmyJPcleSLJ1iQfb+qfTfJ3STY3tw/2jbkuyWSSp5Jc2lc/P8mWZt31SdJV35Kk/R3Z4XPvAT5ZVY8kOQF4OMk9zbo/qao/7N84ydnAWuAc4HTgO0l+oar2AjcC64AHgLuA1cDdHfYuSerT2Z5FVe2oqkea5d3AE8CSGYasAW6vqteq6hlgErggyWLgxKq6v6oKuBW4rKu+JUn7G8k5iyTLgfcCDzaljyV5NMlNSU5uakuAF/qGTTW1Jc3y9LokaUQ6D4skxwNfBT5RVS/TO6T088B5wA7gj/ZtOmB4zVAf9Frrkkwkmdi1a9ebbV2S1Og0LJIcRS8ovlJVXwOoqheram9V/RT4M+CCZvMpYFnf8KXA9qa+dEB9P1W1oapWVdWqRYsWHdo/jCTNY11eDRXgy8ATVfXHffXFfZt9GHisWd4ErE1ydJIVwErgoaraAexOcmHznFcAd3bVtyRpf11eDXUR8FFgS5LNTe3TwEeSnEfvUNKzwO8BVNXWJBuBx+ldSXVNcyUUwNXAzcCx9K6C8kooSRqhzsKiqv6Gwecb7pphzHpg/YD6BHDuoetOkjQbfoJbktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAkteosLJIsS3JfkieSbE3y8aZ+SpJ7kjzd3J/cN+a6JJNJnkpyaV/9/CRbmnXXJ0lXfUuS9tflnsUe4JNV9YvAhcA1Sc4GrgXuraqVwL3NY5p1a4FzgNXADUkWNM91I7AOWNncVnfYtyRpms7Coqp2VNUjzfJu4AlgCbAGuKXZ7BbgsmZ5DXB7Vb1WVc8Ak8AFSRYDJ1bV/VVVwK19YyRJIzCScxZJlgPvBR4ETquqHdALFODUZrMlwAt9w6aa2pJmeXp90OusSzKRZGLXrl2H9M8gSfNZ52GR5Hjgq8AnqurlmTYdUKsZ6vsXqzZU1aqqWrVo0aLZNytJGqjTsEhyFL2g+EpVfa0pv9gcWqK539nUp4BlfcOXAtub+tIBdUnSiHR5NVSALwNPVNUf963aBFzZLF8J3NlXX5vk6CQr6J3Ifqg5VLU7yYXNc17RN0aSNAJHdvjcFwEfBbYk2dzUPg18DtiY5CrgeeBygKrammQj8Di9K6muqaq9zbirgZuBY4G7m5skaUQ6C4uq+hsGn28AuOQAY9YD6wfUJ4BzD113kqTZ8BPckqRWhoUkqZVhIUlqZVhIkloZFpKkVoaFJKmVYSFJamVYSJJaGRaSpFaGhSSplWEhSWplWEiSWg0VFkkuGqYmSXprGnbP4gtD1iRJb0EzfkV5kl8B3g8sSvL7fatOBBZ02Zgkae5o+30WbwOOb7Y7oa/+MvCbXTUlSZpbZgyLqvpr4K+T3FxVz42oJ0nSHDPsb8o7OskGYHn/mKr6tS6akiTNLcOGxX8H/hT4ErC3ZVtJ0lvMsGGxp6pu7LQTSdKcNeyls3+R5N8mWZzklH23TjuTJM0Zw+5ZXNncf6qvVsAZh7YdSdJcNFRYVNWKrhuRJM1dQ4VFkisG1avq1kPbjiRpLhr2MNQv9y0fA1wCPAIYFpI0Dwx7GOrf9T9OchLw3zrpSJI05xzsV5T/A7Bypg2S3JRkZ5LH+mqfTfJ3STY3tw/2rbsuyWSSp5Jc2lc/P8mWZt31SXKQPUuSDtKw5yz+gt7VT9D7AsFfBDa2DLsZ+K/sf6jqT6rqD6c9/9nAWuAc4HTgO0l+oar2AjcC64AHgLuA1cDdw/QtSTo0hj1n0f+f+x7guaqammlAVX03yfIhn38NcHtVvQY8k2QSuCDJs8CJVXU/QJJbgcswLCRppIY6DNV8oeCT9L559mTgH9/Ea34syaPNYaqTm9oS4IW+baaa2pJmeXp9oCTrkkwkmdi1a9ebaFGS1G/Y35T3W8BDwOXAbwEPJjmYryi/Efh54DxgB/BH+15iwLY1Q32gqtpQVauqatWiRYsOoj1J0iDDHob6j8AvV9VOgCSLgO8A/2M2L1ZVL+5bTvJnwDeah1PAsr5NlwLbm/rSAXVJ0ggNezXUEfuCovHDWYx9XZLFfQ8/DOy7UmoTsDbJ0UlW0LvS6qGq2gHsTnJhcxXUFcCds31dSdKbM+yexbeS/CVwW/P4X9O7MumAktwGXAwsTDIFfAa4OMl59A4lPQv8HkBVbU2yEXic3gn0a5oroQCupndl1bH0Tmx7cluSRqztd3CfCZxWVZ9K8q+AX6V3HuF+4Cszja2qjwwof3mG7dcD6wfUJ4BzZ3otSVK32g4lfR7YDVBVX6uq36+qf09vr+Lz3bYmSZor2sJieVU9Or3Y/LS/vJOOJElzTltYHDPDumMPZSOSpLmrLSy+n+R3pxeTXAU83E1LkqS5pu1qqE8AdyT5bd4Ih1XA2+hd+ipJmgdmDIvmQ3TvT/IB3rgi6ZtV9VeddyZJmjOG/X0W9wH3ddyLJGmOOtjfZyFJmkcMC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUqvOwiLJTUl2Jnmsr3ZKknuSPN3cn9y37rokk0meSnJpX/38JFuaddcnSVc9S5IG63LP4mZg9bTatcC9VbUSuLd5TJKzgbXAOc2YG5IsaMbcCKwDVja36c8pSepYZ2FRVd8FfjStvAa4pVm+Bbisr357Vb1WVc8Ak8AFSRYDJ1bV/VVVwK19YyRJIzLqcxanVdUOgOb+1Ka+BHihb7upprakWZ5elySN0Fw5wT3oPETNUB/8JMm6JBNJJnbt2nXImpOk+W7UYfFic2iJ5n5nU58ClvVttxTY3tSXDqgPVFUbqmpVVa1atGjRIW1ckuazUYfFJuDKZvlK4M6++tokRydZQe9E9kPNoardSS5sroK6om+MJGlEjuzqiZPcBlwMLEwyBXwG+BywMclVwPPA5QBVtTXJRuBxYA9wTVXtbZ7qanpXVh0L3N3cJEkj1FlYVNVHDrDqkgNsvx5YP6A+AZx7CFuTJM3SXDnBLUmawwwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVKrsYRFkmeTbEmyOclEUzslyT1Jnm7uT+7b/rokk0meSnLpOHqWpPlsnHsWH6iq86pqVfP4WuDeqloJ3Ns8JsnZwFrgHGA1cEOSBeNoWJLmq7l0GGoNcEuzfAtwWV/99qp6raqeASaBC0bfniTNX+MKiwK+neThJOua2mlVtQOguT+1qS8BXugbO9XU9pNkXZKJJBO7du3qqHVJmn+OHNPrXlRV25OcCtyT5MkZts2AWg3asKo2ABsAVq1aNXAbSdLsjWXPoqq2N/c7gTvoHVZ6McligOZ+Z7P5FLCsb/hSYPvoupUkjTwskhyX5IR9y8C/AB4DNgFXNptdCdzZLG8C1iY5OskKYCXw0Gi7lqT5bRyHoU4D7kiy7/X/vKq+leT7wMYkVwHPA5cDVNXWJBuBx4E9wDVVtXcMfUvSvDXysKiqbcB7BtR/CFxygDHrgfUdtyZJOoC5dOmsJGmOMiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUivDQpLUyrCQJLUyLCRJrQwLSVIrw0KS1MqwkCS1MiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUquR/w5uab7Zu3cvk5OTrz8+88wzWbBgwRg7kmbPsJA6Njk5ybovfpPjFp7Oqy9tZ8M1H+Kss84ad1vSrBgW0ggct/B0Tnznu8fdhnTQPGchSWrlnoUkHcZGdU7MsJhB/fSnbNu27fXHnpiUNNeM6pyYYTGDV3/0f/js15/jHUt+zCs7p7juQ+dwxhlnAAaHDo4/gKgLozgndtiERZLVwH8BFgBfqqrPjeJ1j3tH7y/hlZe289mvb94vOPbu3QvAggULhlqGN/6D8JLKw99s/w77fwDxyigdTg6LsEiyAPgi8M+BKeD7STZV1eOj7GNQcOx6ejNHvP0k3rFkxVDL/UGzbds2/vNdT3D8otNnHUDArINqPiyP+n0Z5u9w27ZtVO0/j6TDyWERFsAFwGRVbQNIcjuwBugkLF59aTsAP/n7nRzx2mu8fMzR+y+//aSDeu6f/Pgl/sOXv8VJpy3hR889yfFLVnL8gPqCY06YcRkYarv5tjyO92WYv8Pjl6wk+dk59epL29m27eDmkbTPtm3bXv8/q3f/3k5e53AJiyXAC32Pp4B/On2jJOuAdc3DV5I8dZCvtxB46SDHdsm+ZmfO9/XdL4y5k58159+vOWZO9vVPvvDJN9vXwN3ewyUsMqBW+xWqNgAb3vSLJRNVterNPs+hZl+zY1+zY1+zM9/6Olw+lDcFLOt7vBTYPqZeJGneOVzC4vvAyiQrkrwNWAtsGnNPkjRvHBaHoapqT5KPAX9J79LZm6pqa4cv+aYPZXXEvmbHvmbHvmZnXvWVqv0O/UuS9DMOl8NQkqQxMiwkSa3mVVgkWZ3kqSSTSa4dsD5Jrm/WP5rkfcOO7biv3276eTTJ95K8p2/ds0m2JNmcZGLEfV2c5MfNa29O8gfDju24r0/19fRYkr1JTmnWdfl+3ZRkZ5LHDrB+XPOrra9xza+2vsY1v9r6Gtf8WpbkviRPJNma5OMDtulujlXVvLjROzH+v4EzgLcBPwDOnrbNB4G76X2u40LgwWHHdtzX+4GTm+V/ua+v5vGzwMIxvV8XA984mLFd9jVt+98A/qrr96t57n8GvA947ADrRz6/huxr5PNryL5GPr+G6WuM82sx8L5m+QTgf43y/7D5tGfx+leGVNU/Avu+MqTfGuDW6nkA+Lkki4cc21lfVfW9qvr75uED9D5n0rU382ce6/s1zUeA2w7Ra8+oqr4L/GiGTcYxv1r7GtP8Gub9OpCxvl/TjHJ+7aiqR5rl3cAT9L7dol9nc2w+hcWgrwyZ/kYfaJthxnbZV7+r6P3ksE8B307ycHpfd3KoDNvXryT5QZK7k5wzy7Fd9kWStwOrga/2lbt6v4Yxjvk1W6OaX8Ma9fwa2jjnV5Ll9L4E6sFpqzqbY4fF5ywOkWG+MuRA2wz1dSMHaejnTvIBev+Yf7WvfFFVbU9yKnBPkiebn4xG0dcjwLur6pUkHwS+DqwccmyXfe3zG8DfVlX/T4ldvV/DGMf8GtqI59cwxjG/ZmMs8yvJ8fQC6hNV9fL01QOGHJI5Np/2LIb5ypADbdPl140M9dxJfgn4ErCmqn64r15V25v7ncAd9HY3R9JXVb1cVa80y3cBRyVZOMzYLvvqs5Zphwg6fL+GMY75NZQxzK9WY5pfszHy+ZXkKHpB8ZWq+tqATbqbY12ciJmLN3p7UduAFbxxguecadt8iJ89OfTQsGM77utdwCTw/mn144AT+pa/B6weYV/v5I0Pdl4APN+8d2N9v5rtTqJ33Pm4Ubxffa+xnAOfsB35/Bqyr5HPryH7Gvn8Gqavcc2v5s9+K/D5GbbpbI7Nm8NQdYCvDEnyb5r1fwrcRe9qgkngH4DfmWnsCPv6A+AdwA1JAPZU71slTwPuaGpHAn9eVd8aYV+/CVydZA/wE2Bt9WbmuN8vgA8D366qV/uGd/Z+ASS5jd4VPAuTTAGfAY7q62vk82vIvkY+v4bsa+Tza8i+YAzzC7gI+CiwJcnmpvZpemHf+Rzz6z4kSa3m0zkLSdJBMiwkSa0MC0lSK8NCktTKsJAktTIsJEmtDAtJUqv/D5l7NqrQn+hMAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(scores, bins =100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "62a7fa87-aa03-4ef1-971f-22b814e71b8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Count'>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEDCAYAAAA4FgP0AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAUZ0lEQVR4nO3df/BddX3n8efLLwGUny1JJYQfgQazVWYUNkXR1qFqdwFp2Xapi+Oo6zrN4GJH1urU1hlLd/aP7e7qtIprJlsZpUOxWpHSFrTuLlXcKUhgE+RnTaMM2SAJoMQAQxt47x/3hF4u9/vNNyHn3u+Xz/Mxc+d7zud87rnv+dyb+8r5cc9JVSFJatdLpl2AJGm6DAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYtyiBIckWS7UnunGf/tyW5O8ldSf6k7/okaTHJYvwdQZI3AruAK6vqtL30PRX4IvCmqvphkp+qqu2TqFOSFoNFuUVQVd8EHh1uS/LTSb6a5LYkNyX5Z92iXwc+XVU/7J5rCEjSkEUZBLNYD/xGVf1z4EPAf+/aXwG8Isn/SXJzknOmVqEkLUAHTbuAAyHJ4cDrgS8l2dN8SPf3IOBU4GzgeOCmJKdV1Y8mXKYkLUgviiBgsGXzo6p6zZhlW4Gbq+ofge8luY9BMNw6wfokacF6UewaqqqdDL7kfw0gA6/uFl8L/ELXvpTBrqIt06hTkhaiRRkESa4G/hZYnWRrkvcC7wDem2QTcBdwQdf9a8AjSe4GbgQ+XFWPTKNuSVqIFuXpo5KkA2dRbhFIkg6cRXeweOnSpbVy5cpplyFJi8ptt932cFUtG7ds0QXBypUr2bBhw7TLkKRFJcn9sy1z15AkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY3rLQiSHJrk20k2dbeI/L0xfc5O8liSjd3jY33VI0kar88flD3F4PaQu5IsAb6V5Iaqunmk301VdX6PdQDw9NNPs3nz5mfnV61axczMTN8vK0kLXm9BUIOr2e3qZpd0j6ld4W7z5s2s/fRfcdjS43j84W2sv+StrF69elrlSNKC0esxgiQzSTYC24GvV9UtY7qd1e0+uiHJq2ZZz9okG5Js2LFjx37Xc9jS4zjy2JM4bOlx+70OSXqx6TUIqurp7q5hxwNnJjltpMvtwElV9WrgUwxuIjNuPeurak1VrVm2bOw1kyRJ+2kiZw119wf+G+CckfadVbWrm74eWNLdRUySNCF9njW0LMnR3fRLgbcA9470OTbd3eaTnNnV493DJGmC+jxraDnw+SQzDL7gv1hVf5nkYoCqWgdcCLwvyW7gSeCi8pZpkjRRfZ41dAdw+pj2dUPTlwOX91WDJGnv/GWxJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIa11sQJDk0ybeTbEpyV5LfG9MnST6ZZHOSO5Kc0Vc9kqTxDupx3U8Bb6qqXUmWAN9KckNV3TzU51zg1O7xWuAz3V9J0oT0tkVQA7u62SXdo0a6XQBc2fW9GTg6yfK+apIkPV+vxwiSzCTZCGwHvl5Vt4x0WQE8MDS/tWsbXc/aJBuSbNixY0dv9UpSi3oNgqp6uqpeAxwPnJnktJEuGfe0MetZX1VrqmrNsmXLeqhUkto1kbOGqupHwN8A54ws2gqcMDR/PLBtEjVJkgb6PGtoWZKju+mXAm8B7h3pdh3wru7sodcBj1XVg33VJEl6vj7PGloOfD7JDIPA+WJV/WWSiwGqah1wPXAesBl4AnhPj/VIksboLQiq6g7g9DHt64amC7ikrxokSXvnL4slqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGtdbECQ5IcmNSe5JcleSD4zpc3aSx5Js7B4f66seSdJ4B/W47t3Ab1bV7UmOAG5L8vWqunuk301VdX6PdUiS5tDbFkFVPVhVt3fTPwbuAVb09XqSpP0zkWMESVYCpwO3jFl8VpJNSW5I8qpZnr82yYYkG3bs2NFnqZLUnN6DIMnhwJeBS6tq58ji24GTqurVwKeAa8eto6rWV9WaqlqzbNmyXuuVpNb0GgRJljAIgauq6prR5VW1s6p2ddPXA0uSLO2zJknSc/V51lCAzwL3VNUnZulzbNePJGd29TzSV02SpOfr86yhNwDvBL6TZGPX9jvAiQBVtQ64EHhfkt3Ak8BFVVU91iRJGtFbEFTVt4Dspc/lwOV91SBJ2jt/WSxJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMbNKwiSvGE+bSPLT0hyY5J7ktyV5ANj+iTJJ5NsTnJHkjPmX7ok6UCY7xbBp+bZNmw38JtV9TPA64BLkrxypM+5wKndYy3wmXnWI0k6QA6aa2GSs4DXA8uSfHBo0ZHAzFzPraoHgQe76R8nuQdYAdw91O0C4MqqKuDmJEcnWd49V5I0AXvbIjgYOJxBYBwx9NgJXDjfF0myEjgduGVk0QrggaH5rV3b6PPXJtmQZMOOHTvm+7KSpHmYc4ugqr4BfCPJ56rq/v15gSSHA18GLq2qnaOLx73smDrWA+sB1qxZ87zlkqT9N2cQDDkkyXpg5fBzqupNcz0pyRIGIXBVVV0zpstW4ISh+eOBbfOsSZJ0AMw3CL4ErAP+CHh6Pk9IEuCzwD1V9YlZul0HvD/JF4DXAo95fECSJmu+QbC7qvb1jJ43AO8EvpNkY9f2O8CJAFW1DrgeOA/YDDwBvGcfX0OS9ALNNwj+Ism/B74CPLWnsaoene0JVfUtxh8DGO5TwCXzrEGS1IP5BsG7u78fHmor4JQDW44kadLmFQRVdXLfhUiSpmNeQZDkXePaq+rKA1uOJGnS5rtr6GeHpg8F3gzcDhgEkrTIzXfX0G8Mzyc5CvjjXiqSJE3U/l6G+gkGF4qTJC1y8z1G8Bf806UfZoCfAb7YV1GSpMmZ7zGC/zY0vRu4v6q29lCPJGnC5rVrqLv43L0Mrjz6E8A/9FmUJGly5nuHsrcB3wZ+DXgbcEuSeV+GWpK0cM1319BHgZ+tqu0ASZYB/xP4s74KkyRNxnzPGnrJnhDoPLIPz5UkLWDz3SL4apKvAVd38/+GwZVDJUmL3N7uWbwKeHlVfTjJrwI/x+CKon8LXDWB+iRJPdvb7p0/AH4MUFXXVNUHq+o/MNga+IN+S5MkTcLegmBlVd0x2lhVGxjctlKStMjtLQgOnWPZSw9kIZKk6dhbENya5NdHG5O8F7itn5IkSZO0t7OGLgW+kuQd/NMX/xrgYOBXeqxLkjQhcwZBVT0EvD7JLwCndc1/VVX/u/fKJEkTMd/7EdwI3LgvK05yBXA+sL2qThuz/Gzgz4HvdU3XVNV/3JfXkCS9cPP9Qdn++BxwOXPfxeymqjq/xxokSXvR22UiquqbwKN9rV+SdGBM+3pBZyXZlOSGJK+aci2S1KQ+dw3tze3ASVW1K8l5wLXMcvvLJGuBtQAnnnjixAqUpBZMbYugqnZW1a5u+npgSZKls/RdX1VrqmrNsmXLJlqnJL3YTS0IkhybJN30mV0tj0yrHklqVW+7hpJcDZwNLE2yFfhdYAlAVa0DLgTel2Q38CRwUVVVX/VIksbrLQiq6u17WX45g9NLJUlTNO2zhiRJU2YQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcb0FQZIrkmxPcucsy5Pkk0k2J7kjyRl91SJJml2fWwSfA86ZY/m5wKndYy3wmR5rkSTNorcgqKpvAo/O0eUC4MoauBk4OsnyvuqRJI03zWMEK4AHhua3dm3Pk2Rtkg1JNuzYsWMixUlSK6YZBBnTVuM6VtX6qlpTVWuWLVvWc1mS1JZpBsFW4ISh+eOBbVOqRZKaNc0guA54V3f20OuAx6rqwSnWI0lNOqivFSe5GjgbWJpkK/C7wBKAqloHXA+cB2wGngDe01ctkqTZ9RYEVfX2vSwv4JK+Xl+SND/+sliSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY3rNQiSnJPkviSbk3xkzPKzkzyWZGP3+Fif9UiSnu+gvlacZAb4NPCLwFbg1iTXVdXdI11vqqrz+6pDkjS33oIAOBPYXFVbAJJ8AbgAGA2CiatnnmHLli3PaVu1ahUzMzNTqkiSpqfPIFgBPDA0vxV47Zh+ZyXZBGwDPlRVd412SLIWWAtw4oknvuDCHn/0B1x27f0cs+KxwfzD21h/yVtZvXr1C163JC02fQZBxrTVyPztwElVtSvJecC1wKnPe1LVemA9wJo1a0bXsV8OO+Y4jjz2pAOxKkla1Po8WLwVOGFo/ngG/+t/VlXtrKpd3fT1wJIkS3usSZI0os8guBU4NcnJSQ4GLgKuG+6Q5Ngk6abP7Op5pMeaJEkjets1VFW7k7wf+BowA1xRVXclubhbvg64EHhfkt3Ak8BFVXVAdv1Ikuanz2MEe3b3XD/Stm5o+nLg8j5rkCTNzV8WS1LjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIa1+uNaRaLeuYZtmzZ8uz8qlWrmJmZmWJFkjQ5BgHw+KM/4LJr7+eYFY/x+MPbWH/JW1m9evW0y5KkiTAIOocdcxxHHnvStMuQpIkzCEa4m0hSawyCEcO7iXZt38pvv/VVnHLKKYChIOnFqdcgSHIO8IfADPBHVfWfR5anW34e8ATwb6vq9j5rmo89u4l2PbyNy67d+LxQePrppwGYmZl5zjQYFpIWn96CIMkM8GngF4GtwK1Jrququ4e6nQuc2j1eC3ym+7tgjAuFHd/dyEtedhTHrDj5OdPzCQtDRNJC0+cWwZnA5qraApDkC8AFwHAQXABcWVUF3Jzk6CTLq+rBPgp6/OFtADz5w+285Kmn2HnoIc+bn3P6ZUfNuf4nH3uY3/rsVznq5St49P57mTn0iHlPP/HD7fynd7752d1QkjSqr7MZ+wyCFcADQ/Nbef7/9sf1WQE8JwiSrAXWdrO7kty3nzUtBR7ez+f2bel5V398Ida2UMfMuvbdQq3NuvbN/tY162mRfQZBxrTVfvShqtYD619wQcmGqlrzQtfTh4Vam3Xtm4VaFyzc2qxr3/RRV5+XmNgKnDA0fzywbT/6SJJ61GcQ3AqcmuTkJAcDFwHXjfS5DnhXBl4HPNbX8QFJ0ni97Rqqqt1J3g98jcHpo1dU1V1JLu6WrwOuZ3Dq6GYGp4++p696Oi9491KPFmpt1rVvFmpdsHBrs659c8DryuCEHUlSq7wMtSQ1ziCQpMY1EwRJzklyX5LNST4yxTpOSHJjknuS3JXkA137ZUn+X5KN3eO8KdT2/STf6V5/Q9f2k0m+nuS73d+fmEJdq4fGZWOSnUkuncaYJbkiyfYkdw61zTpGSX67+8zdl+RfTriu/5rk3iR3JPlKkqO79pVJnhwat3UTrmvW921S4zVHbX86VNf3k2zs2icyZnN8P/T7GauqF/2DwcHqvwdOAQ4GNgGvnFIty4EzuukjgL8DXglcBnxoyuP0fWDpSNt/AT7STX8E+P0F8F7+gMGPYyY+ZsAbgTOAO/c2Rt37ugk4BDi5+wzOTLCufwEc1E3//lBdK4f7TWG8xr5vkxyv2WobWf5x4GOTHLM5vh96/Yy1skXw7OUuquofgD2Xu5i4qnqwugvrVdWPgXsY/Jp6oboA+Hw3/XngX02vFADeDPx9Vd0/jRevqm8Cj440zzZGFwBfqKqnqup7DM6OO3NSdVXVX1fV7m72Zga/05moWcZrNhMbr73V1l0Q823A1X29/iw1zfb90OtnrJUgmO1SFlOVZCVwOnBL1/T+bjP+imnsgmHwq+6/TnJbd1kPgJdX99uO7u9PTaGuYRfx3H+c0x4zmH2MFtLn7t8BNwzNn5zk/yb5RpKfn0I94963hTRePw88VFXfHWqb6JiNfD/0+hlrJQjmdSmLSUpyOPBl4NKq2sngyqs/DbyGwbWWPj6Fst5QVWcwuCrsJUneOIUaZpXBDxN/GfhS17QQxmwuC+Jzl+SjwG7gqq7pQeDEqjod+CDwJ0mOnGBJs71vC2K8Om/nuf/hmOiYjfl+mLXrmLZ9HrNWgmBBXcoiyRIGb/JVVXUNQFU9VFVPV9UzwP+gx03i2VTVtu7vduArXQ0PJVne1b0c2D7puoacC9xeVQ/BwhizzmxjNPXPXZJ3A+cD76hup3K3G+GRbvo2BvuVXzGpmuZ436Y+XgBJDgJ+FfjTPW2THLNx3w/0/BlrJQjmc7mLiej2PX4WuKeqPjHUvnyo268Ad44+t+e6DktyxJ5pBgca72QwTu/uur0b+PNJ1jXiOf9Lm/aYDZltjK4DLkpySJKTGdx349uTKiqDG0P9FvDLVfXEUPuyDO4XQpJTurq2jF9LL3XN9r5NdbyGvAW4t6q27mmY1JjN9v1A35+xvo+CL5QHg0tZ/B2DJP/oFOv4OQabbncAG7vHecAfA9/p2q8Dlk+4rlMYnH2wCbhrzxgBxwD/C/hu9/cnpzRuLwMeAY4aapv4mDEIogeBf2Twv7H3zjVGwEe7z9x9wLkTrmszg/3Hez5n67q+/7p7jzcBtwO/NOG6Zn3fJjVes9XWtX8OuHik70TGbI7vh14/Y15iQpIa18quIUnSLAwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1Lj/Dxb9eRjp5/QwAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot([x for x in zero_and_non_zero_mutation_counts if x < 200], bins = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7e249c42-675d-48f0-a706-68cdd70250a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "132416 13484\n",
      "836537 11101\n",
      "836657 11985\n",
      "837032 10195\n",
      "3943018 13419\n"
     ]
    }
   ],
   "source": [
    "for n, i in enumerate(zero_and_non_zero_mutation_counts):\n",
    "    if i > 10000:\n",
    "        print(n, i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32f6321f-5a4e-481b-80dd-b73ec38fbd18",
   "metadata": {},
   "source": [
    "#### Identify potential ORFS (min length 200) in inter-CDS regions of standard annotation and output to Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fb3b790-e301-4599-a155-74a0e7ae05ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "ORFFinder = orffn.ORF_Finder()\n",
    "trans = util.Translator()\n",
    "min_intergenic_length = 100\n",
    "details =  []\n",
    "results =[]\n",
    "for i, (locus, pseudo, product, start, stop, strand) in enumerate(cds_boundaries):\n",
    "    if i < len(cds_boundaries) - 1:\n",
    "        if cds_boundaries[i+1][3] > stop + min_intergenic_length:\n",
    "            a =ORFFinder.max_orf(stop-40, cds_boundaries[i+1][3]+40, 1e-20, output_all_orfs = False, min_orf_length = 200)\n",
    "            if not(a==(0,0,0)):\n",
    "                ov = 0\n",
    "                info = ('','','','','','','')\n",
    "                for i, (loc, pse, pro, sta, sto, stra) in enumerate(reannotated_cds_boundaries):\n",
    "                    if a[1] > sta and a[0] < sto:\n",
    "                        ov = (min(a[1], sto) - max(a[0], sta)) / (sto - sta)\n",
    "                        if ov > 0.3:\n",
    "                            info =  (loc, pse, pro, sta, sto, stra, ov)\n",
    "                        \n",
    "                ov = 0\n",
    "                myco_info = ('','','','','')\n",
    "                for i, (loc, sta, sto, stra) in enumerate(mycobrowser_features):\n",
    "                    if a[1] > sta and a[0] < sto:\n",
    "                        ov = (min(a[1], sto) - max(a[0], sta)) / (sto - sta)\n",
    "                        if ov > 0.3:\n",
    "                            myco_info =  (loc, sta, sto, stra, ov)\n",
    "                        \n",
    "                sequ = trans.translate_sequence(full_sequence[a[0]:a[1]], a[2], 0)\n",
    "                details.append([a, sequ])\n",
    "                results.append([a[0],a[1],a[2],a[3],info[0],info[1],info[2],info[3],info[4],info[5],info[6],myco_info[0],myco_info[1],myco_info[2],myco_info[3],myco_info[4]])\n",
    "results_df = pd.DataFrame(results, columns = ['start_pos','end_pos','strand','score','PGAP_ref','PGAP_pseudogene','PGAP_product', 'PGAP_start', 'PGAP_end','PGAP_strand', 'PGAP_overlap', 'Mycob_ref','Mycob_start', 'Mycob_end','Mycob_strand', 'Mycob_overlap'])\n",
    "results_df.to_csv(project_dir + '/cds_candidates.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1542c9b-7d8d-4bcc-8104-edf07608f227",
   "metadata": {},
   "source": [
    "#### Calculate probabilities for regions in Smith et al 2021 and plot histogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fe8a8a4-503a-43e1-86a7-48a9cdb0155e",
   "metadata": {},
   "outputs": [],
   "source": [
    "xls = pd.ExcelFile('F:/Datasets/Data_From_Publications/Smith_2021.xlsx')\n",
    "df1 = pd.read_excel(xls, 'Table S3', header=3)\n",
    "co_ords = []\n",
    "for i, r in df1.iterrows():\n",
    "    if r['Classification'] == 'Novel':\n",
    "        if r['Strand'] == '+':\n",
    "            co_ords.append([int(r['Start Coordinate']-1), int(r['Stop Coordinate']), 1])\n",
    "        else:\n",
    "            co_ords.append([int(r['Stop Coordinate'] - 1), int(r['Start Coordinate']),-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58096b14-401d-45d2-bf2f-4a45c6b02730",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "probs = []\n",
    "for x in co_ords:\n",
    "    if x[2] == 1:\n",
    "        a = (mutation_bin_probability(mutation_count_list[x[0]:x[1]]))\n",
    "    else:\n",
    "        a = (mutation_bin_probability(reversed(mutation_count_list[x[0]:x[1]])))\n",
    "    if a == 2:\n",
    "        probs.append(2)\n",
    "    else:\n",
    "        probs.append(a[1])\n",
    "    print(x, abs(x[1]-x[0]), a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6c42903-2248-4181-aa6a-99cb62c354ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(probs, bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "198ce0ae-2b01-482b-83cb-93c720537916",
   "metadata": {},
   "source": [
    "#### Find all (maximal nested) ORFs and filter out ORFS on opposite strand which would have same non-synonymous positions with larger ORF on other strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ea12a609-41c4-402f-a220-c09b6cc10fb1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 63719/63719 [03:57<00:00, 267.85it/s] \n"
     ]
    }
   ],
   "source": [
    "ORFFinder = orffn.ORF_Finder(full_sequence)\n",
    "a = ORFFinder.max_orf(0, 4411532, output_orfs = 'Nested', min_orf_length = 50)\n",
    "a.sort(key = lambda x: x[3], reverse = True)\n",
    "orf_list = [a[0]]\n",
    "for x in tqdm(a[1:]):\n",
    "    matched = 0\n",
    "    for v in orf_list:\n",
    "        if v[0]<=x[0] and v[1]>=x[1]:\n",
    "            if x[2] == v[2]:\n",
    "                if (v[0] - x[0])%3 == 0:\n",
    "                    matched = 1\n",
    "                    break\n",
    "            else:\n",
    "                if (v[0] - x[0])%3 == 1:\n",
    "                    matched = 1\n",
    "                    break\n",
    "    if matched == 0:\n",
    "        orf_list.append(x)\n",
    "orf_list.sort(key = lambda x: x[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "47eaa87f-ac8c-4fa4-9b70-241cfda7bc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = []\n",
    "for (start, stop, strand, length) in orf_list:\n",
    "    if strand == 1:\n",
    "        temp.append(mutation_bin_probability(zero_and_non_zero_mutation_counts[start:stop]))\n",
    "    else:\n",
    "        temp.append(mutation_bin_probability(reversed(zero_and_non_zero_mutation_counts[start:stop])))\n",
    "scores = []\n",
    "for x in temp:\n",
    "    if x == 2:\n",
    "        scores.append(x)\n",
    "    else:\n",
    "        scores.append(x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c310ef26-e928-4ee5-8f96-8f379b56eeff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Count'>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAD4CAYAAADGmmByAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAVvUlEQVR4nO3dfbBc9X3f8ffHF/MkLAJIgJAwEiONYmDqiVEogUzGKZ2xYuqKdIwjTxLUlKmmVHVwm3EL7kzsfzR1ZzKtH8aQ0TgOkLoQ1cGGxMY2ke14WmOIIMQ8hXJ7MaBKRsL28JQUR+LbP/ZcWK7u3rtX5+6uLvf9mtm5Z3/n/PZ89+jofPY87NlUFZIkHam3jLoASdLCZpBIkloxSCRJrRgkkqRWDBJJUivHjLqAYVu2bFmtXr161GVI0oJy//33P1dVy6cbt+iCZPXq1ezevXvUZUjSgpLkqV7jPLQlSWrFIJEktWKQSJJaMUgkSa0YJJKkVgwSSVIrBokkqRWDRJLUyqL7QqJ0tDh06BDj4+OvPV+7di1jY2MjrEg6MgaJNCLj4+Ns/exXWLLsLF5+bi87tl3O+vXrR12WNGcGiTRCS5adxdIzzxl1GVIrniORJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKkVg0SS1IpBIklqxSCRJLUysCBJ8vkk+5M83NV2apK7kzzR/D2la9z1ScaTPJ7kPV3tFyZ5qBn36SRp2o9L8sdN+71JVg/qvUiSehvkHslNwMYpbdcBu6pqHbCreU6S84DNwPlNnxuSTN697kZgK7CueUy+5tXAT6pqLfBfgf88sHciSeppYEFSVd8BfjyleRNwczN8M3BFV/ttVfVKVT0JjAMXJVkBLK2qe6qqgFum9Jl8rS8Cl03urUiShmfY50jOqKp9AM3f05v2lcAzXdPtadpWNsNT29/Qp6oOAs8Dpw2scknStI6Wk+3T7UnUDO0z9Tn8xZOtSXYn2X3gwIEjLFGSNJ1hB8mzzeEqmr/7m/Y9wNld060C9jbtq6Zpf0OfJMcAJ3P4oTQAqmpHVW2oqg3Lly+fp7ciSYLhB8mdwJZmeAtwR1f75uZKrDV0Tqrf1xz+ejHJxc35j6um9Jl8rfcD32zOo0iShmhgP2yV5Fbg3cCyJHuAjwGfAHYmuRp4GrgSoKoeSbITeBQ4CGyrqkPNS11D5wqwE4C7mgfAHwB/lGSczp7I5kG9F0lSbwMLkqr6YI9Rl/WYfjuwfZr23cAF07T/P5ogkiSNztFysl2StEAZJJKkVgZ2aOvN5tChQ4yPj7/2fO3atYyNjc3QQ5IWB4OkT+Pj42z97FdYsuwsXn5uLzu2Xc769etHXZYkjZxBMgdLlp3F0jPPGXUZknRU8RyJJKkVg0SS1IpBIklqxSCRJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKkVg0SS1IpBIklqxSCRJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKkVg0SS1IpBIklqxSCRJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKmVkQRJkn+b5JEkDye5NcnxSU5NcneSJ5q/p3RNf32S8SSPJ3lPV/uFSR5qxn06SUbxfiRpMRt6kCRZCfw2sKGqLgDGgM3AdcCuqloH7Gqek+S8Zvz5wEbghiRjzcvdCGwF1jWPjUN8K5IkRndo6xjghCTHACcCe4FNwM3N+JuBK5rhTcBtVfVKVT0JjAMXJVkBLK2qe6qqgFu6+kiShmToQVJV/xf4PeBpYB/wfFV9AzijqvY10+wDTm+6rASe6XqJPU3bymZ4avthkmxNsjvJ7gMHDszn25GkRW8Uh7ZOobOXsQY4C1iS5Ddm6jJNW83Qfnhj1Y6q2lBVG5YvXz7XkiVJMxjFoa1/DDxZVQeq6u+B24FLgGebw1U0f/c30+8Bzu7qv4rOobA9zfDUdknSEI0iSJ4GLk5yYnOV1WXAY8CdwJZmmi3AHc3wncDmJMclWUPnpPp9zeGvF5Nc3LzOVV19JElDcsywZ1hV9yb5IvAAcBD4K2AHcBKwM8nVdMLmymb6R5LsBB5tpt9WVYeal7sGuAk4AbireUiShmjoQQJQVR8DPjal+RU6eyfTTb8d2D5N+27ggnkvUJLUN7/ZLklqxSCRJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKkVg0SS1IpBIklqxSCRJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKkVg0SS1IpBIklqxSCRJLVikEiSWjFIJEmtGCSSpFYMEklSKwaJJKkVg0SS1IpBIklqxSCRJLVikEiSWukrSJJc2k+bJGnx6XeP5DN9tkmSFpljZhqZ5BeAS4DlSf5d16ilwNiRzjTJzwCfAy4ACvgXwOPAHwOrgR8AH6iqnzTTXw9cDRwCfruqvt60XwjcBJwAfBW4tqrqSOuSJM3dbHskxwIn0Qmct3U9XgDe32K+nwK+VlU/C7wTeAy4DthVVeuAXc1zkpwHbAbOBzYCNySZDLEbga3AuuaxsUVNkqQjMOMeSVX9BfAXSW6qqqfmY4ZJlgK/BPzzZh4/BX6aZBPw7maym4FvA/8B2ATcVlWvAE8mGQcuSvIDYGlV3dO87i3AFcBd81GnJKk/MwZJl+OS7KBz2Om1PlX1j45gnucCB4A/TPJO4H7gWuCMqtrXvO6+JKc3068EvtfVf0/T9vfN8NT2wyTZSmfPhbe//e1HULIkqZd+g+R/AL9P57zGoXmY57uAD1XVvUk+RXMYq4dM01YztB/eWLUD2AGwYcMGz6FI0jzqN0gOVtWN8zTPPcCeqrq3ef5FOkHybJIVzd7ICmB/1/Rnd/VfBext2ldN0y5JGqJ+L//90yT/OsmKJKdOPo5khlX1Q+CZJOubpsuAR4E7gS1N2xbgjmb4TmBzkuOSrKFzUv2+5jDYi0kuThLgqq4+kqQh6XePZHID/5GutqJzvuNIfAj4QpJjgQngt+iE2s4kVwNPA1cCVNUjSXbSCZuDwLaqmjy8dg2vX/57F55ol6Sh6ytIqmrNfM60qh4ENkwz6rIe028Htk/TvpvOd1EkSSPSV5AkuWq69qq6ZX7LkSQtNP0e2vr5ruHj6ew5PAAYJJK0yPV7aOtD3c+TnAz80UAqkiQtKEd6G/m/pXP1lCRpkev3HMmf8vqX/caAdwA7B1WUJGnh6Pccye91DR8EnqqqPb0mliQtHn0d2mpu3vg3dO78ewrw00EWJUlaOPr9hcQPAPfR+ZLgB4B7k7S5jbwk6U2i30Nb/xH4+araD5BkOfDndO6TJUlaxPq9austkyHS+NEc+kqS3sT63SP5WpKvA7c2z3+Nzk/bSpIWudl+s30tnR+c+kiSfwb8Ip3fAbkH+MIQ6pMkHeVm2yP5JPBRgKq6HbgdIMmGZtz7BlibtGjUq68yMTHxhra1a9cyNjY2ooqk/s0WJKur6vtTG6tqd5LVgylJWnxe/vEP+fiXn+K0lc93nj+3lx3bLmf9+vWz9JRGb7YgOX6GcSfMZyHSYrfktLNYeuY5oy5DmrPZrrz6yyT/cmpj8+NT9w+mJEnSQjLbHsmHgS8l+XVeD44NwLHArw6wLknSAjFjkFTVs8AlSX6Z13+J8CtV9c2BVyZJWhD6/T2SbwHfGnAtkqQFyG+nS5JaMUgkSa0YJJKkVgwSSVIrBokkqRWDRJLUikEiSWrFIJEktWKQSJJaMUgkSa0YJJKkVkYWJEnGkvxVkj9rnp+a5O4kTzR/T+ma9vok40keT/KervYLkzzUjPt0kozivUjSYjbKPZJrgce6nl8H7KqqdcCu5jlJzgM2A+cDG4Ebkkz+/uiNwFZgXfPYOJzSJUmTRhIkSVYBlwOf62reBNzcDN8MXNHVfltVvVJVTwLjwEVJVgBLq+qeqirglq4+kqQhGdUeySeBfw+82tV2RlXtA2j+nt60rwSe6ZpuT9O2shme2n6YJFuT7E6y+8CBA/PyBiRJHUMPkiT/BNhfVf3+VO905z1qhvbDG6t2VNWGqtqwfPnyPmcrSepHXz9sNc8uBf5pkvcCxwNLk/w34NkkK6pqX3PYan8z/R7g7K7+q4C9TfuqadolSUM09D2Sqrq+qlZV1Wo6J9G/WVW/AdwJbGkm2wLc0QzfCWxOclySNXROqt/XHP56McnFzdVaV3X1kSQNySj2SHr5BLAzydXA08CVAFX1SJKdwKPAQWBbVR1q+lwD3AScANzVPCRJQzTSIKmqbwPfboZ/BFzWY7rtwPZp2ncDFwyuQknSbPxmuySpFYNEktSKQSJJasUgkSS1YpBIkloxSCRJrRgkkqRWDBJJUisGiSSpFYNEktSKQSJJasUgkSS1YpBIkloxSCRJrRgkkqRWDBJJUisGiSSpFYNEktSKQSJJasUgkSS1YpBIkloxSCRJrRgkkqRWDBJJUisGiSSpFYNEktSKQSJJasUgkSS1YpBIkloZepAkOTvJt5I8luSRJNc27acmuTvJE83fU7r6XJ9kPMnjSd7T1X5hkoeacZ9OkmG/H0la7EaxR3IQ+J2qegdwMbAtyXnAdcCuqloH7Gqe04zbDJwPbARuSDLWvNaNwFZgXfPYOMw3IkkaQZBU1b6qeqAZfhF4DFgJbAJubia7GbiiGd4E3FZVr1TVk8A4cFGSFcDSqrqnqgq4pauPJGlIRnqOJMlq4OeAe4EzqmofdMIGOL2ZbCXwTFe3PU3bymZ4avt089maZHeS3QcOHJjX9yBJi90xo5pxkpOAPwE+XFUvzHB6Y7oRNUP74Y1VO4AdABs2bJh2Gkl6szl06BDj4+OvPV+7di1jY2Mz9DgyIwmSJG+lEyJfqKrbm+Znk6yoqn3NYav9Tfse4Oyu7quAvU37qmnaJUnA+Pg4Wz/7FZYsO4uXn9vLjm2Xs379+nmfzyiu2grwB8BjVfVfukbdCWxphrcAd3S1b05yXJI1dE6q39cc/noxycXNa17V1UeSBCxZdhZLzzyHJcvOGtg8RrFHcinwm8BDSR5s2j4KfALYmeRq4GngSoCqeiTJTuBROld8bauqQ02/a4CbgBOAu5qHJGmIhh4kVfU/mf78BsBlPfpsB7ZP074buGD+qpMkzZXfbJcktWKQSJJaMUgkSa0YJJKkVgwSSVIrBokkqRWDRJLUikEiSWrFIJEktTKyu/9Ki1H33VgnJiYo70WtNwGDRBqi7ruxHnjiQd529jtGXZLUmoe2pCGbvBvrCaecPvvE0gJgkEiSWjFIJEmtGCSSpFYMEklSK161JQ2Yl/zqzc4gkQbMS371ZuehLWkIvORXb2bukUgD4OEsLSYGyRGoV19lYmLitedr165lbGxshBXpaOPhLC0mBskRePnHP+TjX36K01Y+z8vP7WXHtstZv379qMvSiE3dCznxtM7hrJee2zviyqTBMkiO0JJmI6HFpzswDh06BMDY2BgTExP8p68+xknL3QvR4mKQtORhrjevfgPjLSeezGkr17wWHu6FaLExSFrqPsz10v49XH/5+Zx77rnA4g6V7o0wLMxlMfU8R6/AOOak0wwPLWoGyTxY0nUs/ONffnDaUOn+RAsLc8M6nV6B0b0Rbnseaa6h1Gv6XnsYvYannucwMKTpGSTzbLpQAd7wibY7ZHptxGD6DWavjWGv6Xv17Wf6fswUGJPfnZiphn7ef/c8+ll23Yeeuqef6ZBUr2HPc0izM0gGqPuE/NRPtJMh02sj1muD2Wtj2GYD28+n85lCa7rA6KU7FIC+3v/UPYPZlt3UQ0/d0/c6JNVrWNLsDJIRWdLjkEk/YdNrY9hmA9vPp/NeG/nZvmzXfUFCdyjA4QHb6/107xn0s+x6LWtJ888gOYr12mDOdfp+NrD9fDqfbSM/NTAmA6b7goS5HC4yAKSFYcEHSZKNwKeAMeBzVfWJEZe0KEy3kZ8pMAwF6c1rQd+0MckY8FngV4DzgA8mOW+0VS1uk4HhzQmlxWOh75FcBIxX1QRAktuATcCjg5jZy82n6b/7yX7e8sorvHD8cX0NH0mfxTDscum9XF5+bi8TEycPYjXWIjIxMfHadqvz9+cGMp+FHiQrgWe6nu8B/uHUiZJsBbY2T19K8vgRzm8Z8NwR9h0k65qbBVHXdz4zwkoOtyCW2VHkqKzrZz/zO23q6nlp5kIPkkzTdtg1RFW1A9jRembJ7qra0PZ15pt1zY11zd3RWpt1zc2g6lrQ50jo7IGc3fV8FeDZXEkaooUeJH8JrEuyJsmxwGbgzhHXJEmLyoI+tFVVB5P8G+DrdC7//XxVPTLAWbY+PDYg1jU31jV3R2tt1jU3A6kr5W+ASpJaWOiHtiRJI2aQSJJaMUgaSTYmeTzJeJLrphmfJJ9uxn8/ybv67Tvgun69qef7Sb6b5J1d436Q5KEkDybZPeS63p3k+WbeDyb53X77Driuj3TV9HCSQ0lObcYNZHkl+XyS/Uke7jF+VOvWbHWNZN3qs7ZRrV+z1TWK9evsJN9K8liSR5JcO800g13HqmrRP+icqP8/wLnAscBfA+dNmea9wF10vrtyMXBvv30HXNclwCnN8K9M1tU8/wGwbETL693Anx1J30HWNWX69wHfHMLy+iXgXcDDPcYPfd3qs66hr1tzqG3o61c/dY1o/VoBvKsZfhvwv4e9/XKPpOO1W61U1U+ByVutdNsE3FId3wN+JsmKPvsOrK6q+m5V/aR5+j0636UZtDbveaTLa4oPArfO07x7qqrvAD+eYZJRrFuz1jWidWty3rMts15GusymGNb6ta+qHmiGXwQeo3PXj24DXccMko7pbrUy9R+i1zT99B1kXd2upvOpY1IB30hyfzq3iZkv/db1C0n+OsldSc6fY99B1kWSE4GNwJ90NQ9qec1mFOvWXA1r3ZqLYa9ffRvV+pVkNZ0bat07ZdRA17EF/T2SedTPrVZ6TdPXbVqOUN+vneSX6fxn/8Wu5kuram+S04G7k/xN84lqGHU9AJxTVS8leS/wZWBdn30HWdek9wH/q6q6P10OannNZhTrVt+GvG71axTr11wMff1KchKd4PpwVb0wdfQ0XeZtHXOPpKOfW630mmaQt2np67WT/APgc8CmqvrRZHtV7W3+7ge+RGc3dih1VdULVfVSM/xV4K1JlvXTd5B1ddnMlMMOA1xesxnFutWXEaxbfRnR+jUXQ12/kryVToh8oapun2aSwa5j833iZyE+6OyZTQBreP2E0/lTprmcN56suq/fvgOu6+3AOHDJlPYlwNu6hr8LbBxiXWfy+hdeLwKebpbdSJdXM93JdI5zLxnG8mpeczW9TxwPfd3qs66hr1tzqG3o61c/dY1i/Wre9y3AJ2eYZqDrmIe26H2rlST/qhn/+8BX6Vz5MA78LfBbM/UdYl2/C5wG3JAE4GB17u55BvClpu0Y4L9X1deGWNf7gWuSHAT+DthcnTV31MsL4FeBb1TVy13dB7a8ktxK5yqjZUn2AB8D3tpV09DXrT7rGvq6NYfahr5+9VkXDHn9Ai4FfhN4KMmDTdtH6XwQGMo65i1SJEmteI5EktSKQSJJasUgkSS1YpBIkloxSCRJrRgkkqRWDBJJUiv/H2IwN57vuKaRAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(scores, bins =100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dabd8435-6f3a-4389-baf3-46f4f7d741b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "prob = []\n",
    "for x in orf_list:\n",
    "    prob.append(x[4])\n",
    "sns.histplot(prob, bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2769ef-9358-406c-a22a-e88be6b93338",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotated_features = []\n",
    "for genome_record in SeqIO.parse(project_dir + '/annot.gbk', \"genbank\"):\n",
    "    for feature in genome_record.features:\n",
    "        if feature.type != 'source':\n",
    "            annotated_features.append((int(feature.location.start), int(feature.location.end), int(feature.location.strand)))   \n",
    "for genome_record in SeqIO.parse(seq_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "    for feature in genome_record.features:\n",
    "        if feature.type != 'source':\n",
    "            annotated_features.append((int(feature.location.start), int(feature.location.end), int(feature.location.strand)))   \n",
    "annotated_features.sort(key = lambda x: x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fd77402-5afe-4b13-9f1e-1a14e569d8ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_overlapping_orfs = []\n",
    "for i, orf in enumerate(orf_list):\n",
    "    max_ov = 0\n",
    "    for (sta, sto, stra) in annotated_features:\n",
    "        if orf[0] < sto and orf[1] > sta:\n",
    "            ov = (min(orf[1], sto) - max(orf[0], sta)) / (orf[1] - orf[0])\n",
    "            max_ov = max(ov, max_ov)\n",
    "    if max_ov < 0.1:\n",
    "        non_overlapping_orfs.append(orf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2683970-7a45-419b-875a-0ca44f8b7a66",
   "metadata": {},
   "source": [
    "#### Produce FASTA file with CDS candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8564e695-944e-448f-b6e1-4936f0ae0f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_overlapping_orfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "214885e2-505b-42a5-a423-dd784b69cbd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "trans = util.Translator()\n",
    "temp = []\n",
    "for x in non_overlapping_orfs:\n",
    "    if x[4] < 1e-5 or x[4]==2:\n",
    "        if x[2] == 1:\n",
    "            prot = trans.translate_sequence(full_sequence[x[0]:x[1]], 1, 0)\n",
    "        else:\n",
    "            prot = trans.translate_sequence(util.reverse_complement(full_sequence[x[0]:x[1]]), 1, 0)\n",
    "        name = 'Start_'+str(x[0])+'_Stop_'+str(x[1])+'_Strand_'+str(x[2])\n",
    "        temp.append([name, prot[:-1]])\n",
    "util.produce_fasta_file(temp, project_dir + '/' + 'tb_orf_candidates.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02dbfb10-9277-4b21-9446-8fee25d4271a",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.run_tblastn('F:/Datasets/BLAST/actinobacteria_ref_genomes', 'tb_orf_candidates.faa', 'blastdb_sourceseq_actinobacteria', e_value = 1e-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3b5056-e487-4742-8f14-b64008c781ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "prob = []\n",
    "for x in orf_list:\n",
    "    prob.append(x[4])\n",
    "sns.histplot(prob, bins =100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93696eef-27de-4070-b4d8-3a47e7319a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "candidates = [x for x in orf_list if x[4] <1e-3]\n",
    "len(candidates)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0289a929-0d29-4246-a700-44840af34050",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Legacy code (might be useful - this is when tree was defined based on \"optimal splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9ffb9d2-7d32-4666-8b4f-f74a2f4a441b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_tree(species_to_split, position):\n",
    "    temp =  set([x[0] for x in reduced_position_dict[position][1:]]).intersection(species_to_split)\n",
    "    temp_2 = species_to_split - temp\n",
    "    if len(temp_2) > 1 and len(temp) > 1:\n",
    "        return ([(position, temp, len(temp)), (position, temp_2, len(temp_2))])\n",
    "    else:\n",
    "        return([(-1, species_to_split, len(species_to_split))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91cb7982-32e0-4eb1-afab-04b940de7ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def optimal_split_position(species_to_split_list):\n",
    "    best_split_num = 1e20\n",
    "    best_position = 0\n",
    "    splits_achieved_dict = {}\n",
    "    for k, v in reduced_position_dict.items():\n",
    "        mutation_count = 0\n",
    "        worst_count = 0\n",
    "        for species_to_split in species_to_split_list:\n",
    "            num_species_to_split = len(species_to_split)\n",
    "            optimal_split = int(num_species_to_split/2)\n",
    "            mutation_count += abs(len(set([x[0] for x in v[1:]]).intersection(species_to_split)) - optimal_split)\n",
    "            worst_count += optimal_split \n",
    "        splits_achieved_dict[k] = mutation_count\n",
    "        if abs(mutation_count) < abs(best_split_num):\n",
    "            best_position = k\n",
    "            best_split_num = abs(mutation_count)\n",
    "    return((best_position, best_split_num), (worst_count, splits_achieved_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d79dd452-3ecc-4b58-9fa0-1f8950d400fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_to_use = []\n",
    "reduced_position_dict =  {}\n",
    "for k, v in position_dict.items():\n",
    "    if len(v) >=100:\n",
    "        reduced_position_dict[k] = v\n",
    "print(len(position_dict), len(reduced_position_dict))\n",
    "\n",
    "\n",
    "all_species = [[-1,set(x for x in range(len(id_dict))),99,True]]\n",
    "pos = optimal_split_position([x[1] for x in all_species])[0][0]\n",
    "snps_to_use.append(pos)\n",
    "split_results = [split_tree(all_species[0][1], pos)]\n",
    "print( [(x[0], x[2]) for x in split_results[0]] )\n",
    "for i in range(1,50):\n",
    "    if i ==10:\n",
    "        break\n",
    "    start = time.process_time()\n",
    "    split_results.append([])\n",
    "    optimal_split_output = optimal_split_position([x[1] for x in split_results[i-1]])  \n",
    "    splits_achieved_dict = optimal_split_output[1][1]\n",
    "    optimal_split_score = optimal_split_output[1][0]\n",
    "    posn = optimal_split_output[0]\n",
    "    score_list = []\n",
    "    for k, v in splits_achieved_dict.items():\n",
    "        score_list.append(v)\n",
    "    score_list.sort()\n",
    "    score_p50 = score_list[int(len(score_list)/2)]\n",
    "    for k, v in splits_achieved_dict.items():\n",
    "        if v >= score_p50:#optimal_split_score - 1000:\n",
    "            reduced_position_dict.pop(k)\n",
    "    print(len(reduced_position_dict))\n",
    "    pos= posn[0]\n",
    "    snps_to_use.append(pos)\n",
    "    split_score = posn[1]\n",
    "    successful_splits = 0\n",
    "    for x in split_results[i-1]:\n",
    "        temp = split_tree(x[1], pos) \n",
    "        if len(temp) == 1:\n",
    "            split_results[i].append(temp[0])\n",
    "        else:\n",
    "            successful_splits +=1\n",
    "            split_results[i].append(temp[0])\n",
    "            split_results[i].append(temp[1])\n",
    "    print(max([x[2] for x in split_results[i]]))\n",
    "    print (time.process_time() - start)\n",
    "    print([(x[0], x[2]) for x in split_results[i]])\n",
    "    if successful_splits == 0:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a422156-b290-454c-b856-aca6e7fb5bde",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(optimal_split_score)\n",
    "splits_achieved_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7aef85b7-feb1-4670-8967-021f0a358395",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a81c4d9a-7759-4989-a706-0b2a0c868c98",
   "metadata": {},
   "outputs": [],
   "source": [
    "[(x[0],x[2]) for x in split_results[2] if x[2] > 300]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54b74d58-14a0-443a-96ab-8aa78a0fe0f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = []\n",
    "for res in split_results:\n",
    "    for info in res:\n",
    "        temp.append(info[0])\n",
    "snps = set(temp)\n",
    "snps.remove(-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adf9a4de-bd5b-40b8-a247-9a907ed4c342",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutation_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25046046-2644-4c23-aaa4-8176fb887d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mutation_count_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9431024b-eb44-45a3-b7b5-4e5b2a38a672",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(mutation_count_list, bins = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "890f331c-d6de-40d1-9cb5-8362c832e141",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = [x[1] for x in split_results[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47cc3f00-fb7d-4304-a5e1-0d94b4559401",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f684a584-361d-42d3-9cf4-45cbed3e5b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "distinct_ids = []\n",
    "variants = []\n",
    "for k, v in tqdm(variant_dict.items()):\n",
    "    if set(v) in variants:\n",
    "        continue\n",
    "    else:\n",
    "        variants.append(set(v))\n",
    "        distinct_ids.append(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca14ce1-f2b9-4a70-91b0-8e3f0d2f637d",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(variant_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88834d4c-38af-4314-b4ea-9f99491bb946",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(distinct_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "618ef69d-d9a6-4cfd-aa4b-d9248c9aafaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pow(2,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9db1ff2b-c3e3-4011-b49b-fbdb29bd9436",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(position_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f2c251-666a-428b-94bd-ba7bb5f5ab53",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "seq_length = 100\n",
    "\n",
    "with open(project_dir +'/test.faa', 'rb') as f:\n",
    "    sequence_to_score_dict = {}\n",
    "    sequence_to_score_dict['seq1'] = [{x} for x in 'AAATTTT']\n",
    "    sequence_to_score_dict['seq2'] = [{x} for x in 'AAACTTT']\n",
    "    sequence_to_score_dict['seq3'] = [{x} for x in 'AAAGTTT']\n",
    "    sequence_to_score_dict['seq4'] = [{x} for x in 'AAAATTT']\n",
    "master_tree2= ete3.Tree(project_dir + '/' + 'testtree.nwk')\n",
    "for node in master_tree2.traverse(\"postorder\"):\n",
    "    if node.is_leaf():\n",
    "        node.add_features(seq = sequence_to_score_dict[node.name])\n",
    "    else:\n",
    "        children = node.children\n",
    "        node.add_features(seq = fitch_1(children[0].seq, children[1].seq))\n",
    "\n",
    "#for k, v in sequence_to_score_dict.items():              \n",
    "    #seq_length = len(v)\n",
    "    #break\n",
    "mutation_counts = [0 for i in range(seq_length)]\n",
    "for node in master_tree2.traverse(\"preorder\"):\n",
    "    if node.is_leaf():\n",
    "        continue\n",
    "    if node.is_root():\n",
    "        node.seq = [{list(x)[0]} for x in node.seq]\n",
    "    children = node.children\n",
    "    mutations = []\n",
    "    child_sequences = []\n",
    "    for child in children:\n",
    "        (temp_1, temp_2) = fitch_2(node.seq ,child.seq)\n",
    "        child_sequences.append(temp_1)\n",
    "        child.seq = temp_1\n",
    "        mutations.append(temp_2)\n",
    "   \n",
    "    temp = []\n",
    "    for n, (h, i, j) in enumerate(zip(mutation_counts, mutations[0], mutations[1])):\n",
    "        if i + j == 0:\n",
    "            temp.append(h)\n",
    "        elif i + j == 1:\n",
    "            temp.append(h+1)\n",
    "        else:\n",
    "            if child_sequences[0][i] == child_sequences[1][i]:\n",
    "                temp.append(h+1)\n",
    "            else:\n",
    "                temp.append(h+2)\n",
    "    mutation_counts = temp     \n",
    "print(mutation_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "143a476b-f081-4a48-8ee6-b6a7303fa3a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7870105a-c5ba-45a9-8d63-3d735ac3b817",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict = {}\n",
    "for i in range(10000):\n",
    "    for j in range(10000):\n",
    "        dict[(i,j)] = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3d95928-2bc1-42b3-86f9-06690358d69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(range(27, 33))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c2e1a48-9e58-4739-984d-4398c6d93e35",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
