{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50da8fa4-e4d1-4e99-9df9-f2593cc0c057",
   "metadata": {},
   "source": [
    "##### Import modules and set up file locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9417e01-aa2a-4297-8948-c0c526434b04",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_run = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "694b2495-3118-4d6a-9934-5f46d80dfa5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nicho\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\ete3-3.1.2-py3.7.egg\\ete3\\evol\\parser\\codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "C:\\Users\\nicho\\Anaconda3\\envs\\Bioinformatics\\lib\\site-packages\\ete3-3.1.2-py3.7.egg\\ete3\\evol\\parser\\codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n"
     ]
    }
   ],
   "source": [
    "from Comparative_Analysis import Sequence_Analysis_Routines as sar\n",
    "from Comparative_Analysis import HMM as hmm\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from Comparative_Analysis import Alignment_HMM as alignment_hmm\n",
    "from Comparative_Analysis import Alignment_Analysis as alignment_analysis\n",
    "from Comparative_Analysis import Alignment as align\n",
    "from Comparative_Analysis import Master_Alignment_HMM as master_alignment_hmm\n",
    "from Comparative_Analysis import Multi_Species_Master_Alignment_HMM as multi_species_master_alignment_hmm\n",
    "from Comparative_Analysis import Arneson_Ernst_HMM as ae_hmm\n",
    "import random\n",
    "from joblib import Parallel, delayed\n",
    "from scipy import optimize as opt\n",
    "from tqdm import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import logomaker as lm\n",
    "import math\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import ete3;\n",
    "import pickle\n",
    "import copy\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "773a1d9c-8b02-447d-8b4c-41e9d5d24f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'D:/Project_Data/Project_5'\n",
    "genome_datasets_dir = project_dir + '/Datasets/NCBI_Datasets'\n",
    "output_dir = project_dir + '/Output'\n",
    "protein_fasta_output_loc = output_dir + '/Protein_Sequences'\n",
    "outgroup_protein_fasta_output_loc = output_dir + '/Protein_Sequences_With_Outgroup'\n",
    "sonic_paranoid_run_name = 'Run_Without_Outgroup'\n",
    "outgroup_sonic_paranoid_run_name = 'Run_With_Outgroup'\n",
    "sonic_paranoid_output_loc = output_dir + '/Sonic_Paranoid_Output'\n",
    "ortholog_dir = sonic_paranoid_output_loc + '/runs/' + sonic_paranoid_run_name + '/ortholog_groups'\n",
    "outgroup_ortholog_dir = sonic_paranoid_output_loc + '/runs/' + outgroup_sonic_paranoid_run_name + '/ortholog_groups'\n",
    "non_cds_output_dir = output_dir + '/Multiple_Alignment_Data/Downstream_Non_CDS'\n",
    "upstream_non_cds_output_dir = output_dir + '/Multiple_Alignment_Data/Upstream_Non_CDS'\n",
    "cds_output_dir = output_dir + '/Multiple_Alignment_Data/CDS'\n",
    "extended_cds_output_dir = output_dir + '/Multiple_Alignment_Data/Extended_CDS'\n",
    "outgroup_cds_output_dir = output_dir + '/Multiple_Alignment_Data/CDS_With_Outgroup'\n",
    "outgroup_concatenated_cds_output_dir = output_dir + '/Multiple_Alignment_Data/CDS_With_Outgroup_Concatenated'\n",
    "hmm_parameters_output_dir = output_dir +'/HMM_Model_Parameters'\n",
    "conservation_analysis_output_dir = output_dir + '/Conservation_Analysis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56139988-efa1-4bb1-9cd3-4cdd90020453",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "non_cds_offset = 50\n",
    "extended_cds_offset = 100\n",
    "tb_species = 'GCF_000195955.2'\n",
    "outgroup_species = 'GCF_000696675.2'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf829bb3-e529-4170-ad7b-9d885d57f851",
   "metadata": {},
   "source": [
    "##### Determine genomes in ortholog family, generate protein files and run Sonic Paranoid (both with and without outgroup - outgroup needed for tree building)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df7a7627-bb45-4679-97ef-ed8f209447fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_ids_with_outgroup = util.list_dirs(genome_datasets_dir)\n",
    "genome_ids = util.list_dirs(genome_datasets_dir)\n",
    "genome_ids.remove(outgroup_species)\n",
    "non_target_genome_ids = util.list_dirs(genome_datasets_dir)\n",
    "non_target_genome_ids.remove(outgroup_species)\n",
    "non_target_genome_ids.remove(tb_species)\n",
    "num_ids = len(genome_ids)\n",
    "num_ids_with_outgroup = len(genome_ids_with_outgroup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2be93725-4667-49ca-b2bc-430c0c7e9eae",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11/11 [00:08<00:00,  1.31it/s]\n",
      "100%|██████████| 12/12 [00:09<00:00,  1.30it/s]\n"
     ]
    }
   ],
   "source": [
    "if full_run == True:\n",
    "    for folder in sar.tqdm(genome_ids):\n",
    "        sar.generate_protein_file(genome_datasets_dir + '/' + folder + '/genomic.gbff', protein_fasta_output_loc + '/' + folder + '.faa')\n",
    "    for folder in sar.tqdm(genome_ids_with_outgroup):\n",
    "        sar.generate_protein_file(genome_datasets_dir + '/' + folder + '/genomic.gbff', outgroup_protein_fasta_output_loc + '/' + folder + '.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f828d159-34d6-4f7e-878c-e31b598573a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    sar.run_sonic_paranoid(protein_fasta_output_loc, sonic_paranoid_output_loc, sonic_paranoid_run_name)\n",
    "    sar.run_sonic_paranoid(outgroup_protein_fasta_output_loc, sonic_paranoid_output_loc, outgroup_sonic_paranoid_run_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39369771-8cd9-4438-89a5-d426702ac448",
   "metadata": {},
   "source": [
    "##### Generate objects containing orthologs and sequence information for each ortholog group / species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "051a50d7-e6dc-42a7-9675-e58d8ae157b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 75075/75075 [00:03<00:00, 24214.52it/s]\n",
      "100%|██████████| 84312/84312 [00:03<00:00, 23242.28it/s]\n"
     ]
    }
   ],
   "source": [
    "orthologs = sar.Ortholog_Grouping(ortholog_dir)\n",
    "outgroup_orthologs = sar.Ortholog_Grouping(outgroup_ortholog_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5e9977fb-f334-4f33-a820-535d87f4105a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:00<00:00, 212.55it/s]\n",
      "100%|██████████| 16/16 [00:00<00:00, 1022.39it/s]\n",
      "100%|██████████| 16/16 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "seq_data = sar.Ortholog_Sequence_Dataset(orthologs, genome_datasets_dir, genome_ids, non_cds_offset, tb_species) \n",
    "outgroup_seq_data = sar.Ortholog_Sequence_Dataset(outgroup_orthologs, genome_datasets_dir, genome_ids_with_outgroup, non_cds_offset, tb_species) \n",
    "all_copy_seq_data = sar.Ortholog_Sequence_Dataset(orthologs, genome_datasets_dir, genome_ids, non_cds_offset, tb_species, single_copy = False) \n",
    "#print(outgroup_seq_data.species_info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e6d5c8c-fdd4-459f-ad82-b74bb46e0ce7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEGCAYAAACUzrmNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAVOklEQVR4nO3df7DddX3n8eeLEDEtUqFcMCbBYJutQlpDSVla9g8Ut6TWKThbNKwtdJbZdGnc1dbtDtjdse5sZmvXqqsjbGmlxK3CZIsuUaCKKa11hgIXl5IEwpKRXzGRRG0XXLsphPf+cT6ph5vD/d6Ee+65uff5mDlzvuf9/XHeH5jkle+P8/2mqpAkaTLHjLoBSdLsZ1hIkjoZFpKkToaFJKmTYSFJ6nTsqBsYlpNPPrmWL18+6jYk6ahy3333fauqxibW52xYLF++nPHx8VG3IUlHlSSPD6p7GEqS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUqc5+6O8I/Xss8+ybdu2F9RWrlzJwoULR9SRJI2eYTHBtm3b+LVPbOaEV70GgKe/+TjXrIezzjprxJ1J0ugYFgOc8KrXcOJpPzbqNiRp1vCchSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkTkMLiyQvT3JPkr9Osj3JB1r9pCR3JHmkvZ/Yt87VSXYmeTjJhX31s5NsbfM+liTD6luSdKhh7lnsB95UVW8AVgFrkpwLXAVsqaoVwJb2mSRnAGuBM4E1wDVJFrRtXQusA1a015oh9i1JmmBoYVE9320fF7ZXARcBG1t9I3Bxm74IuKmq9lfVo8BO4Jwki4ETququqirgU33rSJJmwFDPWSRZkOR+YC9wR1XdDZxaVXsA2vspbfElwJN9q+9qtSVtemJ90PetSzKeZHzfvn3TOhZJms+GGhZVdaCqVgFL6e0lrJxk8UHnIWqS+qDvu66qVlfV6rGxscPuV5I02IxcDVVVfwv8Ob1zDU+1Q0u0971tsV3Asr7VlgK7W33pgLokaYYM82qosSSvbNOLgDcDO4DNwOVtscuBW9r0ZmBtkuOSnE7vRPY97VDVM0nObVdBXda3jiRpBgzzeRaLgY3tiqZjgE1V9YUkdwGbklwBPAFcAlBV25NsAh4EngPWV9WBtq0rgRuARcDt7SVJmiFDC4uqegA45PFyVfVt4IIXWWcDsGFAfRyY7HyHJGmI/AW3JKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoNLSySLEtyZ5KHkmxP8u5W/+0k30hyf3u9pW+dq5PsTPJwkgv76mcn2drmfSxJhtW3JOlQxw5x288B762qryV5BXBfkjvavI9U1Yf6F05yBrAWOBN4NfDlJP+oqg4A1wLrgL8CbgPWALcPsXdJUp+h7VlU1Z6q+lqbfgZ4CFgyySoXATdV1f6qehTYCZyTZDFwQlXdVVUFfAq4eFh9S5IONSPnLJIsB84C7m6ldyV5IMn1SU5stSXAk32r7Wq1JW16Yl2SNEOGHhZJjgduBt5TVU/TO6T0I8AqYA/wewcXHbB6TVIf9F3rkownGd+3b99LbV2S1Aw1LJIspBcUn66qzwJU1VNVdaCqngf+ADinLb4LWNa3+lJgd6svHVA/RFVdV1Wrq2r12NjY9A5GkuaxYV4NFeCTwENV9eG++uK+xd4GbGvTm4G1SY5LcjqwArinqvYAzyQ5t23zMuCWYfUtSTrUMK+GOg/4ZWBrkvtb7X3ApUlW0TuU9BjwqwBVtT3JJuBBeldSrW9XQgFcCdwALKJ3FZRXQknSDBpaWFTVVxl8vuG2SdbZAGwYUB8HVk5fd5Kkw+EvuCVJnQwLSVInw0KS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUyLCRJnQwLSVInw0KS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUaWlgkWZbkziQPJdme5N2tflKSO5I80t5P7Fvn6iQ7kzyc5MK++tlJtrZ5H0uSYfUtSTrUMPcsngPeW1WvB84F1ic5A7gK2FJVK4At7TNt3lrgTGANcE2SBW1b1wLrgBXttWaIfUuSJhhaWFTVnqr6Wpt+BngIWAJcBGxsi20ELm7TFwE3VdX+qnoU2Amck2QxcEJV3VVVBXyqbx1J0gyYkXMWSZYDZwF3A6dW1R7oBQpwSltsCfBk32q7Wm1Jm55YH/Q965KMJxnft2/ftI5BkuazoYdFkuOBm4H3VNXTky06oFaT1A8tVl1XVauravXY2NjhNytJGmioYZFkIb2g+HRVfbaVn2qHlmjve1t9F7Csb/WlwO5WXzqgLkmaIcO8GirAJ4GHqurDfbM2A5e36cuBW/rqa5Mcl+R0eiey72mHqp5Jcm7b5mV960iSZsCxQ9z2ecAvA1uT3N9q7wN+B9iU5ArgCeASgKranmQT8CC9K6nWV9WBtt6VwA3AIuD29pIkzZChhUVVfZXB5xsALniRdTYAGwbUx4GV09edJOlw+AtuSVInw0KS1GlKYZHkvKnUJElz01T3LD4+xZokaQ6a9AR3kp8GfgYYS/IbfbNOABYMXkuSNNd0XQ31MuD4ttwr+upPA784rKYkSbPLpGFRVX8B/EWSG6rq8RnqSZI0y0z1dxbHJbkOWN6/TlW9aRhNSZJml6mGxf8A/hvwh8CBjmUlSXPMVMPiuaq6dqidSJJmraleOvv5JL+WZHF7LOpJSU4aameSpFljqnsWB+8S+5t9tQJeO73tSJJmoymFRVWdPuxGJEmz15TCIsllg+pV9anpbUeSNBtN9TDUT/VNv5zeLca/BhgWkjQPTPUw1L/u/5zkh4D/PpSOJEmzzpHeovx79B57KkmaB6Z6zuLz9K5+gt4NBF8PbBpWU5Kk2WWq5yw+1Df9HPB4Ve0aQj+SpFloSoeh2g0Fd9C78+yJwN8PsylJ0uwy1SflvR24B7gEeDtwdxJvUS5J88RUD0P9FvBTVbUXIMkY8GXgT4bVmCRp9pjq1VDHHAyK5tuHsa4k6Sg31b/w/zTJF5P8SpJfAW4FbptshSTXJ9mbZFtf7beTfCPJ/e31lr55VyfZmeThJBf21c9OsrXN+1iSHN4QJUkv1aRhkeRHk5xXVb8J/D7wE8AbgLuA6zq2fQOwZkD9I1W1qr1ua99zBrAWOLOtc02Sg8/4vhZYR+93HSteZJuSpCHq2rP4KPAMQFV9tqp+o6p+nd5exUcnW7GqvgJ8Z4p9XATcVFX7q+pRYCdwTpLFwAlVdVdVFb3bi1w8xW1KkqZJV1gsr6oHJharapzeI1aPxLuSPNAOU53YakuAJ/uW2dVqS9r0xPpASdYlGU8yvm/fviNsT5I0UVdYvHySeYuO4PuuBX4EWAXsAX6v1Qedh6hJ6gNV1XVVtbqqVo+NjR1Be5KkQbrC4t4k/3JiMckVwH2H+2VV9VRVHaiq54E/AM5ps3YBy/oWXQrsbvWlA+qSpBnU9TuL9wCfS/JOvh8Oq4GXAW873C9Lsriq9rSPbwMOXim1GfhMkg8Dr6Z3IvueqjqQ5Jkk5wJ3A5cBHz/c75UkvTSThkVVPQX8TJI3Aitb+daq+rOuDSe5ETgfODnJLuD9wPlJVtE7lPQY8Kvte7Yn2QQ8SO/eU+ur6kDb1JX0rqxaBNzeXpKkGTTV51ncCdx5OBuuqksHlD85yfIbgA0D6uN8P6gkSSPgr7AlSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUyLCRJnQwLSVInw0KS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUyLCRJnQwLSVInw0KS1GloYZHk+iR7k2zrq52U5I4kj7T3E/vmXZ1kZ5KHk1zYVz87ydY272NJMqyeJUmDDXPP4gZgzYTaVcCWqloBbGmfSXIGsBY4s61zTZIFbZ1rgXXAivaauE1J0pANLSyq6ivAdyaULwI2tumNwMV99Zuqan9VPQrsBM5Jshg4oaruqqoCPtW3jiRphsz0OYtTq2oPQHs/pdWXAE/2Lber1Za06Yn1gZKsSzKeZHzfvn3T2rgkzWez5QT3oPMQNUl9oKq6rqpWV9XqsbGxaWtOkua7mQ6Lp9qhJdr73lbfBSzrW24psLvVlw6oS5Jm0EyHxWbg8jZ9OXBLX31tkuOSnE7vRPY97VDVM0nObVdBXda3jiRphhw7rA0nuRE4Hzg5yS7g/cDvAJuSXAE8AVwCUFXbk2wCHgSeA9ZX1YG2qSvpXVm1CLi9vSRJM2hoYVFVl77IrAteZPkNwIYB9XFg5TS2Jkk6TLPlBLckaRYzLCRJnQwLSVInw0KS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUyLCRJnQwLSVInw0KS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUqeRhEWSx5JsTXJ/kvFWOynJHUkeae8n9i1/dZKdSR5OcuEoepak+WyUexZvrKpVVbW6fb4K2FJVK4At7TNJzgDWAmcCa4BrkiwYRcOSNF/NpsNQFwEb2/RG4OK++k1Vtb+qHgV2AufMfHuSNH+NKiwK+FKS+5Ksa7VTq2oPQHs/pdWXAE/2rbur1SRJM+TYEX3veVW1O8kpwB1JdkyybAbUauCCveBZB3Daaae99C4lScCI9iyqand73wt8jt5hpaeSLAZo73vb4ruAZX2rLwV2v8h2r6uq1VW1emxsbFjtS9K8M+NhkeQHk7zi4DTws8A2YDNweVvscuCWNr0ZWJvkuCSnAyuAe2a2a0ma30ZxGOpU4HNJDn7/Z6rqT5PcC2xKcgXwBHAJQFVtT7IJeBB4DlhfVQdG0LckzVszHhZV9XXgDQPq3wYueJF1NgAbhtyaJOlFzKZLZyVJs5RhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6zfgzuI82zx94jh07drygtnLlShYuXDiijiRp5hkWHb677xv87q37OWX7fgCe/ubjXLMezjrrrBF3Jkkzx7CYguNPWcaJp/3YqNuQpJHxnIUkqZNhIUnqdNSERZI1SR5OsjPJVaPuR5Lmk6PinEWSBcAngH8K7ALuTbK5qh4cZV/PPvss27Zte8Fn4AVXSnnllKS54KgIC+AcYGdVfR0gyU3ARcBQwuLpbz7+D9P/91t7WPD/9vM3P7DoH+bt2HEcADt27GDDp7/MD/zwqwD49te3s2DRK3jl4tMA+N63v8lvvfPNvO51rxtGm5J0iGFdqXm0hMUS4Mm+z7uAfzxxoSTrgHXt43eTPNyx3ZOBbx1uM3/1yakv+89vvf5wNz8Tjmjcc4Djnl8c95F5zaDi0RIWGVCrQwpV1wHXTXmjyXhVrX4pjR2NHPf84rjnl2GN+2g5wb0LWNb3eSmwe0S9SNK8c7SExb3AiiSnJ3kZsBbYPOKeJGneOCoOQ1XVc0neBXwRWABcX1Xbp2HTUz5kNcc47vnFcc8vQxl3qg459C9J0gscLYehJEkjZFhIkjrNy7CYT7cOSXJ9kr1JtvXVTkpyR5JH2vuJo+xxuiVZluTOJA8l2Z7k3a0+18f98iT3JPnrNu4PtPqcHvdBSRYk+V9JvtA+z5dxP5Zka5L7k4y32rSPfd6FRd+tQ34OOAO4NMkZo+1qqG4A1kyoXQVsqaoVwJb2eS55DnhvVb0eOBdY3/4fz/Vx7wfeVFVvAFYBa5Kcy9wf90HvBh7q+zxfxg3wxqpa1ff7imkf+7wLC/puHVJVfw8cvHXInFRVXwG+M6F8EbCxTW8ELp7JnoatqvZU1dfa9DP0/gJZwtwfd1XVd9vHhe1VzPFxAyRZCvw88Id95Tk/7klM+9jnY1gMunXIkhH1MiqnVtUe6P3FCpwy4n6GJsly4CzgbubBuNuhmPuBvcAdVTUvxg18FPh3wPN9tfkwbuj9g+BLSe5rtzyCIYz9qPidxTSb0q1DdPRLcjxwM/Ceqno6GfS/fm6pqgPAqiSvBD6XZOWIWxq6JG8F9lbVfUnOH3E7o3BeVe1OcgpwR5Idw/iS+bhn4a1D4KkkiwHa+94R9zPtkiykFxSfrqrPtvKcH/dBVfW3wJ/TO18118d9HvALSR6jd1j5TUn+mLk/bgCqand73wt8jt6h9mkf+3wMC28d0hvv5W36cuCWEfYy7dLbhfgk8FBVfbhv1lwf91jboyDJIuDNwA7m+Lir6uqqWlpVy+n9ef6zqvol5vi4AZL8YJJXHJwGfhbYxhDGPi9/wZ3kLfSOcR68dciG0XY0PEluBM6nd9vip4D3A/8T2AScBjwBXFJVE0+CH7WS/BPgL4GtfP8Y9vvonbeYy+P+CXonMxfQ+4fgpqr6j0l+mDk87n7tMNS/raq3zodxJ3ktvb0J6J1W+ExVbRjG2OdlWEiSDs98PAwlSTpMhoUkqZNhIUnqZFhIkjoZFpKkToaFNAskue3gbySk2chLZyVJndyzkKao/Vr21va8iG1J3tGeJfDB9hyJe5L8aFt2LMnNSe5tr/Na/fgkf9SeP/BAkn/W6o8lOblN/1Lb1v1Jfr/dHHBBkhva925N8uuj+y+h+Wg+3khQOlJrgN1V9fMASX4I+CDwdFWdk+QyencGeCvwX4GPVNVXk5wGfBF4PfAfgP9TVT/etvGCh9IkeT3wDno3h3s2yTXAO4HtwJKqWtmWe+WwByv1MyykqdsKfCjJB4EvVNVftjvZ3tjm3wh8pE2/GTij7063J7R7+LyZ3v2LAKiqv5nwHRcAZwP3tnUX0bsJ3OeB1yb5OHAr8KXpHZo0OcNCmqKq+t9JzgbeAvznJAf/wu4/8Xdw+hjgp6vq7/q30W5yONmJwgAbq+rqQ2YkbwAuBNYDbwf+xRENRDoCnrOQpijJq4HvVdUfAx8CfrLNekff+11t+kvAu/rWXfUi9YnPRt4C/GJ7NsHBZym/pp3POKaqbqZ3KOsnkWaQexbS1P048F+SPA88C1wJ/AlwXJK76f3j69K27L8BPpHkAXp/zr4C/CvgP7X6NuAA8AHg4PM2qKoHk/x7ek8+O6Z9z3rg74A/ajWAQ/Y8pGHy0lnpJWgP3FldVd8adS/SMHkYSpLUyT0LSVIn9ywkSZ0MC0lSJ8NCktTJsJAkdTIsJEmd/j+P/zdtsyxv6AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#seq_data.generate_synteny_plot()\n",
    "#seq_data.generate_ortholog_count_plot()\n",
    "all_copy_seq_data.generate_master_count_plot()\n",
    "#seq_data.generate_unassigned_gene_count_plot()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60e3b630-5c42-499e-8389-5532f4dca58c",
   "metadata": {},
   "source": [
    "##### Perform CDS and non-CDS alignments for each full ortholog group and save to folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7cb33f80-a608-425f-bf01-6a0fe7a77226",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:00<00:00, 7995.81it/s]\n",
      "100%|██████████| 16/16 [00:00<?, ?it/s]\n",
      "100%|██████████| 16/16 [00:00<?, ?it/s]\n",
      "100%|██████████| 16/16 [00:00<?, ?it/s]\n",
      "100%|██████████| 16/16 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "min_species = num_ids\n",
    "min_species_with_outgroup = num_ids_with_outgroup\n",
    "groups = random.sample(orthologs.full_single_copy_ortholog_groups, len(orthologs.full_single_copy_ortholog_groups))  #Permutation ensures even distribution of processing speeds\n",
    "outgroup_groups = random.sample(outgroup_orthologs.full_single_copy_ortholog_groups, len(outgroup_orthologs.full_single_copy_ortholog_groups))  #Permutation ensures even distribution of processing speeds\n",
    "if full_run == True:\n",
    "    par = Parallel(n_jobs=-1)(delayed(align.align_and_build)(outgroup_groups, num_cores, core_number, outgroup_seq_data.sequence_data, 'cds_length', 'cds_seq', outgroup_cds_output_dir+'/', min_species_with_outgroup) for core_number in tqdm(core_numbers))\n",
    "    par = Parallel(n_jobs=-1)(delayed(align.align_and_build)(groups, num_cores, core_number, seq_data.sequence_data, 'cds_length', 'cds_seq', cds_output_dir+'/', min_species) for core_number in tqdm(core_numbers))\n",
    "    par = Parallel(n_jobs=-1)(delayed(align.align_and_build)(groups, num_cores, core_number, seq_data.sequence_data, 'non_cds_offset_length', 'non_cds_offset_seq', non_cds_output_dir+'/', min_species) for core_number in tqdm(core_numbers))\n",
    "    par = Parallel(n_jobs=-1)(delayed(align.align_and_build)(groups, num_cores, core_number, seq_data.sequence_data, 'upstream_non_cds_offset_length', 'upstream_non_cds_offset_seq', upstream_non_cds_output_dir+'/', min_species) for core_number in tqdm(core_numbers))\n",
    "    par = Parallel(n_jobs=-1)(delayed(align.align_and_build)(groups, num_cores, core_number, seq_data.sequence_data, 'cds_extended_region_length', 'cds_extended_region_seq', extended_cds_output_dir+'/', min_species) for core_number in tqdm(core_numbers))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4552a5b9-a396-450f-956b-0494d74e13af",
   "metadata": {},
   "source": [
    "##### Run IQTree on concatenated CDS alignments to generate tree and rename with full names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "42928a48-444f-4b46-a8eb-d668dd4565e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    alignment_names = util.list_files(outgroup_cds_output_dir)\n",
    "    util.concatenate_fasta(outgroup_cds_output_dir, alignment_names, outgroup_concatenated_cds_output_dir + '/concatenated_cds.fasta')\n",
    "    subprocess.run('cd \\\\users\\\\nicho\\\\IQTree & bin\\\\iqtree2 -s ' + outgroup_concatenated_cds_output_dir + '/concatenated_cds.fasta' + ' --prefix '+ output_dir + \n",
    "                   '/Trees/Concatenated_JC_Tree -m JC -B 1000 -T AUTO -o ' + outgroup_species, shell=True)\n",
    "    master_tree = ete3.Tree(output_dir + '/Trees/Concatenated_JC_Tree.treefile')\n",
    "    for node in master_tree.traverse():\n",
    "        if node.is_leaf():\n",
    "            node.name = outgroup_seq_data.organism_dict[node.name] \n",
    "    master_tree.write(format=0, outfile= output_dir + '/Trees/Concatenated_JC_Tree_Full_Names.treefile')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1a16ea0b-011d-45b8-9716-609647baa411",
   "metadata": {},
   "outputs": [],
   "source": [
    "master_tree = ete3.Tree(output_dir + '/Trees/Concatenated_JC_Tree.treefile')\n",
    "phylogenetic_distance_dict = {}\n",
    "for gid in non_target_genome_ids:\n",
    "    phylogenetic_distance_dict[gid] = master_tree.get_distance(gid, tb_species)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40396ae4-b80d-4ebd-ab66-e40e6712c758",
   "metadata": {},
   "source": [
    "##### Fit overall Alignment and Pairwise HMMs using EM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4e885b08-1828-4633-9c60-4eeaab22efcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_symbols = 4\n",
    "num_states = 3\n",
    "minimum_fit_length = 10\n",
    "initial_params = [0.95, 0.5, 0.95, 0.5, 0.95, 0.5, 0.56370018, 0.52131172, 0.33906948]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e7da4636-b1db-4d25-a846-72c8fc4190a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    Alignment_HMM_Model = alignment_hmm.Alignment_HMM (num_symbols, num_states, non_cds_output_dir, tb_species, species_order = genome_ids)\n",
    "    parameter_fits = Alignment_HMM_Model.EM_update(num_cores, initial_params, non_cds_offset, minimum_fit_length)\n",
    "    fitted_parameters = parameter_fits[3]\n",
    "    with open(hmm_parameters_output_dir + '/' + 'full_parameter_fit.pkl', 'wb') as f:\n",
    "        pickle.dump(fitted_parameters, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "54b1b4b3-11cc-4e5a-aea4-3d29f0810066",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Alignment_HMM_Model' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_7712/1411489679.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[0mparameter_fits\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mid\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mnon_target_genome_ids\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m         \u001b[0mparameter_fits\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mAlignment_HMM_Model\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mEM_update\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnum_cores\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minitial_params\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnon_cds_offset\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mminimum_fit_length\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mall_species\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcomparison_species\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mid\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m3\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m     \u001b[1;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhmm_parameters_output_dir\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'/'\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'pairwise_parameter_fits.pkl'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'wb'\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m         \u001b[0mpickle\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mparameter_fits\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'Alignment_HMM_Model' is not defined"
     ]
    }
   ],
   "source": [
    "if full_run == True:\n",
    "    parameter_fits = []\n",
    "    for id in non_target_genome_ids:\n",
    "        parameter_fits.append((id, Alignment_HMM_Model.EM_update(num_cores, initial_params, non_cds_offset, minimum_fit_length, all_species = False, comparison_species = id)[3]))\n",
    "    with open(hmm_parameters_output_dir + '/' + 'pairwise_parameter_fits.pkl', 'wb') as f:\n",
    "        pickle.dump(parameter_fits, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75965597-303a-442b-9c61-2f2d1420e1b0",
   "metadata": {},
   "source": [
    "##### Fit overall HMM to pairwise HMMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55473b24-e661-4d7f-a9df-e4a485723fda",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parallel_generate_pairwise_state_probabilities(num_subsets, subset_num, ids, num_states, pairwise_fitted_parameters):\n",
    "    ids = util.chunk_list(ids, num_subsets, subset_num)\n",
    "    pairwise_observation_probabilities = []\n",
    "    for group_id in ids:\n",
    "        temp = []\n",
    "        alignment = align.Alignment(non_cds_output_dir +'/'+str(group_id)+'.fasta', tb_species, 'NT', species_order = genome_ids)\n",
    "        alignment.modify_sequence(1,False,False)\n",
    "        for params in pairwise_fitted_parameters:\n",
    "                transition_probabilities, mutation_probabilities = Alignment_HMM_Model.alignment_hmm_model_inputs(params[1])\n",
    "                observation_probabilities = Alignment_HMM_Model.calculate_observation_probs(mutation_probabilities, alignment.modified_sequence_list, alignment, all_species=False, comparison_species = params[0])\n",
    "                initial_state_probabilities = [1.0/num_states]*num_states\n",
    "                hmm_model = hmm.HMM(initial_state_probabilities, transition_probabilities, observation_probabilities)\n",
    "                hmm_model.calculate_probabilities()\n",
    "                temp.append(hmm_model.state_probabilities)\n",
    "        pairwise_observation_probabilities.append(temp) \n",
    "    return pairwise_observation_probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5182ac-6442-40fc-9286-c32fcb173955",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    with open(hmm_parameters_output_dir + '/' + 'pairwise_parameter_fits.pkl', 'rb') as f:\n",
    "        pairwise_fitted_parameters = pickle.load(f)\n",
    "    file_ids = util.list_files(non_cds_output_dir + '/')\n",
    "    ids = [int(i.split('.')[0]) for i in file_ids] \n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(parallel_generate_pairwise_state_probabilities)(num_cores, core_number, ids, num_states, pairwise_fitted_parameters) for core_number in core_numbers)\n",
    "    pairwise_observation_probabilities = [item for sublist in parallel_output for item in sublist]\n",
    "    with open(hmm_parameters_output_dir + '/' + 'pairwise_observation_probabilities.pkl', 'wb') as f:\n",
    "        pickle.dump(pairwise_observation_probabilities, f)\n",
    "else:\n",
    "    with open(hmm_parameters_output_dir + '/' + 'pairwise_observation_probabilities.pkl', 'rb') as f:\n",
    "        pairwise_observation_probabilities = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "892c503c-dec6-4473-aa3b-38b41ed2b3fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    Master_Alignment_HMM_Model = master_alignment_hmm.Master_Alignment_HMM (pairwise_observation_probabilities)\n",
    "    initial_params = [0.8, 0.5, 0.9, 0.3]\n",
    "    parameter_fits = Master_Alignment_HMM_Model.EM_update(num_cores, initial_params, non_cds_offset, minimum_fit_length)\n",
    "    fitted_parameters = parameter_fits[3]\n",
    "    print(fitted_parameters)\n",
    "    with open(hmm_parameters_output_dir + '/' + 'master_parameter_fit.pkl', 'wb') as f:\n",
    "        pickle.dump(fitted_parameters, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c7ee55c-6761-4018-a55e-5f010e7ff587",
   "metadata": {},
   "source": [
    "#####  Fit overall species specific HMM to pairwise HMMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39bbb69f-fcbd-49f9-b352-459ab22e3038",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    num_comparison_species = len(pairwise_observation_probabilities[0])\n",
    "    Multi_Species_Master_Alignment_HMM_Model = multi_species_master_alignment_hmm.Multi_Species_Master_Alignment_HMM (pairwise_observation_probabilities)\n",
    "    initial_params = [0.8, 0.5, [0.9]*num_comparison_species, [0.3]*num_comparison_species]\n",
    "    parameter_fits = Multi_Species_Master_Alignment_HMM_Model.EM_update(num_cores, initial_params, non_cds_offset, minimum_fit_length)\n",
    "    #print(parameter_fits)\n",
    "    fitted_parameters = parameter_fits[3]\n",
    "    print(fitted_parameters)\n",
    "    with open(hmm_parameters_output_dir + '/' + 'multi_species_master_parameter_fit.pkl', 'wb') as f:\n",
    "        pickle.dump(fitted_parameters, f)\n",
    "    species_conservation_parameter_dict = {}    # Used to plot against phylogenetic distance below\n",
    "    for i, gid in enumerate(non_target_genome_ids):\n",
    "        species_conservation_parameter_dict[gid] = parameter_fits[1][0][i]\n",
    "    with open(hmm_parameters_output_dir + '/' + 'multi_species_conservation_parameter_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(species_conservation_parameter_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27d25ad-5053-4d86-a9d4-c6f8ba51527e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(hmm_parameters_output_dir + '/' + 'multi_species_conservation_parameter_dict.pkl', 'rb') as f:\n",
    "    species_conservation_parameter_dict = pickle.load(f)\n",
    "species_name_labels = []\n",
    "log_conservation = []\n",
    "phylo_distance = []\n",
    "for gid in non_target_genome_ids:\n",
    "    species_name_labels.append(gid)\n",
    "    log_conservation.append(-1*math.log(species_conservation_parameter_dict[gid]))\n",
    "    phylo_distance.append(phylogenetic_distance_dict[gid])\n",
    "conservations_df = pd.DataFrame(\n",
    "    {'Species': species_name_labels,\n",
    "     '-Log conservation probability': log_conservation,\n",
    "     'Phylogenetic distance': phylo_distance\n",
    "    })\n",
    "ax = sns.scatterplot(data = conservations_df, x = 'Phylogenetic distance', y = '-Log conservation probability', s= 20, hue = 'Species')\n",
    "plt.setp(ax.get_legend().get_texts(), fontsize='6') # for legend text\n",
    "plt.setp(ax.get_legend().get_title(), fontsize='8') # for legend title\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016fb960-7994-4247-9feb-154d3609cb65",
   "metadata": {},
   "source": [
    "#####  Fit Arneson Ernst HMM to alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01765f1a-dd3f-4c71-a43d-ad67363c5cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_states = 3\n",
    "num_symbols = 4\n",
    "minimum_fit_length = 10\n",
    "initial_mutation_probs = np.random.rand(num_states,num_ids,3)\n",
    "for i in range(num_states):\n",
    "    for j in range(num_ids):\n",
    "        tot = 0\n",
    "        for k in range(3):\n",
    "            tot += initial_mutation_probs[i, j, k]\n",
    "        for k in range(3):\n",
    "            initial_mutation_probs[i, j, k] = initial_mutation_probs[i, j, k] / tot\n",
    "#print(initial_mutation_probs)\n",
    "initial_params = [0.95, 0.5, 0.95, 0.5, 0.95, 0.5, initial_mutation_probs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f1ba083-9720-4fb3-81a6-49c308b87398",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    AE_HMM_Model = ae_hmm.Alignment_HMM (num_symbols, num_states, non_cds_output_dir, tb_species, species_order = genome_ids)\n",
    "    parameter_fits = AE_HMM_Model.EM_update(num_cores, initial_params, non_cds_offset, minimum_fit_length)\n",
    "    print(parameter_fits)\n",
    "    fitted_parameters = parameter_fits[3]\n",
    "    with open(hmm_parameters_output_dir + '/' + 'AE_parameter_fit.pkl', 'wb') as f:\n",
    "        pickle.dump(fitted_parameters, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3669c845-286e-4aa1-97e4-8a2dee230519",
   "metadata": {},
   "source": [
    "##### Load HMM parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dea7f2a-cd8b-4d52-bf7b-f3b8bde0536e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(hmm_parameters_output_dir + '/' + 'full_parameter_fit.pkl', 'rb') as f:\n",
    "    full_fitted_parameters = pickle.load(f)\n",
    "with open(hmm_parameters_output_dir + '/' + 'pairwise_parameter_fits.pkl', 'rb') as f:\n",
    "    pairwise_fitted_parameters = pickle.load(f)\n",
    "with open(hmm_parameters_output_dir + '/' + 'master_parameter_fit.pkl', 'rb') as f:\n",
    "    master_fitted_parameters = pickle.load(f)\n",
    "with open(hmm_parameters_output_dir + '/' + 'multi_species_master_parameter_fit.pkl', 'rb') as f:\n",
    "    multi_species_master_fitted_parameters = pickle.load(f)\n",
    "with open(hmm_parameters_output_dir + '/' + 'AE_parameter_fit.pkl', 'rb') as f:\n",
    "    AE_fitted_parameters = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d481cd-94cd-4192-9359-cc1e598a6847",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for params in pairwise_fitted_parameters:\n",
    "#    transition_probabilities, mutation_probabilities = copy.deepcopy(Alignment_HMM_Model.alignment_hmm_model_inputs(params[1]))\n",
    "    #print(seq_data.organism_dict[params[0]],transition_probabilities, mutation_probabilities)\n",
    "#transition_probabilities, mutation_probabilities = copy.deepcopy(Master_Alignment_HMM_Model.alignment_hmm_model_inputs(master_fitted_parameters))\n",
    "#print(transition_probabilities, mutation_probabilities)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d3d362-46c0-40e5-9ecc-dbd80525bcfa",
   "metadata": {},
   "source": [
    "##### Analyse ortholog groups for conservation and other features and output to dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c56c44ad-31d8-407f-88b1-591812aa7bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "Alignment_HMM_Model = alignment_hmm.Alignment_HMM (4, 3, non_cds_output_dir, tb_species, species_order = genome_ids)\n",
    "Master_Alignment_HMM_Model = master_alignment_hmm.Master_Alignment_HMM (pairwise_observation_probabilities)\n",
    "Multi_Species_Master_Alignment_HMM_Model = multi_species_master_alignment_hmm.Multi_Species_Master_Alignment_HMM (pairwise_observation_probabilities)\n",
    "AE_HMM_Model = ae_hmm.Alignment_HMM (4, 3, non_cds_output_dir, tb_species, species_order = genome_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "153917ac-dd0e-444e-bbc8-c0f8f9792628",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parallel_build_analysis_dictionary(num_subsets, subset_num, ids, analysis_type, model_type):\n",
    "    ids = util.chunk_list(ids, num_subsets, subset_num)\n",
    "    output_list = []\n",
    "    for group_id in ids:\n",
    "        alignment = align.Alignment(alignment_dir+'/'+str(group_id)+'.fasta', tb_species, 'NT')\n",
    "        if model_type == 'Simple':\n",
    "            analysis = alignment_analysis.Alignment_Analysis(analysis_type, alignment, seq_data, non_cds_offset, group_id, 3, pairwise_fitted_parameters, 'Simple', 2, \n",
    "                                               master_fitted_parameters, non_cds_output_dir, tb_species, genome_ids, pairwise_observation_probabilities, Alignment_HMM_Model, Master_Alignment_HMM_Model)\n",
    "        elif model_type == 'Multi_Species':\n",
    "            analysis = alignment_analysis.Alignment_Analysis(analysis_type, alignment, seq_data, non_cds_offset, group_id, 3, pairwise_fitted_parameters, 'Multi_Species', 2, \n",
    "                                           multi_species_master_fitted_parameters, non_cds_output_dir, tb_species, genome_ids, pairwise_observation_probabilities, \n",
    "                                           Alignment_HMM_Model, Multi_Species_Master_Alignment_HMM_Model)\n",
    "        elif model_type == 'AE':\n",
    "            analysis = alignment_analysis.Alignment_Analysis(analysis_type, alignment, seq_data, non_cds_offset, group_id, 3, pairwise_fitted_parameters, 'AE', 3, \n",
    "                                           AE_fitted_parameters, non_cds_output_dir, tb_species, genome_ids, pairwise_observation_probabilities, \n",
    "                                           Alignment_HMM_Model, AE_HMM_Model)\n",
    "        else:\n",
    "            pass\n",
    "        output_list.append((group_id, analysis))\n",
    "    return output_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c8f2946-1234-4fb8-acdd-e936b8995dec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if full_run == True:\n",
    "if 1==1:\n",
    "    for model_type in ['Simple', 'Multi_Species', 'AE']:\n",
    "        for analysis_type in ['Downstream', 'Upstream']:\n",
    "            if analysis_type == 'Downstream':\n",
    "                alignment_dir = non_cds_output_dir\n",
    "                dict_name = model_type + '_' + 'downstream_conservation_info_dictionary'\n",
    "            else:\n",
    "                alignment_dir = upstream_non_cds_output_dir\n",
    "                dict_name = model_type + '_' + 'upstream_conservation_info_dictionary'\n",
    "            alignment_info_dict = {}\n",
    "            file_ids = util.list_files(alignment_dir+'/')\n",
    "            ids = [int(i.split('.')[0]) for i in file_ids]\n",
    "            #parallel_output = []\n",
    "            #for core_number in core_numbers:\n",
    "            #    parallel_output.append(parallel_build_analysis_dictionary(num_cores, core_number, ids, analysis_type))\n",
    "            parallel_output = Parallel(n_jobs=-1)(delayed(parallel_build_analysis_dictionary)(num_cores, core_number, ids, analysis_type, model_type) for core_number in tqdm(core_numbers))\n",
    "            dictionary_list = [item for sublist in parallel_output for item in sublist]\n",
    "            alignment_info_dict = {}\n",
    "            for (group, analysis) in dictionary_list:\n",
    "                alignment_info_dict[seq_data.master_species_info(group, 'locus_tag')] = analysis\n",
    "            with open(conservation_analysis_output_dir + '/' + dict_name + '.pkl', 'wb') as f:\n",
    "                pickle.dump(alignment_info_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e70cae-421b-48b8-a511-d47b862a8850",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
