{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "898a25bb-a79e-4516-9f4d-7315c4699f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "from tqdm.auto import tqdm\n",
    "from Comparative_Analysis import Sequence_Analysis_Routines as sar\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from Comparative_Analysis import Alignment as align\n",
    "import random\n",
    "import copy\n",
    "from joblib import Parallel, delayed\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import re\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b2e7e6e9-ef16-402f-bd69-b6578ee5f7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_run = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3915b302-59d1-4f6d-9488-d7c9eecaf248",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'D:/Project_Data/Project_8'\n",
    "datasets_dir = project_dir + '/Datasets'\n",
    "output_dir = project_dir + '/Output'\n",
    "seq_dir = 'D:/Actinobacteria_Ref_Rep_Lev_Complete'\n",
    "blast_dir = 'D:/BLAST/actinobacteria_ref_rep_comp'\n",
    "blast_db_name = 'actinobacteria_ref_rep_comp'\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "reference_species_filename = 'GCF_000195955.2_ASM19595v2_genomic.gbff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "258cb331-9ecc-4161-83d6-3f59e95698e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = util.list_files(seq_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3e20e06-57aa-4658-8b67-fd52c03b12d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    reference_list = []\n",
    "    for species in tqdm(species_list):\n",
    "        for genome_record in SeqIO.parse(seq_dir + '/' + species, \"genbank\"):\n",
    "            feature_info = []\n",
    "            organism_name = genome_record.annotations['organism']\n",
    "            accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "            organism_accession = organism_name.replace(' ', '_') + '_' + accession_ver\n",
    "            reference_list.append([organism_accession, str(genome_record.seq)])\n",
    "    util.produce_fasta_file(reference_list, seq_dir + '/all_actinobacteria_ref_rep_comp.faa')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d1e98a5-ef97-45a0-9c6e-ef2ecbaf6155",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    w_d = os.getcwd()\n",
    "    os.chdir(\"D:/\")\n",
    "    subprocess.run('cd '+ seq_dir + ' &  makeblastdb -in all_actinobacteria_ref_rep_comp.faa -dbtype nucl -out ' + blast_db_name, shell=True, capture_output = True)\n",
    "    os.chdir(w_d)\n",
    "    files_to_move = [x for x in util.list_files(seq_dir) if x[:-4] == blast_db_name]\n",
    "    for file in files_to_move:\n",
    "        source = seq_dir + '/' + file\n",
    "        destination = blast_dir + '/' + file\n",
    "        shutil.move(source, destination)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf2d72b7-8100-48f1-8a60-5d3855eceeb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_sequence_info(num_subsets, subset_num, species_master_list):\n",
    "    sequence_info = []\n",
    "    sequence_list = util.chunk_list(species_master_list, num_subsets, subset_num)\n",
    "    for sequence in sequence_list:\n",
    "        for genome_record in SeqIO.parse(seq_dir + '/' + sequence, \"genbank\"):\n",
    "            feature_info = []\n",
    "            organism_name = genome_record.annotations['organism']\n",
    "            accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "            organism_accession = organism_name.replace(' ', '_') + '_' + accession_ver\n",
    "            locus_tag = ''\n",
    "            for feature in genome_record.features:\n",
    "                if locus_tag == '':\n",
    "                    previous_locus_tag = ''\n",
    "                    previous_end = 0\n",
    "                else:\n",
    "                    previous_locus_tag = locus_tag\n",
    "                    previous_end = end\n",
    "                a = feature.qualifiers\n",
    "                \n",
    "                if feature.type != 'gene' and not(feature.type == 'source'):\n",
    "                    if a.get(\"locus_tag\") != None:\n",
    "                        locus_tag = a.get(\"locus_tag\")[0]\n",
    "                    else:\n",
    "                        locus_tag = 'None'\n",
    "                    feature_type = feature.type\n",
    "                    start = int(feature.location.start)\n",
    "                    end = int(feature.location.end)\n",
    "                    strand = int(feature.location.strand)\n",
    "                    if start > previous_end:\n",
    "                        feature_info.append([(previous_locus_tag, locus_tag),'inter_feature', previous_end, start, 0])\n",
    "                    feature_info.append([locus_tag, feature_type, start, end, strand])\n",
    "            sequence_info.append((organism_accession, feature_info))\n",
    "    return sequence_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc56199-7699-4d94-ab35-4a7fcd07fa6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    sequence_info_dict = {}\n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(generate_sequence_info)(num_cores, core_number, species_list) for core_number in core_numbers)\n",
    "    for core_output in parallel_output:\n",
    "        for (accession_ver, feature_info) in core_output:\n",
    "            sequence_info_dict[accession_ver] = feature_info\n",
    "    with open(output_dir + '/' + 'sequence_info_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(sequence_info_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "510ea0b8-bb78-440b-a4fb-d126ecfaa526",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not(full_run == True):\n",
    "    with open(output_dir + '/' + 'sequence_info_dict.pkl', 'rb') as f:\n",
    "        sequence_info_dict = pickle.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "2616d0d9-b446-472c-a198-5ced3637aa18",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "output = []\n",
    "features = []\n",
    "genome_record = next(SeqIO.parse(seq_dir + '/' + reference_species_filename, \"genbank\"))\n",
    "full_sequence = str(genome_record.seq)\n",
    "mycobrowser_df = pd.read_excel(datasets_dir+'/Mycobrowser_Release_4.xlsx')\n",
    "for i, r in mycobrowser_df.iterrows():\n",
    "    if r['Strand'] == '+':\n",
    "        strand = 1\n",
    "    else:\n",
    "        strand = -1\n",
    "    features.append([r['Locus'],r['Start']-1, r['Stop'], strand])\n",
    "features.sort(key=lambda x: x[1])\n",
    "feature_info = []\n",
    "for i, feature in enumerate(features):\n",
    "    feature_sequence = full_sequence[feature[1]: feature[2]]\n",
    "    #feature_info.append([feature[0], feature[1], feature[2], feature_sequence, len(feature_sequence)])\n",
    "    if feature[1] < feature[2]:  \n",
    "        if (i + 1)< len(features) and feature[2] < features[i+1][1]:\n",
    "            utr_coords = (feature[2], features[i+1][1])\n",
    "            utr_sequence = full_sequence[feature[2]: features[i+1][1]]\n",
    "            utr_length = len(utr_sequence)\n",
    "        else:\n",
    "            utr_coords = (0,0)\n",
    "            utr_sequence = ''\n",
    "            utr_length = 0\n",
    "        if utr_length > 50:\n",
    "            feature_info.append([feature[0] + '_IG', utr_coords[0], utr_coords[1], utr_sequence, utr_length])\n",
    "intergenic_df = pd.DataFrame(feature_info, columns = ['Locus', 'Start' , 'End', 'Sequence', 'Length'])\n",
    "intergenic_df.to_csv(output_dir + '/all_regions_df.csv')\n",
    "sequence_list = []\n",
    "records = 0\n",
    "for i, r in intergenic_df.iterrows():\n",
    "    if r['Locus'] == 'Rv0052_IG':\n",
    "        sequence_list.append([r['Locus'], r['Sequence']])\n",
    "util.produce_fasta_file(sequence_list, blast_dir + '/all_regions.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d726ba4-9d93-4ba4-964b-97f5265233ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    w_d = os.getcwd()\n",
    "    os.chdir(\"D:/\")\n",
    "    subprocess.run('cd ' + blast_dir + ' & blastn -query all_regions.faa -db ' + blast_db_name + ' -out blast_hits.csv -evalue 1e-10 -outfmt  \"10 qaccver saccver staxid staxids ssciname qlen slen pident length mismatch gapopen qstart qend sstart send evalue bitscore\" -num_threads 16', shell=True, capture_output = True)\n",
    "    os.chdir(w_d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7306649-6aad-4a97-bc97-58152183873d",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results = pd.read_csv(blast_dir + '/blast_hits.csv', header = None)\n",
    "blast_results.columns = ['query_accession_ver', 'subject_accession_ver', 'subject_tax_id', 'subject_taxids', 'subject_name', 'query_length', 'subject_length', 'percent_identical_matches','alignment_length', 'number_mismatches', 'number_of_gap_openings', 'query_start_alignment', 'query_end_alignment', 'subject_start_alignment', 'subject_end_alignment', 'e_value', 'bit_score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8602070b-d94c-4c70-bc8d-bb71f715e665",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_annotations(accession_ver, start, stop, feature_dict):\n",
    "    if start > stop:\n",
    "        (start, stop) = (stop, start)\n",
    "    annotations_found = []\n",
    "    feature_list = feature_dict[accession_ver]\n",
    "    max_pct_cover = -1\n",
    "    max_feature = []\n",
    "    for feature in feature_list:\n",
    "        if feature[2] < (stop - 1) and feature[3] >= (start - 1):\n",
    "            len_feature = feature[3] - feature[2]\n",
    "            pct_cover = (min(feature[3], stop) - max(feature[2], start)) / (stop - start)\n",
    "            if pct_cover > 0.05 and len_feature < 100000:\n",
    "                if pct_cover > max_pct_cover:\n",
    "                    max_feature = feature\n",
    "                    max_pct_cover = pct_cover\n",
    "                annotations_found.append([feature, pct_cover])\n",
    "    return (annotations_found, [max_feature, max_pct_cover])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2ace8a-51ed-4471-be17-e4e5ac65f954",
   "metadata": {},
   "outputs": [],
   "source": [
    "#blast_results = blast_results[0:30000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a1a41ed-6dde-410f-9930-ad400308ce9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ortholog_hits_dict = {}\n",
    "for i, r in tqdm(blast_results.iterrows(), total=blast_results.shape[0]):\n",
    "        annotations = find_annotations(r['subject_accession_ver'], r['subject_start_alignment'], r['subject_end_alignment'], sequence_info_dict)\n",
    "        blast_results.at[i,'features'] = str(annotations[0])\n",
    "        blast_results.at[i,'closest_cover_feature'] = str(annotations[1])\n",
    "        if annotations[1] == [[],-1]:\n",
    "            blast_results.at[i,'closest_locus'] = []\n",
    "        elif annotations[1][0][1] == 'inter_feature':\n",
    "            blast_results.at[i,'closest_locus'] = [annotations[1][0][0][0], annotations[1][0][0][1]]   \n",
    "        else:\n",
    "            blast_results.at[i,'closest_locus'] = [annotations[1][0][0]]\n",
    "            if annotations[1][0][0] in ortholog_hits_dict:\n",
    "                ortholog_hits_dict[annotations[1][0][0]].append(r['query_accession_ver'])\n",
    "            else:\n",
    "                ortholog_hits_dict[annotations[1][0][0]] = [r['query_accession_ver']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52bbb6f3-50aa-4fa2-a593-c6834972b2d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in tqdm(blast_results.iterrows(), total=blast_results.shape[0]):\n",
    "    closest_locus = r['closest_locus']\n",
    "    if len(closest_locus) == 2:\n",
    "        if closest_locus[0] in ortholog_hits_dict:\n",
    "            tb_1 = ortholog_hits_dict[closest_locus[0]]\n",
    "        else:\n",
    "            tb_1 == ''\n",
    "        if closest_locus[1] in ortholog_hits_dict:\n",
    "            tb_2 = ortholog_hits_dict[closest_locus[1]]\n",
    "        else:\n",
    "            tb_2 == '' \n",
    "        blast_results.at[i,'tb_closest_locus'] = [tb_1, tb_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ca514ef-d66e-49a4-ad9e-1b7f644580f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results.to_csv(output_dir + '/blast_results.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b342b6fd-9eba-427c-be61-f6e1ca99a38e",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_regions_file = blast_dir + '/all_regions.faa'\n",
    "merged_file = seq_dir + '/all_actinobacteria_ref_rep_comp.faa'\n",
    "wsl_merged_file = util.wslname(merged_file)\n",
    "wsl_all_regions_file = util.wslname(all_regions_file)\n",
    "wsl_output_loc = util.wslname(output_dir)\n",
    "#subprocess.run('wsl cd ' + wsl_output_loc + ' ; nhmmer -A align.sto -o hmmer.txt --tblout summary.txt --notextw --cpu 16 --incE 1e-10 '+  wsl_all_regions_file + ' ' + wsl_merged_file, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b80ec15e-87a4-4cef-938f-c7b84c03b2b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def align_keep_top_hit_per_species(output_dir, hit_file, alignment_file, output_alignment_file, package, evalue = 0.01): \n",
    "    wsl_output_loc = util.wslname(output_dir)\n",
    "    dict = {}\n",
    "    if package == 'INFERNAL':\n",
    "        with open(output_dir + '/' + hit_file, 'r') as f:\n",
    "            for l in f:\n",
    "                if not(l[0] == '#'): \n",
    "                    a = l.split()\n",
    "                    if a[16] == '!':\n",
    "                        if a[0] in dict:\n",
    "                            if float(a[15]) < dict[a[0]][1]:\n",
    "                                dict[a[0]] = (a[7]+'-'+a[8],float(a[15]))\n",
    "                        else:\n",
    "                            dict[a[0]] = (a[7]+'-'+a[8],float(a[15]))\n",
    "    elif package == 'HMMER':\n",
    "        with open(output_dir + '/' + hit_file, 'r') as f:\n",
    "            for l in f:\n",
    "                if not(l[0] == '#'): \n",
    "                    a = l.split()\n",
    "                    if float(a[12]) <= evalue:\n",
    "                        if a[0] in dict:\n",
    "                            if float(a[12]) < dict[a[0]][1]:\n",
    "                                dict[a[0]] = (a[6]+'-'+a[7],float(a[12]))\n",
    "                        else:\n",
    "                            dict[a[0]] = (a[6]+'-'+a[7],float(a[12]))\n",
    "    else:\n",
    "        pass\n",
    "    with open(output_dir + '/keep_list.txt', 'w') as f:\n",
    "        lines = []\n",
    "        for k, v in dict.items():\n",
    "            lines.append(k + '/' + v[0] + \"\\n\")\n",
    "        f.write(''.join(lines))\n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ; esl-alimanip -o '+output_alignment_file + ' --seq-r keep_list.txt '+ alignment_file, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "39957c8a-e5da-4c30-aac0-c50fe8126019",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "align_keep_top_hit_per_species(output_dir, 'rcmhits_4.txt', 'rcm_align_4.sto' ,'outtest.sto', 'INFERNAL')\n",
    "align_keep_top_hit_per_species(output_dir, 'summary.txt', 'align_4.sto' ,'outtest2.sto', 'HMMER', 1e-10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c11acb3e-571a-4ef4-901a-a5d725d03d72",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMMER\n",
      "1\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'D:/Project_Data/Project_8/Output/summary.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_22352/90966469.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m5\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m     \u001b[0malign_keep_top_hit_per_species\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0moutput_dir\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'summary.txt'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'align_'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'.sto'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'top_hit_align_'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'.sto'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'HMMER'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mhmmer_evalue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      9\u001b[0m     \u001b[0msubprocess\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'wsl cd '\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mwsl_output_loc\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m' ; hmmbuild --cpu 16 hmm_'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'.hmm top_hit_align_'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'.sto'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     10\u001b[0m     \u001b[0msubprocess\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'wsl cd '\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mwsl_output_loc\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m' ; nhmmer -A align_'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'.sto -o hmmer.txt --tblout summary.txt --notextw --cpu 16 --incE '\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhmmer_evalue\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m\u001b[1;34m' hmm_'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'.hmm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_22352/3093384962.py\u001b[0m in \u001b[0;36malign_keep_top_hit_per_species\u001b[1;34m(output_dir, hit_file, alignment_file, output_alignment_file, package, evalue)\u001b[0m\n\u001b[0;32m     14\u001b[0m                             \u001b[0mdict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m7\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'-'\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m8\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mfloat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m15\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     15\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mpackage\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'HMMER'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 16\u001b[1;33m         \u001b[1;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0moutput_dir\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'/'\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mhit_file\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'r'\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     17\u001b[0m             \u001b[1;32mfor\u001b[0m \u001b[0ml\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     18\u001b[0m                 \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ml\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'#'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'D:/Project_Data/Project_8/Output/summary.txt'"
     ]
    }
   ],
   "source": [
    "hmmer_evalue = 1e-10\n",
    "infernal_evalue = 1e-10\n",
    "subprocess.run('wsl cd ' + wsl_output_loc + ' ; nhmmer -A align_0.sto -o hmmer_0.txt --tblout summary.txt --notextw --cpu 16 --incE 1e-10 /mnt/d/BLAST/actinobacteria_ref_rep_comp/all_regions.faa /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa', shell=True)\n",
    "print ('HMMER')\n",
    "\n",
    "for i in range(1, 5):\n",
    "    print(i)\n",
    "    align_keep_top_hit_per_species(output_dir, 'summary.txt', 'align_'+str(i-1)+'.sto', 'top_hit_align_'+str(i-1)+'.sto', 'HMMER', hmmer_evalue)\n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ; hmmbuild --cpu 16 hmm_'+str(i)+'.hmm top_hit_align_'+str(i-1)+'.sto', shell=True)\n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ; nhmmer -A align_'+str(i)+'.sto -o hmmer.txt --tblout summary.txt --notextw --cpu 16 --incE ' + str(hmmer_evalue) +' hmm_'+str(i)+'.hmm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa', shell=True)\n",
    "print('INFERNAL')\n",
    "for i in range(1, 5):\n",
    "    print(i)\n",
    "    if i == 1:\n",
    "        align_keep_top_hit_per_species(output_dir, 'summary.txt', 'align_4.sto', 'top_hit_align_4.sto', 'HMMER', hmmer_evalue)\n",
    "        subprocess.run('wsl cd ' + wsl_output_loc + ' ; ~/rscape_v2.0.0.g/bin/R-scape  --cacofold --outname struc_'+str(i)+' top_hit_align_4.sto ', shell=True)\n",
    "  \n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ; cmbuild -F rcm_'+str(i)+'.cm struc_'+str(i)+'.cacofold.sto', shell=True)\n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ;  cmcalibrate rcm_'+str(i)+'.cm', shell=True)\n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ;  cmsearch --tblout rcmhits_'+str(i)+'.txt -A rcm_align_'+str(i)+'.sto --cpu 16 --incE ' + str(infernal_evalue) + ' rcm_'+str(i)+'.cm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa', shell=True)\n",
    "    align_keep_top_hit_per_species(output_dir, 'rcmhits_'+str(i)+'.txt', 'rcm_align_'+str(i)+'.sto', 'top_hit_rcm_align_'+str(i)+'.sto', 'INFERNAL')\n",
    "    subprocess.run('wsl cd ' + wsl_output_loc + ' ; ~/rscape_v2.0.0.g/bin/R-scape --cacofold --outname struc_'+str(i + 1)+' top_hit_rcm_align_'+str(i)+'.sto ', shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "14c894c5-0592-424d-b571-a9b45ba8c849",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mycobacterium_canettii_CIPT_140010059_NC_015848.1 3.7e-121 1279060 1279619\n",
      "Mycobacterium_tuberculosis_H37Rv_NC_000962.3 4.3e-120 1261351 1261910\n",
      "Mycobacterium_tuberculosis_variant_bovis_AF2122/97_LT708304.1 4.3e-120 1262750 1263309\n",
      "Mycobacterium_tuberculosis_variant_bovis_BCG_str._Pasteur_1173P2_AM408590.1 4.3e-120 1292766 1293325\n",
      "Mycobacterium_shinjukuense_NZ_AP022575.1 3.4e-108 373878 374464\n",
      "Mycobacterium_shottsii_NZ_AP022572.1 2e-104 1641123 1641612\n",
      "Mycobacterium_marinum_NZ_CP058277.1 5.9e-104 165016 164527\n",
      "Mycobacterium_pseudoshottsii_JCM_15466_NZ_AP018410.1 2e-103 4885404 4884915\n",
      "Mycobacterium_ulcerans_NZ_CP085200.1 9e-103 129360 128871\n",
      "Mycobacterium_noviomagense_NZ_AP022583.1 7.7e-101 663992 663541\n",
      "Mycobacterium_conspicuum_NZ_AP022613.1 2.2e-100 4529374 4528931\n",
      "Mycobacterium_mantenii_NZ_AP022590.1 5.4e-100 3128711 3128251\n",
      "Mycobacterium_malmoense_NZ_CP080999.1 1.2e-99 4111384 4110893\n",
      "Mycobacterium_stomatepiae_NZ_AP022587.1 1.4e-99 5227870 5228342\n",
      "Mycobacterium_simiae_NZ_AP022568.1 5e-99 3983566 3984029\n",
      "Mycobacterium_cookii_NZ_AP022569.1 9.8e-99 2149030 2148529\n",
      "Mycobacterium_leprae_NZ_CP029543.1 4.7e-98 1102568 1103053\n",
      "Mycobacterium_marseillense_NZ_CP023147.1 6.9e-98 1132290 1132775\n",
      "Mycobacterium_paraseoulense_NZ_AP022619.1 8.2e-98 2840155 2839687\n",
      "Mycobacterium_spongiae_NZ_CP046600.1 2.3e-97 4265319 4264860\n",
      "Mycobacterium_florentinum_NZ_AP022576.1 5.7e-97 4970662 4971116\n",
      "Mycobacterium_seoulense_NZ_AP022582.1 6.1e-97 2911099 2910638\n",
      "Mycobacterium_vicinigordonae_NZ_CP059165.1 1.6e-96 4944784 4944246\n",
      "Mycobacterium_lepromatosis_NZ_CP083405.1 6.9e-96 1394974 1395463\n",
      "Mycobacterium_shigaense_NZ_AP018164.1 2.3e-94 4156636 4156086\n",
      "Mycobacterium_paraintracellulare_NZ_AP024251.1 1.3e-93 1325001 1325538\n",
      "Mycobacterium_intracellulare_ATCC_13950_NC_016946.1 2.2e-93 1164143 1164680\n",
      "Mycobacterium_branderi_NZ_AP022606.1 3.5e-93 2702583 2702989\n",
      "Mycobacterium_lacus_NZ_AP022581.1 3.7e-93 753139 753527\n",
      "Mycobacterium_novum_NZ_AP022562.1 6.6e-93 821206 820723\n",
      "Mycolicibacter_terrae_NZ_LT906469.1 6.9e-93 1128832 1129323\n",
      "Mycobacterium_basiliense_NZ_LR130759.1 3.1e-92 4357417 4356937\n",
      "Mycolicibacter_sinensis_NC_015576.1 3.4e-92 1163205 1163688\n",
      "Mycolicibacter_minnesotensis_NZ_AP022589.1 7.3e-92 2931858 2932344\n",
      "Mycobacterium_virginiense_NZ_CP092430.1 3.4e-91 1284843 1285344\n",
      "Mycobacterium_ostraviense_NZ_CP089224.1 3.9e-91 724261 724630\n",
      "Mycobacterium_haemophilum_DSM_44634_NZ_CP011883.2 6e-91 3224358 3223918\n",
      "Mycobacterium_paragordonae_NZ_CP025546.1 8e-91 5310758 5310226\n",
      "Mycobacterium_kansasii_ATCC_12478_NC_022663.1 8.3e-91 1743727 1743354\n",
      "Mycobacterium_heidelbergense_NZ_AP022615.1 1.4e-90 3284109 3284532\n",
      "Mycolicibacter_hiberniae_NZ_AP022609.1 4.5e-89 670840 671315\n",
      "Mycobacterium_paraterrae_NZ_CP092488.1 1.1e-87 1202042 1202531\n",
      "Mycobacterium_koreense_NZ_AP022594.1 2.3e-86 4117199 4117668\n",
      "Mycolicibacterium_fortuitum_NZ_CP011269.1 6.2e-85 4220663 4221091\n",
      "Mycolicibacillus_parakoreensis_NZ_CP092365.1 1.4e-84 890890 891363\n",
      "Mycobacterium_heckeshornense_NZ_AP024310.1 2e-83 1402656 1402996\n",
      "Mycolicibacterium_helvum_NZ_AP022596.1 6e-83 4107633 4108064\n",
      "Mycolicibacterium_boenickei_NZ_AP022579.1 5.7e-82 3404328 3403931\n",
      "Mycobacteroides_chelonae_CCUG_47445_NZ_CP007220.1 1.5e-81 4746508 4746913\n",
      "[Mycobacterium]_stephanolepidis_NZ_AP018165.1 6.3e-81 4760801 4761206\n",
      "Mycobacterium_dioxanotrophicus_NZ_CP020809.1 1e-80 2325258 2324871\n",
      "Mycolicibacterium_pulveris_NZ_AP022599.1 1.2e-80 2579484 2579128\n",
      "Mycobacterium_holsaticum_NZ_CP080998.1 1.6e-80 253399 253046\n",
      "Mycolicibacterium_mageritense_NZ_AP022567.1 2.9e-80 7961234 7961634\n",
      "Mycolicibacterium_alvei_NZ_AP022565.1 4.3e-80 5603698 5603315\n",
      "Mycobacteroides_immunogenum_NZ_CP011530.1 1e-78 5253469 5253859\n",
      "Mycolicibacterium_mucogenicum_DSM_44124_NZ_CP062008.1 1.9e-78 5316241 5316602\n",
      "Mycobacteroides_abscessus_NZ_CP034181.1 1.7e-77 1845908 1846309\n",
      "Mycobacteroides_salmoniphilum_NZ_CP024633.1 2.5e-77 4519339 4519733\n",
      "Mycolicibacterium_neoaurum_NZ_CP074376.1 2.9e-77 113078 112690\n",
      "Mycolicibacterium_neoaurum_NZ_CP074376.1 4.2e-77 117271 117634\n",
      "Mycolicibacterium_celeriflavum_NZ_AP022591.1 5.8e-77 4009693 4009316\n",
      "Mycolicibacterium_smegmatis_NZ_CP102342.1 3.1e-76 6705452 6705885\n",
      "Mycobacteroides_saopaulense_NZ_CP010271.1 3.4e-76 4392471 4392886\n",
      "Mycolicibacterium_chitae_NZ_LR134355.1 3.2e-75 5185212 5185576\n",
      "Mycobacterium_shinjukuense_NZ_AP022575.1 1.1e-74 1409087 1408676\n",
      "Mycolicibacterium_sarraceniae_NZ_AP022595.1 2.4e-74 494376 494739\n",
      "Mycolicibacterium_phocaicum_NZ_AP022616.1 2.8e-74 4443225 4442871\n",
      "Mycobacterium_basiliense_NZ_LR130759.1 6.2e-74 3067619 3068100\n",
      "Mycobacterium_pseudoshottsii_JCM_15466_NZ_AP018410.1 6.6e-74 3289936 3290341\n",
      "Mycobacterium_ulcerans_NZ_CP085200.1 6.6e-74 2508495 2508900\n",
      "Mycobacterium_marinum_NZ_CP058277.1 9.1e-74 4931090 4931495\n",
      "Streptomyces_pratensis_NZ_CP051486.1 1.5e-73 4205428 4205099\n",
      "Rhodococcus_triatomae_NZ_CP048813.1 1.7e-73 390132 390507\n",
      "Rhodococcus_hoagii_NZ_CP027793.1 2e-73 1137031 1137495\n",
      "Streptomyces_phaeolivaceus_NZ_CP045096.1 7.4e-73 9732554 9732841\n",
      "Streptomyces_roseirectus_NZ_CP060828.1 2.9e-72 8227763 8228145\n",
      "Streptomyces_tuirus_NZ_AP023439.1 3.1e-72 155947 155648\n",
      "Mycolicibacterium_thermoresistibile_NZ_LT906483.1 3.5e-72 4747425 4747081\n",
      "Streptomyces_gilvosporeus_NZ_CP020569.1 7e-72 1718755 1718471\n",
      "Mycobacterium_goodii_NZ_CP092364.1 8.4e-72 2388928 2389302\n",
      "Streptomyces_microflavus_NZ_CP054926.1 8.6e-72 6771629 6771306\n",
      "Streptacidiphilus_bronchialis_NZ_CP031264.1 1.2e-71 6893552 6893841\n",
      "Streptomyces_chartreusis_NZ_CP023689.1 1.7e-71 2041828 2041538\n",
      "Kitasatospora_albolonga_NZ_CP020563.1 1.9e-71 7144487 7144179\n",
      "Streptomyces_liangshanensis_NZ_CP050177.1 2.1e-71 303721 303442\n",
      "Streptomyces_seoulensis_NZ_CP032229.1 2.3e-71 6221172 6221448\n",
      "Mycobacterium_shottsii_NZ_AP022572.1 2.5e-71 4212051 4212456\n",
      "Streptomyces_griseorubiginosus_NZ_CP032427.1 3.6e-71 6499719 6500019\n",
      "Streptomyces_californicus_NZ_CP070242.1 4.4e-71 6871659 6871372\n",
      "Streptomyces_violaceoruber_NZ_CP020570.1 4.4e-71 6914373 6914086\n",
      "Streptomyces_griseochromogenes_NZ_CP016279.1 7.1e-71 2597033 2597330\n",
      "Mycobacterium_spongiae_NZ_CP046600.1 7.1e-71 2670396 2669997\n",
      "Mycolicibacterium_insubricum_NZ_AP022618.1 8e-71 2213472 2213876\n",
      "Streptomyces_ficellus_NZ_CP034279.1 8.1e-71 299851 299567\n",
      "Streptomyces_sp._ZYC-3_NZ_CP066831.1 1e-70 2645087 2644786\n",
      "Streptomyces_fagopyri_NZ_CP045643.1 1e-70 8682454 8682738\n",
      "Streptomyces_anulatus_NZ_CP086102.1 1.1e-70 1794169 1793818\n",
      "Streptomyces_murinus_NZ_CP066774.1 1.2e-70 3351687 3351969\n",
      "Streptomyces_venezuelae_NZ_CP018074.1 1.5e-70 6071398 6071686\n",
      "Streptomyces_griseoviridis_NZ_CP034687.1 1.8e-70 504005 503691\n",
      "Nocardia_terpenica_NZ_CP023778.1 2.4e-70 6385191 6384747\n",
      "Rhodococcus_koreensis_NZ_CP070609.1 2.4e-70 3823162 3822775\n",
      "Streptomyces_griseofuscus_NZ_CP051006.1 2.5e-70 1110932 1110646\n",
      "Streptomyces_formicae_NZ_CP071872.1 2.6e-70 2260999 2261276\n",
      "Streptomyces_aquilus_NZ_CP034463.1 2.8e-70 9853396 9853697\n",
      "Streptomyces_spectabilis_NZ_CP023690.1 3e-70 9063071 9063352\n",
      "Rhodococcus_globerulus_NZ_CP079698.1 3.1e-70 6667345 6666945\n",
      "Streptomyces_cyaneochromogenes_NZ_CP034539.1 3.6e-70 8573716 8574023\n",
      "Rhodococcus_qingshengii_NZ_AP023172.1 4e-70 4921762 4921365\n",
      "Streptomyces_yatensis_NZ_CP072941.1 4.4e-70 8651958 8652295\n",
      "Streptomyces_globosus_NZ_CP030862.1 4.5e-70 2558572 2558284\n",
      "Gordonia_pseudoamarae_NZ_CP045806.1 5.2e-70 1709890 1710272\n",
      "Streptomyces_avermitilis_MA-4680_=_NBRC_14893_NC_003155.5 7.4e-70 2496280 2495999\n",
      "Streptomyces_albidoflavus_NC_020990.1 8.6e-70 1040091 1040370\n",
      "Streptomyces_ferrugineus_NZ_CP063373.1 8.9e-70 6921368 6921084\n",
      "Streptomyces_subrutilus_NZ_CP023701.1 1e-69 6667142 6667432\n",
      "Mycobacterium_parmense_NZ_AP022614.1 1e-69 4316845 4317245\n",
      "Mycolicibacterium_fallax_NZ_AP022603.1 1.1e-69 351808 352194\n",
      "Streptomyces_cadmiisoli_NZ_CP030073.1 1.2e-69 2658498 2658199\n",
      "Streptomyces_griseus_subsp._griseus_NBRC_13350_NC_010572.1 1.3e-69 1431160 1431434\n",
      "Streptomyces_nodosus_NZ_CP071465.1 1.4e-69 1006930 1006588\n",
      "Streptomyces_clavuligerus_NZ_CP065254.1 1.8e-69 452358 452652\n",
      "Streptomyces_kanamyceticus_NZ_CP023699.1 2.3e-69 384623 384337\n",
      "Streptomyces_cavourensis_NZ_CP024957.1 3.1e-69 6578318 6578034\n",
      "Gordonia_polyisoprenivorans_VH2_NC_016906.1 3.9e-69 1954786 1955104\n",
      "Streptomyces_qinzhouensis_NZ_CP042266.1 4.3e-69 399081 398789\n",
      "Streptomyces_mobaraensis_NZ_CP083590.1 5.2e-69 1124320 1124012\n",
      "Nocardia_cyriacigeorgica_GUH-2_NC_016887.1 6.4e-69 4301112 4300794\n",
      "Streptomyces_tsukubensis_NZ_CP020700.1 6.9e-69 7637543 7637847\n",
      "Streptomyces_tirandamycinicus_NZ_CP029188.1 7.5e-69 5838442 5838766\n",
      "Gordonia_amarae_NZ_CP045805.1 9.4e-69 1806474 1806854\n",
      "Streptomyces_vinaceus_NZ_CP023692.1 9.7e-69 320969 320659\n",
      "Streptomyces_koyangensis_NZ_CP049945.1 1.2e-68 5667213 5666933\n",
      "Streptomyces_lydicus_NZ_CP019457.1 1.6e-68 6467937 6468204\n",
      "Salinispora_tropica_CNB-440_NC_009380.1 1.7e-68 4966588 4966847\n",
      "Rhodococcus_erythropolis_R138_NZ_CP007255.1 1.7e-68 2032241 2032637\n",
      "Streptomyces_solisilvae_NZ_CP065050.1 1.7e-68 2174028 2173737\n",
      "Streptomyces_olivaceus_NZ_CP043317.1 2.3e-68 7486007 7486342\n",
      "Streptomyces_gardneri_NZ_CP059991.1 2.4e-68 7299348 7299624\n",
      "Nocardioides_okcheonensis_NZ_CP087710.1 3.2e-68 1668418 1668139\n",
      "Streptomyces_parvulus_NZ_CP015866.1 4.1e-68 6392487 6392788\n",
      "Rhodococcus_opacus_PD630_NZ_CP080954.1 5.8e-68 1465122 1465508\n",
      "Streptomyces_nojiriensis_NZ_CP071139.1 7.5e-68 1977841 1978122\n",
      "Actinocatenispora_sera_NZ_AP023354.1 9.5e-68 2896277 2896531\n",
      "Streptomyces_inhibens_NZ_CP084681.1 9.7e-68 7733887 7734226\n",
      "Streptomyces_aquilus_NZ_CP034463.1 1.5e-67 8092096 8092383\n",
      "Streptomyces_rectiverticillatus_NZ_CP040752.1 2.3e-67 569350 569079\n",
      "Nocardia_wallacei_NZ_AP023396.1 3.4e-67 2301553 2301932\n",
      "Streptomyces_paludis_NZ_CP031194.1 5.8e-67 6321519 6321806\n",
      "Nocardia_otitidiscaviarum_NZ_CP041695.1 1.2e-66 4820318 4819860\n",
      "Pimelobacter_simplex_NZ_CP009896.1 1.4e-66 5560925 5561208\n",
      "Gordonia_otitidis_NZ_CP085974.1 1.4e-66 357967 357564\n",
      "Streptomyces_rimosus_subsp._rimosus_ATCC_10970_NZ_CP048261.1 1.5e-66 820127 819849\n",
      "Streptomyces_genisteinicus_NZ_CP060825.1 2.3e-66 5624581 5624869\n",
      "Micromonospora_carbonacea_NZ_CP058322.1 5.3e-66 4049510 4049225\n",
      "Streptomyces_angustmyceticus_NZ_CP082945.1 5.9e-66 1775998 1775734\n",
      "Rhodococcus_fascians_D188_NZ_CP015235.1 6.1e-66 800172 799801\n",
      "Nocardia_brasiliensis_NZ_CP022088.2 1.1e-65 8851998 8852461\n",
      "Luteipulveratus_mongoliensis_NZ_CP011112.1 1.2e-65 946581 946277\n",
      "Rhodococcus_pyridinivorans_SB3094_NC_023150.1 1.6e-65 2227705 2227300\n",
      "Rhodococcus_pyridinivorans_SB3094_NC_023150.1 1.6e-65 3120867 3120462\n",
      "Nocardia_yamanashiensis_NZ_CP089226.1 2.3e-65 230331 230714\n",
      "Natronosporangium_hydrolyticum_NZ_CP070499.1 2.4e-65 4738721 4738978\n",
      "Actinoplanes_ianthinogenes_NZ_AP023356.1 3.1e-65 5226122 5225869\n",
      "Nocardia_mangyaensis_NZ_CP018082.1 3.1e-65 4862488 4862174\n",
      "Actinocatenispora_thailandica_NZ_AP023355.1 7.2e-65 4318083 4317819\n",
      "Streptomyces_decoyicus_NZ_CP082301.1 7.8e-65 6636277 6636548\n",
      "Skermania_piniformis_NZ_CP079105.1 9.6e-65 1077290 1076868\n",
      "Polymorphospora_rubra_NZ_AP023359.1 1.3e-64 6205299 6205017\n",
      "Micromonospora_craniellae_NZ_CP061725.1 2.2e-64 6593395 6593650\n",
      "Nocardia_asteroides_NZ_LR134352.1 2.4e-64 5743060 5742740\n",
      "Actinoplanes_friuliensis_DSM_7358_NC_022657.1 2.4e-64 7787083 7786825\n",
      "Streptomyces_platensis_NZ_CP023691.1 2.5e-64 6880708 6880984\n",
      "Phytohabitans_suffuscus_NZ_AP022871.1 3.2e-64 7744947 7744675\n",
      "Micromonospora_aurantiaca_ATCC_27029_NC_014391.1 3.3e-64 2994156 2994456\n",
      "Gordonia_bronchialis_DSM_43247_NC_013441.1 3.5e-64 1908798 1909136\n",
      "Nocardia_tengchongensis_NZ_CP074371.1 5.3e-64 7630519 7630900\n",
      "Micromonospora_maris_AB-18-032_NC_015434.1 2.1e-63 2552659 2552342\n",
      "Streptomyces_auratus_AGR0001_NZ_CP072931.1 2.5e-63 6388864 6389186\n",
      "Streptomyces_lincolnensis_NZ_CP022744.1 2.7e-63 9111075 9111351\n",
      "Gordonia_insulae_NZ_CP033972.1 3.8e-63 1558567 1558139\n",
      "Nocardia_iowensis_NZ_CP078145.1 3.9e-63 6596508 6596152\n",
      "Tomitella_fengzijianii_NZ_CP041765.1 5.3e-63 1235136 1234774\n",
      "Actinomadura_graeca_NZ_CP059572.1 6.8e-63 3735582 3735334\n",
      "Micromonospora_tulbaghiae_NZ_CP024087.1 7.8e-63 2653943 2653641\n",
      "Streptomyces_sp._1_25_NZ_CP086120.1 8.9e-63 2690143 2689883\n",
      "Streptomyces_xiamenensis_NZ_CP009922.3 1.1e-62 5302740 5303010\n",
      "Thermobifida_alba_NZ_CP051627.1 1.4e-62 3724082 3723793\n",
      "Stackebrandtia_nassauensis_DSM_44728_NC_013947.1 2.8e-62 290304 289961\n",
      "Micromonospora_maris_AB-18-032_NC_015434.1 4.7e-62 3036632 3036381\n",
      "Streptosporangium_roseum_DSM_43021_NC_013595.1 5.1e-62 7969060 7968698\n",
      "Saccharopolyspora_erythraea_NRRL_2338_NC_009142.1 6.6e-62 5303358 5302977\n",
      "Nocardia_otitidiscaviarum_NZ_CP041695.1 7.3e-62 3337844 3337479\n",
      "Streptomyces_nodosus_NZ_CP071465.1 1.3e-61 747732 747419\n",
      "Streptomyces_harbinensis_NZ_CP054938.1 1.3e-61 5453233 5453496\n",
      "Spiractinospora_alimapuensis_NZ_CP072467.1 1.5e-61 2395556 2395294\n",
      "Actinosynnema_mirum_DSM_43827_NC_013093.1 2.4e-61 3644402 3644028\n",
      "Micromonospora_maris_AB-18-032_NC_015434.1 2.4e-61 2768900 2768626\n",
      "Micromonospora_carbonacea_NZ_CP058322.1 2.4e-61 766192 765906\n",
      "Nocardia_gipuzkoensis_NZ_CP089225.1 3.7e-61 5082264 5082636\n",
      "Saccharopolyspora_spinosa_NZ_CP061007.1 4.1e-61 7677109 7676746\n",
      "Polymorphospora_rubra_NZ_AP023359.1 2.3e-60 7988294 7988552\n",
      "Nocardia_huaxiensis_NZ_CP059399.1 2.4e-60 5610641 5610274\n",
      "Saccharopolyspora_pogona_NZ_CP031142.1 2.8e-60 1523573 1523208\n",
      "Nocardia_seriolae_NZ_AP017900.1 2.9e-60 2289054 2289367\n",
      "Tomitella_gaofuii_NZ_CP059900.1 4.2e-60 1487212 1486862\n",
      "Streptomyces_fradiae_ATCC_10745_=_DSM_40063_NZ_CP023696.1 6.2e-60 5465289 5465628\n",
      "Mycobacterium_diernhoferi_NZ_CP080332.1 6.6e-60 642216 642506\n",
      "Micromonospora_aurantiaca_ATCC_27029_NC_014391.1 1.6e-59 2996691 2996958\n",
      "Kutzneria_albida_DSM_43870_NZ_CP007155.1 1.8e-59 2256799 2256503\n",
      "Micromonospora_tulbaghiae_NZ_CP024087.1 1.8e-59 2651437 2651170\n",
      "Nocardioides_humi_NZ_CP041146.1 1.9e-59 5337008 5336715\n",
      "Micromonospora_craniellae_NZ_CP061725.1 2.2e-59 151898 152162\n",
      "Streptomyces_buecherae_NZ_CP060404.1 2.5e-59 7305330 7305602\n",
      "Streptomyces_prasinus_NZ_CP023697.1 2.7e-59 6154300 6154620\n",
      "Actinosynnema_pretiosum_NZ_CP023445.1 6.8e-59 3731697 3731409\n",
      "Actinoplanes_missouriensis_431_NC_017093.1 7.5e-59 2020623 2020878\n",
      "Luteipulveratus_mongoliensis_NZ_CP011112.1 1e-58 954479 954794\n",
      "Saccharopolyspora_gregorii_NZ_CP059557.1 1.1e-58 1802142 1802443\n",
      "Natronosporangium_hydrolyticum_NZ_CP070499.1 1.1e-58 4734956 4735210\n",
      "Streptomyces_nitrosporeus_NZ_CP023702.1 1.3e-58 66873 66519\n",
      "Aeromicrobium_yanjiei_NZ_CP045737.1 1.5e-58 1674108 1673705\n",
      "Streptomyces_olivoreticuli_subsp._olivoreticuli_NZ_CP031455.1 2.4e-58 2686238 2686466\n",
      "Micromonospora_carbonacea_NZ_CP058322.1 2.9e-58 777229 776951\n",
      "Kibdelosporangium_phytohabitans_NZ_CP012752.1 3.9e-58 4679695 4679962\n",
      "Catenulispora_acidiphila_DSM_44928_NC_013131.1 4.3e-58 2314295 2314612\n",
      "Microlunatus_phosphovorus_NM-1_NC_015635.1 7.6e-58 3625044 3625361\n",
      "Streptomyces_huasconensis_NZ_CP086119.1 8.5e-58 679969 679614\n",
      "Streptomyces_ficellus_NZ_CP034279.1 9e-58 6636247 6636522\n",
      "Kutzneria_albida_DSM_43870_NZ_CP007155.1 1.1e-57 3630224 3630476\n",
      "Pseudonocardia_autotrophica_NZ_AP018920.1 1.2e-57 6599116 6599401\n",
      "Phytohabitans_suffuscus_NZ_AP022871.1 3.5e-57 7074446 7074193\n",
      "Aeromicrobium_erythreum_NZ_CP011502.1 4e-57 3078573 3078283\n",
      "Pseudonocardia_autotrophica_NZ_AP018920.1 4.1e-57 6548296 6548030\n",
      "Streptomyces_genisteinicus_NZ_CP060825.1 5e-57 7215503 7215187\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 5.7e-57 2192264 2192526\n",
      "Saccharothrix_espanaensis_DSM_44229_NC_019673.1 7.1e-57 5498231 5497958\n",
      "Amycolatopsis_orientalis_NZ_CP016174.1 8.6e-57 3406844 3407109\n",
      "Saccharopolyspora_gregorii_NZ_CP059557.1 8.8e-57 5652900 5652580\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 1.1e-56 3252646 3252904\n",
      "Streptomyces_alfalfae_NZ_CP060742.1 2.1e-56 293224 292870\n",
      "Amycolatopsis_keratiniphila_NC_021252.1 2.5e-56 3047106 3047372\n",
      "Gordonia_iterans_NZ_CP027433.1 4.1e-56 1522888 1523377\n",
      "Micromonospora_craniellae_NZ_CP061725.1 4.7e-56 146331 145993\n",
      "Propioniciclava_coleopterorum_NZ_CP049865.1 8e-56 2182377 2182655\n",
      "Amycolatopsis_japonica_NZ_CP008953.1 9.9e-56 5264743 5264477\n",
      "Streptomyces_angustmyceticus_NZ_CP082945.1 1.6e-55 7889516 7889818\n",
      "Kutzneria_albida_DSM_43870_NZ_CP007155.1 1.8e-55 4432884 4432596\n",
      "Nonomuraea_gerenzanensis_NZ_CP084058.1 2.2e-55 8391555 8391830\n",
      "Amycolatopsis_albispora_NZ_CP015163.1 5.9e-55 1082961 1083296\n",
      "Nocardiopsis_gilva_YIM_90087_NZ_CP022753.1 6.5e-55 4317456 4317806\n",
      "Streptomyces_ambofaciens_ATCC_23877_NZ_CP012382.1 8.6e-55 6502556 6502902\n",
      "Allosaccharopolyspora_coralli_NZ_CP045929.1 1.2e-54 3837024 3836711\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 2.8e-54 808797 809068\n",
      "Saccharopolyspora_spinosa_NZ_CP061007.1 7e-54 7275161 7274880\n",
      "Acidothermus_cellulolyticus_11B_NC_008578.1 9.6e-54 1942321 1942065\n",
      "Streptomyces_globosus_NZ_CP030862.1 1.3e-53 3383871 3384148\n",
      "Streptomyces_calvus_NZ_CP022310.1 1.3e-53 7174796 7175084\n",
      "Streptomyces_bathyalis_NZ_CP048882.1 1.4e-53 5929431 5929098\n",
      "Prauserella_marina_NZ_CP016353.1 1.6e-53 3973352 3973638\n",
      "Nocardioides_ungokensis_NZ_CP059164.1 3e-53 2114450 2114163\n",
      "Saccharopolyspora_erythraea_NRRL_2338_NC_009142.1 6.3e-53 1414344 1414670\n",
      "Acidothermus_cellulolyticus_11B_NC_008578.1 8.2e-53 1939416 1939161\n",
      "Lentzea_guizhouensis_NZ_CP016793.1 2.6e-52 9767826 9767476\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 3.5e-52 2551592 2551868\n",
      "Saccharothrix_espanaensis_DSM_44229_NC_019673.1 4.1e-52 5497457 5497708\n",
      "Tessaracoccus_aquimaris_NZ_CP019606.1 4.9e-52 2386683 2386922\n",
      "Nonomuraea_nitratireducens_NZ_CP045572.1 5.1e-52 7574090 7574350\n",
      "Actinoalloteichus_hymeniacidonis_NZ_CP014859.1 6.8e-52 2310272 2309945\n",
      "Prauserella_marina_NZ_CP016353.1 8.2e-52 4746303 4746578\n",
      "Streptosporangium_roseum_DSM_43021_NC_013595.1 1e-51 6877600 6877802\n",
      "Serinicoccus_hydrothermalis_NZ_CP014989.1 1.7e-51 573760 573466\n",
      "Gordonia_phthalatica_NZ_CP011853.1 3e-51 1449499 1449921\n",
      "Microlunatus_phosphovorus_NM-1_NC_015635.1 3.2e-51 2622840 2623130\n",
      "Microlunatus_elymi_NZ_CP041692.1 1.1e-50 2493398 2493693\n",
      "Kibdelosporangium_phytohabitans_NZ_CP012752.1 1.3e-50 2929256 2929012\n",
      "Tomitella_gaofuii_NZ_CP059900.1 3.4e-50 3540410 3540140\n",
      "Saccharomonospora_viridis_DSM_43017_NC_013159.1 3.8e-50 3821443 3821697\n",
      "Actinosynnema_mirum_DSM_43827_NC_013093.1 4.3e-50 3556190 3555915\n",
      "Streptomonospora_sp._M2_NZ_CP036455.1 4.5e-50 3922131 3922397\n",
      "Amycolatopsis_aidingensis_NZ_CP076538.1 7.5e-50 3105687 3106028\n",
      "Actinoalloteichus_fjordicus_NZ_CP016076.1 7.9e-50 2541856 2541531\n",
      "Serinicoccus_chungangensis_NZ_CP040887.1 1e-49 1479228 1479526\n",
      "Actinopolyspora_erythraea_NZ_CP022752.1 2.6e-49 1086902 1087202\n",
      "Nonomuraea_coxensis_DSM_45129_NZ_CP068985.1 3.9e-49 6113730 6113952\n",
      "Saccharopolyspora_pogona_NZ_CP031142.1 4.6e-49 5224025 5223741\n",
      "Saccharopolyspora_spinosa_NZ_CP061007.1 5.3e-49 4828212 4828495\n",
      "Saccharomonospora_viridis_DSM_43017_NC_013159.1 6.9e-49 1684758 1685016\n",
      "Actinopolyspora_erythraea_NZ_CP022752.1 7.8e-49 1230527 1230242\n",
      "Actinoalloteichus_hymeniacidonis_NZ_CP014859.1 2.6e-48 2300557 2300864\n",
      "Nocardiopsis_gilva_YIM_90087_NZ_CP022753.1 6e-48 164573 164871\n",
      "Micromonospora_tulbaghiae_NZ_CP024087.1 1.2e-47 1152648 1152388\n",
      "Tessaracoccus_defluvii_NZ_CP060789.1 1.8e-47 607284 607073\n",
      "Actinoalloteichus_hoggarensis_NZ_CP022521.1 1.1e-46 2416732 2416403\n",
      "Actinoalloteichus_fjordicus_NZ_CP016076.1 2.9e-46 2531160 2531465\n",
      "Nakamurella_multipartita_DSM_44233_NC_013235.1 5.5e-46 5830914 5830628\n",
      "Cutibacterium_acnes_HL096PA1_NC_021085.1 1.2e-45 2228946 2228704\n",
      "Cutibacterium_avidum_44067_NC_021064.1 1.4e-45 106997 107222\n",
      "Spiractinospora_alimapuensis_NZ_CP072467.1 1.6e-45 3217959 3217669\n",
      "Allosaccharopolyspora_coralli_NZ_CP045929.1 2.1e-45 1188076 1188336\n",
      "Corynebacterium_xerosis_NZ_CP032788.1 2.7e-45 2526101 2526334\n",
      "Actinoalloteichus_hoggarensis_NZ_CP022521.1 4.4e-45 2406150 2406455\n",
      "Rhodococcus_fascians_D188_NZ_CP015235.1 5.3e-45 3795790 3796026\n",
      "Cutibacterium_modestum_NZ_CP017040.1 8.3e-45 1666058 1666294\n",
      "Stackebrandtia_nassauensis_DSM_44728_NC_013947.1 3e-44 5371183 5370969\n",
      "Nocardiopsis_dassonvillei_NZ_LR134501.1 3.4e-44 5891590 5891304\n",
      "Actinosynnema_pretiosum_NZ_CP023445.1 3.2e-43 3633812 3633613\n",
      "Serinicoccus_hydrothermalis_NZ_CP014989.1 4.1e-43 2279828 2279593\n",
      "Streptomyces_pratensis_NZ_CP051486.1 6.9e-43 4218388 4218088\n",
      "Mycobacteroides_chelonae_CCUG_47445_NZ_CP007220.1 2.4e-41 506606 506833\n",
      "Mycobacteroides_saopaulense_NZ_CP010271.1 5.3e-41 469557 469796\n",
      "Microlunatus_elymi_NZ_CP041692.1 6e-41 2487287 2487080\n",
      "[Mycobacterium]_stephanolepidis_NZ_AP018165.1 6.6e-41 498007 498255\n",
      "Streptomyces_buecherae_NZ_CP060404.1 1.2e-40 1866027 1865748\n",
      "Mycobacteroides_salmoniphilum_NZ_CP024633.1 1.2e-40 454831 455060\n",
      "Mycobacteroides_abscessus_NZ_CP034181.1 1.3e-40 2754969 2755212\n",
      "Tessaracoccus_timonensis_NZ_LT996886.1 4.6e-40 620628 620860\n",
      "Cutibacterium_avidum_44067_NC_021064.1 5.3e-40 1700318 1700661\n",
      "Arachnia_rubra_NZ_CP072384.1 1.9e-39 243551 243800\n",
      "Egicoccus_halophilus_NZ_CP036250.1 2.2e-39 2447331 2447000\n",
      "Mycobacteroides_immunogenum_NZ_CP011530.1 2.2e-38 664991 665231\n",
      "Raineyella_fluvialis_NZ_CP045725.1 2.6e-38 1647122 1646858\n",
      "Corynebacterium_sphenisci_DSM_44792_NZ_CP009248.1 1.2e-37 2130781 2130493\n",
      "Cutibacterium_modestum_NZ_CP017040.1 1.4e-37 2479367 2478969\n",
      "Propionibacterium_acidifaciens_NZ_CP033719.1 6e-37 208659 208885\n",
      "Gordonia_polyisoprenivorans_VH2_NC_016906.1 6.6e-37 1751341 1751051\n",
      "Propionibacterium_australiense_NZ_LR134442.1 1.1e-36 2199301 2199544\n",
      "Cutibacterium_acnes_HL096PA1_NC_021085.1 1.6e-36 1703181 1703476\n",
      "Acidipropionibacterium_acidipropionici_NZ_CP031057.1 1.7e-36 2781364 2781646\n",
      "Streptomyces_broussonetiae_NZ_CP047020.1 3.7e-36 9509215 9509477\n",
      "Cutibacterium_acnes_HL096PA1_NC_021085.1 1.1e-35 2251737 2251977\n",
      "Cutibacterium_granulosum_NZ_LT906441.1 2.9e-35 665062 664913\n",
      "Cutibacterium_avidum_44067_NC_021064.1 5.3e-35 79470 79233\n",
      "Propionibacterium_australiense_NZ_LR134442.1 8.6e-35 2963336 2963537\n",
      "Acidipropionibacterium_virtanenii_NZ_CP025198.1 1.2e-34 2400903 2401174\n",
      "Propionibacterium_freudenreichii_NZ_LT593929.1 3.1e-34 1667038 1666837\n",
      "Acidipropionibacterium_jensenii_NZ_CP040635.1 4.2e-34 219741 220070\n",
      "Cutibacterium_modestum_NZ_CP017040.1 6.5e-34 1643852 1643607\n",
      "Propionibacterium_acidifaciens_NZ_CP033719.1 7.3e-34 2073973 2073772\n",
      "Propionibacterium_acidifaciens_NZ_CP033719.1 4.1e-33 1276668 1276878\n",
      "Epidermidibacterium_keratini_NZ_CP047156.1 8e-33 2584560 2584303\n",
      "Microlunatus_elymi_NZ_CP041692.1 9.3e-32 1346119 1346374\n",
      "Cutibacterium_granulosum_NZ_LT906441.1 1.5e-31 1088257 1087983\n",
      "Propionibacterium_australiense_NZ_LR134442.1 1.6e-31 2388400 2388607\n",
      "Corynebacterium_lactis_RW2-5_NZ_CP006841.1 2.5e-30 2241626 2241286\n",
      "Saccharopolyspora_spinosa_NZ_CP061007.1 1.8e-29 7279948 7279780\n",
      "Natronoglycomyces_albus_NZ_CP070496.1 2.1e-29 1028917 1028562\n",
      "Streptomyces_qaidamensis_NZ_CP015098.1 7.9e-27 1491750 1491908\n",
      "Streptomyces_marincola_NZ_CP021121.1 1.1e-26 4907646 4907801\n",
      "Saccharomonospora_viridis_DSM_43017_NC_013159.1 1.8e-26 1829469 1829612\n",
      "Corynebacterium_amycolatum_NZ_CP068168.1 7.3e-26 1281696 1281382\n",
      "Streptomyces_huasconensis_NZ_CP086119.1 7.7e-26 2113882 2114031\n",
      "Nocardioides_aromaticivorans_NZ_CP022295.1 1.1e-25 5049438 5049297\n",
      "Streptomyces_hawaiiensis_NZ_CP021978.1 1.5e-25 1452200 1452354\n",
      "Gordonia_polyisoprenivorans_VH2_NC_016906.1 1.8e-25 1748130 1747923\n",
      "Streptomyces_sp._ZYC-3_NZ_CP066831.1 1.9e-25 9117806 9117672\n",
      "Streptomyces_venezuelae_NZ_CP018074.1 2.2e-25 171604 171466\n",
      "Streptomyces_alboniger_NZ_CP023695.1 3.6e-25 2190564 2190712\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 4e-25 2595313 2595444\n",
      "Streptomyces_nitrosporeus_NZ_CP023702.1 4.8e-25 7232272 7232117\n",
      "Streptomyces_griseus_subsp._griseus_NBRC_13350_NC_010572.1 5.5e-25 7932619 7932470\n",
      "Nocardioides_anomalus_NZ_CP049257.1 5.9e-25 2675564 2675711\n",
      "Streptomyces_kanamyceticus_NZ_CP023699.1 6.9e-25 7265541 7265395\n",
      "Aeromicrobium_chenweiae_NZ_CP026952.1 7e-25 1854137 1854268\n",
      "Prauserella_marina_NZ_CP016353.1 8.3e-25 4810528 4810668\n",
      "Janibacter_melonis_NZ_CP044548.2 1.3e-24 2934342 2934185\n",
      "Streptomyces_californicus_NZ_CP070242.1 1.5e-24 328669 328823\n",
      "Streptomyces_atratus_NZ_CP027306.1 1.9e-24 867407 867541\n",
      "Streptomyces_violaceoruber_NZ_CP020570.1 2e-24 329471 329625\n",
      "Mycolicibacterium_thermoresistibile_NZ_LT906483.1 2e-24 2565107 2565231\n",
      "Streptomyces_paludis_NZ_CP031194.1 2.6e-24 5497017 5496888\n",
      "Streptomyces_alfalfae_NZ_CP060742.1 3.6e-24 2539612 2539750\n",
      "Streptomyces_gardneri_NZ_CP059991.1 4e-24 6917535 6917690\n",
      "Mycobacterium_holsaticum_NZ_CP080998.1 4.3e-24 2505598 2505720\n",
      "Gordonia_pseudoamarae_NZ_CP045806.1 4.9e-24 3625265 3625457\n",
      "Streptomyces_coeruleorubidus_NZ_CP023694.1 5.6e-24 1528164 1528303\n",
      "Streptomyces_pratensis_NZ_CP051486.1 8.2e-24 4458300 4458455\n",
      "Janibacter_indicus_NZ_CP013290.1 8.3e-24 288982 288852\n",
      "Streptomyces_parvulus_NZ_CP015866.1 1.1e-23 6758426 6758279\n",
      "Streptomyces_xinghaiensis_S187_NZ_CP023202.1 1.1e-23 5298025 5297897\n",
      "Amycolatopsis_albispora_NZ_CP015163.1 1.2e-23 2123709 2123859\n",
      "Thermobifida_halotolerans_NZ_CP063196.1 1.9e-23 158331 158179\n",
      "Mycolicibacterium_chitae_NZ_LR134355.1 2.3e-23 5262362 5262220\n",
      "Micromonospora_tulbaghiae_NZ_CP024087.1 2.4e-23 1076743 1076614\n",
      "Streptomyces_aurantiacus_NZ_AP023440.1 3.2e-23 8067221 8067087\n",
      "Pimelobacter_simplex_NZ_CP009896.1 3.3e-23 5561482 5561331\n",
      "Streptomyces_anulatus_NZ_CP086102.1 3.5e-23 7987395 7987245\n",
      "Streptomyces_olivaceus_NZ_CP043317.1 3.7e-23 406311 406460\n",
      "Streptomyces_ferrugineus_NZ_CP063373.1 5.3e-23 6920451 6920600\n",
      "Nocardioides_sambongensis_NZ_CP041091.1 5.4e-23 1460297 1460157\n",
      "Streptomyces_griseorubiginosus_NZ_CP032427.1 6.4e-23 1218672 1218821\n",
      "Tessaracoccus_lapidicaptus_NZ_LR214441.1 8.3e-23 350064 350192\n",
      "Nocardioides_dokdonensis_FR1436_NZ_CP015079.1 9.9e-23 4041453 4041308\n",
      "Propioniciclava_coleopterorum_NZ_CP049865.1 1.1e-22 1884848 1884709\n",
      "Corynebacterium_kroppenstedtii_DSM_44385_NC_012704.1 1.1e-22 2022904 2022645\n",
      "Streptomyces_chartreusis_NZ_CP023689.1 1.2e-22 2035838 2035972\n",
      "Acidipropionibacterium_virtanenii_NZ_CP025198.1 1.3e-22 53174 53335\n",
      "Nocardioides_sambongensis_NZ_CP041091.1 1.3e-22 3566100 3565963\n",
      "Nocardiopsis_dassonvillei_NZ_LR134501.1 1.5e-22 331867 331668\n",
      "Kitasatospora_albolonga_NZ_CP020563.1 1.7e-22 320506 320641\n",
      "Mycolicibacterium_mageritense_NZ_AP022567.1 2.2e-22 1903132 1903263\n",
      "Streptomyces_griseoviridis_NZ_CP034687.1 2.6e-22 7521204 7521071\n",
      "Thermobifida_alba_NZ_CP051627.1 4.8e-22 31203 31047\n",
      "Mycolicibacterium_mucogenicum_DSM_44124_NZ_CP062008.1 5.5e-22 250602 250472\n",
      "Streptomyces_cavourensis_NZ_CP024957.1 5.6e-22 389356 389505\n",
      "Streptomyces_fungicidicus_NZ_CP023407.1 6.7e-22 341481 341679\n",
      "Gordonia_phthalatica_NZ_CP011853.1 7.1e-22 3794546 3794396\n",
      "Nocardioides_aquaticus_NZ_CP075371.1 9.1e-22 1687641 1687792\n",
      "Streptomyces_vietnamensis_NZ_CP010407.1 1.1e-21 6657647 6657509\n",
      "Streptomyces_cinereoruber_NZ_CP023693.1 1.1e-21 6065462 6065324\n",
      "Streptomyces_albidoflavus_NC_020990.1 1.2e-21 807753 807620\n",
      "Gordonia_amicalis_NZ_CP091855.1 1.2e-21 1805973 1806112\n",
      "Nocardioides_humi_NZ_CP041146.1 1.3e-21 5587886 5588019\n",
      "Streptomyces_leeuwenhoekii_NZ_LN831790.1 2.1e-21 936564 936739\n",
      "Arachnia_propionica_NZ_LR134406.1 2.2e-21 3056315 3056491\n",
      "Nocardioides_piscis_NZ_CP049866.1 2.3e-21 2041943 2041795\n",
      "Gordonia_pseudoamarae_NZ_CP045806.1 2.4e-21 5167628 5167759\n",
      "Mycolicibacterium_phocaicum_NZ_AP022616.1 2.5e-21 3517229 3517359\n",
      "Gordonia_zhaorongruii_NZ_CP041763.1 2.6e-21 2100127 2100254\n",
      "Streptomyces_spectabilis_NZ_CP023690.1 2.6e-21 2966486 2966635\n",
      "Dietzia_timorensis_NZ_CP015961.1 2.8e-21 3091144 3091424\n",
      "Amycolatopsis_aidingensis_NZ_CP076538.1 3e-21 1446763 1446631\n",
      "Gordonia_alkanivorans_NZ_CP027114.1 3.1e-21 4840480 4840338\n",
      "Gordonia_rubripertincta_NZ_CP059694.1 3.1e-21 2197803 2197945\n",
      "Nocardioides_aromaticivorans_NZ_CP022295.1 3.4e-21 3256884 3256753\n",
      "Streptomyces_microflavus_NZ_CP054926.1 3.4e-21 486966 487105\n",
      "Streptomyces_nigra_NZ_CP029043.1 4.8e-21 6134892 6134769\n",
      "Nocardiopsis_dassonvillei_NZ_LR134501.1 6.9e-21 355246 355431\n",
      "Streptomyces_yatensis_NZ_CP072941.1 7.4e-21 8588482 8588282\n",
      "Streptomyces_koyangensis_NZ_CP049945.1 7.6e-21 5932944 5933077\n",
      "Nocardioides_panacisoli_NZ_CP080409.1 9.1e-21 1458728 1458564\n",
      "Streptomyces_marincola_NZ_CP021121.1 1.2e-20 4885921 4885790\n",
      "Nocardioides_humi_NZ_CP041146.1 1.4e-20 2529610 2529476\n",
      "Gordonia_pseudoamarae_NZ_CP045806.1 1.4e-20 85481 85353\n",
      "Rhodococcus_erythropolis_R138_NZ_CP007255.1 1.4e-20 5678784 5678657\n",
      "Streptomyces_solisilvae_NZ_CP065050.1 1.6e-20 2217647 2217847\n",
      "Rhodococcus_qingshengii_NZ_AP023172.1 1.8e-20 1221405 1221532\n",
      "Pimelobacter_simplex_NZ_CP009896.1 2e-20 3456196 3456063\n",
      "Streptomyces_sp._CRXT-Y-14_NZ_CP061281.1 2.5e-20 540599 540457\n",
      "Arachnia_propionica_NZ_LR134406.1 2.9e-20 3320261 3320120\n",
      "Nocardioides_seonyuensis_NZ_CP038436.1 5.1e-20 3446350 3446210\n",
      "Arachnia_rubra_NZ_CP072384.1 5.7e-20 3228901 3228746\n",
      "Gordonia_phthalatica_NZ_CP011853.1 6.2e-20 2944028 2944160\n",
      "Mycolicibacterium_neoaurum_NZ_CP074376.1 6.5e-20 3527901 3527762\n",
      "Propioniciclava_coleopterorum_NZ_CP049865.1 7.4e-20 3589984 3590120\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 9.9e-20 812457 812672\n",
      "Acidipropionibacterium_virtanenii_NZ_CP025198.1 1.1e-19 46431 46560\n",
      "Nocardioides_aquaticus_NZ_CP075371.1 1.2e-19 282320 282465\n",
      "Streptomyces_genisteinicus_NZ_CP060825.1 2.7e-19 5683688 5683533\n",
      "Mycolicibacterium_mengxianglii_NZ_CP065373.1 8.5e-19 5646592 5646743\n",
      "Streptomyces_griseofuscus_NZ_CP051006.1 1.1e-18 1113560 1113419\n",
      "Aeromicrobium_erythreum_NZ_CP011502.1 1.3e-18 3074533 3074401\n",
      "Gephyromycinifex_aptenodytis_NZ_CP051155.1 2.7e-18 541248 541415\n",
      "Actinopolyspora_erythraea_NZ_CP022752.1 4.6e-18 4253642 4253774\n",
      "Propionibacterium_acidifaciens_NZ_CP033719.1 6.4e-18 1813634 1813491\n",
      "Mycolicibacterium_pulveris_NZ_AP022599.1 1.1e-17 4659966 4659821\n",
      "Dermatophilus_congolensis_NZ_LT906453.1 1.9e-17 694971 694845\n",
      "Arachnia_rubra_NZ_CP072384.1 2.1e-17 7817 7701\n",
      "Salinispora_tropica_CNB-440_NC_009380.1 2.3e-17 2236180 2236310\n",
      "Propionibacterium_acidifaciens_NZ_CP033719.1 2.7e-17 2271566 2271469\n",
      "Mycolicibacterium_hassiacum_DSM_44199_NZ_LR026975.1 4.7e-17 2127817 2127947\n",
      "Cutibacterium_avidum_44067_NC_021064.1 5.4e-17 1063179 1063341\n",
      "Mycolicibacterium_thermoresistibile_NZ_LT906483.1 7.2e-17 4737767 4737636\n",
      "Arachnia_propionica_NZ_LR134406.1 7.9e-17 119104 119286\n",
      "Streptomyces_venezuelae_NZ_CP018074.1 9.6e-17 168147 168041\n",
      "Arachnia_propionica_NZ_LR134406.1 1.3e-16 19201 19085\n",
      "Propionibacterium_australiense_NZ_LR134442.1 2.9e-16 438329 438188\n",
      "Cutibacterium_acnes_HL096PA1_NC_021085.1 3.3e-16 487683 487569\n",
      "Nocardia_brasiliensis_NZ_CP022088.2 4e-16 1741223 1741379\n",
      "Nocardia_brasiliensis_NZ_CP022088.2 4.1e-16 1741379 1741537\n",
      "Cutibacterium_granulosum_NZ_LT906441.1 5e-16 929901 929788\n",
      "Arachnia_propionica_NZ_LR134406.1 1e-15 3056935 3056748\n",
      "Ornithinimicrobium_pratense_NZ_CP044427.1 1.3e-15 2785474 2785344\n",
      "Nocardioides_baekrokdamisoli_NZ_AP019307.1 2.3e-15 544595 544750\n",
      "Propionibacterium_acidifaciens_NZ_CP033719.1 3.5e-15 1314778 1314652\n",
      "Propionibacterium_freudenreichii_NZ_LT593929.1 5.4e-15 2483872 2483991\n",
      "Saccharopolyspora_pogona_NZ_CP031142.1 6.8e-15 4402012 4402152\n",
      "Propionibacterium_australiense_NZ_LR134442.1 6.9e-15 3002665 3002538\n",
      "Cutibacterium_modestum_NZ_CP017040.1 7.4e-15 915698 915811\n",
      "Spiractinospora_alimapuensis_NZ_CP072467.1 2.7e-14 778782 778917\n",
      "Saccharopolyspora_erythraea_NRRL_2338_NC_009142.1 3e-14 4900015 4900183\n",
      "Cutibacterium_modestum_NZ_CP017040.1 7e-14 1317809 1317938\n",
      "Saccharopolyspora_gregorii_NZ_CP059557.1 4e-13 4197425 4197558\n",
      "Arachnia_rubra_NZ_CP072384.1 7.7e-13 297428 297545\n",
      "Streptomyces_decoyicus_NZ_CP082301.1 9.3e-13 5585938 5586114\n",
      "Tsukamurella_paurometabola_DSM_20162_NC_014158.1 1.1e-12 2413124 2413017\n",
      "Streptomyces_inhibens_NZ_CP084681.1 1.5e-12 6557900 6558073\n",
      "Allosaccharopolyspora_coralli_NZ_CP045929.1 1.5e-12 3007301 3007143\n",
      "Corynebacterium_mustelae_NZ_CP011542.1 2.6e-12 1389454 1389319\n",
      "Streptomyces_angustmyceticus_NZ_CP082945.1 5.1e-12 2732942 2732766\n",
      "Gordonia_hongkongensis_NZ_CP095552.1 7e-12 1182256 1182395\n",
      "Gordonia_otitidis_NZ_CP085974.1 1.1e-11 2455637 2455498\n",
      "Streptomyces_platensis_NZ_CP023691.1 1.6e-11 5844372 5844549\n",
      "Conexibacter_woesei_DSM_14684_NC_013739.1 1.9e-11 1815067 1814888\n",
      "Actinomadura_graeca_NZ_CP059572.1 1.1e-10 4514849 4514948\n",
      "Cutibacterium_avidum_44067_NC_021064.1 2e-10 1927168 1927053\n",
      "Cutibacterium_acnes_HL096PA1_NC_021085.1 1.3e-09 408128 408260\n",
      "Kitasatospora_aureofaciens_NZ_CP020567.1 1.5e-09 1753186 1753093\n",
      "Corynebacterium_urogenitale_NZ_CP045032.1 1.8e-09 358107 358227\n",
      "Cutibacterium_avidum_44067_NC_021064.1 2e-09 1659683 1659805\n",
      "Cutibacterium_granulosum_NZ_LT906441.1 3e-09 1128915 1128793\n",
      "Propionimicrobium_sp._Marseille-P3275_NZ_LT706985.1 3.5e-09 1397319 1397441\n",
      "Streptomyces_sp._1_25_NZ_CP086120.1 4.1e-09 3990645 3990785\n",
      "Corynebacterium_choanae_NZ_CP033896.1 4.9e-09 548663 548837\n",
      "Cutibacterium_acnes_HL096PA1_NC_021085.1 5.1e-09 1640011 1640133\n",
      "Cutibacterium_modestum_NZ_CP017040.1 1.6e-08 2529416 2529295\n",
      "Streptomyces_hawaiiensis_NZ_CP021978.1 1.8e-08 5836627 5836754\n",
      "Streptomyces_fodineus_NZ_CP017248.1 2.1e-08 3623678 3623547\n",
      "Streptacidiphilus_bronchialis_NZ_CP031264.1 2.5e-08 4515673 4515717\n",
      "Streptomyces_roseirectus_NZ_CP060828.1 3.8e-08 8304170 8304217\n",
      "Corynebacterium_ulcerans_NZ_CP009716.1 4.2e-08 1003968 1003836\n",
      "Corynebacterium_pseudotuberculosis_NZ_CP016826.1 5.1e-08 1922370 1922504\n",
      "Corynebacterium_argentoratense_DSM_44202_NC_022198.1 5.7e-08 1734884 1734840\n",
      "Corynebacterium_sphenisci_DSM_44792_NZ_CP009248.1 7.1e-08 1763393 1763434\n",
      "Streptomyces_vinaceus_NZ_CP023692.1 8.9e-08 2622907 2622859\n",
      "Corynebacterium_diphtheriae_bv._mitis_str._ISS_3319_NZ_CP025209.1 9.8e-08 990945 990806\n",
      "Streptomyces_pluripotens_NZ_CP021080.1 1e-07 4967725 4967866\n",
      "Streptomyces_auratus_AGR0001_NZ_CP072931.1 1e-07 5358666 5358709\n",
      "Streptomyces_rimosus_subsp._rimosus_ATCC_10970_NZ_CP048261.1 1.1e-07 6098452 6098493\n",
      "Streptomyces_roseirectus_NZ_CP060828.1 1.4e-07 3228758 3228626\n",
      "Streptomyces_aquilus_NZ_CP034463.1 1.9e-07 6948034 6948081\n",
      "Streptomyces_tuirus_NZ_AP023439.1 1.9e-07 5087304 5087343\n",
      "Streptomyces_yatensis_NZ_CP072941.1 2e-07 3864260 3864218\n",
      "Streptomyces_nojiriensis_NZ_CP071139.1 2.1e-07 5935920 5935960\n",
      "Streptomyces_aurantiacus_NZ_AP023440.1 2.1e-07 3185878 3185830\n",
      "Cutibacterium_modestum_NZ_CP017040.1 2.2e-07 999318 999195\n",
      "Streptomyces_parvulus_NZ_CP015866.1 2.4e-07 4846364 4846470\n",
      "Streptomyces_broussonetiae_NZ_CP047020.1 2.5e-07 6209971 6210018\n",
      "Streptomyces_griseofuscus_NZ_CP051006.1 2.5e-07 3162304 3162265\n",
      "Streptomyces_murinus_NZ_CP066774.1 2.5e-07 1432320 1432359\n",
      "Streptomyces_phaeolivaceus_NZ_CP045096.1 2.5e-07 7170535 7170678\n",
      "Kitasatospora_albolonga_NZ_CP020563.1 3e-07 5642508 5642548\n",
      "Streptomyces_griseus_subsp._griseus_NBRC_13350_NC_010572.1 3e-07 2717412 2717372\n",
      "Streptomyces_microflavus_NZ_CP054926.1 3e-07 5420422 5420462\n",
      "Streptomyces_qaidamensis_NZ_CP015098.1 3e-07 6017689 6017795\n",
      "Streptomyces_gilvosporeus_NZ_CP020569.1 3.1e-07 6026900 6026942\n",
      "Streptomyces_cyaneochromogenes_NZ_CP034539.1 3.5e-07 6745600 6745639\n",
      "Frankia_inefficax_NC_014666.1 4.3e-07 1735902 1736038\n",
      "Streptomyces_avermitilis_MA-4680_=_NBRC_14893_NC_003155.5 4.3e-07 3783220 3783181\n",
      "Streptomyces_chromofuscus_NZ_CP063374.1 4.3e-07 5320427 5320466\n",
      "Streptomyces_dengpaensis_NZ_CP026652.1 4.3e-07 2953236 2953197\n",
      "Streptomyces_galilaeus_NZ_CP023703.1 4.3e-07 5299365 5299404\n",
      "Streptomyces_nigra_NZ_CP029043.1 4.3e-07 4986461 4986500\n",
      "Streptomyces_viridosporus_T7A_NZ_CP023700.1 4.3e-07 5499447 5499486\n",
      "Streptomyces_coeruleorubidus_NZ_CP023694.1 4.3e-07 6194038 6194077\n",
      "Streptomyces_ferrugineus_NZ_CP063373.1 4.3e-07 8430684 8430645\n",
      "Streptomyces_nodosus_NZ_CP071465.1 4.3e-07 2908860 2908821\n",
      "Streptomyces_chartreusis_NZ_CP023689.1 4.7e-07 3465710 3465671\n",
      "Streptomyces_griseorubiginosus_NZ_CP032427.1 5.2e-07 3355416 3355377\n",
      "Streptomyces_lincolnensis_NZ_CP022744.1 5.2e-07 3424867 3424828\n",
      "Streptomyces_leeuwenhoekii_NZ_LN831790.1 5.2e-07 3051469 3051430\n",
      "Streptomyces_xiamenensis_NZ_CP009922.3 7e-07 4245233 4245283\n",
      "Corynebacterium_rouxii_NZ_LR738855.1 7.1e-07 1024600 1024460\n",
      "Streptomyces_harbinensis_NZ_CP054938.1 7.2e-07 4345777 4345819\n",
      "Streptomyces_sp._CRXT-Y-14_NZ_CP061281.1 1.8e-06 2584510 2584468\n",
      "Kitasatospora_setae_KM-6054_NC_016109.1 2e-06 5452117 5452161\n",
      "Streptomyces_fradiae_ATCC_10745_=_DSM_40063_NZ_CP023696.1 2.1e-06 4618353 4618393\n",
      "Kribbella_qitaiheensis_NZ_CP043661.1 2.2e-06 1468994 1468950\n",
      "Streptomyces_rimosus_subsp._rimosus_ATCC_10970_NZ_CP048261.1 3.2e-06 6098340 6098437\n",
      "Corynebacterium_aquilae_DSM_44791_NZ_CP009245.1 4.2e-06 2352140 2352083\n",
      "Corynebacterium_silvaticum_NZ_CP021417.1 4.8e-06 1082836 1082702\n",
      "Streptomyces_deccanensis_NZ_CP092431.1 5e-06 6867680 6867719\n",
      "Streptomyces_lydicus_NZ_CP019457.1 6e-06 5667241 5667415\n",
      "Spiractinospora_alimapuensis_NZ_CP072467.1 1.1e-05 779169 779220\n",
      "Streptomyces_roseirectus_NZ_CP060828.1 1.2e-05 8303904 8303981\n",
      "Corynebacterium_lactis_RW2-5_NZ_CP006841.1 1.6e-05 2414427 2414469\n",
      "Streptomyces_alfalfae_NZ_CP060742.1 3.9e-05 5605377 5605415\n",
      "Streptomyces_sp._1_25_NZ_CP086120.1 9.1e-05 3990556 3990623\n",
      "Corynebacterium_argentoratense_DSM_44202_NC_022198.1 0.00011 784065 784027\n",
      "Streptomyces_auratus_AGR0001_NZ_CP072931.1 0.00012 5358535 5358646\n",
      "Kribbella_qitaiheensis_NZ_CP043661.1 0.00013 1469115 1469004\n",
      "Streptomyces_alfalfae_NZ_CP060742.1 0.00024 5605267 5605351\n",
      "Corynebacterium_amycolatum_NZ_CP068168.1 0.00032 1429132 1429174\n",
      "Gordonia_otitidis_NZ_CP085974.1 0.00032 2455720 2455647\n",
      "Natronoglycomyces_albus_NZ_CP070496.1 0.00076 560906 561199\n",
      "Corynebacterium_anserum_NZ_CP046883.1 0.0014 303769 303870\n",
      "Streptomyces_griseus_subsp._griseus_NBRC_13350_NC_010572.1 0.0023 2717520 2717426\n",
      "Rhodococcus_globerulus_NZ_CP079698.1 0.0039 688767 688728\n",
      "Streptomyces_sp._CRXT-Y-14_NZ_CP061281.1 0.0046 2584617 2584532\n",
      "Rhodococcus_globerulus_NZ_CP079698.1 0.0051 6650555 6650514\n",
      "Corynebacterium_gerontici_NZ_CP033897.1 0.011 990467 990427\n",
      "Streptomyces_chartreusis_NZ_CP023689.1 0.043 3465836 3465734\n",
      "Streptomyces_phaeolivaceus_NZ_CP045096.1 0.058 7170411 7170496\n",
      "Propionimicrobium_sp._Marseille-P3275_NZ_LT706985.1 0.073 1605530 1605566\n",
      "Corynebacterium_kutscheri_NZ_CP011312.1 0.074 851139 851013\n",
      "Streptomyces_deccanensis_NZ_CP092431.1 0.086 6867555 6867641\n",
      "Streptomyces_aurantiacus_NZ_AP023440.1 0.086 3186025 3185939\n",
      "Cutibacterium_modestum_NZ_CP017040.1 0.088 2145152 2145118\n",
      "Corynebacterium_pelargi_NZ_CP035299.1 0.089 1402550 1402590\n",
      "Corynebacterium_pseudopelargi_NZ_CP033898.1 0.089 1402939 1402979\n",
      "Streptomyces_leeuwenhoekii_NZ_LN831790.1 0.093 3051609 3051523\n",
      "Streptomyces_chromofuscus_NZ_CP063374.1 0.11 5320313 5320399\n",
      "Streptomyces_ferrugineus_NZ_CP063373.1 0.12 8430814 8430727\n",
      "Streptomyces_griseorubiginosus_NZ_CP032427.1 0.14 3355539 3355448\n",
      "Streptomyces_nigra_NZ_CP029043.1 0.14 4986347 4986433\n",
      "Streptacidiphilus_bronchialis_NZ_CP031264.1 0.15 4515574 4515668\n",
      "Streptomyces_tuirus_NZ_AP023439.1 0.17 5087188 5087274\n",
      "Streptomyces_hawaiiensis_NZ_CP021978.1 0.17 5836500 5836585\n",
      "Streptomyces_yatensis_NZ_CP072941.1 0.17 3864355 3864267\n",
      "Streptomyces_viridosporus_T7A_NZ_CP023700.1 0.19 5499317 5499403\n",
      "Streptomyces_xiamenensis_NZ_CP009922.3 0.19 4245140 4245208\n",
      "Streptomyces_lincolnensis_NZ_CP022744.1 0.21 3424984 3424899\n",
      "Streptomyces_fodineus_NZ_CP017248.1 0.25 3623795 3623693\n",
      "Streptomyces_fradiae_ATCC_10745_=_DSM_40063_NZ_CP023696.1 0.31 4618242 4618332\n",
      "Streptomyces_coeruleorubidus_NZ_CP023694.1 0.37 6193922 6194006\n",
      "Streptomyces_harbinensis_NZ_CP054938.1 0.39 4345684 4345752\n",
      "Cutibacterium_modestum_NZ_CP017040.1 0.43 1881981 1882015\n",
      "Phytohabitans_suffuscus_NZ_AP022871.1 0.44 6622869 6622727\n",
      "Rhodoluna_limnophila_NZ_CP040508.1 0.77 926607 926565\n",
      "Streptomyces_broussonetiae_NZ_CP047020.1 1.2 6209853 6209943\n",
      "Kitasatospora_albolonga_NZ_CP020563.1 1.5 5642371 5642484\n",
      "Streptomyces_gilvosporeus_NZ_CP020569.1 1.7 6026771 6026868\n",
      "Corynebacterium_callunae_DSM_20147_NC_020506.1 1.7 548836 548872\n",
      "Corynebacterium_argentoratense_DSM_44202_NC_022198.1 3.5 1855253 1855228\n",
      "Streptomyces_microflavus_NZ_CP054926.1 3.7 5420314 5420398\n",
      "Streptomyces_cyaneochromogenes_NZ_CP034539.1 6 6745457 6745542\n",
      "Streptomyces_dengpaensis_NZ_CP026652.1 8.5 2953363 2953276\n",
      "Corynebacterium_rouxii_NZ_LR738855.1 9.4 572482 572341\n"
     ]
    }
   ],
   "source": [
    "dict = {}\n",
    "with open(output_dir + '/' + 'summary.txt', 'r') as f:\n",
    "    for l in f:\n",
    "        if not(l[0] == '#'): \n",
    "            a = l.split()\n",
    "            print(a[0], a[12], a[6], a[7])\n",
    "        #    if a[16] == '!':\n",
    "        #        if a[0] in dict:\n",
    "        #            if float(a[15]) < dict[a[0]][1]:\n",
    "        #                dict[a[0]] = (a[7]+'-'+a[8],float(a[15]))\n",
    "        #        else:\n",
    "        #            dict[a[0]] = (a[7]+'-'+a[8],float(a[15]))\n",
    "#with open(output_dir + '/keep_list.txt', 'w') as f:\n",
    "#    lines = []\n",
    "#    for k, v in dict.items():\n",
    "#        lines.append(k + '/' + v[0] + \"\\n\")\n",
    "#    f.write(''.join(lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bbc6031a-0417-40c1-abeb-d42cc97abb51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='wsl cd /mnt/d/Project_Data/Project_8/Output ; esl-alimanip -o unique_aligns.sto --seq-r keep_list.txt rcm_align_4.sto ', returncode=0)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run('wsl cd ' + wsl_output_loc + ' ; esl-alimanip -o unique_aligns.sto --seq-r keep_list.txt rcm_align_4.sto ', shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7bbbf8d4-a16c-4953-b89d-36b6463c6187",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mycobacterium_noviomagense_NZ_AP022583.1                                    -         struc_4.cacofold     -          cm        1      247   663916   663628      -    no    1 0.66   5.4  190.6   2.2e-31 !   -\n",
      "\n",
      "Mycobacterium_noviomagense_NZ_AP022583.1                                    -         struc_4.cacofold     -          cm        1      247  1344376  1344211      -    no    1 0.63   0.0   36.0      0.48 ?   -\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(output_dir + '/' + 'rcmhits_4.txt', 'r') as f:\n",
    "    for l in f:\n",
    "        if 'Mycobacterium_noviomagens' in l:\n",
    "            print(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e29d987b-afc0-4c48-8e81-5ca658c466d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('663916-663628', 2.2e-31)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict['Mycobacterium_noviomagense_NZ_AP022583.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bb06f0b-ba3e-4f6a-83cf-7997ca085a8b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55c2e3db-7c3a-4ad6-8705-1c2bcb7095fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmbuild --noss -F cm1.cm align_mask5.sto\n",
    "cmcalibrate cm1.cm\n",
    "cmsearch --tblout cmhits1.txt -A cmalign.sto --incE 1e-10 cm1.cm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa\n",
    "esl-alimask -g -o cm_mask1.sto cmalign.sto\n",
    "\n",
    "\n",
    "cmbuild -F cm2.cm cm_mask1.sto\n",
    "cmcalibrate cm2.cm\n",
    "cmsearch --tblout cmhits2.txt -A cmalign1.sto --cpu 16 --incE 1e-10 cm2.cm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa\n",
    "esl-alimask -g -o cm_mask2.sto cmalign1.sto\n",
    "\n",
    "cmbuild --noss -F cm3.cm cm_mask2.sto\n",
    "cmcalibrate cm3.cm\n",
    "cmsearch --tblout cmhits3.txt -A cmalign2.sto --incE 1e-5 cm3.cm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa\n",
    "esl-alimask -g -o cm_mask3.sto cmalign2.sto\n",
    "esl-alimask -g -o cm_mask3.sto cmalign2.sto\n",
    "\n",
    "~/rscape_v2.0.0.g/bin/R-scape --cacofold --outname struc1 align_mask5.sto\n",
    "cmbuild -F rcm2.cm struc1.cacofold.sto \n",
    "cmcalibrate rcm2.cm\n",
    "cmsearch --tblout rcmhits2.txt -A rcmalign1.sto --cpu 16 --incE 1e-10 rcm2.cm /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa\n",
    "esl-alimask -g -o rcm_mask2.sto rcmalign1.sto\n",
    "\n",
    "~/rscape_v2.0.0.g/bin/R-scape --cacofold --outname struc2_t rcm_mask2.sto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535cde22-6303-4e73-a609-83411b63d486",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 1\n",
    "a= 'nhmmer -A align_'+str(i)+'.sto -o hmmer.txt --tblout summary.txt --notextw --cpu 16 --incE 1e-10 hmm_'+str(i)+'.hmm /mnt/d/BLAST/actinobacteria_ref_rep_comp/all_regions.faa /mnt/d/Actinobacteria_Ref_Rep_Lev_Complete/all_actinobacteria_ref_rep_comp.faa'\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b286e106-a4ba-4d5e-8242-21f8480c484e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5141bc1-e7d8-4b26-94d3-9517a352f157",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
