{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1e30ed4-8a0e-4a6f-96eb-dfe116dfb2de",
   "metadata": {},
   "source": [
    "#### Import packages, set directories and parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf9c6377-5080-4c22-94db-feabe17cc47c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import math\n",
    "from tqdm import tqdm\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from joblib import Parallel, delayed\n",
    "from Comparative_Analysis import Blast_Functions as blastfn\n",
    "import time\n",
    "import os\n",
    "import copy\n",
    "from scipy.stats import chi2, binom\n",
    "from Comparative_Analysis import Alignment as alignfn\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import subprocess\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "79c1ca3d-a68f-4a9d-a5c4-25df7385b3ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = 'F:/Project_Data/mabR_Project'\n",
    "mycobacteria_seq_dir = 'F:/Datasets/NCBI_Refseq_Mycobacteriaceae_Complete_Annot_20230511/data'\n",
    "actinobacteria_seq_dir = 'F:/Datasets/NCBI_Refseq_Actinomycetes_Complete_Annot_20230511/data'\n",
    "tb_species = 'AL123456.3' \n",
    "tb_annotation_dirname = 'GCA_000195955.2'\n",
    "min_region_length = 7 \n",
    "full_build = False\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "muscle_exe = 'C:/Users/nicho/Muscle/muscle3.8.31_i86win32.exe'\n",
    "full_run = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f3bcbac1-03df-43f0-813e-0ca8f97a28bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "mycobacteria_dirs = []\n",
    "for dir in util.list_dirs(mycobacteria_seq_dir):\n",
    "    if os.path.exists(mycobacteria_seq_dir + '/' + dir + '/genomic.gbff'):\n",
    "        mycobacteria_dirs.append(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "161765c6-2dbb-4994-9fe0-5df9f817a1b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "actinobacteria_dirs = []\n",
    "for dir in util.list_dirs(actinobacteria_seq_dir):\n",
    "    if os.path.exists(actinobacteria_seq_dir + '/' + dir + '/genomic.gbff'):\n",
    "        actinobacteria_dirs.append(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98d6cb4b-f436-4866-8be5-fa1232f4d8dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tb_mycobacteria_rbh = pd.read_csv(project_dir + '/tb_mycobacteria_reciprocal_best_hits.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "63f58a21-4009-4b7e-b289-7766b64797b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_dict = {}\n",
    "dirname_dict = {}\n",
    "for dirname in (mycobacteria_dirs):\n",
    "        for genome_record in SeqIO.parse(mycobacteria_seq_dir + '/' + dirname + '/genomic.gbff', \"genbank\"):\n",
    "            accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "            names_dict[accession_ver] = genome_record.annotations['organism']\n",
    "        dirname_dict[dirname] = genome_record.annotations['organism']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8e79089d-e833-4b2c-9d19-54875fb3525e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(seq_string):\n",
    "    temp = []\n",
    "    for char in reversed(seq_string):\n",
    "        if char == 'A':\n",
    "            comp = 'T'\n",
    "        elif char == 'T':\n",
    "            comp = 'A'\n",
    "        elif char == 'C':\n",
    "            comp = 'G'\n",
    "        elif char == 'G':\n",
    "            comp = 'C'\n",
    "        else:\n",
    "            comp = char\n",
    "        temp.append(comp)\n",
    "    return ''.join(temp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c997be0-04a1-4685-b8ba-67c71a3373c2",
   "metadata": {},
   "source": [
    "#### Lookup function to find orthologs in M.tb based on reciprocal best hit outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "270bbd3f-598e-428d-8b30-7d9fc6fc9a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ortholog_in_mtb(accession_ver, locus):\n",
    "    gene = accession_ver + '@' + locus\n",
    "    temp = tb_mycobacteria_rbh.query('target_ref == @gene and query_species == @tb_species')\n",
    "    if len(temp) == 1:\n",
    "        for i, r in temp.iterrows():\n",
    "            temp2 = r['query_ref']\n",
    "            return temp2.split('@')[1]\n",
    "    else:\n",
    "        return 'No_orth'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e687d1b4-31ac-4d5c-a571-d05535336de9",
   "metadata": {},
   "source": [
    "#### Produce reference FASTA files (a) TB and (b, c) all mycobacteria, actionobacteria for searching against motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2a34e5ee-8b1b-4a69-85e9-360339a0e3f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4031\n"
     ]
    }
   ],
   "source": [
    "genome_record = next(SeqIO.parse(mycobacteria_seq_dir + '/' + tb_annotation_dirname + '/genomic.gbff', \"genbank\"))\n",
    "tot = 0\n",
    "for feature in genome_record.features:\n",
    "    a = feature.qualifiers\n",
    "    if feature.type not in ['source','gene'] and (int(feature.location.start) < int(feature.location.end)) and (int(feature.location.end) - int(feature.location.start)) < 1000000:\n",
    "        if feature.type == 'CDS':\n",
    "            tot+=1\n",
    "print(tot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c78a8a2a-f363-48b7-a30e-8d4f0dbd81e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 11.82it/s]\n"
     ]
    }
   ],
   "source": [
    "genome_record = next(SeqIO.parse(mycobacteria_seq_dir + '/' + tb_annotation_dirname + '/genomic.gbff', \"genbank\"))\n",
    "mtb_sequence = str(genome_record.seq)\n",
    "accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "util.produce_fasta_file([[accession_ver, mtb_sequence]], project_dir + '/mtb_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7d173f9d-f5bd-49d4-8ae1-789fae6e2a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    temp = []\n",
    "    for dir in util.list_dirs(mycobacteria_seq_dir): \n",
    "        if os.path.exists(mycobacteria_seq_dir + '/' + dir + '/genomic.gbff'):\n",
    "            for genome_record in (SeqIO.parse(mycobacteria_seq_dir + '/' + dir + '/genomic.gbff', \"genbank\")):\n",
    "                accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "                full_sequence = str(genome_record.seq)\n",
    "                temp.append([accession_ver, full_sequence])\n",
    "    util.produce_fasta_file(temp, project_dir + '/all_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d0d10d6d-10c8-49f5-9e7b-838f3fb960f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 900/900 [01:03<00:00, 14.18it/s]\n"
     ]
    }
   ],
   "source": [
    "if 1==1:\n",
    "#if full_run == True:\n",
    "    temp = []\n",
    "    for dir in util.list_dirs(actinobacteria_seq_dir): \n",
    "        if os.path.exists(actinobacteria_seq_dir + '/' + dir + '/genomic.gbff'):\n",
    "            for genome_record in (SeqIO.parse(actinobacteria_seq_dir + '/' + dir + '/genomic.gbff', \"genbank\")):\n",
    "                accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "                full_sequence = str(genome_record.seq)\n",
    "                temp.append([accession_ver, full_sequence])\n",
    "    util.produce_fasta_file(temp, project_dir + '/all_actinobacteria_seq')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50fbf555-1082-451c-827f-492e9f228f5b",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Functions to locate arbitrary region in organism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "be6df13d-3ee1-485b-b5c7-8d788256b55e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotated_regions_dataset(num_subsets, subset_num, dir_list, seqdir, cds_only = False): \n",
    "    output = []\n",
    "    sequence_dirs = util.chunk_list(dir_list, num_subsets, subset_num)\n",
    "    for dirname in sequence_dirs:\n",
    "        annotated_regions = []\n",
    "        intergenic_regions = []\n",
    "        for record in (SeqIO.parse(seqdir + '/'+dirname+'/genomic.gbff', \"genbank\")):\n",
    "            accession_ver = record.annotations['accessions'][0] + '.' + str(record.annotations['sequence_version'])\n",
    "            for feature in record.features:\n",
    "                a = feature.qualifiers\n",
    "                if feature.type not in ['source','gene'] and (int(feature.location.start) < int(feature.location.end)) and (int(feature.location.end) - int(feature.location.start)) < 1000000:\n",
    "                    if cds_only == True and not(feature.type == 'CDS'):\n",
    "                        continue\n",
    "                    else:\n",
    "                        if not(a.get(\"product\") == None):\n",
    "                               product = a.get(\"product\")[0]\n",
    "                        else:\n",
    "                               product = ''\n",
    "                        if not(a.get(\"locus_tag\")==None):\n",
    "                            locus_tag = a.get(\"locus_tag\")[0]\n",
    "                        else:\n",
    "                            locus_tag = feature.type\n",
    "                        annotated_regions.append((locus_tag, product, feature.type, int(feature.location.start), int(feature.location.end), str(feature.location.strand)))\n",
    "            annotated_regions.sort(key = lambda x: x[4])\n",
    "            prev_strand = 0\n",
    "            prev_locus = ''\n",
    "            prev_product = ''\n",
    "            max_stop = 0\n",
    "            for n, (locus, product, feature_type, start, stop, strand) in enumerate(annotated_regions):\n",
    "                if start > max_stop:\n",
    "                    intergenic_regions.append([prev_locus+':'+locus, prev_product + ':' + product, 'Inter-feature',max_stop, start, str(prev_strand)+':'+str(strand)])\n",
    "                if stop > max_stop:\n",
    "                    prev_locus = locus\n",
    "                    prev_product = product\n",
    "                    prev_strand = strand\n",
    "                max_stop = max(max_stop, stop)    \n",
    "            for x in intergenic_regions:\n",
    "                annotated_regions.append(x)\n",
    "            annotated_regions.sort(key = lambda x : x[4])\n",
    "            output.append([accession_ver, annotated_regions])\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "41cad641-7b47-4eff-b637-970c4593b41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "parallel_output = Parallel(n_jobs=-1)(delayed(annotated_regions_dataset)(num_cores, core_number, mycobacteria_dirs, mycobacteria_seq_dir) for core_number in core_numbers)\n",
    "annotated_regions_dict = {}\n",
    "for x in parallel_output:\n",
    "    for y in x:\n",
    "        annotated_regions_dict[y[0]] = y[1]\n",
    "parallel_output = Parallel(n_jobs=-1)(delayed(annotated_regions_dataset)(num_cores, core_number, mycobacteria_dirs, mycobacteria_seq_dir, True) for core_number in core_numbers)\n",
    "cds_annotated_regions_dict = {}\n",
    "for x in parallel_output:\n",
    "    for y in x:\n",
    "        cds_annotated_regions_dict[y[0]] = y[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "486e3d2c-4c89-4649-9e43-457a28c4a3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def location(accession_ver, start1, end1, annotation_lookup_dictionary):\n",
    "    feature_matches = []\n",
    "    for (locus, product, feature, start, stop, strand) in annotation_lookup_dictionary[accession_ver]:\n",
    "            if start< end1 and stop > start1:\n",
    "                overlap = str(int(100*(min(end1, stop) - max(start1, start))/ (end1-start1)))+'%'\n",
    "                feature_matches.append([locus, product, feature, overlap, strand])\n",
    "    return feature_matches"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd747b2c-841f-40e5-982d-45d812215121",
   "metadata": {},
   "source": [
    "#### Create dictionary of upstream sequences - sequence continue upstream to the next CDS not next feature (e.g. if intervening repeat region this is included"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "88d4c9a3-0301-4492-a123-78ee275dd0ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_upstream_sequence_dataset(num_subsets, subset_num, dir_list, seqdir, additional_offset = 0):     # Set additional offset = 3 if you want start codon of downstream CDS to be included - useful for alignment \n",
    "    sequence_dirs = util.chunk_list(dir_list, num_subsets, subset_num)\n",
    "    upstream_cds_regions = []\n",
    "    for dirname in (sequence_dirs):\n",
    "        all_features = []\n",
    "        cds_features = []\n",
    "        for genome_record in SeqIO.parse(mycobacteria_seq_dir + '/' + dirname + '/genomic.gbff', \"genbank\"):\n",
    "            accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "            full_sequence = str(genome_record.seq)\n",
    "            len_full_sequence = len(full_sequence)\n",
    "            for feature in genome_record.features:\n",
    "                if feature.type in ['gene', 'source']:\n",
    "                    continue\n",
    "                a = feature.qualifiers\n",
    "                feature_type = feature.type\n",
    "                if a.get(\"locus_tag\") != None:\n",
    "                    locus_tag = a.get(\"locus_tag\")[0]\n",
    "                    accession_locus = accession_ver + '@' + locus_tag\n",
    "                else:\n",
    "                    accession_locus  = ''\n",
    "                (start, stop, strand) = (int(feature.location.start), int(feature.location.end), int(feature.location.strand))\n",
    "                all_features.append([accession_locus, feature_type, start, stop, strand])\n",
    "                if feature_type == 'CDS':\n",
    "                    cds_features.append([accession_locus, feature_type, start, stop, strand])\n",
    "            \n",
    "            features = cds_features  # Replace with line below if you want to go upstream only to next annotated feature and not next CDS\n",
    "            #features = all_features\n",
    "            \n",
    "            # Positive strand upstream \n",
    "            features.sort(key = lambda x: x[2])\n",
    "            max_stop = 0\n",
    "            for (accession_locus, feature_type, start, stop, strand) in features:\n",
    "                if max_stop < start and feature_type == 'CDS' and strand == 1 and start - max_stop < 100000:    #Avoid joins where biopython interprets inconsistently \n",
    "                    upstream_cds_regions.append([accession_locus, max_stop, start, strand, full_sequence[max_stop: start+additional_offset]])\n",
    "                max_stop = max(max_stop, stop)\n",
    "            # Negative strand upstream\n",
    "            features.sort(key = lambda x: x[3], reverse = True)\n",
    "            min_start = len(full_sequence)-1\n",
    "            for (accession_locus, feature_type, start, stop, strand) in features:\n",
    "                if stop < min_start and feature_type == 'CDS' and strand == -1 and min_start - stop < 100000:\n",
    "                    upstream_cds_regions.append([accession_locus, stop, min_start, strand, reverse_complement(full_sequence[stop-additional_offset: min_start])])\n",
    "                min_start = min(min_start, start)\n",
    "    return (upstream_cds_regions)           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a8075f18-ddc6-43c2-aaec-d25000f9d589",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_build == True:\n",
    "#if 1==1:\n",
    "    mycobacteria_upstream_dict = {}\n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(generate_upstream_sequence_dataset)(num_cores, core_number, mycobacteria_dirs, mycobacteria_seq_dir) for core_number in core_numbers)\n",
    "    for x in parallel_output:\n",
    "        for n in x:\n",
    "            mycobacteria_upstream_dict[n[0]] = [n[1], n[2], n[3], n[4]]\n",
    "    with open(project_dir + '/mycobacteria_upstream_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(mycobacteria_upstream_dict, f) \n",
    "else:\n",
    "    with open(project_dir + '/mycobacteria_upstream_dict.pkl', 'rb') as f:\n",
    "        mycobacteria_upstream_dict = pickle.load(f)    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b4bf2c-e1bb-4623-aed5-c0da8fb1f9c7",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Functions to run MEME and FIMO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "66efc6cc-4049-4566-bca3-404079e7a691",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_meme(search_regions_loc, output_dir, min_width, min_sites):\n",
    "    subprocess.run('wsl export PATH=$HOME/meme/bin:$HOME/meme/libexec/meme-5.4.1:usr/bin:$PATH ; meme '+ util.wslname(search_regions_loc) + ' -oc '+ util.wslname(output_dir) +' -dna -evt 0.001 -p 8 -revcomp -mod anr -brief 4000 -minw ' + str(min_width) +' -maxw 200 -minsites ' + str(min_sites)\n",
    "               , shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9faaf599-3275-429b-ad53-99f29695e146",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_fimo(motif_file, sequence_to_search_file, output_dir):\n",
    "    subprocess.run('wsl export PATH=$HOME/meme/bin:$HOME/meme/libexec/meme-5.4.1:usr/bin:$PATH ; fimo -oc ' + util.wslname(output_dir) + ' ' + util.wslname(motif_file) + ' ' + util.wslname(sequence_to_search_file)\n",
    "               , shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a21ce3b0-d00a-46fa-a299-95b88e2d0533",
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_matched_sequence_file(hit_df, output_file, q_value):\n",
    "    matched_sequences = []\n",
    "    for i, r in hit_df.iterrows():\n",
    "        if '#' in r['motif_id']:\n",
    "            continue\n",
    "        if float(r['q-value']) < q_value:\n",
    "            matched_sequences.append([str(r['start'])+'_'+str(r['stop']), r['matched_sequence']])\n",
    "    util.produce_fasta_file(matched_sequences, output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcbb740d-f4e8-4842-b6f6-0084271d62e2",
   "metadata": {},
   "source": [
    "#### Import list of significantly up/downregulated genes and produce fasta file of upstream sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2dfe78c9-fd29-4b8c-bc7d-3ce16e4fb4ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "regulation_type_dict = {}\n",
    "p_val_dict = {}\n",
    "downreg_genes = pd.read_excel(project_dir + '/2022-10-02_data_NU_NAs_renamed.xlsx', sheet_name = 'significant genes downregulated')\n",
    "upreg_genes = pd.read_excel(project_dir + '/2022-10-02_data.xlsx', sheet_name = 'significant genes upregulated')\n",
    "all_genes = pd.read_excel(project_dir + '/2022-10-02_data.xlsx', sheet_name = 'all genes')\n",
    "gene_list = []\n",
    "for i, r in downreg_genes.iterrows():\n",
    "    gene_list.append(r['Locus'])\n",
    "    regulation_type_dict[r['Locus']] = 'Downregulated'\n",
    "for i, r in upreg_genes.iterrows():\n",
    "    gene_list.append(r['Locus'])\n",
    "    regulation_type_dict[r['Locus']] = 'Upregulated'\n",
    "for i, r in all_genes.iterrows():\n",
    "    if r['padj'] == 'NA':\n",
    "        p_val_dict[r['Locus']] = 999\n",
    "    else:    \n",
    "        p_val_dict[r['Locus']] = r['padj']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "92f57e96-7a0f-48eb-8d5b-008b3a27e0a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 175/175 [00:00<00:00, 175137.96it/s]\n"
     ]
    }
   ],
   "source": [
    "significant_upstream_tb_regions = []\n",
    "temp_dict = {}\n",
    "for gene in gene_list:\n",
    "    locus = tb_species + '@' + str(gene)\n",
    "    if locus in mycobacteria_upstream_dict:\n",
    "        upstream_region = mycobacteria_upstream_dict[locus][3]  \n",
    "        (start, stop) = (mycobacteria_upstream_dict[locus][0],  mycobacteria_upstream_dict[locus][1])\n",
    "        if (start, stop) in temp_dict:\n",
    "            pass  # Eliminate upstream region if already there (e.g if neighbouring divergent pair)\n",
    "        else:\n",
    "            temp_dict[(start, stop)] = 1      \n",
    "            if len(upstream_region) >= min_region_length:\n",
    "                significant_upstream_tb_regions.append([locus, upstream_region])\n",
    "util.produce_fasta_file(significant_upstream_tb_regions, project_dir + '/significant_upstream_tb_regions.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8399606-8ab1-4292-ae09-29791de85a15",
   "metadata": {},
   "source": [
    "#### Run MEME, search against tb using FIMO and process hits into final MEME Run = Second_MEME_Upstream_Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "7bb56b27-9fff-483b-906e-dd001e482986",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 110/110 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "run_fimo(project_dir + '/MEME_Upstream_Output/meme.txt', project_dir + '/mtb_seq' , project_dir + '/FIMO_Upstream_Output')\n",
    "fimo_upstream_hits = pd.read_csv(project_dir + '/FIMO_Upstream_Output/fimo.tsv', sep='\\t')\n",
    "produce_matched_sequence_file(fimo_upstream_hits, project_dir + '/fimo_upstream_hit_sequences.faa', q_value = 1e-3)\n",
    "run_meme(project_dir + '/fimo_upstream_hit_sequences.faa', project_dir + '/Second_MEME_Upstream_Output',3,5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9004f80f-1ec8-4de5-b96e-9f8e4b6724c2",
   "metadata": {},
   "source": [
    "##### Produce FIMO hits for TB and all mycobacteria and all actinobacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "53c0fe66-0118-427f-b2e4-d74f419e8f1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_fimo(project_dir + '/Second_MEME_Upstream_Output/meme.txt', project_dir + '/mtb_seq' , project_dir + '/Second_FIMO_Upstream_Output')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "4fc9df09-edea-48ac-8508-a91e50e06b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_fimo(project_dir + '/Second_MEME_Upstream_Output/meme.txt', project_dir + '/all_seq' , project_dir + '/All_Species_Second_FIMO_Upstream_Output')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "96da801d-1396-45c4-9909-4e5b295c95c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_fimo(project_dir + '/Second_MEME_Upstream_Output/meme.txt', project_dir + '/all_actinobacteria_seq' , project_dir + '/All_Actinobacteria_Species_Second_FIMO_Upstream_Output')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d2bd314-c7a1-4a66-aa64-fce2906a5c8b",
   "metadata": {},
   "source": [
    "#### Use all species hits as a template for MEME (All_Species_Output) and search these motifs in other species (then refine into motif and search again),  Hits in /Single_Species_MEME_FIMO Final motif in /Temp_Second_Meme_Upstream_Output, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "e95e0f5a-6f98-44e1-bb0a-76066c7b3f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "fimo_upstream_hits = pd.read_csv(project_dir + '/All_Species_Second_FIMO_Upstream_Output/fimo.tsv', sep='\\t')\n",
    "produce_matched_sequence_file(fimo_upstream_hits, project_dir + '/all_species_hit_sequences.faa', q_value = 1e-3)\n",
    "run_meme(project_dir + '/all_species_hit_sequences.faa', project_dir + '/All_Species_Output',3,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed9e744-fb90-4407-ba87-ce7de82df2ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_run == True:\n",
    "    for dirname in tqdm(mycobacteria_dirs):\n",
    "        temp = []\n",
    "        for genome_record in (SeqIO.parse(mycobacteria_seq_dir + '/' + dirname + '/genomic.gbff', \"genbank\")):\n",
    "            ntsequence = str(genome_record.seq)\n",
    "            accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "            temp.append([accession_ver, ntsequence])\n",
    "        util.produce_fasta_file(temp, project_dir + '/temp_seq')\n",
    "        run_fimo(project_dir + '/All_Species_Output/meme.txt', project_dir + '/temp_seq' , project_dir + '/Single_Species_MEME_FIMO/' + dirname)\n",
    "        fimo_upstream_hits = pd.read_csv(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/fimo.tsv', sep='\\t')\n",
    "        produce_matched_sequence_file(fimo_upstream_hits, project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/temp_fimo_upstream_hit_sequences.faa', q_value = 1e-3)\n",
    "        run_meme(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/temp_fimo_upstream_hit_sequences.faa', project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/Temp_Second_MEME_Upstream_Output',3,5)\n",
    "\n",
    "        run_fimo(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/Temp_Second_MEME_Upstream_Output/meme.txt', project_dir + '/temp_seq' , project_dir + '/Single_Species_MEME_FIMO/' + dirname)\n",
    "        fimo_upstream_hits = pd.read_csv(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/fimo.tsv', sep='\\t')\n",
    "        produce_matched_sequence_file(fimo_upstream_hits, project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/temp_fimo_upstream_hit_sequences.faa', q_value = 1e-3)\n",
    "        run_meme(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/temp_fimo_upstream_hit_sequences.faa', project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/Temp_Second_MEME_Upstream_Output',3,5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4733095-8bf8-40cb-8cc0-9ea222a62087",
   "metadata": {},
   "source": [
    "#### Produce dictionary of hits for each species based on ortholog of hits in M.tb "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "648a416d-6518-43e3-b8ae-a86e95165884",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [05:26<00:00,  3.26s/it]\n"
     ]
    }
   ],
   "source": [
    "appearance_dict = {}\n",
    "for dirname in tqdm(mycobacteria_dirs):\n",
    "    temp_set = set()\n",
    "    temp_set_2 = set()\n",
    "    temp_list_3 = []\n",
    "    temp_fimo_hits = pd.read_csv(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/fimo.tsv', sep='\\t')\n",
    "    for i, r in temp_fimo_hits.iterrows():\n",
    "        if '#' in r['motif_id']:\n",
    "            continue\n",
    "        organism = r['sequence_name']\n",
    "        motif_id = r['motif_alt_id']\n",
    "        if float(r['q-value']) < 1e-5:\n",
    "            temp = location(r['sequence_name'],int(r['start']), int(r['stop']), cds_annotated_regions_dict)\n",
    "            ortholog_orientations = []\n",
    "            for x in temp:\n",
    "                if len(x[0].split(':')) == 1:\n",
    "                    ortholog_orientations.append((ortholog_in_mtb(organism, x[0]), 'I'))\n",
    "                else:\n",
    "                    a = x[0].split(':')\n",
    "                    b = x[4].split(':')\n",
    "                    if b[0] == '1':\n",
    "                        ortholog_orientations.append((ortholog_in_mtb(organism, a[0]), 'U'))\n",
    "                    else:\n",
    "                        ortholog_orientations.append((ortholog_in_mtb(organism, a[0]), 'D'))\n",
    "                    if b[1] == '1':\n",
    "                        ortholog_orientations.append((ortholog_in_mtb(organism, a[1]), 'D'))\n",
    "                    else:\n",
    "                        ortholog_orientations.append((ortholog_in_mtb(organism, a[1]), 'U'))\n",
    "            for x in ortholog_orientations:\n",
    "                temp_list_3.append(x)                                          \n",
    "            temp2 = ([x[0].split(':') for x in temp])\n",
    "            for x in temp2:\n",
    "                temp_set = temp_set.union({ortholog_in_mtb(organism, w) for w in set(x)})\n",
    "                temp_set_2 = temp_set_2.union(set(x))\n",
    "    #print(temp_set)\n",
    "    temp_set_3 = set(temp_list_3)\n",
    "    \n",
    "    for ortholog in temp_set_3:\n",
    "        if ortholog in appearance_dict:\n",
    "            appearance_dict[ortholog].append(dirname_dict[dirname])\n",
    "        else:\n",
    "            appearance_dict[ortholog] = [dirname_dict[dirname]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "60e201d6-b3b8-45b8-a92f-98cc7481c6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file = []\n",
    "temp = []\n",
    "for k, v in appearance_dict.items():\n",
    "    temp.append((k[0], k[1], len(v), v))\n",
    "temp.sort(key = lambda x: x[2], reverse = True)\n",
    "for (x, y, l, m) in temp:\n",
    "    if 1==1:\n",
    "        out_file.append([str(x)+'_'+str(y), l])\n",
    "    #print(x, y, l, m)\n",
    "out_df = pd.DataFrame(out_file, columns = ['Orthologue_region', 'Species_hits'])\n",
    "out_df.to_csv(project_dir + '/out_df.csv')\n",
    "#sns.barplot(data = out_df, x = 'Orthologue_region', y = 'Species_hits')\n",
    "#plt.xticks(rotation=45)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "773b053b-70ba-4412-a34b-749612756719",
   "metadata": {},
   "source": [
    "#### Output motifs per species for display in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "14624834-c9cf-4e20-9e54-efb7f6d74ecb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 66/66 [00:00<00:00, 65973.32it/s]\n"
     ]
    }
   ],
   "source": [
    "temp = []\n",
    "from Bio import motifs\n",
    "dirnames = ['GCA_010726505.1']\n",
    "for dirname in mycobacteria_dirs:\n",
    "    filename = project_dir + '/Single_Species_MEME_FIMO/'+dirname +'/Temp_Second_MEME_Upstream_Output/meme.xml'\n",
    "    if os.path.exists(filename):\n",
    "        with open(filename) as f:\n",
    "            record = motifs.parse(f, 'MEME')\n",
    "        for motif in record:\n",
    "            if motif.id == 'motif_1':\n",
    "                temp.append([dirname_dict[dirname].replace(' ','_').replace('[','').replace(']',''), str(motif.consensus)])\n",
    "util.produce_fasta_file(temp, project_dir + '/unaligned_motifs.faa')\n",
    "cline = MuscleCommandline(muscle_exe, input= project_dir + '/unaligned_motifs.faa', out = project_dir + '/aligned_motifs.faa')\n",
    "stdout, stderr = cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "21db49a5-32fd-4c13-b7b0-213f736bdbdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = util.read_fasta_to_array(project_dir + '/aligned_motifs.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a75f94e3-4040-4a20-ba5c-934aef52c2e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "for dirname in mycobacteria_dirs:\n",
    "    species = dirname_dict[dirname].replace(' ','_').replace('[','').replace(']','')\n",
    "    if species in temp[0]:\n",
    "        pass\n",
    "    else:\n",
    "        temp[0].append(species)\n",
    "        temp[1].append('-'*len(temp[1][0]))\n",
    "temp_2 = []\n",
    "for (x, y) in zip(temp[0],temp[1]):\n",
    "    temp_2.append([x,y])\n",
    "util.produce_fasta_file(temp_2, project_dir + '/aligned_motifs_plus_blanks.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a1dadb3-0799-4a79-834b-e430c1715702",
   "metadata": {},
   "source": [
    "#### Final motif and searches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "23ed590f-b6b9-4206-b591-b6ca506b3041",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [07:06<00:00,  4.26s/it]\n"
     ]
    }
   ],
   "source": [
    "#temp_fimo_hits = pd.read_csv(project_dir + '/Second_FIMO_Upstream_Output/fimo.tsv', sep='\\t')\n",
    "hit_dict = {}\n",
    "for dirname in tqdm(mycobacteria_dirs):\n",
    "    temp_fimo_hits = pd.read_csv(project_dir + '/Single_Species_MEME_FIMO/' + dirname + '/fimo.tsv', sep='\\t')\n",
    "\n",
    "    for i, r in temp_fimo_hits.iterrows():\n",
    "        if '#' in r['motif_id']:\n",
    "            continue\n",
    "        organism = r['sequence_name']\n",
    "        motif_id = r['motif_alt_id']\n",
    "        temp = location(r['sequence_name'],int(r['start']), int(r['stop']), annotated_regions_dict)\n",
    "        if len(temp) > 0:\n",
    "            temp.sort(key = lambda x: float(x[3].strip('%')), reverse = True)\n",
    "            main_location = temp[0] #temp[0]\n",
    "        else:\n",
    "            main_location = ''\n",
    "        info = [int(r['start']), int(r['stop']), r['strand'], float(r['q-value']), main_location, r['matched_sequence'], r['motif_alt_id']]\n",
    "        if organism in hit_dict:\n",
    "            hit_dict[organism].append(info)\n",
    "        else:\n",
    "            hit_dict[organism] = [info]\n",
    "\n",
    "for k, v in hit_dict.items():\n",
    "    v.sort(key = lambda x: x[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd64039e-5a79-426c-8f78-9df2595a1ce4",
   "metadata": {},
   "source": [
    "#### Print tb hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d55ccb2-e2e0-46a7-9c83-fd71b6197361",
   "metadata": {},
   "outputs": [],
   "source": [
    "for n, (start, stop, strand, q, loc, matched_seq, motif_id) in enumerate(hit_dict[tb_species]):\n",
    "    if q > 1e-2 or not(motif_id == 'MEME-1'):\n",
    "        continue\n",
    "    print((start, stop, strand, q, loc, matched_seq, motif_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6ad8a68-7826-42ce-b4dd-46426713930d",
   "metadata": {},
   "source": [
    "#### Print tb hits split by whether they are near (the motif + 17 complement motif pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "755dc968-fce2-451a-96a5-903dcfd95d46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "69 23\n",
      "69\n",
      "[17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 98, 17, 17, 17, 17, 19, 17, 17, 17, 17, 17, 17, 16, 17, 17, 17, 17, 17, 18, 17, 17, 17, 17, 17, 17, 18, 17, 17, 17, 14, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17]\n",
      "162\n"
     ]
    }
   ],
   "source": [
    "paired_motif_sequences = []\n",
    "unpaired_motif_sequences = []\n",
    "temp_df_list =[]\n",
    "gene_pairs = []\n",
    "sole_genes = []\n",
    "inter_distances = []\n",
    "occurrences = 0\n",
    "pairs = 0\n",
    "no_pairs = 0\n",
    "v = hit_dict[tb_species]\n",
    "last_stop = -999\n",
    "pair_poss = 0\n",
    "for n, (start, stop, strand, q, loc, matched_seq, motif_id) in enumerate(v):\n",
    "    if q > 1e-2 or not(motif_id == 'MEME-1'):\n",
    "        continue\n",
    "    else:\n",
    "        occurrences +=1\n",
    "        if start - last_stop < 100 and pair_poss == 1:\n",
    "            inter_distances.append(start-last_stop)\n",
    "            last_temp = copy.deepcopy(v[last_n])\n",
    "            last_temp.append(\"paired\")\n",
    "            latest_temp = copy.deepcopy(v[n])\n",
    "            latest_temp.append(\"paired\")\n",
    "            temp_df_list.append(last_temp)\n",
    "            temp_df_list.append(latest_temp)\n",
    "            if start - last_stop < 20:\n",
    "                genes = v[n][4][0].split(\":\")\n",
    "                genes2 = v[last_n][4][0].split(\":\")\n",
    "                pref = 'H37Rv_'\n",
    "                for gene in genes:\n",
    "                    if gene in set(gene_list):\n",
    "                        pref = v[n][4][0] + '___H37Rv_'\n",
    "                for gene in genes2:\n",
    "                    if gene in set(gene_list):\n",
    "                        pref = v[last_n][4][0] + '___H37Rv_'\n",
    "                paired_motif_sequences.append([pref+str(v[last_n][0])+'_'+str(v[n][1]), mtb_sequence[v[last_n][0]:v[n][1]]])\n",
    "            pair_poss = 0\n",
    "            pairs+=1\n",
    "            for x in v[n][4][0].split(':'):\n",
    "                gene_pairs.append(x)\n",
    "            for x in v[last_n][4][0].split(':'):\n",
    "                gene_pairs.append(x)\n",
    "\n",
    "        elif pair_poss == 1:\n",
    "            last_temp = copy.deepcopy(v[last_n])\n",
    "            last_temp.append(\"Unpaired\")\n",
    "            temp_df_list.append(last_temp)\n",
    "            #print(\"no_pair\", v[last_n])\n",
    "            no_pairs+=1#\n",
    "            for x in v[last_n][4][0].split(':'):\n",
    "                sole_genes.append(x)\n",
    "            genes2 = v[last_n][4][0].split(\":\")\n",
    "            pref = 'NP_H37Rv_'\n",
    "            for gene in genes2:\n",
    "                if gene in set(gene_list):\n",
    "                    pref = v[last_n][4][0] + '___NP_H37Rv_'\n",
    "            if v[last_n][2] == '+':\n",
    "                unpaired_motif_sequences.append([pref+str(v[last_n][0]-30)+'_'+str(v[last_n][1]+40), mtb_sequence[v[last_n][0]-30:v[last_n][1]+40]])\n",
    "            else:\n",
    "                unpaired_motif_sequences.append([pref+str(v[last_n][0]-30)+'_'+str(v[last_n][1]+40), util.reverse_complement(mtb_sequence[v[last_n][0]-30:v[last_n][1]+40])])\n",
    "        else:\n",
    "            pair_poss=1\n",
    "        last_stop = stop\n",
    "        last_n = n\n",
    "print(pairs, no_pairs)\n",
    "print(len(inter_distances))\n",
    "print(inter_distances)\n",
    "print(occurrences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "ba512279-67ee-464e-8019-e7ce077bfad1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 68/68 [00:00<?, ?it/s]\n",
      "100%|██████████| 23/23 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "util.produce_fasta_file(paired_motif_sequences, project_dir + '/unaligned_pair_motifs.faa')\n",
    "cline = MuscleCommandline(muscle_exe, input= project_dir + '/unaligned_pair_motifs.faa', out = project_dir + '/aligned_pair_motifs.faa')\n",
    "stdout, stderr = cline()\n",
    "util.produce_fasta_file(unpaired_motif_sequences, project_dir + '/unaligned_unpaired_motifs.faa')\n",
    "cline = MuscleCommandline(muscle_exe, input= project_dir + '/unaligned_unpaired_motifs.faa', out = project_dir + '/aligned_unpaired_motifs.faa')\n",
    "stdout, stderr = cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "aff4ea7c-564f-439c-98bc-f4df2ae6be55",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastfn.convert_fasta_to_stockholm(project_dir, 'aligned_pair_motifs.faa', 'aligned_pair_motifs.sto')\n",
    "blastfn.run_rscape(project_dir, 'aligned_pair_motifs.sto', 'rscape_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "3558e4c3-210d-4969-86e7-fe0d212ce010",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp= pd.DataFrame(temp_df_list, columns = ['start','stop','strand','q_val','region_info','matched_seq','motif','pair'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "e2007531-6b3d-4403-81c6-38235f7f8456",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp['study_association'] = ''\n",
    "for i, r in temp.iterrows():\n",
    "    temp.at[i, 'type'] = r['region_info'][2]\n",
    "    temp.at[i, 'product'] = r['region_info'][1]\n",
    "    temp.at[i,'pct_occupancy'] = r['region_info'][3]\n",
    "    genes = r['region_info'][0].split(\":\")\n",
    "    associated_genes = []\n",
    "    for gene in genes:\n",
    "        if gene in set(gene_list):\n",
    "            associated_genes.append(gene)\n",
    "    temp.at[i, 'study_association'] = associated_genes\n",
    "temp.to_csv(project_dir + '/motif_hits.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "23059b03-8146-4299-a3a9-38e924bca854",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>q_val</th>\n",
       "      <th>region_info</th>\n",
       "      <th>matched_seq</th>\n",
       "      <th>motif</th>\n",
       "      <th>pair</th>\n",
       "      <th>study_association</th>\n",
       "      <th>type</th>\n",
       "      <th>product</th>\n",
       "      <th>pct_occupancy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>93380</td>\n",
       "      <td>93401</td>\n",
       "      <td>+</td>\n",
       "      <td>3.650000e-03</td>\n",
       "      <td>[Rv0085, Possible hydrogenase HycP, CDS, 100%, 1]</td>\n",
       "      <td>GGGCCATTGTGCGGCTGCTGGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>Unpaired</td>\n",
       "      <td>[]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Possible hydrogenase HycP</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>183232</td>\n",
       "      <td>183253</td>\n",
       "      <td>-</td>\n",
       "      <td>1.030000e-03</td>\n",
       "      <td>[Rv0154c:Rv0155, Probable acyl-CoA dehydrogena...</td>\n",
       "      <td>AGGGACTTATGCGTCTGCTAGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>paired</td>\n",
       "      <td>[Rv0154c]</td>\n",
       "      <td>Inter-feature</td>\n",
       "      <td>Probable acyl-CoA dehydrogenase FadE2:Probable...</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183270</td>\n",
       "      <td>183291</td>\n",
       "      <td>+</td>\n",
       "      <td>9.690000e-07</td>\n",
       "      <td>[Rv0154c:Rv0155, Probable acyl-CoA dehydrogena...</td>\n",
       "      <td>AGGGGGTTTTGCGTCTGCTCGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>paired</td>\n",
       "      <td>[Rv0154c]</td>\n",
       "      <td>Inter-feature</td>\n",
       "      <td>Probable acyl-CoA dehydrogenase FadE2:Probable...</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>231365</td>\n",
       "      <td>231386</td>\n",
       "      <td>+</td>\n",
       "      <td>5.810000e-03</td>\n",
       "      <td>[Rv0195, Possible two component transcriptiona...</td>\n",
       "      <td>TGGCGGTGCTGCGTCTGCTAGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>Unpaired</td>\n",
       "      <td>[]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Possible two component transcriptional regulat...</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>234447</td>\n",
       "      <td>234468</td>\n",
       "      <td>-</td>\n",
       "      <td>7.910000e-08</td>\n",
       "      <td>[Rv0197, Possible oxidoreductase, CDS, 100%, 1]</td>\n",
       "      <td>GGGCGATTCTGCGTCTGCTCGG</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>Unpaired</td>\n",
       "      <td>[]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Possible oxidoreductase</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>4237935</td>\n",
       "      <td>4237956</td>\n",
       "      <td>-</td>\n",
       "      <td>8.640000e-08</td>\n",
       "      <td>[Rv3792, Arabinofuranosyltransferase AftA, CDS...</td>\n",
       "      <td>GGGGGCTTTTGCGTCTGCTCGG</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>paired</td>\n",
       "      <td>[Rv3792]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Arabinofuranosyltransferase AftA</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>4237973</td>\n",
       "      <td>4237994</td>\n",
       "      <td>+</td>\n",
       "      <td>2.280000e-07</td>\n",
       "      <td>[Rv3792, Arabinofuranosyltransferase AftA, CDS...</td>\n",
       "      <td>GGGGGCTTTTACGTCTGCTCGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>paired</td>\n",
       "      <td>[Rv3792]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Arabinofuranosyltransferase AftA</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>4246526</td>\n",
       "      <td>4246547</td>\n",
       "      <td>-</td>\n",
       "      <td>1.110000e-07</td>\n",
       "      <td>[Rv3795, Integral membrane indolylacetylinosit...</td>\n",
       "      <td>GGGTGCTTTTGCGTCTGCTCGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>paired</td>\n",
       "      <td>[Rv3795]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Integral membrane indolylacetylinositol arabin...</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>4246564</td>\n",
       "      <td>4246585</td>\n",
       "      <td>+</td>\n",
       "      <td>3.930000e-08</td>\n",
       "      <td>[Rv3795, Integral membrane indolylacetylinosit...</td>\n",
       "      <td>GGGGGCTTTTGCGTCTGCTCGC</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>paired</td>\n",
       "      <td>[Rv3795]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Integral membrane indolylacetylinositol arabin...</td>\n",
       "      <td>100%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>4252853</td>\n",
       "      <td>4252874</td>\n",
       "      <td>-</td>\n",
       "      <td>5.600000e-07</td>\n",
       "      <td>[Rv3797, Probable acyl-CoA dehydrogenase FadE3...</td>\n",
       "      <td>GGGGGCTTTTACGTCTGCTCGG</td>\n",
       "      <td>MEME-1</td>\n",
       "      <td>Unpaired</td>\n",
       "      <td>[]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Probable acyl-CoA dehydrogenase FadE35</td>\n",
       "      <td>61%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       start     stop strand         q_val  \\\n",
       "0      93380    93401      +  3.650000e-03   \n",
       "1     183232   183253      -  1.030000e-03   \n",
       "2     183270   183291      +  9.690000e-07   \n",
       "3     231365   231386      +  5.810000e-03   \n",
       "4     234447   234468      -  7.910000e-08   \n",
       "..       ...      ...    ...           ...   \n",
       "156  4237935  4237956      -  8.640000e-08   \n",
       "157  4237973  4237994      +  2.280000e-07   \n",
       "158  4246526  4246547      -  1.110000e-07   \n",
       "159  4246564  4246585      +  3.930000e-08   \n",
       "160  4252853  4252874      -  5.600000e-07   \n",
       "\n",
       "                                           region_info  \\\n",
       "0    [Rv0085, Possible hydrogenase HycP, CDS, 100%, 1]   \n",
       "1    [Rv0154c:Rv0155, Probable acyl-CoA dehydrogena...   \n",
       "2    [Rv0154c:Rv0155, Probable acyl-CoA dehydrogena...   \n",
       "3    [Rv0195, Possible two component transcriptiona...   \n",
       "4      [Rv0197, Possible oxidoreductase, CDS, 100%, 1]   \n",
       "..                                                 ...   \n",
       "156  [Rv3792, Arabinofuranosyltransferase AftA, CDS...   \n",
       "157  [Rv3792, Arabinofuranosyltransferase AftA, CDS...   \n",
       "158  [Rv3795, Integral membrane indolylacetylinosit...   \n",
       "159  [Rv3795, Integral membrane indolylacetylinosit...   \n",
       "160  [Rv3797, Probable acyl-CoA dehydrogenase FadE3...   \n",
       "\n",
       "                matched_seq   motif      pair study_association  \\\n",
       "0    GGGCCATTGTGCGGCTGCTGGC  MEME-1  Unpaired                []   \n",
       "1    AGGGACTTATGCGTCTGCTAGC  MEME-1    paired         [Rv0154c]   \n",
       "2    AGGGGGTTTTGCGTCTGCTCGC  MEME-1    paired         [Rv0154c]   \n",
       "3    TGGCGGTGCTGCGTCTGCTAGC  MEME-1  Unpaired                []   \n",
       "4    GGGCGATTCTGCGTCTGCTCGG  MEME-1  Unpaired                []   \n",
       "..                      ...     ...       ...               ...   \n",
       "156  GGGGGCTTTTGCGTCTGCTCGG  MEME-1    paired          [Rv3792]   \n",
       "157  GGGGGCTTTTACGTCTGCTCGC  MEME-1    paired          [Rv3792]   \n",
       "158  GGGTGCTTTTGCGTCTGCTCGC  MEME-1    paired          [Rv3795]   \n",
       "159  GGGGGCTTTTGCGTCTGCTCGC  MEME-1    paired          [Rv3795]   \n",
       "160  GGGGGCTTTTACGTCTGCTCGG  MEME-1  Unpaired                []   \n",
       "\n",
       "              type                                            product  \\\n",
       "0              CDS                          Possible hydrogenase HycP   \n",
       "1    Inter-feature  Probable acyl-CoA dehydrogenase FadE2:Probable...   \n",
       "2    Inter-feature  Probable acyl-CoA dehydrogenase FadE2:Probable...   \n",
       "3              CDS  Possible two component transcriptional regulat...   \n",
       "4              CDS                            Possible oxidoreductase   \n",
       "..             ...                                                ...   \n",
       "156            CDS                   Arabinofuranosyltransferase AftA   \n",
       "157            CDS                   Arabinofuranosyltransferase AftA   \n",
       "158            CDS  Integral membrane indolylacetylinositol arabin...   \n",
       "159            CDS  Integral membrane indolylacetylinositol arabin...   \n",
       "160            CDS             Probable acyl-CoA dehydrogenase FadE35   \n",
       "\n",
       "    pct_occupancy  \n",
       "0            100%  \n",
       "1            100%  \n",
       "2            100%  \n",
       "3            100%  \n",
       "4            100%  \n",
       "..            ...  \n",
       "156          100%  \n",
       "157          100%  \n",
       "158          100%  \n",
       "159          100%  \n",
       "160           61%  \n",
       "\n",
       "[161 rows x 12 columns]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d4207bdb-6ab8-4f3a-b720-fdba8c152c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = []\n",
    "v = hit_dict[tb_species]\n",
    "for (start, stop, strand, q, loc, matched_seq, motif_id) in v:\n",
    "    if q > 1e-2 or not(motif_id == 'MEME-1'):\n",
    "        continue\n",
    "    for x in loc[0].split(':'):\n",
    "        genes.append(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "0a39a7b5-e5a6-47e8-879a-86c647abe861",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "163 20 236\n",
      "0.0005947770818272025\n"
     ]
    }
   ],
   "source": [
    "b = set(genes)\n",
    "inters = b.intersection(set(gene_list))\n",
    "print(len(b), len(inters), len(gene_list))\n",
    "print(1- binom.cdf(len(inters), len(b),(len(gene_list))/4031))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "ba96407e-e557-4a3d-b678-4762a27b18c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Rv0154c',\n",
       " 'Rv0211',\n",
       " 'Rv0467',\n",
       " 'Rv0468',\n",
       " 'Rv0711',\n",
       " 'Rv0860',\n",
       " 'Rv0896',\n",
       " 'Rv0926c',\n",
       " 'Rv0927c',\n",
       " 'Rv1058',\n",
       " 'Rv1180',\n",
       " 'Rv2845c',\n",
       " 'Rv2846c',\n",
       " 'Rv3028c',\n",
       " 'Rv3140',\n",
       " 'Rv3319',\n",
       " 'Rv3383c',\n",
       " 'Rv3416',\n",
       " 'Rv3792',\n",
       " 'Rv3795']"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = list(b.intersection(set(gene_list)))\n",
    "a.sort()\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b945319-9ec1-4afd-94ff-0183bd680502",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in hit_dict.items():\n",
    "    for x in v:\n",
    "        print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de7bd340-a386-4878-87d0-0a6ac14e2aea",
   "metadata": {},
   "source": [
    "#### Number of motif hits in different species and output so can be merged with R to display info against phylo tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "507f1c29-7df7-4422-a35d-33eb7022e6e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "hit_info = []\n",
    "for k, v in hit_dict.items():\n",
    "    temp = []\n",
    "    for (start, stop, strand, q, loc, matched_seq) in v:\n",
    "        for x in loc[0].split(':'):\n",
    "            temp.append(x)\n",
    "    hit_info.append([names_dict[k].replace(' ','_'),'Num_hits', len(set(temp))])\n",
    "pd.DataFrame(hit_info, columns = ['name', 'variable', 'value']).to_csv(project_dir + '/Hit_Counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ff15ebbf-b15e-4791-8a9a-d1e2962f2ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "tb_mycobacteria_rbh = pd.read_csv(project_dir + '/tb_mycobacteria_reciprocal_best_hits.csv')\n",
    "ortholog_info = []\n",
    "Rv_2242_orthologs = []\n",
    "temp = tb_species+\"@\"+\"Rv2242\"\n",
    "temp = tb_mycobacteria_rbh.query('query_ref == @temp')\n",
    "for i, r in temp.iterrows():\n",
    "    Rv_2242_orthologs.append(r['target_species_name'].replace(' ','_'))\n",
    "for dirname in mycobacteria_dirs:\n",
    "    species = dirname_dict[dirname].replace(' ','_').replace('[','').replace(']','')\n",
    "    if species in Rv_2242_orthologs:\n",
    "        ortholog_info.append([species, 'Ortholog', 1])\n",
    "    else:\n",
    "        ortholog_info.append([species, 'Ortholog', 0])\n",
    "pd.DataFrame(ortholog_info, columns = ['name', 'variable', 'value']).to_csv(project_dir + '/Rv_2242_Hits.csv')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa6c316c-4077-4dc7-8e85-eba63af93f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for dirname in mycobacteria_dirs:\n",
    "    species = dirname_dict[dirname].replace(' ','_').replace('[','').replace(']','')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b6d4de2c-2826-411d-8724-69c2f474ad5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>query_ref</th>\n",
       "      <th>target_ref</th>\n",
       "      <th>query_length</th>\n",
       "      <th>subject_length</th>\n",
       "      <th>percent_identical_matches</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>number_mismatches</th>\n",
       "      <th>number_of_gap_openings</th>\n",
       "      <th>query_start_alignment</th>\n",
       "      <th>...</th>\n",
       "      <th>target_start_alignment</th>\n",
       "      <th>target_end_alignment</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>query_species</th>\n",
       "      <th>target_species</th>\n",
       "      <th>query_species_name</th>\n",
       "      <th>target_species_name</th>\n",
       "      <th>species_count</th>\n",
       "      <th>reciprocal_best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146346</th>\n",
       "      <td>452995</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>CP097320.1@M5I08_12060</td>\n",
       "      <td>414</td>\n",
       "      <td>437</td>\n",
       "      <td>83.902</td>\n",
       "      <td>410</td>\n",
       "      <td>55</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>399</td>\n",
       "      <td>0.0</td>\n",
       "      <td>683.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>CP097320.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Candidatus Mycobacterium methanotrophicum</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146347</th>\n",
       "      <td>452966</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>LR130759.1@MB901379_03007</td>\n",
       "      <td>414</td>\n",
       "      <td>436</td>\n",
       "      <td>92.067</td>\n",
       "      <td>416</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>413</td>\n",
       "      <td>0.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>LR130759.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycobacterium basiliense</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146348</th>\n",
       "      <td>452992</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>AP022606.1@MBRA_35650</td>\n",
       "      <td>414</td>\n",
       "      <td>435</td>\n",
       "      <td>84.300</td>\n",
       "      <td>414</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>406</td>\n",
       "      <td>0.0</td>\n",
       "      <td>700.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>AP022606.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycobacterium branderi</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146349</th>\n",
       "      <td>452960</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>HE572590.1@MCAN_22641</td>\n",
       "      <td>414</td>\n",
       "      <td>414</td>\n",
       "      <td>100.000</td>\n",
       "      <td>414</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>414</td>\n",
       "      <td>0.0</td>\n",
       "      <td>832.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>HE572590.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycobacterium canettii CIPT 140010059</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146350</th>\n",
       "      <td>452981</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>AP022613.1@MCNS_32520</td>\n",
       "      <td>414</td>\n",
       "      <td>446</td>\n",
       "      <td>88.889</td>\n",
       "      <td>414</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>413</td>\n",
       "      <td>0.0</td>\n",
       "      <td>715.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>AP022613.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycobacterium conspicuum</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146440</th>\n",
       "      <td>453042</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>LT906483.1@SAMEA4412656_01789</td>\n",
       "      <td>414</td>\n",
       "      <td>428</td>\n",
       "      <td>75.550</td>\n",
       "      <td>409</td>\n",
       "      <td>89</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>584.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>LT906483.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycolicibacterium thermoresistibile</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146441</th>\n",
       "      <td>453006</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>AP022600.1@MTOK_56240</td>\n",
       "      <td>414</td>\n",
       "      <td>429</td>\n",
       "      <td>80.549</td>\n",
       "      <td>401</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>619.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>AP022600.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycolicibacterium tokaiense</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146442</th>\n",
       "      <td>453051</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>CP011491.1@MYVA_3575</td>\n",
       "      <td>414</td>\n",
       "      <td>427</td>\n",
       "      <td>74.572</td>\n",
       "      <td>409</td>\n",
       "      <td>93</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>399</td>\n",
       "      <td>0.0</td>\n",
       "      <td>561.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>CP011491.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycolicibacterium vaccae 95051</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146443</th>\n",
       "      <td>453029</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>CP000511.1@Mvan_3753</td>\n",
       "      <td>414</td>\n",
       "      <td>428</td>\n",
       "      <td>75.245</td>\n",
       "      <td>408</td>\n",
       "      <td>90</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>590.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>CP000511.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>Mycolicibacterium vanbaalenii PYR-1</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146444</th>\n",
       "      <td>453054</td>\n",
       "      <td>AL123456.3@Rv2242</td>\n",
       "      <td>AP018165.1@MSTE_01798</td>\n",
       "      <td>414</td>\n",
       "      <td>426</td>\n",
       "      <td>70.202</td>\n",
       "      <td>396</td>\n",
       "      <td>109</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>402</td>\n",
       "      <td>0.0</td>\n",
       "      <td>555.0</td>\n",
       "      <td>AL123456.3</td>\n",
       "      <td>AP018165.1</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>[Mycobacterium] stephanolepidis</td>\n",
       "      <td>101</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0          query_ref                     target_ref  \\\n",
       "146346      452995  AL123456.3@Rv2242         CP097320.1@M5I08_12060   \n",
       "146347      452966  AL123456.3@Rv2242      LR130759.1@MB901379_03007   \n",
       "146348      452992  AL123456.3@Rv2242          AP022606.1@MBRA_35650   \n",
       "146349      452960  AL123456.3@Rv2242          HE572590.1@MCAN_22641   \n",
       "146350      452981  AL123456.3@Rv2242          AP022613.1@MCNS_32520   \n",
       "...            ...                ...                            ...   \n",
       "146440      453042  AL123456.3@Rv2242  LT906483.1@SAMEA4412656_01789   \n",
       "146441      453006  AL123456.3@Rv2242          AP022600.1@MTOK_56240   \n",
       "146442      453051  AL123456.3@Rv2242           CP011491.1@MYVA_3575   \n",
       "146443      453029  AL123456.3@Rv2242           CP000511.1@Mvan_3753   \n",
       "146444      453054  AL123456.3@Rv2242          AP018165.1@MSTE_01798   \n",
       "\n",
       "        query_length  subject_length  percent_identical_matches  \\\n",
       "146346           414             437                     83.902   \n",
       "146347           414             436                     92.067   \n",
       "146348           414             435                     84.300   \n",
       "146349           414             414                    100.000   \n",
       "146350           414             446                     88.889   \n",
       "...              ...             ...                        ...   \n",
       "146440           414             428                     75.550   \n",
       "146441           414             429                     80.549   \n",
       "146442           414             427                     74.572   \n",
       "146443           414             428                     75.245   \n",
       "146444           414             426                     70.202   \n",
       "\n",
       "        alignment_length  number_mismatches  number_of_gap_openings  \\\n",
       "146346               410                 55                       2   \n",
       "146347               416                 27                       2   \n",
       "146348               414                 57                       1   \n",
       "146349               414                  0                       0   \n",
       "146350               414                 45                       1   \n",
       "...                  ...                ...                     ...   \n",
       "146440               409                 89                       2   \n",
       "146441               401                 70                       1   \n",
       "146442               409                 93                       2   \n",
       "146443               408                 90                       2   \n",
       "146444               396                109                       2   \n",
       "\n",
       "        query_start_alignment  ...  target_start_alignment  \\\n",
       "146346                      1  ...                       1   \n",
       "146347                      1  ...                       1   \n",
       "146348                      1  ...                       1   \n",
       "146349                      1  ...                       1   \n",
       "146350                      1  ...                       1   \n",
       "...                       ...  ...                     ...   \n",
       "146440                      1  ...                       1   \n",
       "146441                      8  ...                       6   \n",
       "146442                      1  ...                       2   \n",
       "146443                      1  ...                       1   \n",
       "146444                     13  ...                      16   \n",
       "\n",
       "        target_end_alignment  e_value  bit_score  query_species  \\\n",
       "146346                   399      0.0      683.0     AL123456.3   \n",
       "146347                   413      0.0      736.0     AL123456.3   \n",
       "146348                   406      0.0      700.0     AL123456.3   \n",
       "146349                   414      0.0      832.0     AL123456.3   \n",
       "146350                   413      0.0      715.0     AL123456.3   \n",
       "...                      ...      ...        ...            ...   \n",
       "146440                   398      0.0      584.0     AL123456.3   \n",
       "146441                   398      0.0      619.0     AL123456.3   \n",
       "146442                   399      0.0      561.0     AL123456.3   \n",
       "146443                   397      0.0      590.0     AL123456.3   \n",
       "146444                   402      0.0      555.0     AL123456.3   \n",
       "\n",
       "       target_species                query_species_name  \\\n",
       "146346     CP097320.1  Mycobacterium tuberculosis H37Rv   \n",
       "146347     LR130759.1  Mycobacterium tuberculosis H37Rv   \n",
       "146348     AP022606.1  Mycobacterium tuberculosis H37Rv   \n",
       "146349     HE572590.1  Mycobacterium tuberculosis H37Rv   \n",
       "146350     AP022613.1  Mycobacterium tuberculosis H37Rv   \n",
       "...               ...                               ...   \n",
       "146440     LT906483.1  Mycobacterium tuberculosis H37Rv   \n",
       "146441     AP022600.1  Mycobacterium tuberculosis H37Rv   \n",
       "146442     CP011491.1  Mycobacterium tuberculosis H37Rv   \n",
       "146443     CP000511.1  Mycobacterium tuberculosis H37Rv   \n",
       "146444     AP018165.1  Mycobacterium tuberculosis H37Rv   \n",
       "\n",
       "                              target_species_name species_count  \\\n",
       "146346  Candidatus Mycobacterium methanotrophicum           101   \n",
       "146347                   Mycobacterium basiliense           101   \n",
       "146348                     Mycobacterium branderi           101   \n",
       "146349      Mycobacterium canettii CIPT 140010059           101   \n",
       "146350                   Mycobacterium conspicuum           101   \n",
       "...                                           ...           ...   \n",
       "146440        Mycolicibacterium thermoresistibile           101   \n",
       "146441                Mycolicibacterium tokaiense           101   \n",
       "146442             Mycolicibacterium vaccae 95051           101   \n",
       "146443        Mycolicibacterium vanbaalenii PYR-1           101   \n",
       "146444            [Mycobacterium] stephanolepidis           101   \n",
       "\n",
       "        reciprocal_best_hit  \n",
       "146346                    Y  \n",
       "146347                    Y  \n",
       "146348                    Y  \n",
       "146349                    Y  \n",
       "146350                    Y  \n",
       "...                     ...  \n",
       "146440                    Y  \n",
       "146441                    Y  \n",
       "146442                    Y  \n",
       "146443                    Y  \n",
       "146444                    Y  \n",
       "\n",
       "[99 rows x 21 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = tb_species+\"@\"+\"Rv2242\"\n",
    "temp = tb_mycobacteria_rbh.query('query_ref == @temp')\n",
    "temp2 = tb_mycobacteria_rbh\n",
    "set((list(temp['target_species_name'].unique()))).symmetric_difference(set((list(temp2['target_species_name'].unique()))))\n",
    "temp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7876e157-c075-4044-b5eb-18e182da0f55",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Build phylo tree "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e913452-4203-42f1-9a0c-eb6f50c13353",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_len = len(tb_mycobacteria_rbh['target_species_name'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ae63ded-7e41-4437-9069-1fc771ab8a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_ortholog_refs = []\n",
    "temp = tb_mycobacteria_rbh.groupby('query_ref').agg({'target_ref': \"count\", 'percent_identical_matches': \"min\"}).reset_index()\n",
    "temp = temp.query('target_ref == @max_len and percent_identical_matches > 85')\n",
    "for i, r in temp.iterrows():\n",
    "    full_ortholog_refs.append(r['query_ref'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a2f29d1-311f-41f8-aec4-10d742a249e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_protein_dataset(num_subsets, subset_num, dir_list, seqdir): \n",
    "    sequence_dirs = util.chunk_list(dir_list, num_subsets, subset_num)\n",
    "    all_cds = []\n",
    "    for dirname in (sequence_dirs):\n",
    "        for genome_record in SeqIO.parse(seqdir + '/' + dirname + '/genomic.gbff', \"genbank\"):\n",
    "            accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "            full_sequence = str(genome_record.seq)\n",
    "            for feature in genome_record.features:\n",
    "                a = feature.qualifiers\n",
    "                if feature.type == 'CDS' and a.get(\"translation\") != None and a.get(\"locus_tag\") != None:\n",
    "                    locus_tag = a.get(\"locus_tag\")[0]\n",
    "                    accession_locus = accession_ver + '@' + locus_tag\n",
    "                    translation = a.get(\"translation\")[0]\n",
    "                    all_cds.append([accession_locus, translation])\n",
    "    return (all_cds)           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b870157-c0aa-499e-a1e9-2fed6b97592a",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_build = True\n",
    "if full_build == True:\n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(generate_protein_dataset)(num_cores, core_number, mycobacteria_dirs, mycobacteria_seq_dir) for core_number in core_numbers)\n",
    "    protein_dict = {}\n",
    "    for x in parallel_output:\n",
    "        for temp in x[0]:\n",
    "            protein_dict[temp[0]] = temp[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ef5d679-1c5b-4600-886a-a393f4e1cb98",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_alignment_dict = {}\n",
    "for ref in tqdm(full_ortholog_refs):\n",
    "    temp_seq = []\n",
    "    temp = tb_mycobacteria_rbh[tb_mycobacteria_rbh['query_ref'] == ref]\n",
    "    for i, r in temp.iterrows():\n",
    "        temp_seq.append([r['target_species_name'].replace(' ','_'),protein_dict[r['target_ref']]])\n",
    "    util.produce_fasta_file(temp_seq, project_dir +'/temp_seq.fasta')    \n",
    "    cline = MuscleCommandline(muscle_exe, input= project_dir +'/temp_seq.fasta', out=project_dir +'/temp_seq_alignment.fasta')\n",
    "    result = cline();    \n",
    "    alignment = util.read_fasta_to_array(project_dir +'/temp_seq_alignment.fasta')    \n",
    "    for (name, sequence) in zip(alignment[0], alignment[1]):\n",
    "        if name in concatenated_alignment_dict:\n",
    "            temp2 = concatenated_alignment_dict[name]\n",
    "            concatenated_alignment_dict[name] = temp2 + sequence\n",
    "        else:\n",
    "            concatenated_alignment_dict[name] = sequence\n",
    "temp  = []\n",
    "for k, v in concatenated_alignment_dict.items():\n",
    "    temp.append([k, v])\n",
    "util.produce_fasta_file(temp, project_dir + '/concatenated_alignment.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29f1cedc-4c63-4c6a-ae79-69fa70649da6",
   "metadata": {},
   "source": [
    "####  Actinobacteria hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b43b90b9-c31e-4810-8e94-68ab20b31dd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 645/645 [05:23<00:00,  1.99it/s]\n"
     ]
    }
   ],
   "source": [
    "actinobacteria_names_dict = {}\n",
    "for dirname in tqdm(actinobacteria_dirs):\n",
    "    for genome_record in (SeqIO.parse(actinobacteria_seq_dir + '/' + dirname + '/genomic.gbff', \"genbank\")):\n",
    "        accession_ver = genome_record.annotations['accessions'][0] + '.' + str(genome_record.annotations['sequence_version'])\n",
    "        actinobacteria_names_dict[accession_ver] = genome_record.annotations['organism']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "56489b05-9d48-4590-9969-ed638cc1abab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corynebacterium singulare MEME-1 0.000512\n",
      "Gordonia iterans MEME-1 0.000793\n",
      "Gordonia amicalis MEME-1 0.000835\n",
      "Salinispora tropica CNB-440 MEME-1 0.000835\n",
      "Streptomyces tirandamycinicus MEME-1 0.000869\n",
      "Corynebacterium epidermidicanis MEME-1 0.000873\n",
      "Rathayibacter rathayi MEME-1 0.00094\n",
      "Rhodococcus opacus PD630 MEME-1 0.000946\n",
      "Streptomyces genisteinicus MEME-1 0.000993\n"
     ]
    }
   ],
   "source": [
    "hit_dict = {}\n",
    "temp_fimo_hits = pd.read_csv(project_dir + '/All_Actinobacteria_Species_Second_FIMO_Upstream_Output/fimo.tsv', sep='\\t')\n",
    "for i, r in temp_fimo_hits.iterrows():\n",
    "    if '#' in r['motif_id']:\n",
    "        continue\n",
    "    organism = actinobacteria_names_dict[r['sequence_name']]\n",
    "    motif_id = r['motif_alt_id']\n",
    "    q_val = float(r['q-value'])\n",
    "    if q_val < 1e-3 and motif_id == 'MEME-1' and not('Myco' in organism):\n",
    "        print(organism, motif_id, q_val)\n",
    "  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f6b7715f-78c6-40f1-ae2c-b8f715c1219a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'LR134406.1': 'Arachnia propionica',\n",
       " 'LR134442.1': 'Propionibacterium australiense',\n",
       " 'LR134476.1': 'Trueperella bialowiezensis',\n",
       " 'LR134477.1': 'Actinomyces viscosus',\n",
       " 'LR134479.1': 'Rothia aeria',\n",
       " 'LR134501.1': 'Nocardiopsis dassonvillei',\n",
       " 'LS483427.1': 'Arcanobacterium haemolyticum',\n",
       " 'LS483464.1': 'Corynebacterium renale',\n",
       " 'LS483460.1': 'Corynebacterium minutissimum',\n",
       " 'LS483468.1': 'Rhodococcus coprophilus',\n",
       " 'LR026975.1': 'Mycolicibacterium hassiacum DSM 44199',\n",
       " 'LR134350.1': 'Actinomyces howellii',\n",
       " 'LR134352.1': 'Nocardia asteroides',\n",
       " 'LR134356.1': 'Mycolicibacterium aurum',\n",
       " 'LT906441.1': 'Cutibacterium granulosum',\n",
       " 'LT906453.1': 'Dermatophilus congolensis',\n",
       " 'LT906483.1': 'Mycolicibacterium thermoresistibile',\n",
       " 'LT906469.1': 'Mycolicibacter terrae',\n",
       " 'LT906467.1': 'Corynebacterium imitans',\n",
       " 'LT906481.1': 'Corynebacterium urealyticum',\n",
       " 'LT906473.1': 'Corynebacterium cystitidis',\n",
       " 'LT985188.1': 'Micropruina glycogenica',\n",
       " 'LR130759.1': 'Mycobacterium basiliense',\n",
       " 'LS483423.1': 'Jonesia denitrificans',\n",
       " 'CP116355.1': 'Streptomyces xanthophaeus',\n",
       " 'CP116356.1': 'Streptomyces xanthophaeus',\n",
       " 'CP116357.1': 'Streptomyces xanthophaeus',\n",
       " 'CP063191.1': 'Corynebacterium heidelbergense',\n",
       " 'CP063192.1': 'Corynebacterium heidelbergense',\n",
       " 'CP063194.1': 'Corynebacterium jeddahense',\n",
       " 'CP117869.1': 'Arthrobacter agilis',\n",
       " 'CP117870.1': 'Arthrobacter agilis',\n",
       " 'CP117872.1': 'Arthrobacter agilis',\n",
       " 'CP117871.1': 'Arthrobacter agilis',\n",
       " 'CP118233.1': 'Micromonospora chalcea',\n",
       " 'LT593929.1': 'Propionibacterium freudenreichii',\n",
       " 'CP113264.1': 'Streptomonospora nanhaiensis',\n",
       " 'CP113265.1': 'Streptomonospora nanhaiensis',\n",
       " 'CP114413.1': 'Streptomyces cinnabarinus',\n",
       " 'CP114203.1': 'Streptomyces nigrescens',\n",
       " 'CP114204.1': 'Streptomyces nigrescens',\n",
       " 'CP073767.1': 'Dactylosporangium aurantiacum',\n",
       " 'CP107020.1': 'Brachybacterium huguangmaarense',\n",
       " 'CP107021.1': 'Brachybacterium huguangmaarense',\n",
       " 'CP107222.1': 'Microbacterium terricola',\n",
       " 'CP084204.1': 'Streptomyces tanashiensis',\n",
       " 'CP103060.1': 'Streptomyces albus',\n",
       " 'CP103061.1': 'Streptomyces albus',\n",
       " 'CP073720.1': 'Dactylosporangium fulvum',\n",
       " 'CP104302.1': 'Mycolicibacterium brumae',\n",
       " 'CP073721.1': 'Dactylosporangium roseum',\n",
       " 'CP101471.1': 'Microbacterium liquefaciens',\n",
       " 'CP072934.1': 'Corynebacterium pseudogenitalium',\n",
       " 'CP102415.1': 'Amycolatopsis roodepoortensis',\n",
       " 'CP102514.1': 'Streptomyces yangpuensis',\n",
       " 'CP102517.1': 'Streptomyces yangpuensis',\n",
       " 'CP102515.1': 'Streptomyces yangpuensis',\n",
       " 'CP102516.1': 'Streptomyces yangpuensis',\n",
       " 'CP102850.1': 'Gordonia mangrovi',\n",
       " 'CP099837.1': 'Nocardiopsis exhalans',\n",
       " 'CP099838.1': 'Nocardiopsis exhalans',\n",
       " 'CP046605.1': 'Corynebacterium accolens',\n",
       " 'CP097650.1': 'Micrococcus yunnanensis',\n",
       " 'CP099468.1': 'Streptomyces phaeoluteigriseus',\n",
       " 'CP099469.1': 'Streptomyces phaeoluteigriseus',\n",
       " 'CP099470.1': 'Streptomyces phaeoluteigriseus',\n",
       " 'CP099547.1': 'Arcanobacterium pinnipediorum',\n",
       " 'CP051627.1': 'Thermobifida alba',\n",
       " 'CP097218.1': 'Brachybacterium kimchii',\n",
       " 'CP097219.1': 'Brachybacterium kimchii',\n",
       " 'CP097289.1': 'Streptomyces durmitorensis',\n",
       " 'CP097290.1': 'Streptomyces durmitorensis',\n",
       " 'CP097320.1': 'Candidatus Mycobacterium methanotrophicum',\n",
       " 'CP097321.1': 'Candidatus Mycobacterium methanotrophicum',\n",
       " 'CP097332.1': 'Jatrophihabitans telluris',\n",
       " 'CP093326.1': 'Arthrobacter sulfonylureivorans',\n",
       " 'CP093327.1': 'Arthrobacter sulfonylureivorans',\n",
       " 'CP071872.1': 'Streptomyces formicae',\n",
       " 'CP071873.1': 'Streptomyces formicae',\n",
       " 'CP093846.1': 'Streptomyces tubbatahanensis',\n",
       " 'CP094533.1': 'Agromyces soli',\n",
       " 'CP095552.1': 'Gordonia hongkongensis',\n",
       " 'CP095553.1': 'Gordonia hongkongensis',\n",
       " 'CP092427.2': 'Mycobacterium rufum',\n",
       " 'CP092428.2': 'Mycobacterium rufum',\n",
       " 'CP103313.1': 'Mycobacterium rufum',\n",
       " 'CP092430.2': 'Mycolicibacter virginiensis',\n",
       " 'CP092431.1': 'Streptomyces deccanensis',\n",
       " 'CP092488.2': 'Mycobacterium paraterrae',\n",
       " 'CP092364.2': 'Mycobacterium goodii',\n",
       " 'CP092365.1': 'Mycolicibacillus parakoreensis',\n",
       " 'CP092366.2': 'Mycolicibacillus parakoreensis',\n",
       " 'CP040786.1': 'Clavibacter phaseoli',\n",
       " 'CP040787.1': 'Clavibacter phaseoli',\n",
       " 'CP040792.1': 'Clavibacter californiensis',\n",
       " 'CP040793.1': 'Clavibacter californiensis',\n",
       " 'CP091855.1': 'Gordonia amicalis',\n",
       " 'CP018019.1': 'Mycobacterium avium subsp. hominissuis',\n",
       " 'CP084681.1': 'Streptomyces inhibens',\n",
       " 'CP092362.2': 'Mycolicibacterium crocinum',\n",
       " 'CP092363.2': 'Mycolicibacterium crocinum',\n",
       " 'CP103314.1': 'Mycolicibacterium crocinum',\n",
       " 'CP090063.1': 'Amycolatopsis acidiphila',\n",
       " 'CP071516.1': 'Arthrobacter polaris',\n",
       " 'CP090347.1': 'Microbacterium binotii',\n",
       " 'CP087879.1': 'Agromyces marinus',\n",
       " 'CP062221.1': 'Cellulomonas palmilytica',\n",
       " 'CP089293.1': 'Paenarthrobacter nicotinovorans',\n",
       " 'CP089294.1': 'Paenarthrobacter nicotinovorans',\n",
       " 'CP089224.1': 'Mycobacterium ostraviense',\n",
       " 'CP086120.1': 'Streptomyces gobiensis',\n",
       " 'CP088017.1': 'Ruania halotolerans',\n",
       " 'CP088969.1': 'Rhodococcus aetherivorans',\n",
       " 'CP082781.1': 'Microbacterium resistens',\n",
       " 'CP089226.1': 'Nocardia yamanashiensis',\n",
       " 'CP089225.1': 'Nocardia gipuzkoensis',\n",
       " 'CP085945.1': 'Mycobacterium intracellulare',\n",
       " 'CP085946.1': 'Mycobacterium intracellulare',\n",
       " 'CP086254.1': 'Brachybacterium halotolerans subsp. kimchii',\n",
       " 'CP086255.1': 'Brachybacterium halotolerans subsp. kimchii',\n",
       " 'AP024525.1': 'Corynebacterium cyclohexanicum',\n",
       " 'CP087710.1': 'Nocardioides okcheonensis',\n",
       " 'CP086119.1': 'Streptomyces huasconensis',\n",
       " 'CP088018.1': 'Ruania suaedae',\n",
       " 'CP084585.1': 'Cellulomonas iranensis',\n",
       " 'CP080203.1': 'Mycetocola spongiae',\n",
       " 'CP079233.1': 'Bifidobacterium pseudocatenulatum',\n",
       " 'CP079234.1': 'Bifidobacterium pseudocatenulatum',\n",
       " 'CP080954.1': 'Rhodococcus opacus PD630',\n",
       " 'CP080955.1': 'Rhodococcus opacus PD630',\n",
       " 'CP080956.1': 'Rhodococcus opacus PD630',\n",
       " 'CP080957.1': 'Rhodococcus opacus PD630',\n",
       " 'CP085051.1': 'Corynebacterium uberis',\n",
       " 'CP085200.1': 'Mycobacterium ulcerans',\n",
       " 'CP085201.1': 'Mycobacterium ulcerans',\n",
       " 'CP085974.1': 'Gordonia otitidis',\n",
       " 'CP083647.1': 'Corynebacterium falsenii',\n",
       " 'CP083590.1': 'Streptomyces mobaraensis',\n",
       " 'CP083591.1': 'Streptomyces mobaraensis',\n",
       " 'CP084058.1': 'Nonomuraea gerenzanensis',\n",
       " 'CP084059.1': 'Nonomuraea gerenzanensis',\n",
       " 'CP084060.1': 'Nonomuraea gerenzanensis',\n",
       " 'CP082273.1': 'Nocardioides coralli',\n",
       " 'CP082331.1': 'Micrococcus luteus',\n",
       " 'CP082332.1': 'Micrococcus luteus',\n",
       " 'CP061913.1': 'Dactylosporangium vinaceum',\n",
       " 'CP082922.1': 'Nocardioides rotundus',\n",
       " 'CP082945.1': 'Streptomyces angustmyceticus',\n",
       " 'CP080998.1': 'Mycolicibacterium holsaticum DSM 44478 = JCM 12374',\n",
       " 'CP080999.1': 'Mycobacterium malmoense',\n",
       " 'CP081000.1': 'Mycolicibacterium senegalense',\n",
       " 'CP081673.1': 'Mycolicibacterium farcinogenes',\n",
       " 'CP081674.1': 'Mycolicibacterium farcinogenes',\n",
       " 'CP081675.1': 'Mycolicibacterium farcinogenes',\n",
       " 'CP081676.1': 'Mycolicibacterium farcinogenes',\n",
       " 'CP082301.1': 'Streptomyces decoyicus',\n",
       " 'CP080332.1': 'Mycolicibacterium diernhoferi',\n",
       " 'CP080333.1': 'Mycolicibacterium pallens',\n",
       " 'CP080334.1': 'Mycolicibacterium pallens',\n",
       " 'CP080335.1': 'Mycolicibacterium pallens',\n",
       " 'CP080647.1': 'Streptomyces akebiae',\n",
       " 'CP080648.1': 'Streptomyces akebiae',\n",
       " 'CP080997.1': 'Mycolicibacter heraklionensis',\n",
       " 'CP079105.1': 'Skermania piniformis',\n",
       " 'CP079216.1': 'Tessaracoccus palaemonis',\n",
       " 'CP079698.1': 'Rhodococcus globerulus',\n",
       " 'CP068985.1': 'Nonomuraea coxensis DSM 45129',\n",
       " 'CP080409.1': 'Nocardioides panacisoli',\n",
       " 'CP077258.1': 'Corynebacterium freneyi',\n",
       " 'CP059572.1': 'Actinomadura graeca',\n",
       " 'CP059573.1': 'Actinomadura graeca',\n",
       " 'CP059574.1': 'Actinomadura graeca',\n",
       " 'CP078145.1': 'Nocardia iowensis',\n",
       " 'CP076456.1': 'Arthrobacter sunyaminii',\n",
       " 'CP077062.1': 'Nocardioides panacis',\n",
       " 'CP072467.1': 'Spiractinospora alimapuensis',\n",
       " 'CP075371.1': 'Nocardioides aquaticus',\n",
       " 'CP076022.1': 'Arthrobacter jiangjiafuii',\n",
       " 'CP076023.1': 'Cellulomonas dongxiuzhuiae',\n",
       " 'CP074371.1': 'Nocardia tengchongensis',\n",
       " 'CP074376.1': 'Mycolicibacterium neoaurum',\n",
       " 'CP074405.1': 'Cellulomonas wangleii',\n",
       " 'CP074404.1': 'Cellulomonas fengjieae',\n",
       " 'CP074132.1': 'Nocardiopsis akebiae',\n",
       " 'AP023356.1': 'Actinoplanes ianthinogenes',\n",
       " 'AP023359.1': 'Polymorphospora rubra',\n",
       " 'AP023354.1': 'Actinocatenispora sera',\n",
       " 'CP060742.1': 'Streptomyces alfalfae',\n",
       " 'CP060743.1': 'Streptomyces alfalfae',\n",
       " 'CP073778.1': 'Streptomyces lavendulae subsp. lavendulae',\n",
       " 'CP046600.1': 'Mycobacterium spongiae',\n",
       " 'CP074133.1': 'Nocardiopsis changdeensis',\n",
       " 'CP074134.1': 'Nocardiopsis changdeensis',\n",
       " 'CP074135.1': 'Nocardiopsis changdeensis',\n",
       " 'CP074136.1': 'Nocardiopsis changdeensis',\n",
       " 'CP071974.1': 'Trueperella pecoris',\n",
       " 'CP071139.1': 'Streptomyces nojiriensis',\n",
       " 'CP071883.1': 'Curtobacterium flaccumfaciens pv. flaccumfaciens',\n",
       " 'CP071884.1': 'Curtobacterium flaccumfaciens pv. flaccumfaciens',\n",
       " 'CP071696.1': 'Agromyces archimandritae',\n",
       " 'CP072387.1': 'Cellulosimicrobium cellulans',\n",
       " 'CP072388.1': 'Cellulosimicrobium cellulans',\n",
       " 'CP072384.1': 'Arachnia rubra',\n",
       " 'CP039124.1': 'Thermobispora bispora',\n",
       " 'CP039125.1': 'Thermobispora bispora',\n",
       " 'CP022295.1': 'Nocardioides aromaticivorans',\n",
       " 'CP014574.1': 'Paenarthrobacter ureafaciens',\n",
       " 'CP014575.1': 'Paenarthrobacter ureafaciens',\n",
       " 'CP014576.1': 'Paenarthrobacter ureafaciens',\n",
       " 'CP071839.1': 'Streptomyces cyanogenus',\n",
       " 'CP071868.1': 'Pengzhenrongella sicca',\n",
       " 'CP070609.1': 'Rhodococcus koreensis',\n",
       " 'CP070610.1': 'Rhodococcus koreensis',\n",
       " 'CP070611.1': 'Rhodococcus koreensis',\n",
       " 'CP070612.1': 'Rhodococcus koreensis',\n",
       " 'CP070613.1': 'Rhodococcus koreensis',\n",
       " 'CP070619.1': 'Rhodococcus pseudokoreensis',\n",
       " 'CP070618.1': 'Rhodococcus pseudokoreensis',\n",
       " 'CP070617.1': 'Rhodococcus pseudokoreensis',\n",
       " 'CP070616.1': 'Rhodococcus pseudokoreensis',\n",
       " 'CP070615.1': 'Rhodococcus pseudokoreensis',\n",
       " 'CP070614.1': 'Rhodococcus pseudokoreensis',\n",
       " 'CP070496.1': 'Natronoglycomyces albus',\n",
       " 'CP070498.1': 'Natronoglycomyces albus',\n",
       " 'CP070497.1': 'Natronoglycomyces albus',\n",
       " 'CP070499.1': 'Natronosporangium hydrolyticum',\n",
       " 'AP024310.1': 'Mycobacterium heckeshornense',\n",
       " 'AP023355.1': 'Actinocatenispora thailandica',\n",
       " 'CP069806.1': 'Glutamicibacter protophormiae',\n",
       " 'CP069807.1': 'Glutamicibacter protophormiae',\n",
       " 'CP069808.1': 'Glutamicibacter protophormiae',\n",
       " 'CP070228.1': 'Arcanobacterium phocisimile',\n",
       " 'CP070242.1': 'Streptomyces californicus',\n",
       " 'CP070241.1': 'Streptomyces californicus',\n",
       " 'CP070243.1': 'Streptomyces californicus',\n",
       " 'CP070326.1': 'Streptomyces noursei',\n",
       " 'CP068156.1': 'Corynebacterium tuberculostearicum',\n",
       " 'CP068162.1': 'Corynebacterium glucuronolyticum',\n",
       " 'CP068163.1': 'Corynebacterium glucuronolyticum',\n",
       " 'CP068161.1': 'Corynebacterium propinquum',\n",
       " 'CP068168.1': 'Corynebacterium amycolatum',\n",
       " 'AP024251.1': 'Mycobacterium paraintracellulare',\n",
       " 'AP024252.1': 'Mycobacterium paraintracellulare',\n",
       " 'CP049945.1': 'Streptomyces koyangensis',\n",
       " 'CP049946.1': 'Streptomyces koyangensis',\n",
       " 'CP044211.1': 'Rhodococcus ruber',\n",
       " 'CP044212.1': 'Rhodococcus ruber',\n",
       " 'CP044213.1': 'Rhodococcus ruber',\n",
       " 'CP066802.1': 'Actinomyces weissii',\n",
       " 'CP067012.1': 'Corynebacterium kefirresidentii',\n",
       " 'CP067011.1': 'Corynebacterium kefirresidentii',\n",
       " 'CP068292.1': 'Corynebacterium macginleyi',\n",
       " 'CP068158.1': 'Corynebacterium striatum',\n",
       " 'CP048882.1': 'Streptomyces bathyalis',\n",
       " 'CP065682.1': 'Brevibacterium casei',\n",
       " 'CP066049.1': 'Actinomyces naeslundii',\n",
       " 'CP066060.1': 'Actinomyces oris',\n",
       " 'CP066061.1': 'Actinomyces oris',\n",
       " 'CP066065.1': 'Schaalia meyeri',\n",
       " 'CP066831.1': 'Streptomyces liliifuscus',\n",
       " 'CP066832.1': 'Streptomyces liliifuscus',\n",
       " 'CP064760.1': 'Microbacterium schleiferi',\n",
       " 'CP065050.1': 'Streptomyces solisilvae',\n",
       " 'CP059991.1': 'Streptomyces gardneri',\n",
       " 'CP065253.1': 'Streptomyces clavuligerus',\n",
       " 'CP065254.1': 'Streptomyces clavuligerus',\n",
       " 'CP063232.1': 'Thermobifida fusca',\n",
       " 'AP023172.1': 'Rhodococcus qingshengii',\n",
       " 'AP023173.1': 'Rhodococcus qingshengii',\n",
       " 'CP063373.1': 'Streptomyces ferrugineus',\n",
       " 'CP063374.1': 'Streptomyces chromofuscus',\n",
       " 'CP061274.1': 'Clavibacter zhangzhiyongii',\n",
       " 'CP119182.1': 'Streptomyces caniscabiei',\n",
       " 'CP062938.1': 'Bifidobacterium eulemuris',\n",
       " 'CP062948.1': 'Bifidobacterium lemurum',\n",
       " 'CP062939.1': 'Bifidobacterium subtile',\n",
       " 'CP063169.1': 'Ruania alkalisoli',\n",
       " 'CP061344.1': 'Microbacterium hominis',\n",
       " 'CP061725.1': 'Micromonospora craniellae',\n",
       " 'AP023439.1': 'Streptomyces tuirus',\n",
       " 'AP023440.1': 'Streptomyces aurantiacus',\n",
       " 'CP061539.1': 'Rothia terrae',\n",
       " 'CP062960.1': 'Rothia terrae',\n",
       " 'CP062961.1': 'Rothia terrae',\n",
       " 'CP062962.1': 'Rothia terrae',\n",
       " 'CP062963.1': 'Rothia terrae',\n",
       " 'CP061538.1': 'Rothia amarae',\n",
       " 'CP062957.1': 'Rothia amarae',\n",
       " 'CP062958.1': 'Rothia amarae',\n",
       " 'CP062959.1': 'Rothia amarae',\n",
       " 'CP060828.1': 'Streptomyces roseirectus',\n",
       " 'CP061032.1': 'Corynebacterium lujinxingii',\n",
       " 'CP061033.1': 'Corynebacterium zhongnanshanii',\n",
       " 'CP046884.1': 'Corynebacterium poyangense',\n",
       " 'CP063989.1': 'Actinomyces respiraculi',\n",
       " 'CP061281.1': 'Streptomyces xanthii',\n",
       " 'CP061282.1': 'Streptomyces xanthii',\n",
       " 'CP061283.1': 'Streptomyces xanthii',\n",
       " 'CP051006.1': 'Streptomyces griseofuscus',\n",
       " 'CP051007.1': 'Streptomyces griseofuscus',\n",
       " 'CP051008.1': 'Streptomyces griseofuscus',\n",
       " 'CP051009.1': 'Streptomyces griseofuscus',\n",
       " 'CP060789.1': 'Tessaracoccus defluvii',\n",
       " 'CP060825.1': 'Streptomyces genisteinicus',\n",
       " 'CP060826.1': 'Streptomyces genisteinicus',\n",
       " 'CP060827.1': 'Streptomyces genisteinicus',\n",
       " 'CP060713.1': 'Nocardioides mesophilus',\n",
       " 'CP060716.1': 'Leucobacter denitrificans',\n",
       " 'AP023396.1': 'Nocardia wallacei',\n",
       " 'AP023397.1': 'Nocardia wallacei',\n",
       " 'CP046883.1': 'Corynebacterium anserum',\n",
       " 'CP060404.1': 'Streptomyces buecherae',\n",
       " 'CP059694.1': 'Gordonia rubripertincta',\n",
       " 'CP059695.1': 'Gordonia rubripertincta',\n",
       " 'CP045704.1': 'Streptomyces rutgersensis',\n",
       " 'CP045705.1': 'Streptomyces rutgersensis',\n",
       " 'CP043661.1': 'Kribbella qitaiheensis',\n",
       " 'CP048813.1': 'Rhodococcus triatomae',\n",
       " 'CP060131.1': 'Pseudonocardia petroleophila',\n",
       " 'CP058322.1': 'Micromonospora carbonacea',\n",
       " 'CP040752.1': 'Streptomyces rectiverticillatus',\n",
       " 'CP059165.1': 'Mycobacterium vicinigordonae',\n",
       " 'CP059399.1': 'Nocardia huaxiensis',\n",
       " 'CP059491.1': 'Gordonia jinghuaiqii',\n",
       " 'CP050804.1': 'Arcanobacterium buesumense',\n",
       " 'CP053564.1': 'Pseudonocardia broussonetiae',\n",
       " 'CP053565.1': 'Pseudonocardia broussonetiae',\n",
       " 'CP053566.1': 'Pseudonocardia broussonetiae',\n",
       " 'CP053567.1': 'Pseudonocardia broussonetiae',\n",
       " 'CP053642.1': 'Actinomyces marmotae',\n",
       " 'CP054056.1': 'Aquiluna borgnonia',\n",
       " 'CP054938.1': 'Streptomyces harbinensis',\n",
       " 'CP054926.1': 'Streptomyces microflavus',\n",
       " 'CP054927.1': 'Streptomyces microflavus',\n",
       " 'CP054928.1': 'Streptomyces microflavus',\n",
       " 'CP035810.1': 'Brevibacterium luteolum',\n",
       " 'CP050177.1': 'Streptomyces liangshanensis',\n",
       " 'CP061169.1': 'Microbacterium chengjingii',\n",
       " 'AP022870.1': 'Phytohabitans flavus',\n",
       " 'AP022871.1': 'Phytohabitans suffuscus',\n",
       " 'CP064955.1': 'Corynebacterium qintianiae',\n",
       " 'CP064956.1': 'Corynebacterium qintianiae',\n",
       " 'CP064954.1': 'Corynebacterium lizhenjunii',\n",
       " 'CP049257.1': 'Nocardioides anomalus',\n",
       " 'CP049863.1': 'Leucobacter viscericola',\n",
       " 'CP049866.1': 'Nocardioides piscis',\n",
       " 'CP049934.1': 'Leucobacter insecticola',\n",
       " 'CP049933.1': 'Leucobacter coleopterorum',\n",
       " 'CP049865.1': 'Propioniciclava coleopterorum',\n",
       " 'AP022593.1': 'Mycolicibacterium arabiense',\n",
       " 'AP022592.1': 'Mycolicibacterium arabiense',\n",
       " 'AP022594.1': 'Mycolicibacillus koreensis',\n",
       " 'AP022620.1': 'Mycolicibacterium anyangense',\n",
       " 'AP022595.1': 'Mycolicibacterium sarraceniae',\n",
       " 'AP022596.1': 'Mycolicibacterium helvum',\n",
       " 'AP022590.1': 'Mycobacterium mantenii',\n",
       " 'AP022591.1': 'Mycolicibacterium celeriflavum',\n",
       " 'AP022588.1': 'Mycolicibacterium sediminis',\n",
       " 'AP022589.1': 'Mycolicibacter minnesotensis',\n",
       " 'AP022582.1': 'Mycobacterium seoulense',\n",
       " 'AP022618.1': 'Mycolicibacterium insubricum',\n",
       " 'AP022583.1': 'Mycobacterium noviomagense',\n",
       " 'AP022619.1': 'Mycobacterium paraseoulense',\n",
       " 'AP022586.1': 'Mycolicibacterium litorale',\n",
       " 'AP022587.1': 'Mycobacterium stomatepiae',\n",
       " 'AP022615.1': 'Mycobacterium heidelbergense',\n",
       " 'AP022616.1': 'Mycolicibacterium phocaicum',\n",
       " 'AP022579.1': 'Mycolicibacterium boenickei',\n",
       " 'AP022580.1': 'Mycolicibacterium boenickei',\n",
       " 'AP022581.1': 'Mycobacterium lacus',\n",
       " 'AP022617.1': 'Mycolicibacterium monacense',\n",
       " 'AP022613.1': 'Mycobacterium conspicuum',\n",
       " 'AP022576.1': 'Mycobacterium florentinum',\n",
       " 'AP022614.1': 'Mycobacterium parmense',\n",
       " 'AP022610.1': 'Mycolicibacterium madagascariense',\n",
       " 'AP022611.1': 'Mycolicibacterium madagascariense',\n",
       " 'AP022612.1': 'Mycolicibacterium confluentis',\n",
       " 'AP022575.1': 'Mycobacterium shinjukuense',\n",
       " 'AP022606.1': 'Mycobacterium branderi',\n",
       " 'AP022607.1': 'Mycobacterium branderi',\n",
       " 'AP022608.1': 'Mycolicibacterium gadium',\n",
       " 'AP022573.1': 'Mycobacterium saskatchewanense',\n",
       " 'AP022574.1': 'Mycolicibacterium psychrotolerans',\n",
       " 'AP022609.1': 'Mycolicibacter hiberniae',\n",
       " 'AP022569.1': 'Mycobacterium cookii',\n",
       " 'AP022605.1': 'Mycobacterium doricum',\n",
       " 'AP022570.1': 'Mycolicibacterium poriferae',\n",
       " 'AP022571.1': 'Mycolicibacterium poriferae',\n",
       " 'AP022572.1': 'Mycobacterium shottsii',\n",
       " 'AP022562.1': 'Mycobacterium novum',\n",
       " 'AP022563.1': 'Mycolicibacterium duvalii',\n",
       " 'AP022603.1': 'Mycolicibacterium fallax',\n",
       " 'AP022565.1': 'Mycolicibacterium alvei',\n",
       " 'AP022566.1': 'Mycolicibacterium alvei',\n",
       " 'AP022567.1': 'Mycolicibacterium mageritense',\n",
       " 'AP022568.1': 'Mycobacterium simiae',\n",
       " 'CP017146.1': 'Marisediminicola antarctica',\n",
       " 'AP022598.1': 'Mycolicibacterium parafortuitum',\n",
       " 'AP022599.1': 'Mycolicibacterium pulveris',\n",
       " 'AP022600.1': 'Mycolicibacterium tokaiense',\n",
       " 'AP022560.1': 'Mycolicibacterium moriokaense',\n",
       " 'AP022561.1': 'Mycolicibacterium aichiense',\n",
       " 'CP047156.1': 'Epidermidibacterium keratini',\n",
       " 'CP047186.1': 'Rathayibacter tanaceti',\n",
       " 'CP047180.1': 'Rathayibacter festucae',\n",
       " 'CP047181.1': 'Rathayibacter festucae',\n",
       " 'CP047182.1': 'Rathayibacter festucae',\n",
       " 'CP047898.1': 'Pseudarthrobacter psychrotolerans',\n",
       " 'CP047899.1': 'Pseudarthrobacter psychrotolerans',\n",
       " 'CP047900.1': 'Pseudarthrobacter psychrotolerans',\n",
       " 'CP045805.1': 'Gordonia amarae',\n",
       " 'CP045806.1': 'Gordonia pseudamarae',\n",
       " 'CP045807.1': 'Gordonia pseudamarae',\n",
       " 'CP046452.1': 'Corynebacterium kalinowskii',\n",
       " 'CP046453.1': 'Corynebacterium comes',\n",
       " 'CP046454.1': 'Corynebacterium comes',\n",
       " 'CP046455.1': 'Corynebacterium occultum',\n",
       " 'CP046456.1': 'Corynebacterium occultum',\n",
       " 'CP033724.1': 'Clavibacter michiganensis subsp. michiganensis',\n",
       " 'CP033725.1': 'Clavibacter michiganensis subsp. michiganensis',\n",
       " 'CP033726.1': 'Clavibacter michiganensis subsp. michiganensis',\n",
       " 'CP034279.1': 'Streptomyces ficellus',\n",
       " 'CP047020.1': 'Streptomyces broussonetiae',\n",
       " 'CP045725.1': 'Raineyella fluvialis',\n",
       " 'CP045737.1': 'Aeromicrobium yanjiei',\n",
       " 'CP045736.1': 'Aeromicrobium yanjiei',\n",
       " 'CP045929.1': 'Allosaccharopolyspora coralli',\n",
       " 'CP045309.1': 'Micromonospora terminaliae',\n",
       " 'CP045310.1': 'Micromonospora terminaliae',\n",
       " 'CP052757.1': 'Cellulosimicrobium protaetiae',\n",
       " 'CP052758.1': 'Cellulosimicrobium protaetiae',\n",
       " 'CP034550.1': 'Saccharothrix syringae',\n",
       " 'CP045643.1': 'Streptomyces fagopyri',\n",
       " 'CP045032.1': 'Corynebacterium urogenitale',\n",
       " 'CP045096.1': 'Streptomyces phaeolivaceus',\n",
       " 'CP045095.1': 'Streptomyces phaeolivaceus',\n",
       " 'CP023693.1': 'Streptomyces cinereoruber',\n",
       " 'CP045529.1': 'Luteimicrobium xylanilyticum',\n",
       " 'CP045572.1': 'Nonomuraea phyllanthi',\n",
       " 'CP023692.1': 'Streptomyces vinaceus',\n",
       " 'CP023694.1': 'Streptomyces coeruleorubidus',\n",
       " 'CP044231.1': 'Microbacterium caowuchunii',\n",
       " 'CP044232.1': 'Microbacterium lushaniae',\n",
       " 'CP044427.1': 'Ornithinimicrobium pratense',\n",
       " 'CP023703.1': 'Streptomyces galilaeus',\n",
       " 'CP023689.1': 'Streptomyces chartreusis',\n",
       " 'CP023690.1': 'Streptomyces spectabilis',\n",
       " 'CP023691.1': 'Streptomyces platensis',\n",
       " 'CP023699.1': 'Streptomyces kanamyceticus',\n",
       " 'CP023700.1': 'Streptomyces viridosporus T7A',\n",
       " 'CP023701.1': 'Streptomyces subrutilus',\n",
       " 'CP023702.1': 'Streptomyces nitrosporeus',\n",
       " 'CP043474.1': 'Mycobacterium grossiae',\n",
       " 'CP043475.1': 'Mycobacterium grossiae',\n",
       " 'CP043504.1': 'Protaetiibacter larvae',\n",
       " 'CP023695.1': 'Streptomyces alboniger',\n",
       " 'CP023696.1': 'Streptomyces fradiae ATCC 10745 = DSM 40063',\n",
       " 'CP023697.1': 'Streptomyces prasinus',\n",
       " 'CP042305.1': 'Humibacter ginsenosidimutans',\n",
       " 'CP041692.1': 'Microlunatus elymi',\n",
       " 'CP041765.1': 'Tomitella fengzijianii',\n",
       " 'CP041766.1': 'Tomitella fengzijianii',\n",
       " 'CP038157.1': 'Corynebacterium sanguinis',\n",
       " 'CP042266.1': 'Streptomyces qinzhouensis',\n",
       " 'CP040635.1': 'Acidipropionibacterium jensenii',\n",
       " 'CP034412.1': 'Glutamicibacter creatinolyticus',\n",
       " 'CP040899.1': 'Georgenia wutianyii',\n",
       " 'CP040915.1': 'Georgenia yuyongxinii',\n",
       " 'CP022310.1': 'Streptomyces calvus',\n",
       " 'CP041616.1': 'Ornithinimicrobium ciconiae',\n",
       " 'CP041695.1': 'Nocardia otitidiscaviarum',\n",
       " 'CP039247.1': 'Corynebacterium endometrii',\n",
       " 'CP039291.1': 'Cellulomonas shaoxiangyii',\n",
       " 'CP021978.1': 'Streptomyces hawaiiensis',\n",
       " 'CP024633.1': 'Mycobacteroides salmoniphilum',\n",
       " 'CP062008.1': 'Mycolicibacterium mucogenicum DSM 44124',\n",
       " 'CP038266.1': 'Microbacterium wangchenii',\n",
       " 'CP038267.1': 'Nocardioides euryhalodurans',\n",
       " 'CP038436.1': 'Nocardioides seonyuensis',\n",
       " 'CP035493.1': 'Xylanimonas protaetiae',\n",
       " 'CP035491.1': 'Agromyces protaetiae',\n",
       " 'CP035806.1': 'Leucobacter triazinivorans',\n",
       " 'CP036455.1': 'Streptomonospora litoralis',\n",
       " 'CP036456.1': 'Streptomonospora litoralis',\n",
       " 'CP035504.1': 'Kocuria indica',\n",
       " 'CP035505.1': 'Kocuria indica',\n",
       " 'CP032229.1': 'Streptomyces seoulensis',\n",
       " 'CP032230.1': 'Streptomyces seoulensis',\n",
       " 'CP038256.1': 'Microbacterium sediminis',\n",
       " 'CP038257.1': 'Microbacterium sediminis',\n",
       " 'CP027114.1': 'Gordonia alkanivorans',\n",
       " 'CP027115.1': 'Gordonia alkanivorans',\n",
       " 'CP034181.1': 'Mycobacteroides abscessus',\n",
       " 'CP034180.1': 'Mycobacteroides abscessus',\n",
       " 'CP035037.1': 'Leucobacter muris',\n",
       " 'CP035299.1': 'Corynebacterium pelargi',\n",
       " 'CP035494.1': 'Microbacterium protaetiae',\n",
       " 'CP035495.1': 'Xylanimonas allomyrinae',\n",
       " 'CP035496.1': 'Xylanimonas allomyrinae',\n",
       " 'CP034593.1': 'Flaviflexus ciconiae',\n",
       " 'CP031423.1': 'Microbacterium lemovicicum',\n",
       " 'CP034687.1': 'Streptomyces griseoviridis',\n",
       " 'CP026734.1': 'Brevibacterium linens',\n",
       " 'CP026735.1': 'Brevibacterium linens',\n",
       " 'CP028129.1': 'Rathayibacter rathayi',\n",
       " 'CP028130.1': 'Rathayibacter iranicus',\n",
       " 'AP018920.1': 'Pseudonocardia autotrophica',\n",
       " 'AP018921.1': 'Pseudonocardia autotrophica',\n",
       " 'CP034438.1': 'Flaviflexus salsibiostraticola',\n",
       " 'CP034463.1': 'Streptomyces aquilus',\n",
       " 'CP034539.1': 'Streptomyces cyaneochromogenes',\n",
       " 'CP025333.1': 'Brevibacterium aurantiacum',\n",
       " 'AP019307.1': 'Nocardioides baekrokdamisoli',\n",
       " 'CP044548.2': 'Janibacter melonis',\n",
       " 'CP046475.1': 'Janibacter melonis',\n",
       " 'CP033719.1': 'Propionibacterium acidifaciens',\n",
       " 'CP033896.1': 'Corynebacterium choanae',\n",
       " 'CP033897.1': 'Corynebacterium gerontici',\n",
       " 'CP033898.1': 'Corynebacterium pseudopelargi',\n",
       " 'CP033899.1': 'Corynebacterium pseudopelargi',\n",
       " 'CP033972.1': 'Gordonia insulae',\n",
       " 'CP034170.1': 'Nakamurella antarctica',\n",
       " 'CP020700.1': 'Streptomyces tsukubensis',\n",
       " 'CP020701.1': 'Streptomyces tsukubensis',\n",
       " 'CP020702.1': 'Streptomyces tsukubensis',\n",
       " 'CP025546.1': 'Mycobacterium paragordonae',\n",
       " 'CP025547.1': 'Mycobacterium paragordonae',\n",
       " 'CP025548.1': 'Mycobacterium paragordonae',\n",
       " 'CP025549.1': 'Mycobacterium paragordonae',\n",
       " 'CP025550.1': 'Mycobacterium paragordonae',\n",
       " 'CP032624.1': 'Gryllotalpicola protaetiae',\n",
       " 'CP032625.1': 'Gryllotalpicola protaetiae',\n",
       " 'CP032630.1': 'Protaetiibacter intestinalis',\n",
       " 'CP032698.1': 'Streptomyces hundungensis',\n",
       " 'CP032788.1': 'Corynebacterium xerosis',\n",
       " 'CP023407.1': 'Streptomyces fungicidicus',\n",
       " 'CP023408.1': 'Streptomyces fungicidicus',\n",
       " 'CP033073.1': 'Streptomyces dangxiongensis',\n",
       " 'CP033072.1': 'Streptomyces dangxiongensis',\n",
       " 'CP031425.1': 'Microbacterium foliorum',\n",
       " 'CP063196.1': 'Thermobifida halotolerans',\n",
       " 'CP063197.1': 'Thermobifida halotolerans',\n",
       " 'AP018410.1': 'Mycobacterium pseudoshottsii JCM 15466',\n",
       " 'CP032427.1': 'Streptomyces griseorubiginosus',\n",
       " 'CP024087.1': 'Micromonospora tulbaghiae',\n",
       " 'CP022744.1': 'Streptomyces lincolnensis',\n",
       " 'CP031194.1': 'Streptomyces paludis',\n",
       " 'CP031229.1': 'Ornithinimicrobium avium',\n",
       " 'CP031356.1': 'Brachybacterium saurashtrense',\n",
       " 'CP015163.1': 'Amycolatopsis albispora',\n",
       " 'CP030862.1': 'Streptomyces globosus',\n",
       " 'CP030863.1': 'Streptomyces globosus',\n",
       " 'CP030864.1': 'Streptomyces globosus',\n",
       " 'CP025198.1': 'Acidipropionibacterium virtanenii',\n",
       " 'CP027306.1': 'Streptomyces atratus',\n",
       " 'CP021354.1': 'Rhodococcus oxybenzonivorans',\n",
       " 'CP021357.1': 'Rhodococcus oxybenzonivorans',\n",
       " 'CP021356.1': 'Rhodococcus oxybenzonivorans',\n",
       " 'CP021355.1': 'Rhodococcus oxybenzonivorans',\n",
       " 'CP023994.1': 'Aurantimicrobium photophilum',\n",
       " 'CP029543.1': 'Mycobacterium leprae',\n",
       " 'CP031264.1': 'Peterkaempfera bronchialis',\n",
       " 'CP030073.1': 'Streptomyces cadmiisoli',\n",
       " 'CP030074.1': 'Streptomyces cadmiisoli',\n",
       " 'CP026948.1': 'Corynebacterium liangguodongii',\n",
       " 'CP028913.1': 'Agromyces badenianii',\n",
       " 'CP029043.1': 'Streptomyces nigra',\n",
       " 'CP015449.1': 'Dietzia lutea',\n",
       " 'CP015450.1': 'Dietzia lutea',\n",
       " 'CP015451.1': 'Dietzia lutea',\n",
       " 'CP015452.1': 'Dietzia lutea',\n",
       " 'CP015453.1': 'Dietzia psychralcaliphila',\n",
       " 'CP015454.1': 'Dietzia psychralcaliphila',\n",
       " 'CP029188.1': 'Streptomyces tirandamycinicus',\n",
       " 'CP029254.1': 'Streptomyces spongiicola',\n",
       " 'CP028341.1': 'Bifidobacterium adolescentis',\n",
       " 'CP026304.1': 'Streptomyces lunaelactis',\n",
       " 'CP026305.1': 'Streptomyces lunaelactis',\n",
       " 'CP026306.1': 'Streptomyces lunaelactis',\n",
       " 'CP026947.1': 'Corynebacterium yudongzhengii',\n",
       " 'CP026949.1': 'Mycetocola zhujimingii',\n",
       " 'CP026950.1': 'Mycetocola zhujimingii',\n",
       " 'CP026952.1': 'Aeromicrobium chenweiae',\n",
       " 'CP026923.1': 'Pontimonas salivibrio',\n",
       " 'CP027433.1': 'Gordonia iterans',\n",
       " 'CP028426.1': 'Gulosibacter molinativorax',\n",
       " 'CP028427.1': 'Gulosibacter molinativorax',\n",
       " 'CP027793.1': 'Prescottella equi',\n",
       " 'CP027794.1': 'Prescottella equi',\n",
       " 'CP027795.1': 'Prescottella equi',\n",
       " 'CP026746.1': 'Nocardia cyriacigeorgica',\n",
       " 'CP026745.1': 'Nocardia cyriacigeorgica',\n",
       " 'CP023564.1': 'Brachybacterium ginsengisoli',\n",
       " 'CP023563.1': 'Brachybacterium vulturis',\n",
       " 'CP023778.1': 'Nocardia terpenica',\n",
       " 'CP023779.1': 'Nocardia terpenica',\n",
       " 'CP018044.1': 'Bifidobacterium choerinum',\n",
       " 'CP018045.1': 'Bifidobacterium choerinum',\n",
       " 'CP024957.1': 'Streptomyces cavourensis',\n",
       " 'CP018863.1': 'Arthrobacter crystallopoietes',\n",
       " 'CP018864.1': 'Arthrobacter crystallopoietes',\n",
       " 'CP018865.1': 'Arthrobacter crystallopoietes',\n",
       " 'CP026652.1': 'Streptomyces dengpaensis',\n",
       " 'CP026653.1': 'Streptomyces dengpaensis',\n",
       " 'CP026654.1': 'Streptomyces dengpaensis',\n",
       " 'CP016780.1': 'Candidatus Planktophila lacus',\n",
       " 'CP016782.1': 'Candidatus Planktophila limnetica',\n",
       " 'CP023445.1': 'Actinosynnema pretiosum',\n",
       " 'AP017369.1': 'Corynebacterium suranareeae',\n",
       " 'AP017457.1': 'Aurantimicrobium minutum',\n",
       " 'AP017900.1': 'Nocardia seriolae',\n",
       " 'AP018164.1': 'Mycobacterium shigaense',\n",
       " 'AP018165.1': '[Mycobacterium] stephanolepidis',\n",
       " 'CP022544.1': 'Bifidobacterium pseudolongum',\n",
       " 'CP023147.1': 'Mycobacterium marseillense',\n",
       " 'CP023148.1': 'Mycobacterium marseillense',\n",
       " 'CP019688.1': 'Corynebacterium glaucum',\n",
       " 'CP016768.2': 'Candidatus Nanopelagicus limnes',\n",
       " 'CP016770.1': 'Candidatus Planktophila dulcis',\n",
       " 'CP016771.1': 'Candidatus Nanopelagicus hibericus',\n",
       " 'CP016773.1': 'Candidatus Planktophila sulfonica',\n",
       " 'CP016775.1': 'Candidatus Planktophila versatilis',\n",
       " 'CP016776.1': 'Candidatus Planktophila vernalis',\n",
       " 'CP016779.1': 'Candidatus Nanopelagicus abundans',\n",
       " 'CP021417.2': 'Corynebacterium silvaticum',\n",
       " 'CP022088.2': 'Nocardia brasiliensis',\n",
       " 'CP022316.1': 'Brachybacterium avium',\n",
       " 'CP022317.1': 'Brachybacterium avium',\n",
       " 'CP022318.1': 'Brachybacterium avium',\n",
       " 'CP022521.1': 'Actinoalloteichus hoggarensis',\n",
       " 'CP016353.1': 'Prauserella marina',\n",
       " 'CP016354.1': 'Prauserella marina',\n",
       " 'CP022753.1': 'Nocardiopsis gilva YIM 90087',\n",
       " 'CP022752.1': 'Actinopolyspora erythraea',\n",
       " 'CP020569.1': 'Streptomyces gilvosporeus',\n",
       " 'CP020567.1': 'Kitasatospora aureofaciens',\n",
       " 'CP020568.1': 'Kitasatospora aureofaciens',\n",
       " 'CP020715.1': 'Cnuibacter physcomitrellae',\n",
       " 'CP020716.1': 'Cnuibacter physcomitrellae',\n",
       " 'CP020717.1': 'Cnuibacter physcomitrellae',\n",
       " 'CP021121.1': 'Streptomyces marincola',\n",
       " 'CP020809.1': 'Mycobacterium dioxanotrophicus',\n",
       " 'CP020810.1': 'Mycobacterium dioxanotrophicus',\n",
       " 'CP020811.1': 'Mycobacterium dioxanotrophicus',\n",
       " 'CP020812.1': 'Mycobacterium dioxanotrophicus',\n",
       " 'CP020813.1': 'Mycobacterium dioxanotrophicus',\n",
       " 'CP009249.1': 'Corynebacterium phocae',\n",
       " 'CP009250.1': 'Corynebacterium phocae',\n",
       " 'CP016076.1': 'Actinoalloteichus fjordicus',\n",
       " 'CP018762.1': 'Microbacterium aurum',\n",
       " 'CP031421.1': 'Microbacterium oleivorans',\n",
       " 'CP019605.1': 'Tessaracoccus flavus',\n",
       " 'CP019606.1': 'Tessaracoccus aquimaris',\n",
       " 'CP019607.1': 'Tessaracoccus flavescens',\n",
       " 'CP019608.1': 'Tessaracoccus flavescens',\n",
       " 'CP020468.1': 'Actinomyces gaoshouyii',\n",
       " 'CP020570.1': 'Streptomyces violaceoruber',\n",
       " 'CP018082.1': 'Nocardia mangyaensis',\n",
       " 'CP018135.1': 'Neomicrococcus aestuarii',\n",
       " 'CP018134.1': 'Microbacterium paludicola',\n",
       " 'CP013290.1': 'Janibacter indicus',\n",
       " 'CP009251.1': 'Corynebacterium stationis',\n",
       " 'CP009252.1': 'Corynebacterium stationis',\n",
       " 'CP009245.1': 'Corynebacterium aquilae DSM 44791',\n",
       " 'CP009246.1': 'Corynebacterium flavescens',\n",
       " 'CP009247.1': 'Corynebacterium frankenforstense DSM 45800',\n",
       " 'CP009248.1': 'Corynebacterium sphenisci DSM 44792',\n",
       " 'CP017040.1': 'Cutibacterium modestum',\n",
       " 'CP017041.1': 'Cutibacterium modestum',\n",
       " 'CP017248.1': 'Streptomyces fodineus',\n",
       " 'CP014859.1': 'Actinoalloteichus hymeniacidonis',\n",
       " 'CP015208.1': 'Candidatus Rhodoluna planktonica',\n",
       " 'CP017812.1': 'Boudabousia tangfeifanii',\n",
       " 'CP016826.1': 'Corynebacterium pseudotuberculosis',\n",
       " 'CP018074.1': 'Streptomyces venezuelae',\n",
       " 'CP018075.1': 'Streptomyces venezuelae',\n",
       " 'CP015961.1': 'Dietzia timorensis',\n",
       " 'CP015866.1': 'Streptomyces parvulus',\n",
       " 'CP015867.1': 'Streptomyces parvulus',\n",
       " 'CP013979.1': 'Agromyces aureus',\n",
       " 'CP012117.1': 'Dermabacter vaginalis',\n",
       " 'CP014989.1': 'Serinicoccus hydrothermalis',\n",
       " 'CP016793.1': 'Lentzea guizhouensis',\n",
       " 'CP015235.1': 'Rhodococcus fascians D188',\n",
       " 'CP015237.1': 'Rhodococcus fascians D188',\n",
       " 'CP015236.1': 'Rhodococcus fascians D188',\n",
       " 'CP007220.1': 'Mycobacteroides chelonae CCUG 47445',\n",
       " 'CP014209.1': 'Isoptericola dokdonensis DS-3',\n",
       " 'CP015515.1': 'Rathayibacter tritici',\n",
       " 'CP015622.1': 'Corynebacterium crudilactis',\n",
       " 'CP015623.1': 'Corynebacterium crudilactis',\n",
       " 'CP015624.1': 'Corynebacterium crudilactis',\n",
       " 'CP015079.1': 'Nocardioides dokdonensis FR1436',\n",
       " 'CP011491.1': 'Mycolicibacterium vaccae 95051',\n",
       " 'CP011502.1': 'Aeromicrobium erythreum',\n",
       " 'CP016279.1': 'Streptomyces griseochromogenes',\n",
       " 'CP014228.1': 'Actinomyces radicidentis',\n",
       " 'CP014635.1': 'Corynebacterium simulans',\n",
       " 'CP011530.1': 'Mycobacteroides immunogenum',\n",
       " 'CP015098.1': 'Streptomyces qaidamensis',\n",
       " 'CP011853.1': 'Gordonia phthalatica',\n",
       " 'CP011269.1': 'Mycolicibacterium fortuitum',\n",
       " 'CP013126.1': 'Acidipropionibacterium acidipropionici',\n",
       " 'CP010271.1': 'Mycobacteroides saopaulense',\n",
       " 'CP013254.1': 'Kocuria flava',\n",
       " 'CP013255.1': 'Kocuria flava',\n",
       " 'CP013256.1': 'Kocuria flava',\n",
       " 'CP013257.1': 'Kocuria flava',\n",
       " 'CP013258.1': 'Kocuria flava',\n",
       " 'CP006841.1': 'Corynebacterium lactis RW2-5',\n",
       " 'CP012507.1': 'Kocuria palustris',\n",
       " 'CP009220.1': 'Corynebacterium deserti GIMN1.010',\n",
       " 'CP009221.1': 'Corynebacterium deserti GIMN1.010',\n",
       " 'CP009222.1': 'Corynebacterium deserti GIMN1.010',\n",
       " 'CP011340.1': 'Streptomyces pristinaespiralis',\n",
       " 'CP010413.1': 'Bifidobacterium breve',\n",
       " 'CP010414.1': 'Bifidobacterium breve',\n",
       " 'CP009312.1': 'Lawsonella clevelandensis',\n",
       " 'CP012750.1': 'Glutamicibacter halophytocola',\n",
       " 'CP012751.1': 'Glutamicibacter halophytocola',\n",
       " 'CP012752.1': 'Kibdelosporangium phytohabitans',\n",
       " 'AP012331.1': 'Bifidobacterium scardovii JCM 12489 = DSM 13734',\n",
       " 'AP012333.1': 'Parascardovia denticolens DSM 10105 = JCM 12538',\n",
       " 'AP012334.1': 'Scardovia inopinata JCM 12537',\n",
       " 'CP011112.1': 'Luteipulveratus mongoliensis',\n",
       " 'CP011786.1': 'Bifidobacterium actinocoloniiforme DSM 22766',\n",
       " 'CP012382.1': 'Streptomyces ambofaciens ATCC 23877',\n",
       " 'CP012383.1': 'Streptomyces ambofaciens ATCC 23877',\n",
       " 'LN831790.1': 'Streptomyces leeuwenhoekii',\n",
       " 'LN831788.1': 'Streptomyces leeuwenhoekii',\n",
       " 'LN831789.1': 'Streptomyces leeuwenhoekii',\n",
       " 'CP011542.1': 'Corynebacterium mustelae',\n",
       " 'CP011543.1': 'Corynebacterium mustelae',\n",
       " 'CP011544.1': 'Corynebacterium mustelae',\n",
       " 'CP011541.1': 'Corynebacterium epidermidicanis',\n",
       " 'CP011545.1': 'Corynebacterium testudinoris',\n",
       " 'CP011546.1': 'Corynebacterium uterequi',\n",
       " 'AP012322.1': 'Bifidobacterium angulatum DSM 20098 = JCM 7096',\n",
       " 'AP012325.1': 'Bifidobacterium catenulatum DSM 16992 = JCM 1194 = LMG 11043',\n",
       " 'CP016174.1': 'Amycolatopsis orientalis',\n",
       " 'CP011005.1': 'Psychromicrobium lacuslunae',\n",
       " 'CP011006.1': 'Psychromicrobium lacuslunae',\n",
       " 'CP083405.1': 'Mycobacterium lepromatosis',\n",
       " 'CP011311.1': 'Corynebacterium camporealensis',\n",
       " 'CP011312.1': 'Corynebacterium kutscheri',\n",
       " 'CP009922.3': 'Streptomyces xiamenensis',\n",
       " 'CP006764.1': 'Corynebacterium doosanense CAU 212 = DSM 45436',\n",
       " 'CP006765.1': 'Corynebacterium doosanense CAU 212 = DSM 45436',\n",
       " 'CP009896.1': 'Pimelobacter simplex',\n",
       " 'CP021080.1': 'Streptomyces pluripotens',\n",
       " 'CP010407.1': 'Streptomyces vietnamensis',\n",
       " 'CP010408.1': 'Streptomyces vietnamensis',\n",
       " 'CP010827.1': 'Corynebacterium singulare',\n",
       " 'CP007790.1': 'Corynebacterium marinum DSM 44953',\n",
       " 'CP007792.1': 'Corynebacterium marinum DSM 44953',\n",
       " 'CP007791.1': 'Corynebacterium marinum DSM 44953',\n",
       " 'HG917972.2': 'Mycobacterium marinum E11',\n",
       " 'HG917973.1': 'Mycobacterium marinum E11',\n",
       " 'CP008802.1': 'Actinotignum schaalii',\n",
       " 'CP008889.1': 'Dermacoccus nishinomiyaensis',\n",
       " 'CP008890.1': 'Dermacoccus nishinomiyaensis',\n",
       " 'CP008953.1': 'Amycolatopsis japonica',\n",
       " 'CP008954.1': 'Amycolatopsis japonica',\n",
       " 'CP008944.1': 'Corynebacterium atypicum',\n",
       " 'CP008945.1': 'Corynebacterium atypicum',\n",
       " 'CP007287.1': 'Bifidobacterium coryneforme',\n",
       " 'CP009215.1': 'Corynebacterium ureicelerivorans',\n",
       " 'CP009216.1': 'Corynebacterium ureicelerivorans',\n",
       " 'CP013745.1': 'Arthrobacter alpinus',\n",
       " 'CP013746.1': 'Arthrobacter alpinus',\n",
       " 'CP004353.1': 'Corynebacterium vitaeruminis DSM 20294',\n",
       " 'CP006365.1': 'Corynebacterium argentoratense DSM 44202',\n",
       " 'CP007519.1': 'Trueperella pyogenes',\n",
       " 'CP006842.1': 'Corynebacterium glyciniphilum AJ 3170',\n",
       " 'CP006843.1': 'Corynebacterium glyciniphilum AJ 3170',\n",
       " 'CP007255.1': 'Rhodococcus erythropolis R138',\n",
       " 'CP007256.1': 'Rhodococcus erythropolis R138',\n",
       " 'CP007257.1': 'Rhodococcus erythropolis R138',\n",
       " 'CP007490.1': 'Rhodoluna lacicola',\n",
       " 'CP006018.1': 'Bifidobacterium indicum LMG 11587 = DSM 20214',\n",
       " 'CP006734.1': 'Leifsonia xyli subsp. cynodontis DSM 46306',\n",
       " 'CP006272.1': 'Actinoplanes friuliensis DSM 7358',\n",
       " 'CP006996.1': 'Rhodococcus pyridinivorans SB3094',\n",
       " 'CP006997.1': 'Rhodococcus pyridinivorans SB3094',\n",
       " 'CP006998.1': 'Rhodococcus pyridinivorans SB3094',\n",
       " 'CP007155.1': 'Kutzneria albida DSM 43870',\n",
       " 'CP004350.1': 'Corynebacterium casei LMG S-19264',\n",
       " 'CP004351.1': 'Corynebacterium casei LMG S-19264',\n",
       " 'CP004352.1': 'Corynebacterium casei LMG S-19264',\n",
       " 'CP003410.1': 'Amycolatopsis keratiniphila',\n",
       " 'CP003411.1': 'Amycolatopsis keratiniphila',\n",
       " 'CP004048.1': 'Corynebacterium glutamicum SCgG2',\n",
       " 'CP003696.1': 'Corynebacterium terpenotabidum Y-11',\n",
       " 'CP003924.1': 'Corynebacterium maris DSM 45190',\n",
       " 'CP003925.1': 'Corynebacterium maris DSM 45190',\n",
       " 'CP006259.1': 'Streptomyces collinus Tu 365',\n",
       " 'CP006260.1': 'Streptomyces collinus Tu 365',\n",
       " 'CP006261.1': 'Streptomyces collinus Tu 365',\n",
       " 'CP003777.1': 'Amycolatopsis mediterranei RB',\n",
       " 'CP048261.1': 'Streptomyces rimosus subsp. rimosus ATCC 10970',\n",
       " 'CP048262.1': 'Streptomyces rimosus subsp. rimosus ATCC 10970',\n",
       " 'CP011883.2': 'Mycobacterium haemophilum DSM 44634',\n",
       " 'CP003697.1': 'Corynebacterium halotolerans YIM 70093 = DSM 44683',\n",
       " 'CP003698.1': 'Corynebacterium halotolerans YIM 70093 = DSM 44683',\n",
       " 'CP004354.1': 'Corynebacterium callunae DSM 20147',\n",
       " 'CP004355.1': 'Corynebacterium callunae DSM 20147',\n",
       " 'CP004356.1': 'Corynebacterium callunae DSM 20147',\n",
       " 'CP004370.1': 'Streptomyces albidoflavus',\n",
       " 'CP005287.1': 'Cutibacterium avidum 44067',\n",
       " 'CP003293.1': 'Cutibacterium acnes HL096PA1',\n",
       " 'CP003294.1': 'Cutibacterium acnes HL096PA1',\n",
       " 'AP012204.1': 'Microlunatus phosphovorus NM-1',\n",
       " 'CP072931.1': 'Streptomyces auratus AGR0001',\n",
       " 'AP012319.1': 'Actinoplanes missouriensis 431',\n",
       " 'CP003325.1': 'Bifidobacterium asteroides PRL2011',\n",
       " 'HE804045.1': 'Saccharothrix espanaensis DSM 44229',\n",
       " 'CP003039.2': 'Bifidobacterium animalis subsp. lactis BLC1',\n",
       " 'FQ859185.1': 'Streptantibioticus cattleyicolor NRRL 8057 = DSM 46488',\n",
       " 'FQ859184.1': 'Streptantibioticus cattleyicolor NRRL 8057 = DSM 46488',\n",
       " 'CP003119.1': 'Gordonia polyisoprenivorans VH2',\n",
       " 'CP003120.1': 'Gordonia polyisoprenivorans VH2',\n",
       " 'HE572590.1': 'Mycobacterium canettii CIPT 140010059',\n",
       " 'AP010968.1': 'Kitasatospora setae KM-6054',\n",
       " 'CP002638.1': 'Micromonospora maris AB-18-032',\n",
       " 'CP002639.1': 'Micromonospora maris AB-18-032',\n",
       " 'CP002666.1': 'Cellulomonas fimi ATCC 484',\n",
       " 'CP002329.1': 'Mycolicibacter sinensis',\n",
       " 'CP002786.1': 'Hoyosella subflava DQS3-9A1',\n",
       " 'CP002787.1': 'Hoyosella subflava DQS3-9A1',\n",
       " 'CP002788.1': 'Hoyosella subflava DQS3-9A1',\n",
       " 'CP002810.1': 'Isoptericola variabilis 225',\n",
       " 'CP002790.1': 'Corynebacterium ulcerans 809',\n",
       " 'CP002665.1': 'Cellulomonas gilvus ATCC 13127',\n",
       " 'CP002857.1': 'Corynebacterium resistens DSM 45100',\n",
       " 'CP002385.1': 'Mycolicibacterium gilvum Spyr1',\n",
       " 'CP002386.1': 'Mycolicibacterium gilvum Spyr1',\n",
       " 'CP002387.1': 'Mycolicibacterium gilvum Spyr1',\n",
       " 'CP002343.1': 'Intrasporangium calvum DSM 43043',\n",
       " 'AL123456.3': 'Mycobacterium tuberculosis H37Rv',\n",
       " 'BX072543.1': 'Tropheryma whipplei TW08/27',\n",
       " 'CP001992.1': 'Mobiluncus curtisii ATCC 43063',\n",
       " 'AP010888.1': 'Bifidobacterium longum subsp. longum JCM 1217',\n",
       " 'CP001958.1': 'Segniliparus rotundus DSM 44985',\n",
       " 'CP001964.1': 'Cellulomonas flavigena DSM 20109',\n",
       " 'CP002162.1': 'Micromonospora aurantiaca ATCC 27029',\n",
       " 'CP006835.1': 'Mycobacterium kansasii ATCC 12478',\n",
       " 'CP006836.1': 'Mycobacterium kansasii ATCC 12478',\n",
       " 'CP002280.1': 'Rothia dentocariosa ATCC 17931',\n",
       " 'CP002299.1': 'Frankia inefficax',\n",
       " 'AM420293.1': 'Saccharopolyspora erythraea NRRL 2338',\n",
       " 'FN554889.1': 'Streptomyces scabiei 87.22',\n",
       " 'CP001966.1': 'Tsukamurella paurometabola DSM 20162',\n",
       " 'CP001967.1': 'Tsukamurella paurometabola DSM 20162',\n",
       " 'CP001802.1': 'Gordonia bronchialis DSM 43247',\n",
       " 'CP001803.1': 'Gordonia bronchialis DSM 43247',\n",
       " 'CP001814.1': 'Streptosporangium roseum DSM 43021',\n",
       " 'CP001815.1': 'Streptosporangium roseum DSM 43021',\n",
       " 'CP001819.1': 'Sanguibacter keddieii DSM 10542',\n",
       " 'CP001821.1': 'Xylanimonas cellulosilytica DSM 15894',\n",
       " 'CP001822.1': 'Xylanimonas cellulosilytica DSM 15894',\n",
       " 'CP001867.1': 'Geodermatophilus obscurus DSM 43160',\n",
       " 'CT573213.2': 'Frankia alni ACN14a',\n",
       " 'CP001736.1': 'Kribbella flavida DSM 17836',\n",
       " 'CP001737.1': 'Nakamurella multipartita DSM 44233',\n",
       " 'CP001738.1': 'Thermomonospora curvata DSM 43183',\n",
       " 'CP001778.1': 'Stackebrandtia nassauensis DSM 44728',\n",
       " 'CP001601.1': 'Corynebacterium aurimucosum ATCC 700975',\n",
       " 'FM164414.1': 'Corynebacterium aurimucosum ATCC 700975',\n",
       " 'CP001618.1': 'Beutenbergia cavernae DSM 12333',\n",
       " 'CP001620.1': 'Corynebacterium kroppenstedtii DSM 44385',\n",
       " 'CP001630.1': 'Actinosynnema mirum DSM 43827',\n",
       " 'CP001643.1': 'Brachybacterium faecium DSM 4810',\n",
       " 'CP001683.1': 'Saccharomonospora viridis DSM 43017',\n",
       " 'CP001700.1': 'Catenulispora acidiphila DSM 44928',\n",
       " 'CP000481.1': 'Acidothermus cellulolyticus 11B',\n",
       " 'CP000511.1': 'Mycolicibacterium vanbaalenii PYR-1',\n",
       " 'CP000667.1': 'Salinispora tropica CNB-440',\n",
       " 'CP000750.2': 'Kineococcus radiotolerans SRS30216 = ATCC BAA-149',\n",
       " 'CP000751.1': 'Kineococcus radiotolerans SRS30216 = ATCC BAA-149',\n",
       " 'CP000752.1': 'Kineococcus radiotolerans SRS30216 = ATCC BAA-149',\n",
       " 'CP000910.1': 'Renibacterium salmoninarum ATCC 33209',\n",
       " 'CP001341.1': 'Pseudarthrobacter chlorophenolicus A6',\n",
       " 'CP001342.1': 'Pseudarthrobacter chlorophenolicus A6',\n",
       " 'CP001343.1': 'Pseudarthrobacter chlorophenolicus A6',\n",
       " 'CR931997.1': 'Corynebacterium jeikeium K411',\n",
       " 'AF401314.1': 'Corynebacterium jeikeium K411',\n",
       " 'BA000030.4': 'Streptomyces avermitilis MA-4680 = NBRC 14893',\n",
       " 'AP005645.1': 'Streptomyces avermitilis MA-4680 = NBRC 14893',\n",
       " 'AP006618.1': 'Nocardia farcinica IFM 10152',\n",
       " 'AP006619.1': 'Nocardia farcinica IFM 10152',\n",
       " 'AP006620.1': 'Nocardia farcinica IFM 10152',\n",
       " 'AP009493.1': 'Streptomyces griseus subsp. griseus NBRC 13350',\n",
       " 'BA000035.2': 'Corynebacterium efficiens YS-314',\n",
       " 'AP005225.1': 'Corynebacterium efficiens YS-314',\n",
       " 'AP005226.1': 'Corynebacterium efficiens YS-314',\n",
       " 'CP000249.1': 'Frankia casuarinae',\n",
       " 'LR214441.1': 'Tessaracoccus lapidicaptus',\n",
       " 'LR738855.1': 'Corynebacterium rouxii',\n",
       " 'LR134355.1': 'Mycolicibacterium chitae',\n",
       " 'LR134363.1': 'Actinomyces slackii'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "actinobacteria_names_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fc08fc8-7948-4b47-8ae8-d055db255e76",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
