{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "0cb2f38a-3c12-4916-8913-b1644afe105e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  \n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, AlignIO, pairwise2, Align, Seq, motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from scipy.stats import binom\n",
    "import math\n",
    "import time\n",
    "from tqdm.auto import tqdm\n",
    "import random\n",
    "import copy\n",
    "from joblib import Parallel, delayed\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import re\n",
    "import shutil\n",
    "from Comparative_Analysis import Sequence_Analysis_Routines as sar\n",
    "from Comparative_Analysis import Utilities as util\n",
    "from Comparative_Analysis import Alignment as align\n",
    "from Comparative_Analysis import Blast_Functions as blastfn\n",
    "from scipy.stats import chi2\n",
    "from IPython import display\n",
    "\n",
    "num_cores = 16\n",
    "core_numbers = list(range(1, num_cores+1))\n",
    "project_dir = 'D:/Project_Data/Project_9/Mycobacteriaceae'\n",
    "output_dir = project_dir + '/Output'\n",
    "seq_dir = 'F:/Datasets/ncbi_dataset/ncbi_dataset/data'\n",
    "#seq_dir = 'D:/Project_Data/Project_8/Datasets/Actinobacteria_Ref_Rep_Lev_Complete'\n",
    "tb_species = 'NC_000962.3' \n",
    "tb_genome_filename = 'GCA_000195955.2/genomic.gbff'\n",
    "\n",
    "e_val = 1e-10\n",
    "produce_data = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "c54006ed-6479-49a8-9e50-546d25dd8b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(seq_string):\n",
    "    complement_dict = {'A':'T','C':'G','G':'C','T':'A'}\n",
    "    temp = []\n",
    "    for char in reversed(seq_string):\n",
    "        if char in complement_dict:\n",
    "            temp.append(complement_dict[char])\n",
    "        else:\n",
    "            return('')\n",
    "    return ''.join(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "b7058cb0-8529-4a01-af5d-58760d251c00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10669\n",
      "10669\n"
     ]
    }
   ],
   "source": [
    "file_list_temp = util.list_dirs(seq_dir)\n",
    "file_list = [x for x in file_list_temp if os.path.exists(seq_dir + '/' + x + '/genomic.gbff.temp')]\n",
    "print(len(file_list_temp))\n",
    "print(len(file_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "0094af3e-f50e-4641-809f-641cd3045754",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_rprotein_list(num_subsets, subset_num, id_list):\n",
    "    ids = util.chunk_list(id_list, num_subsets, subset_num)\n",
    "    temp = []\n",
    "    for filename in ids:\n",
    "        for record in SeqIO.parse(seq_dir + '/' + filename + '/genomic.gbff.temp', \"genbank\"):\n",
    "            organism_name = record.annotations['organism']\n",
    "            taxonomy = record.annotations['taxonomy']\n",
    "            accession_ver = record.annotations['accessions'][0] + '.' + str(record.annotations['sequence_version'])\n",
    "            organism_accession = organism_name.replace(' ', '_') + '_' + accession_ver\n",
    "            full_sequence = str(record.seq)\n",
    "            features = []\n",
    "            for feature in record.features:\n",
    "                a = feature.qualifiers\n",
    "                if a.get(\"product\")!= None and a.get(\"locus_tag\")!= None:\n",
    "                    locus_tag = a.get(\"locus_tag\")[0]\n",
    "                    accession_locus = accession_ver + '@' + locus_tag\n",
    "                    product = a.get(\"product\")[0]\n",
    "                    features.append([(accession_locus, product), int(feature.location.start), int(feature.location.end), int(feature.location.strand)])\n",
    "            features.sort(key=lambda x: x[1])\n",
    "\n",
    "            for i, feature in enumerate(features):\n",
    "                if 'ibosomal' in feature[0][1] and 'protein' in feature[0][1]:\n",
    "                    exception = 1\n",
    "                    if feature[1] < feature[2]:  \n",
    "                        if feature[3] == 1:\n",
    "                            if i > 0 and feature[1] > features[i-1][2]:\n",
    "                                exception = 0\n",
    "                                utr_coords = (features[i-1][2], feature[1])\n",
    "                                utr_sequence = full_sequence[features[i-1][2]: feature[1]]\n",
    "                                utr_length = len(utr_sequence)\n",
    "                        else:\n",
    "                            if i + 1 < len(features) and feature[2] < features[i+1][1]:\n",
    "                                exception = 0\n",
    "                                utr_coords = (feature[2], features[i+1][1])\n",
    "                                utr_sequence =  reverse_complement(full_sequence[feature[2]: features[i+1][1]])\n",
    "                                utr_length = len(utr_sequence)\n",
    "                    if exception == 1:\n",
    "                        utr_coords = (0,0)\n",
    "                        utr_sequence = ''\n",
    "                        utr_length = 0 \n",
    "                    temp.append([organism_name, feature[0], taxonomy, (feature[1], feature[2], feature[3]), (utr_coords, utr_sequence)])\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b7c8b61-c1da-4521-ba3f-c0c76f7284e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if produce_data == True:\n",
    "    parallel_output = Parallel(n_jobs=-1)(delayed(generate_rprotein_list)(num_cores, core_number, file_list) for core_number in core_numbers)\n",
    "    temp = [item for sublist in parallel_output for item in sublist]\n",
    "    rprotein_dict = {}\n",
    "    name_dict = {}\n",
    "    for x in temp:\n",
    "        if not(x[1][0].split('@')[0] in name_dict):\n",
    "            name_dict[x[1][0].split('@')[0]] = (x[0], x[2])\n",
    "        rprotein_dict[x[1][0]] = [x[1][1], x[3], x[4]]\n",
    "    with open(output_dir + '/name_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(name_dict, f) \n",
    "    with open(output_dir + '/rprotein_dict.pkl', 'wb') as f:\n",
    "        pickle.dump(rprotein_dict, f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "e8c54ef0-8e30-4c77-bc7d-feacf5b5e917",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_dir + '/name_dict.pkl', 'rb') as f:\n",
    "    name_dict = pickle.load(f) \n",
    "with open(output_dir + '/rprotein_dict.pkl', 'rb') as f:\n",
    "    rprotein_dict = pickle.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "5fe90a46-1503-4bcd-8632-b9165051e2d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Rv0053']\n",
      "['Rv0055']\n",
      "['Rv0056']\n",
      "['Rv0105c']\n",
      "['Rv0634B']\n",
      "['Rv0640']\n",
      "['Rv0641']\n",
      "['Rv0651']\n",
      "['Rv0652']\n",
      "['Rv0682']\n",
      "['Rv0683']\n",
      "['Rv0700']\n",
      "['Rv0701']\n",
      "['Rv0702']\n",
      "['Rv0703']\n",
      "['Rv0704']\n",
      "['Rv0705']\n",
      "['Rv0706']\n",
      "['Rv0707']\n",
      "['Rv0708']\n",
      "['Rv0709']\n",
      "['Rv0710']\n",
      "['Rv0714']\n",
      "['Rv0715']\n",
      "['Rv0716']\n",
      "['Rv0717']\n",
      "['Rv0718']\n",
      "['Rv0719']\n",
      "['Rv0720']\n",
      "['Rv0721']\n",
      "['Rv0722']\n",
      "['Rv0723']\n",
      "['Rv0979A']\n",
      "['Rv0995']\n",
      "['Rv1015c']\n",
      "['Rv1298']\n",
      "['Rv1630']\n",
      "['Rv1642']\n",
      "['Rv1643']\n",
      "['Rv2055c']\n",
      "['Rv2056c']\n",
      "['Rv2057c']\n",
      "['Rv2058c']\n",
      "['Rv2412']\n",
      "['Rv2441c']\n",
      "['Rv2442c']\n",
      "['Rv2785c']\n",
      "['Rv2890c']\n",
      "['Rv2904c']\n",
      "['Rv2909c']\n",
      "['Rv3420c']\n",
      "['Rv3442c']\n",
      "['Rv3443c']\n",
      "['Rv3456c']\n",
      "['Rv3458c']\n",
      "['Rv3459c']\n",
      "['Rv3460c']\n",
      "['Rv3461c']\n",
      "['Rv3924c']\n",
      "['RpsF', 'RpsR1', 'RplI', 'RpmB1', 'RpmG2', 'RplK', 'RplA', 'RplJ', 'L7', 'RpsL', 'RpsG', 'NusE)', 'RplC', 'RplD', 'RplW', 'RplB', 'RpsS', 'RplV', 'RpsC', 'RplP', 'RpmC', 'RpsQ', 'RplN', 'RplX', 'RplE', 'RpsN1', 'RpsH', 'RplF', 'RplR', 'RpsE', 'RpmD', 'RplO', 'RpmF', 'S5)', 'RplY', 'RpmE', 'RpsA', 'RpmI', 'RplT', 'RpsR2', 'RpsN2', 'RpmG1', 'RpmB2', 'RpsT', 'RpmA', 'RplU', 'RpsO', 'RpsB', 'RplS', 'RpsP', 'S18)', 'RpsI', 'RplM', 'RplQ', 'RpsD', 'RpsK', 'RpsM', 'RpmJ', 'RpmH']\n"
     ]
    }
   ],
   "source": [
    "ribosomal_protein_ids = []\n",
    "for record in SeqIO.parse(seq_dir + '/' + tb_genome_filename, \"genbank\"):\n",
    "            for feature in record.features:\n",
    "                if 1==1:\n",
    "                #if not(feature.type == 'gene'):\n",
    "                    a = feature.qualifiers\n",
    "                    if a.get(\"product\") != None:\n",
    "                        if 'ibosomal' in a.get(\"product\")[0] and 'protein' in a.get(\"product\")[0]:\n",
    "                            ribosomal_protein_ids.append(a.get(\"product\")[0].split('/')[0].split(' ')[-1:][0])\n",
    "                            print(a.get(\"locus_tag\"))\n",
    "print(ribosomal_protein_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "91b630c4-2b33-4844-8890-893c95f8681e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ribosomal_protein_ids = ['L10']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "34e63b2d-cf35-45f8-92b3-6ab6e1ab1685",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4892\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4892/4892 [00:00<00:00, 206073.53it/s]\n"
     ]
    }
   ],
   "source": [
    "for rprotein_id in ribosomal_protein_ids:    \n",
    "    temp_results_dir = output_dir + '/' + rprotein_id\n",
    "    results_dir = temp_results_dir.replace(' ','_')\n",
    "    wsl_results_dir = util.wslname(results_dir)\n",
    "    malign_output_filename = 'malign.faa'\n",
    "    sequence_list = []\n",
    "    \n",
    "    if not os.path.exists(results_dir):\n",
    "        os.makedirs(results_dir)\n",
    "    \n",
    "    for k, v in rprotein_dict.items():\n",
    "        taxonomy_info = name_dict[k.split('@')[0]][1]\n",
    "        if 'Corynebacteriales' in taxonomy_info and rprotein_id in v[0].split('/')[0]:\n",
    "            sequence_list.append([k,v[2][1]])\n",
    "    print (len(sequence_list))\n",
    "    region_file_name = 'all_upstream_regions.faa'\n",
    "    util.produce_fasta_file(sequence_list, results_dir + '/' + region_file_name)\n",
    "\n",
    "    result = subprocess.run('wsl echo magpie | sudo -S mafft ' + wsl_results_dir + '/' + region_file_name + ' > ' + wsl_results_dir + '/' + malign_output_filename, capture_output=True, text=True)\n",
    "    #muscle_exe = 'C:/Users/nicho/Muscle/muscle3.8.31_i86win32.exe'\n",
    "    #cline = MuscleCommandline(muscle_exe, input=results_dir+'/'+ region_file_name, out=results_dir + '/'+malign_output_filename)\n",
    "    #result = cline();\n",
    " \n",
    "    blastfn.hmmer_build(results_dir, malign_output_filename, 'hmm.hmm')\n",
    "\n",
    "    for i in range(3):\n",
    "        blastfn.nhmmer_search_model(results_dir, 'hmm.hmm', results_dir, region_file_name, 'align_sto', 'output.txt', 'hsummary,txt', e_val)\n",
    "        blastfn.hmmer_build(results_dir, 'align_sto', 'hmm.hmm')\n",
    "\n",
    "\n",
    "    blastfn.nhmmer_search_model(results_dir, 'hmm.hmm', results_dir, region_file_name, 'align.sto', 'output.txt', 'hsummary,txt', e_val)\n",
    "\n",
    "    blastfn.infernal_build_and_calib(results_dir, 'align.sto' ,'cm_1.cm', False)\n",
    "    blastfn.infernal_search(results_dir, 'cm_1.cm', results_dir, region_file_name, 'search_1.sto', 'search_hits_1.txt', 'summary_1.txt', e_val) \n",
    "    if not os.path.exists(results_dir + '/search_hits_1.txt'):\n",
    "        continue\n",
    "    if os.path.getsize(results_dir + '/search_hits_1.txt') > 1e8:\n",
    "        continue\n",
    "    blastfn.run_rscape(results_dir, 'search_1.sto', 'rscape_1')\n",
    "\n",
    "\n",
    "    #2\n",
    "    blastfn.infernal_build_and_calib(results_dir, 'rscape_1.cacofold.R2R.sto' ,'cm_2.cm')\n",
    "    blastfn.infernal_search(results_dir, 'cm_2.cm', results_dir, region_file_name, 'search_2.sto', 'search_hits_2.txt', 'summary_2.txt', e_val) \n",
    "    if not os.path.exists(results_dir + '/search_hits_2.txt'):\n",
    "        continue\n",
    "    if os.path.getsize(results_dir + '/search_hits_2.txt') > 1e8:\n",
    "        continue\n",
    "    blastfn.run_rscape(results_dir, 'search_2.sto', 'rscape_2')\n",
    "\n",
    "    #3\n",
    "\n",
    "    blastfn.infernal_build_and_calib(results_dir, 'rscape_2.cacofold.R2R.sto' ,'cm_3.cm')\n",
    "    blastfn.infernal_search(results_dir, 'cm_3.cm', results_dir, region_file_name, 'search_3.sto', 'search_hits_3.txt', 'summary_3.txt', e_val) \n",
    "    if not os.path.exists(results_dir + '/search_hits_3.txt'):\n",
    "        continue\n",
    "    if os.path.getsize(results_dir + '/search_hits_3.txt') > 1e8:\n",
    "        continue\n",
    "    blastfn.run_rscape(results_dir, 'search_3.sto', 'rscape_3')\n",
    "\n",
    "    #Final\n",
    "\n",
    "    #blastfn.infernal_build_and_calib(output_dir, 'rscape_3.cacofold.R2R.sto' ,'cm_4.cm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "af71d84e-479e-4a74-8a38-3ba6410f74aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000017.1@H1J40_RS05720',\n",
       "   '50S ribosomal protein L25/general stress protein Ctc'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (37447, 38095, -1),\n",
       "  ((38095, 38311),\n",
       "   'MAKSASNQLRVTVRTETGKGASRRARRAGKIPAVLYGHGAEPQHLELPGHDYAAVLRHSGTNAVLTLDIAGKEQLALTKALHIHPIRRTIQHADLLVVRRGEKVVVEVSVVVEGQAGPDTLVTQETNSIEIEAEALSIPEQLTVSIEGAEPGTQLTAGQIALPAGVSLISDPDLLVVNVVKAPTAEELEGEVAGAEEAEEAAVEAGEAEAAGESE*')]]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translator = util.Translator()\n",
    "temp = []\n",
    "for record in SeqIO.parse('C:/Users/nicho/Downloads/ncbi_dataset/ncbi_dataset/data/GCF_013697105.1/genomic.gbff', \"genbank\"):\n",
    "        organism_name = record.annotations['organism']\n",
    "        taxonomy = record.annotations['taxonomy']\n",
    "        accession_ver = record.annotations['accessions'][0] + '.' + str(record.annotations['sequence_version'])\n",
    "        organism_accession = organism_name.replace(' ', '_') + '_' + accession_ver\n",
    "        full_sequence = str(record.seq)\n",
    "        features = []\n",
    "        for feature in record.features:\n",
    "            a = feature.qualifiers\n",
    "            if a.get(\"product\")!= None and a.get(\"locus_tag\")!= None:\n",
    "                locus_tag = a.get(\"locus_tag\")[0]\n",
    "                accession_locus = accession_ver + '@' + locus_tag\n",
    "                product = a.get(\"product\")[0]\n",
    "                features.append([(accession_locus, product), int(feature.location.start), int(feature.location.end), int(feature.location.strand)])\n",
    "        features.sort(key=lambda x: x[1])\n",
    "        for i, feature in enumerate(features):\n",
    "            if 'ibosomal' in feature[0][1] and 'protein' in feature[0][1] and 'L25' in feature[0][1]:\n",
    "                exception = 1\n",
    "                if feature[1] < feature[2]:  \n",
    "                    if feature[3] == 1:\n",
    "                        if i > 0 and feature[1] > features[i-1][2]:\n",
    "                            exception = 0\n",
    "                            utr_coords = (features[i-1][2], feature[1])\n",
    "                            utr_sequence = full_sequence[features[i-1][2]: feature[1]]\n",
    "                            utr_length = len(utr_sequence)\n",
    "                    else:\n",
    "                        if i + 1 < len(features) and feature[2] < features[i+1][1]:\n",
    "                            seq = translator.translate_sequence(full_sequence[feature[1]:feature[2]], -1, 0)\n",
    "                            exception = 0\n",
    "                            utr_coords = (feature[2], features[i+1][1])\n",
    "                            utr_sequence =  reverse_complement(full_sequence[feature[2]: features[i+1][1]])\n",
    "                            utr_length = len(utr_sequence)\n",
    "                if exception == 1:\n",
    "                    utr_coords = (0,0)\n",
    "                    utr_sequence = ''\n",
    "                    utr_length = 0 \n",
    "                temp.append([organism_name, feature[0], taxonomy, (feature[1], feature[2], feature[3]), (utr_coords, seq)])\n",
    "temp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "cb5388be-970a-4521-96bc-47b4d71a0154",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04060', '30S ribosomal protein S10'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (740, 1046, 1),\n",
       "  ((0, 0), '')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04085', '30S ribosomal protein S19'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (3818, 4100, 1),\n",
       "  ((3778, 3818), 'CCGCGCAATCAGATCTAGGGAGTTTCAGGAGTAGCCAACC')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04095', '30S ribosomal protein S3'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (4722, 5514, 1),\n",
       "  ((4690, 4722), 'TGGGCCAGAAGATCAATCCGCACGGCTTCCGG')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04110', '30S ribosomal protein S17'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (6166, 6574, 1),\n",
       "  ((0, 0), '')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04145', 'type Z 30S ribosomal protein S14'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (13034, 13220, 1),\n",
       "  ((13030, 13034), 'GCAG')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04150', '30S ribosomal protein S8'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (13516, 13915, 1),\n",
       "  ((13220, 13516),\n",
       "   'CGGGACACGGGGACTAGAACATATGACCGCGCTGACGACGATGCAGTGGGGGTACCCCCNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNAGACGCGCAGCGGCGAGGGGGCCGCAAGCGATGAGGAGGAGTAGCGCTCAATGACCGCGCTGACGACGATGCAGTGGGGGTACCCCCAGACGCGCAGCGGCGAGGGGGCCGCAAGCGATGAGGAGGAGTAGCGCTCA')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000011.1@H1J40_RS04165', '30S ribosomal protein S5'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (14868, 15531, 1),\n",
       "  ((14849, 14868), 'TGAACGGGAGGACCGCATA')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000047.1@H1J40_RS11070', '30S ribosomal protein S18'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (50480, 50750, -1),\n",
       "  ((0, 0), '')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000047.1@H1J40_RS11075', '30S ribosomal protein S14'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (50750, 51056, -1),\n",
       "  ((51056, 51057), 'C')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000001.1@H1J40_RS00690', '30S ribosomal protein S6'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (148690, 148981, 1),\n",
       "  ((148472, 148690),\n",
       "   'TCTGTTCGCGATAGGCCTCGCCGTTCGCGACACTGACATTGCGCACACGACACGCCGCGGATCGTCGCACCGGGTTAAGCCTGGAGTGCGGTGGTGCCTGGTCGGCATTTTCGCAGTCGAGGGCTCTCGTGTAGCCTGGGCGAGTTGCCGACGCAGGCGACCCTCCTGCCACGGATCGACCGTGGCCGCACACGACCACAGGAGGTGATGAGGTTCCT')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000001.1@H1J40_RS00700', '30S ribosomal protein S18'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (149620, 149875, 1),\n",
       "  ((149579, 149620), 'CCCCAAGAACTGCAAATCAAGAAACGGAAAGATAGACACTC')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000084.1@H1J40_RS18255', '30S ribosomal protein S9'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (21509, 21965, -1),\n",
       "  ((0, 0), '')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000084.1@H1J40_RS18330', '30S ribosomal protein S4'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (38651, 39257, -1),\n",
       "  ((39257, 39265), 'GAGAGAAG')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000084.1@H1J40_RS18335', '30S ribosomal protein S11'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (39265, 39685, -1),\n",
       "  ((39685, 39688), 'CCG')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000084.1@H1J40_RS18340', '30S ribosomal protein S13'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (39688, 40063, -1),\n",
       "  ((40063, 40392),\n",
       "   'TTCGGCGACGATGCGGGCCGCGTAGTGGTCCGAGGAGGAGCCGGACAGTTGGATTTGGTCGGCGACGATGCGGGCCGCGTAGTGGTCCGAGGAGGAGCCGGACAGTTGGATTTGGTCGGCGACGATGCGGGCCGCGTAGTGGTCCGAGGAGGAGCCGGCCGCCCCGACAACTAAATGCAGACCCCCAGCACCAATGAGCGGCTGGCATCGATGGAGATGGGCCAGCTCATTCACGCCCGGACGGAGGCCGGGCCCCACACCCAAAGAGCTTTTTTGGGCAGGGAACGGACTGGGACCATACCTCCGCAGAGGAAGAGGAAACGCCACCT')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000062.1@H1J40_RS14880', '30S ribosomal protein S15'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (20365, 20629, -1),\n",
       "  ((20629, 20792),\n",
       "   'TGGTTCTTGCGGTCGTCACCCCCGGCTGGCACCTGCGGCGTTTGTGTTTGGCCCCTCAACCCGCTGCTAGACTGCCCAACGACATCGGTGCGTGCTGCGGTTCGCGGTGGCCGTGCCCTGATATCCATGATCGCGAACCCGTTGATGGAGACATTTCGTGGCG')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000064.1@H1J40_RS15475', '30S ribosomal protein S2'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (76932, 77796, -1),\n",
       "  ((77796, 78079),\n",
       "   'TCGCCGCCGCCCGTTTTGTGGACAATGTCGTCGCTGTGGATTCCGCTGCCGGGAAAATCACCAGCTGGGTCGGTGTAAACTACTCGGCGCAGCTCGCATCGGCGGGCTGACTTCGCGTGTCTGCACCGCGGCCCGCCTTTCCGGGATCCGCAACAGCATGCCGGGGCGGTCCCGCGGCGGTTCCCTTCGGGGATTTCCTCCCGGGTCCGGCATCGCGGCAGACGCCAGGGCCCGGCTTCAGCCGATGCCGGGCGAACAACCGACACATGAGTAAGGCATAGGT')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000064.1@H1J40_RS15570', '30S ribosomal protein S16'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (94968, 95457, -1),\n",
       "  ((95457, 95640),\n",
       "   'AGCCCGCTGGGCGAATTACTGCCCCAGCATCACGTTCTGGCACAATTGGCGGCTGTCCACCGCGCGACGACTTGTCCTGTTCGGGGTTGCCGCCCGGTGGTCATACACGCGAGGCAAAACCGGATCCGGGCATCCCGCCCGGGTCGCTGAATTGCAGCGTGACACCAATAGGAGCGCTCAATC')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000010.1@H1J40_RS03955', '30S ribosomal protein S12'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (5209, 5584, 1),\n",
       "  ((4961, 5209),\n",
       "   'CGCCCTGCTTCGGCCGCAACGCCCGCTTTGACCTGCCAGACTGGCGGCGGGTATTGTGGTTGCTCGTGCCTGGCGGCTTACGCCTGATGTAGGGGCGTGGATGCCGGGCCAATTCGCATGTCCGCGATGCCTCGGATGAGACGAATCGAGTTTGAGGCAAGCTATGCGACACACCCGGCCGCGGGTAACCGTGGCGGGGCATGGCCGACAAACAGAACGTGAAAGCGCCCAAGATAGAAAGCCGGTAG')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000010.1@H1J40_RS03960', '30S ribosomal protein S7'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (5583, 6054, 1),\n",
       "  ((0, 0), '')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000056.1@H1J40_RS12970', '30S ribosomal protein S20'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (9649, 9916, 1),\n",
       "  ((9546, 9649),\n",
       "   'GCTGCCTCAAATTCGACATTCCGGCAGACCACAGATTCCGCTTTTGGGCGAAAACGTAACCGACTGATAACCTGGGCAGCCGAATCACACCGACAAAGGGAAC')],\n",
       " ['Mycobacterium canettii',\n",
       "  ('NZ_JACEGU010000037.1@H1J40_RS08835', '30S ribosomal protein S1'),\n",
       "  ['Bacteria',\n",
       "   'Actinobacteria',\n",
       "   'Corynebacteriales',\n",
       "   'Mycobacteriaceae',\n",
       "   'Mycobacterium',\n",
       "   'Mycobacterium tuberculosis complex'],\n",
       "  (58898, 60344, 1),\n",
       "  ((58736, 58898),\n",
       "   'GCGCCGAGCGTGCATCTGGGGCGGGAATTCGGCGATTTTTCCGCCCTGAGTTCACACTCGGCGCAATCGGGACCGAGTTTGTCCAGCGTGTACCCCGTCGAGTAGCCTCGTCAGGTACCAATCTGTCCCTACGACCAACCCTGTCCGGAGCAACCCAACAAT')]]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb4fbf95-38fd-447a-8de8-5b30262661df",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
